---
title: "spatialGE: Basic tutorial for visualization and spatial heterogeneity analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{spatialGE_tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

The package `spatialGE` provides a set of tools for the visualization of gene expression from 
spatially-resolved arrays, such as those generated by the 10X Visium platform. 

## Installation
The `spatialGE` repository is available at GitHub and can be installed via `devtools`.
To install `devtools` in case is not installed in your R console, please run the 
following code:
```{r}
if("devtools" %in% rownames(installed.packages()) == FALSE) {
  install.packages("devtools")
}
```

After making sure `devtools` is installed, proceed to install `spatialGE`:
```{r}
#devtools::install_github("fridleylab/spatialGEdev")
```

To use `spatialGE`, load the package using the command:
```{r setup}
library(spatialGEdev)
```

```{r echo=FALSE}
# fx_dir <- list.files(path="~/OneDrive - Moffitt Cancer Center/SPATIAL_TRANSCRIPTOMICS/code/spatialGEdev/R", 
#            full.names=T)
# fx_dir <- grep(".R$", fx_dir, value=T)
# sapply(fx_dir, FUN=source)
```

## Spatially-resolved expression of melanoma stage III tumor biopsies
To show the utility of `spatialGE`, we use the Spatial Transcriptomics data set
generated by Thrane et al. (2018) that included lymph node biopsies from four stage
IIIc melanoma patients, with two arrays per biopsy (DOI: 10.1158/0008-5472.CAN-18-0747).
The spatial slides used in this melanoma comprised 1,007 'spots' with 200μM of distance
between their centers. The spots had a diameter of 100μM, covering 5-40 cells according
to the authors.

In the data provided as example for `spatialGE`, we have removed probe names and duplicate
gene names (resulting from different probes). The user can find the original raw data at
the authors website (https://www.spatialresearch.org/resources-published-datasets/doi-10-1158-0008-5472-can-18-0747/).

## Creating an STList (Spatial Transcriptomics List)
Raw and processed data are stored in an STList (an R S4 class object). The STList
can be created with the function `STList()`, which takes minimally two files:
1. A text file containing file paths of gene raw count matrices, one per line. The gene
count matrices must be tab-separated, with gene names in the first column, and 
spots in subsequent columns. Matrices of different for each tissue slice are allowed, but
duplicated gene names are not.
2. A text file containing file paths of x,y coordinates for each spot, one per line.
The order of the coordinate file paths must match that of the raw count matrices.

Optionally, the user can provide a table with clinical/phenotype data associated with
each spatial array. This table is a comma-separated file with sample ID in the first
column, and relevant variables in subsequent columns. The order of the rows in the 
clinical file must match that of the file paths. 

The example files in this tutorial can be downloaded from the GitHub repository (https://github.com/FridleyLab/spatialGEdev/tree/main/inst/extdata). They are also
stored in you local `spatialGE` installation:
```{r}
count_files <- system.file("extdata", "fpaths_count_files.txt", package = "spatialGEdev")
coord_files <- system.file("extdata", "fpaths_mapping_files.txt", package = "spatialGEdev")
clin_file <- system.file("extdata", "thrane_clinical.csv", package = "spatialGEdev")
```

```{r echo=F}
# count_files <- "~/OneDrive - Moffitt Cancer Center/SPATIAL_TRANSCRIPTOMICS/code/spatialGEdev/inst/extdata/short_test_version_3subjs/fpaths_count_files.txt"
# coord_files <- "~/OneDrive - Moffitt Cancer Center/SPATIAL_TRANSCRIPTOMICS/code/spatialGEdev/inst/extdata/short_test_version_3subjs/fpaths_mapping_files.txt"
# clin_file <- "~/OneDrive - Moffitt Cancer Center/SPATIAL_TRANSCRIPTOMICS/code/spatialGEdev/inst/extdata/short_test_version_3subjs/thrane_clinical.csv"
```

We can load the files into an STList object like so:
```{r}
melanoma <- STList(countfiles = count_files, 
                   coordfiles = coord_files,
                   clinical = clin_file
)
```

The `melanoma` object is an STList, and contains the count data, spot coordinates,
and clinical data.
```{r}
melanoma
```

## Exploring variation between spatial arrays
The function `STbulk_pca()` allows for a quick snapshot of the variation in gene
expression among samples. The function creates 'bulk' RNA-Seq data sets by combining
all counts from a spatial array. Then, it applies limma-voom normalization to these
'bulk' RNA-Seq libraries and performs a PCA. Note that spatial (x,y) coordinate 
information is not included in this analysis, which is intended as a tool for 
data exploration. In this case, we apply the function to look for agreement between
samples from the same patient: it is expected that tissue slices from the same patient
are more similar between them, than slices from other patients. The `STbulk_pca()`
allows to map a clinical variable to the PCA (gender in this example), by including
the exact name of the column from the clinical/phenotype file.
```{r}
bulk_pca(melanoma, clinvar='Gender')
```

## Data normalization of spatially-resolved transcriptomic data
Many normalization methods are available for both bulk and single-cell
RNA-Seq data. In `spatialGE`, we apply limma-voom normalization to RNA-Seq libraries 
from each of the spots in order to obtain a gaussian-shape normal distribution
required by most methods in this package. Many of the methods implemented in `spatialGE`
have been borrowed from the field of geostatistics, which often require this kind
of data normalization.

Normalization is achieved by applying the function `voom_norm()` to the `melanoma`
object.
```{r}
melanoma <- voom_norm(melanoma)
```

## Visualization of gene expression from spatially-resolved transcriptomics data
After normalization, expression of specific genes can be visualized using 'quilt' 
plots. The function `plot_gene_quilt()` re-creates the shape of the spatial array
and shows the scaled normalized expression of a gene. We have adopted the color 
palettes from the package `khroma`. The color palette can be passed using the 
argument `color_pal`. The default behavior of the function produces plots for all 
spatial arrays in the STList, but we can define who should be plotted using the 
argument `plot_who`.

Let's produce a quilt plot for the gene CAPG and SOX10, two genes known to be 
expressed in melanoma.
```{r}
quilts1 <- plot_gene_quilt(melanoma, genes=c('CAPG', 'SOX10'), color_pal="YlOrBr", plot_who=1, purity = T)
ggpubr::ggarrange(plotlist=quilts1, nrow=1, ncol=2)
```

We can see that gene expression patterns of both genes are different, with a more
dispersed pattern in CAPG, compared to SOX10. The large area of lower CAPG
expression corresponds to the melanoma pocket detected by Thrane et al. (see Figure 3
of their study).

### Spatial interpolation ('kriging') of gene expression
We can predict a smooth surface of the gene expression in a spatial array. In 
`spatialGE`, this prediction is achieved by a type of spatial interpolation widely
used on geographic data. The method named 'kriging' allows for the estimation of 
the normalized gene expression value of the unsampled areas between spots, or spots 
for which a RNA-Seq library failed to provide data. Estimating a transcriptomic 
surface via kriging assumes that gene expression of two given points is
correlated with the physical distance between them.

The function `gene_krige()` performs kriging of gene expression via the `geoR`
package:
```{r}
melanoma <- gene_krige(melanoma, genes=c('CAPG', 'SOX10'), who=c(1, 3))
```

The surface can be visualized using the `plot_gene_krige()` function:
```{r}
kriges1 <- plot_gene_krige(melanoma, genes=c('CAPG', 'SOX10'), plot_who=1)
ggpubr::ggarrange(plotlist=kriges1, nrow=1, ncol=2)
```

By looking at the transcriptomic surfaces of the tho genes, it is easier to detect
pockets of higher (or lower) expression within the tissue. It is now more evident that
expression of CAPG is lower on the left (melanoma) portion of the tumor slice, compared
to the rest of the slice.

In the transcriptomic surface plots, `spatialGE` outputs three spatial heterogeneity 
statistics: Moran's I, Geary's C, and Getis-Ord Gi.

```{r echo=F}
sphet_info <- tibble::tibble("Low"=c('Dispersion', 'Clustering', '"Cold spots"'),
"Statistic"=c('Moran’s I', 'Geary’s C', 'Getis-Ord Gi'),
"High"=c('Clustering', 'Dispersion', '“Hot spots”')
)

knitr::kable(sphet_info, align='c', caption='Heterogeneity statitstics in `spatialGE`', format = 'html') 
```

In this case, the two genes showed almost the same values for all three statistics.
A slightly lower Moran's I for CAPG is a result of higher dispersion of spots with
high expression values through the slice surface. Both Moran's I and Geary's C showed
statistical significance (_p_<0.05) after a permutation test (performed with `spdep`). 

A higher difference in Moran's I can be seen for one of the tissue slices of patient 2.
With a higher Moran's I, CAPG shows expression concentrated in the center right and
lower left of the tissue. For SOX10, expression is almost uniform for the lower half
of the tissue slice.
```{r}
kriges2 <- plot_gene_krige(melanoma, genes=c('CAPG', 'SOX10'), plot_who=3)
ggpubr::ggarrange(plotlist=kriges2, nrow=1, ncol=2)
```

## Gene expression deconvolution in spatial arrays
Many gene expression deconvolution methods are available, and their use will depend
largely on the question and biological system to be studied. In `spatialGE`, we have
implemented xCell to provide a qualitatuve snapshot of the cell composition within a
spatial array. Decovolution in xCell focuses on immunological cell types, which is of
crucial importance in the study of tumor microenvironment. 

Let's apply xCell to the spatial arrays:
```{r}
melanoma <- spatial_deconv(melanoma, method='xcell')
```

With the function `plot_cell_quilt()`, we can visualize the cell type scores 
inferred by xCell. For example, B cells and lymphatic endothelial cells.
```{r}
quilts2 <- plot_cell_quilt(melanoma, cells=c('b_cells', 'ly_endothelial_cells'), color_pal="YlOrBr", plot_who=1, purity=T)
ggpubr::ggarrange(plotlist=quilts2, nrow=1, ncol=2)
```


