---
title: "spatialGE: Visualization and spatial heterogeneity analysis of melanoma"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{spatialGE_tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r vig_settings, include = FALSE}
knitr::opts_chunk$set(
  collapse=TRUE,
  comment="#>",
  fig.width=10, fig.height=7
)
```

The package `spatialGE` provides a set of tools for the visualization of gene expression from 
spatially-resolved arrays, such as those generated by the 10X Visium platform. 

## Installation
The `spatialGE` repository is available at GitHub and can be installed via `devtools`.
To install `devtools`, in case is not installed in your R console, please run the 
following code:
```{r install_devtools}
if("devtools" %in% rownames(installed.packages()) == FALSE) {
  install.packages("devtools")
}
```

After making sure `devtools` is installed, proceed to install `spatialGE`:
```{r install_spatialGE}
# devtools::install_github("fridleylab/spatialGE")
```

To use `spatialGE`, load the package using the command:
```{r load_spatialGE}
library(spatialGE)
```

## Spatially-resolved expression of melanoma stage III tumor biopsies
To show the utility of `spatialGE`, we use the Spatial Transcriptomics data set
generated by Thrane et al. (2018). This data set includes lymph node biopsies from four stage
IIIc melanoma patients, with two spatial arrays per biopsy (https:doi.org/10.1158/0008-5472.CAN-18-0747).
The spatial arrays comprised 1,007 'spots' with 200μM of distance between their centers. 
The spots had a diameter of 100μM, covering 5-40 cells according to the authors.

In the version of this data set provided along `spatialGE`, we have removed probe 
names and duplicate gene names (resulting from different probes). The user can 
find the original raw data at the authors website (https://www.spatialresearch.org/resources-published-datasets/doi-10-1158-0008-5472-can-18-0747/).

## Creating an STList (Spatial Transcriptomics List)
Raw and processed data are stored in an STList (an R S4 class object). The STList
can be created with the function `STList()`, which can take minimally inputs in 
different ways. Minimally, raw RNA counts and spot coordinates must be provided.
For this example, we provide to the STList function the following inputs:

1. Comma- or tab-deimited files with raw RNA counts, one per spatial array.
The raw count files must have gene names in the first column, and spots in subsequent 
columns. Duplicated gene names are not allowed.

2. Comma- or tab-deimited files with y and x coordinates for each spot. The first
column of this file contains spot IDs matching column names in the raw count files.
The seond and third columns must match the Y and X coordinates of each spot respectively.

We also input a table with clinical data associated with each spatial array. This 
table is optional, and must be a comma- or tab-delimited file with IDs for each spatial
array in the first column, and relevant variables in subsequent columns. The array IDs
should match the file paths of the RNA count and coordinate files (for example, the 
sample ID "ST_mel4_rep2" matches the RNA count file "./inst/extdata/ST_mel4_rep2_counts_dedup_genes.tsv"
and coordinate file "./inst/extdata/ST_mel4_rep2_counts_mapping.txt".

The example files in this tutorial can be downloaded from the GitHub repository (https://github.com/FridleyLab/spatialGEdev/tree/main/inst/extdata). They are also
stored in your local `spatialGE` installation:
```{r data_fpaths1}
data_files <- system.file("extdata", package="spatialGE")
data_files <- list.files(data_files, full.names=T)
```

We then find the necessary files within the data repository in `spatialGE`. Those two
files will be the input to later create the Spatial Transcriptomics List (STList).
```{r data_fpaths2}
count_files <- grep("genes", data_files, value=T)
coord_files <- grep("mapping", data_files, value=T)
clin_file <- grep("clinical", data_files, value=T)
```

We can load the files into an STList object like so:
```{r create_stlist, warning=F}
melanoma <- STList(rnacounts=count_files, spotcoords=coord_files, samples=clin_file)
```

* Note: If you receive a warning indicating that an input file has a incomplete final
line, you can ingnore it. Alternatively, modify the input files with a text editor by
adding an empty line at the end of the files.

The `melanoma` object is an STList, and contains the count data, spot coordinates,
and clinical data.
```{r call_stlist}
melanoma
```

## Exploring variation between spatial arrays
The function `bulk_pca()` allows for a quick snapshot of the variation in gene
expression among samples. The function creates 'bulk' RNA-Seq data sets by combining
all counts from a spatial array. Then, it applies voom transformation to the 'bulk' 
RNA-Seq libraries and performs a PCA. Note that spatial (x,y) coordinate information 
is not included in this analysis, which is intended only as a tool for data exploration. 
In this case, we apply the function to look for agreement between samples from 
the same patient: it is expected that tissue slices from the same patient are more 
similar between them, than tissue slices from other patients. The `bulk_pca()` 
allows to map a clinical variable to the PCA (gender in this example), by including 
the exact, case-sensitive name of the column from the clinical/sample file.
```{r pca_chunk}
bulk_pca(melanoma, clinvar='gender')
```

## Data normalization of spatially-resolved transcriptomic data
Many transformation methods are available for both bulk and single-cell RNA-Seq 
count data. In `spatialGE`, we apply voom transformation to RNA-Seq libraries 
from each of the spots in order to obtain a gaussian-shape normal distribution
required by most methods in this package. Many of the methods implemented in `spatialGE`
have been borrowed from the field of geostatistics, which often require this kind
of data distribution.

Normalization is achieved by applying the function `voom_norm()` to the `melanoma`
object.
```{r norm_chunk}
melanoma <- voom_norm(melanoma)
```

## Visualization of gene expression from spatially-resolved transcriptomics data
After data transformation, expression of specific genes can be visualized using 'quilt' 
plots. The function `plot_gene_quilt()` shows the transformed expression of one or 
several genes. We have adopted the color palettes from the package `khroma`. The 
color palette can be passed using the argument `color_pal`. Other palettes from the
`RColorBrewer` package can also be specified. The default behavior of the function 
produces plots for allspatial arrays in the STList, but we can define which arrays 
will be plotted using the argument `plot_who`.

Let's produce a quilt plot for the genes CD74 and SOX10, an immune gene and a 
melanoma marker.
```{r genequilt_chunk}
quilts1 <- plot_gene_quilt(melanoma, genes=c('CD74', 'SOX10'), plot_who=2, visium=F)
```

Because `spatialGE` functions output ggplot objects, we can plot the results 
side-by-side using functions such as `ggarrange`.
```{r}
ggpubr::ggarrange(plotlist=quilts1, nrow=1, ncol=2, common.legend=T, legend='bottom')
```

We can see that gene expression patterns of both genes are different, with a more
dispersed pattern in SOX10, compared to CD74. The "pocket" of lower CD74 expression 
on the left side of the tissue corresponds to the melanoma pocket detected by 
Thrane et al. (see Figure 3 of their study).

### Spatial interpolation ('kriging') of gene expression
We can predict a smooth transcriptomic surface of the gene expression in an spatial 
array. In `spatialGE`, this prediction is achieved by a widely used spatial interpolation 
method in geostatistics. The method known as 'kriging' allows for the estimation of 
the transfromed gene expression values in the unsampled areas between spots, or spots 
for which a RNA-Seq library failed to provide data. Estimating a transcriptomic 
surface via kriging assumes that gene expression of two given points is correlated with 
the physical distance between them.

The function `gene_krige()` performs kriging of gene expression via the `geoR`
package. We specify that kriging will be performed for two os the spatial arrays:
```{r genekrige_chunk}
melanoma <- gene_krige(melanoma, genes=c('CD74', 'SOX10'), who=c(2, 3), python=F)
```

The surfaces can be visualized using the `plot_gene_krige()` function:
```{r plotgenekrige_chunk1}
kriges1 <- plot_gene_krige(melanoma, genes=c('CD74', 'SOX10'), plot_who=2, visium=F)
ggpubr::ggarrange(plotlist=kriges1, nrow=1, ncol=2, common.legend=T, legend='bottom')
```

By looking at the transcriptomic surfaces of the tho genes, it is easier to detect
pockets of higher (or lower) expression within the tissue. It is now more evident that
expression of CD74 is lower on the left (melanoma) region of the tumor slice, compared
to the rest of the slice.

In the transcriptomic surface plots, `spatialGE` outputs three spatial heterogeneity 
statistics: Moran's I, Geary's C, and Getis-Ord Gi.

```{r sphet_table, echo=F}
sphet_info <- tibble::tibble("Low"=c('Dispersion', 'Clustering', '"Cold spots"'),
"Statistic"=c('Moran’s I', 'Geary’s C', 'Getis-Ord Gi'),
"High"=c('Clustering', 'Dispersion', '“Hot spots”')
)

kableExtra::kbl(sphet_info, align='c', caption='Heterogeneity statitstics in `spatialGE`') %>%
  kableExtra::kable_styling(position="center", full_width=F)
```

A higher Moran's I for CD74 indicates that there are clusters of high expression. In 
other words, gene expression of CD74 is not uniform across the tissue. Geary's C was 
significantly different than zero in CD74, but not SOX10 (indicated by asterisk),
further validating the higher gene expression clustering in IGLL5. Statistical 
significance (_p_<0.05) in `spatialGE` is obtained via permutation (performed with `spdep`). 

We can control which tissue slice to plot with the `plot_who` argument. This
argument is also included in the quilt plot functions.
```{r plotgenekrige_chunk2}
kriges2 <- plot_gene_krige(melanoma, genes=c('CD74', 'SOX10'), plot_who=3, visium=F)
ggpubr::ggarrange(plotlist=kriges2, nrow=1, ncol=2, common.legend=T, legend='bottom')
```

## Gene expression deconvolution in spatial arrays
Many gene expression deconvolution methods are available, and their use will depend
largely on the question and biological system to be studied. In `spatialGE`, we have
implemented xCell to provide a qualitative snapshot of a tissue's cell composition. 
Decovolution in xCell focuses on immunological cell types, which is of crucial 
importance in the study of the tumor microenvironment.

In addition to xCell, the `spatial_deconv` function automatically applies ESTIMATE
to obtain tumor purity scores. The purity scores are internally clustered using
model-based clustering (`mclust`) to detect tumor/stroma compartments.

Let's apply xCell to the spatial arrays:
```{r deconv_chunk}
melanoma <- spatial_deconv(melanoma, method='xcell')
```

With the function `plot_deconv_quilt()`, we can visualize the cell type scores 
inferred by xCell. For example, B cells and dendritic cells.
```{r deconvquilts_chunk}
quilts2 <- plot_deconv_quilt(melanoma, cells=c('b_cells', 'i_dc'), plot_who=2, 
                             purity=T, visium=F)
ggpubr::ggarrange(plotlist=quilts2, nrow=1, ncol=3, common.legend=T, legend='bottom')
```

By plotting the tumor/stroma classification of the spots, the differences in cell
scores can quiclky be detected. Melanoma areas (squares) show lower B cell and 
dendritic cell scores, as expected. Nonetheless, B cells were abundant in the 
lymphoid pocket (upper right part of the tissue).

Similar plots can be generated for gene expression. These plots can also be made
interactive by making `inter=T`, which converts plots to Plotly objects.
```{r genepurity_chunk}
quilts3 <- plot_gene_quilt(melanoma, genes='CD74', plot_who=2, purity=T, visium=F,
                           inter=T)
quilts3$CD74_2
```

Spatial interpolation of cell scores is also possible with the `deconv_krige()`
function:
```{r deconvkrige_chunk}
melanoma <- deconv_krige(melanoma, cells=c('b_cells', 'i_dc'), who=2, python=F)
```

We plot the surfaces using the `plot_deconv_krige()` function.
```{r plotdeconvkrige_chunk}
kriges3 <- plot_deconv_krige(melanoma,  cells=c('b_cells', 'i_dc'), plot_who=2, visium=F)
ggpubr::ggarrange(plotlist=kriges3, nrow=1, ncol=2, common.legend=T, legend='bottom')
```

Cell scores in xCell are accompanied by p-values (cell scores not equal to zero). 
The p-values are calculated automatically during deconvolution in `spatialGE`, and
we can plot them using the `pvalue` option:
```{r plotkrigepvals_chunk}
kriges4 <- plot_deconv_krige(melanoma,  cells=c('b_cells', 'i_dc'), plot_who=2, 
                             pvalues=T, visium=F)
ggpubr::ggarrange(plotlist=kriges4, nrow=1, ncol=2, common.legend=T, legend='bottom')
```

As observed in the surfaces, B cells are present throughout the tissue, but less 
abundant in the melanoma pocket. The estimates were significantly different from 
zero as sugggested by xCell's permutation method. The xCell scores for the dendritic 
cells were only significantly different from zero only for a few spots (gray hollow 
circles). 

The `plot_deconv_krige()` also allows annotation of tumor spots:
```{r plotkrigepurity_chunk}
kriges5 <- plot_deconv_krige(melanoma,  cells=c('b_cells', 'i_dc'), plot_who=2, 
                             purity=T, visium=F)
ggpubr::ggarrange(plotlist=kriges5, nrow=1, ncol=3, common.legend=T, legend='bottom')
```

## Unsupervised spatially-informed clustering
We can now perform unsupervised spatially-informed clustering to detect compartments in 
the tissue. In `spatialGE`, we use weighted average matrices to capture the 
transcriptomic differences among the spots. A transcriptomic (euclidean) 
distance matrix is calculated and scaled from PCA. Following, scaled euclidean distances are
computed for the spatial distances among spots. The user defines a weight, from 0 to 1,
to apply to the spatial distances. The higher the weight, the less biologically 
meaningful the clustering is given that the groups will only reflect the physical
distances between the spots and less information on the transcriptomic profiles will be
used. After many tests, we have found that weights between 0.1 - 0.25 should be 
enough to capture the tissue heterogeneity. This approach to spatial clustering can
be performed using the `cluster_STspot()` function:
```{r clustespots_chunk}
melanoma <- cluster_STspot(melanoma, ks='dtc', weights=0.1)
```

Results of clustering can be plotted via the `plot_STclusters()` function:
```{r plotclustspots_chunk}
cluster_p <- plot_STclusters(melanoma, purity=T, plot_who=c(2,3), visium=F)

# In case specific k values are selected:
# for(i in 1:2){
#  print(ggpubr::ggarrange(plotlist=cluster_p[[i]], nrow=2, ncol=2, common.legend=T, legend='bottom'))
# }

ggpubr::ggarrange(plotlist=cluster_p, nrow=1, ncol=2, common.legend=T, legend='bottom')
```

We have used here the dynamicTreeCut (`dtc` option) to select the number of clusters,
but users can define their own range of k to evaluate. The two tissues plotted here, 
show compartmentalization, but in the sample on the left, the lymphoid tissue on the
right portion of the tissue is shown in blue possibly indicating spots with immune 
activity, which also surround the melanoma pocket.

## Association between spatial heterogeneity and clinical variables
To explore the relationship between a clinical variable of interest and the 
level of gene expression uniformity within a tissue section, we can use the 
function `plot_phenovar()`:
```{r clinplot_chunk, fig.width=10, fig.height=3}
pheno_p <- plot_phenovar(melanoma, phenovar='survival_months', gene='CD74')
ggpubr::ggarrange(plotlist=pheno_p, nrow=1, ncol=3, common.legend=T, legend='bottom')
```

In the case of the gene CD74, an immune gene, samples with a less uniform expression
(i.e., clustered gene expression) came from subjects with a higher survival time 
(high Moran's I; low Geary's C). The survival of the subjects did not show an association
with the tendency of CD74 to generate expression "hotspots" (Getis-Ord Gi was intermediate
in subject 1)

## Summary
`spatialGE` provides a suite of tools for the study of tissue heterogeneity via
spatial transcriptomics. These tools are designed for exploratory analysis, and
can be specially useful to understand the immune landscape of tumors.
