---
title: "spatialGE: Guide Spatial heterogeneity analysis in melanoma (Thrane et al. 2018)"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{spatialGE_tutorial_main}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r vig_settings, include = FALSE}
knitr::opts_chunk$set(
  collapse=TRUE,
  comment="#>",
  fig.width=10, fig.height=7
)
```

The package `spatialGE` provides a set of tools for the visualization of gene expression from 
spatially-resolved transcriptomic experiments, such as those generated by the 10X Visium platform. 

## Installation
The `spatialGE` repository is available at GitHub and can be installed via `devtools`.
To install `devtools`, in case is not already installed in your R console, please run the 
following code:
```{r install_devtools}
if("devtools" %in% rownames(installed.packages()) == FALSE) {
  install.packages("devtools")
}
```

After making sure `devtools` is installed, proceed to install `spatialGE`:
```{r install_spatialGE}
# devtools::install_github("fridleylab/spatialGE")
```

To use `spatialGE`, load the package using:
```{r load_spatialGE}
library(spatialGE)
```

## Spatially-resolved expression of melanoma stage III tumor biopsies
To show the utility of `spatialGE`, we use the Spatial Transcriptomics (ST) data set
generated by Thrane et al. (2018). This data set includes lymph node biopsies from 
four patients with stage IIIc melanoma, two spatial arrays (samples) per biopsy (https:doi.org/10.1158/0008-5472.CAN-18-0747).
The technology used in this study (ST) allows for the collection of up to 1,007 "spots" 
with 200μM of distance between their centers. Data for each sample, however, has a
variable number of spots due to differences in size of the tissue slices. The spots 
had a diameter of 100μM, covering 5-40 cells according to the authors.

The data set is included in the spatialGE repository, but users can find the data in 
the authors' website (https://www.spatialresearch.org/resources-published-datasets/doi-10-1158-0008-5472-can-18-0747/).

## Creating an STList (Spatial Transcriptomics List)
Raw and processed data are stored in an STList (S4 class object). The STList can be 
created with the function `STList()`, which can take different inputs (see [here](https://fridleylab.github.io/spatialGE/reference/STList.html) 
for more info). For this example, we provide to the STList function the following inputs:

1. Files with raw RNA counts, one per sample. The raw count files have gene names 
in the first column, and spots in subsequent columns.

2. Files with y and x coordinates for each spot. The first column of this file contains
spot IDs matching column names in the raw count files. The second and third columns 
contain the Y and X coordinates of each spot respectively.

3. Names of the samples. These names should match, at least partially, the names of
the count and coordinate files. In this example, sample names are included in the 
first column of the metadata file.

We also input a table with clinical data associated with each sample. This  table is 
optional, and must be a comma- or tab-delimited file with IDs for each sample in the 
first column, and relevant variables in subsequent columns. The array IDs should match, 
at least partially, the file names of the RNA count and coordinate files (for example, the 
sample ID "ST_mel4_rep2" matches the RNA count file "ST_mel4_rep2_counts.tsv" and 
coordinate file "ST_mel4_rep2_mapping.tsv".

The example files in this tutorial can be downloaded from the GitHub repository 
(https://github.com/FridleyLab/spatialGEdev/tree/main/inst/extdata). They are also
stored in your local `spatialGE` installation:
```{r data_fpaths1}
data_files <- system.file("extdata", package="spatialGE")
```

Then, we look for the find the necessary files within the data repository in `spatialGE`.
These files contain the raw gene counts (`count_files`), coordinates (`coord_files`),
and `clinical` metadata.
```{r data_fpaths2}
count_files <- list.files(data_files, full.names=T, pattern='counts')
coord_files <- list.files(data_files, full.names=T, pattern='mapping')
clin_file <- list.files(data_files, full.names=T, pattern='clinical')
```

We can load the files into an STList object like so:
```{r create_stlist, warning=F}
melanoma <- STList(rnacounts=count_files, spotcoords=coord_files, samples=clin_file)
```

* Note: If you receive a warning indicating that an input file has a incomplete final
line, you can ignore it. Alternatively, modify the input files with a text editor by
adding an empty line at the end of the files.

The `melanoma` object is an STList, and contains the count data, spot coordinates,
and clinical data.
```{r call_stlist}
melanoma
```

## Exploring variation between spatial arrays
The function `bulk_pca()` allows for a quick snapshot of the variation in gene
expression among samples. The function creates 'bulk' RNAseq data sets by combining
all counts from a spatial array. Then, it log transform the 'bulk' RNAseq counts
and performs a PCA. Note that spatial (x,y) coordinate information is not considered
in this analysis, which is intended only as a tool for exploratory analysis. 
In this case, we apply the function to look for agreement between samples from 
the same patient: It is expected that tissue slices from the same patient are more 
similar between them, than tissue slices from other patients. The `bulk_pca()` 
allows to map a clinical variable to the PCA (gender in this example), by including 
the exact, case-sensitive name of the column from the sample metadata file.
```{r pca_chunk}
bulk_pca(melanoma, clinvar='gender')
```

## Transformation of spatially-resolved transcriptomic data
Many transformation methods are available for both bulk and single-cell RNA-Seq 
count data. In `spatialGE`, the function `spatialTransform()` applies log-transformation 
to the data, after library size normalization. Similar to Seurat, it applies a scaling
factor (`scale_f=10000` by default). `spatialGE` borrows methods from the field of
spatial statistics, often requiring gaussian-shaped data distributions. Hence, users
also can apply voom transformation, which could in certain situations, provide more 
gaussian-shaped distributions than log-transformed data.
```{r norm_chunk}
melanoma <- spatialTransform(melanoma)
```

## Visualization of gene expression from spatially-resolved transcriptomics data
After data transformation, expression of specific genes can be visualized using 'quilt' 
plots. The function `plot_gene_quilt()` shows the transformed expression of one or 
several genes. We have adopted the color palettes from the packages `khroma` and
`RColorBrewer`. The name of a color palette can be passed using the argument `color_pal`.
The default behavior of the function produces plots for all samples within the STList, 
but we can define specific samples to be plotted using the argument `plot_who`.

Let's produce a quilt plot for the genes *CD19* and *MLANA* (an immune gene and a 
melanoma marker), for sample number 2 of patient 1 (`plot_who=2`).
```{r genequilt_chunk}
quilts1 <- plot_gene_quilt(melanoma, genes=c('CD19', 'MLANA'), plot_who=2, visium=F, 
                           color_pal='BuRd', ptsize=2)
```

Because `spatialGE` functions output lists of ggplot objects, we can plot the results 
side-by-side using functions such as `ggarrange`.
```{r}
ggpubr::ggarrange(plotlist=quilts1, nrow=1, ncol=2, legend='bottom')
```

We can see that gene expression patterns of both genes are almost opposite: *CD19*
is expressed in the upper right portion of the tissue, whereas *MLANA* is expressed
in the left portion. The left portion of the tissue corresponds to the melanoma 
compartment detected by Thrane et al., and the upper right portion corresponds to
the a lymphoid tissue area (see Figure 3 of their study).

### Spatial interpolation ("kriging") of gene expression
We can predict a smooth gene expression surface for each sample. In `spatialGE`, 
this prediction is achieved by a widely used spatial interpolation method in
spatial statistics. The method known as 'kriging' allows for the estimation of gene 
expression values in the unsampled areas between spots, or spots for which a the 
sample preparation failed to provide data. Estimating a transcriptomic surface via 
kriging assumes that gene expression of two given points is correlated with 
the physical distance between them.

The function `gene_krige()` performs kriging of gene expression via the `geoR`
package. The argument `res=0.2`, sets the resolution of kriging, the smaller the
value of `res`, the more resolution and computing time rquired. We specify that 
kriging will be performed for two of the spatial arrays (`who=c(2, 3)`):
```{r genekrige_chunk}
melanoma <- gene_krige(melanoma, genes=c('CD19', 'MLANA'), who=c(2, 3), python=F, 
                       res=0.2)
```

The surfaces can be visualized using the `plot_gene_krige()` function:
```{r plotgenekrige_chunk1}
kriges1 <- plot_gene_krige(melanoma, genes=c('CD19', 'MLANA'), plot_who=2, visium=F)
ggpubr::ggarrange(plotlist=kriges1, nrow=1, ncol=2, legend='bottom')
```

By looking at the transcriptomic surfaces of the tho genes, it is easier to detect
"pockets" of high (or low) expression within the tissue. It is now more evident that
expression of CD19 is lower on the left (melanoma) region of the tumor slice, compared
to the rest of the slice.

Just like in `plot_gene_quilt()`, we can control which tissue slice to plot with 
the `plot_who` argument.
```{r plotgenekrige_chunk2}
kriges2 <- plot_gene_krige(melanoma, genes=c('CD19', 'MLANA'), plot_who=3, visium=F)
ggpubr::ggarrange(plotlist=kriges2, nrow=1, ncol=2, common.legend=T, legend='bottom')
```

## Gene expression deconvolution in spatial arrays
In `spatialGE`, we have implemented xCell to provide a qualitative snapshot of a 
tissue's cell composition. Decovolution in xCell focuses on immunological cell types, 
which is of crucial importance in the study of the tumor microenvironment. Other
deconvolution methods will be implemented in future versions of `spatialGE`. The 
user may also want to look at other methods such as [SPOTlight](https://academic.oup.com/nar/article/49/9/e50/6129341).

In addition to xCell, the `spatialDeconv` function automatically applies ESTIMATE
to obtain tumor purity scores. The purity scores are internally clustered using
model-based clustering (`mclust`) to detect tumor/stroma compartments.

Let's apply xCell to the samples:
```{r deconv_chunk, warning=F}
melanoma <- spatialDeconv(melanoma, method='xcell')
```

With the function `plot_deconv_quilt()`, we can visualize the cell type scores 
inferred by xCell. For example, B cells.
```{r deconvquilts_chunk}
quilts2 <- plot_deconv_quilt(melanoma, cells='b_cells', plot_who=2, purity=T, 
                             color_pal='BuRd', visium=F)
ggpubr::ggarrange(plotlist=quilts2, nrow=1, ncol=2, legend='bottom')
```

By plotting the tumor/stroma classification of the spots, the differences in cell
scores can quickly be detected. Melanoma areas (squares) show lower B cell as expected. 
Nonetheless, B cells were abundant in the lymphoid pocket (upper right part of the tissue).

Similar plots can be generated for gene expression. These plots can also be made
interactive by making `inter=T`, which converts plots to Plotly objects. For example,
let us make an interactive plot for gene *CD74*:
```{r genepurity_chunk}
quilts3 <- plot_gene_quilt(melanoma, genes='CD74', plot_who=2, purity=T, visium=F,
                           inter=T)
quilts3[[1]]
```

Spatial interpolation of cell scores is also possible with the `deconv_krige()`
function:
```{r deconvkrige_chunk}
melanoma <- deconv_krige(melanoma, cells='b_cells', who=2, python=F, res=0.2)
```

We plot the surfaces using the `plot_deconv_krige()` function.
```{r plotdeconvkrige_chunk}
kriges3 <- plot_deconv_krige(melanoma,  cells='b_cells', plot_who=2, visium=F)
ggpubr::ggarrange(plotlist=kriges3, nrow=1, ncol=1, legend='bottom')
```

Cell scores in xCell are accompanied by p-values (test for cell scores different
to zero). The p-values are calculated automatically during deconvolution in `spatialGE`, 
and we can plot them using the `pvalue` option:
```{r plotkrigepvals_chunk}
kriges4 <- plot_deconv_krige(melanoma,  cells='b_cells', plot_who=2, pvalues=T, 
                             visium=F)
ggpubr::ggarrange(plotlist=kriges4, nrow=1, ncol=1, legend='bottom')
```

As observed in the surfaces, B cells are present throughout the tissue, but less 
abundant in the melanoma pocket. Most of the spots had B cell scores significantly
different to zero, as suggested by xCell's permutation method (gray hollow 
circles). Most of the non-sognificant scores were observed in the melanoma area,
reinforcing the observation of lower infiltration in that area.

The `plot_deconv_krige()` also allows annotation of tumor spots (hollow squares):
```{r plotkrigepurity_chunk}
kriges5 <- plot_deconv_krige(melanoma,  cells='b_cells', plot_who=2, purity=T, 
                             visium=F)
ggpubr::ggarrange(plotlist=kriges5, nrow=1, ncol=2, common.legend=T, legend='bottom')
```

## Unsupervised spatially-informed clustering (*STclust*)
We can also perform *STclust*, the spatially informed clustering method in `spatialGE`
to detect compartments in the tissue. In STclust, we use weighted average matrices 
to capture the transcriptomic differences among the spots. A transcriptomic (Manhattan) 
distance matrix is calculated and scaled from PCA. Following, scaled euclidean distances are
computed for the spatial distances among spots. The user defines a weight, from 0 to 1,
to apply to the spatial distances. The higher the weight, the less biologically 
meaningful the clustering is given that the groups will only reflect the physical
distances between the spots and less information on the transcriptomic profiles will be
used. After many tests, we have found that weights between 0.025 - 0.25 is enough to 
capture the tissue heterogeneity. This approach to spatial clustering can
be performed using the `STclust()` function:
```{r clustespots_chunk}
melanoma <- STclust(melanoma, ks='dtc', weights=c(0, 0.025, 0.05, 0.075))
```

Results of clustering can be plotted via the `plot_STclusters()` function:
```{r plotclustspots_chunk}
cluster_p <- plot_STclusters(melanoma, plot_who=2, visium=F)
ggpubr::ggarrange(plotlist=cluster_p, nrow=2, ncol=2, legend='right')
```

We can see that at `w=0.025`, tissue compartments are observed for the lymphoid tissue 
(upper right portion of the tissue in red), which also sureounds the melnoma area
(yellow), possibly indicating immune activity around the tumor spots. Blue likely
corresponds to stromal spots. We have used here the dynamicTreeCut (`dtc` option) 
to select the number of clusters, but users can define their own range of k to 
be evaluated, allowing to further isolate cell types.

## Association between spatial heterogeneity and clinical variables (*SThet*)
To explore the relationship between a clinical variable of interest and the 
level of gene expression uniformity within a tissue section, we can use the 
function `plot_phenovar()`:
```{r clinplot_chunk}
melanoma <- spatial_stat_plot(melanoma, samplevar='survival_months', gene='CD74')
```

In the case of the gene CD74, an immune gene, samples with a less uniform expression
(i.e., clustered gene expression) also were from a patient  with a higher survival time 
(high Moran's I; low Geary's C). The survival of the patients did not show an association
with the tendency of CD74 to generate expression "hot spots" (Getis-Ord Gi was intermediate
in patient 1). See the table below for a simplistic interpretation of the spatial
autocorrelation statistics calculated in `spatialGE`:

```{r sphet_table, echo=F}
sphet_info <- tibble::tibble("Low"=c('Dispersion', 'Clustering', '"Cold spots"'),
"Statistic"=c('Moran’s I', 'Geary’s C', 'Getis-Ord Gi'),
"High"=c('Clustering', 'Dispersion', '“Hot spots”')
)

kableExtra::kbl(sphet_info, align='c', caption='Heterogeneity statitstics in `spatialGE`') %>%
  kableExtra::kable_styling(position="center", full_width=F)
```

Estimating these statistics can be computationally expensive. Hence, plots can be
shown again by accessing the appropriate STList slot (`melanoma@pheno_plots[['CD74']]`).

<details>
  <summary>**Session Info**</summary>
```{r}
sessionInfo()
```
</details>
