---
title: "spatialGE: Basic tutorial for visualization and spatial heterogeneity analysis"
output: rmarkdown::melanoma_ST_vignette
vignette: >
  %\VignetteIndexEntry{spatialGE_tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

The package `spatialGE` provides a set of tools for the visualization of gene expression from 
spatially-resolved arrays, such as those generated by the 10X Visium platform. 

## Installation
The `spatialGE` repository is available at GitHub and can be installed via `devtools`.
To install `devtools` in case is not installed in your R console, please run the 
following code:
```{r}
if("devtools" %in% rownames(installed.packages()) == FALSE) {
  install.packages("devtools")
}
```

After making sure `devtools` is installed, proceed to install `spatialGE`:
```{r}
devtools::install_github("fridleylab/spatialGEdev")
```

To use `spatialGE`, load the package using the command:
```{r setup}
library(spatialGEdev)
```

## Spatially-resolved expression of melanoma stage III tumor biopsies
To show the utility of `spatialGE`, we use the Spatial Transcriptomics data set
generated by Thrane et al. (2018) that included lymph node biopsies from four stage
IIIc melanoma patients, with two arrays per biopsy (DOI: 10.1158/0008-5472.CAN-18-0747).
The spatial slides used in this melanoma comprised 1,007 'spots' with 200μM of distance
between their centers. The spots had a diameter of 100μM, covering 5-40 cells according
to the authors.

In the data provided as example for `spatialGE`, we have removed probe names and duplicate
gene names (resulting from different probes). The user can find the original raw data at
the authors website (https://www.spatialresearch.org/resources-published-datasets/doi-10-1158-0008-5472-can-18-0747/).

## Creating an STList (Spatial Transcriptomics List)
Raw and processed data are stored in an STList (an R S4 class object). The STList
can be created with the function `STList()`, which takes minimally two files:
1. A text file containing file paths of gene raw count matrices, one per line. The gene
count matrices must be tab-separated, with gene names in the first column, and 
spots in subsequent columns. Matrices of different for each tissue slice are allowed, but
duplicated gene names are not.
2. A text file containing file paths of x,y coordinates for each spot, one per line.
The order of the coordinate file paths must match that of the raw count matrices.

Optionally, the user can provide a table with clinical/phenotype data associated with
each spatial array. This table is a comma-separated file with sample ID in the first
column, and relevant variables in subsequent columns. The order of the rows in the 
clinical file must match that of the file paths. 

The example files in this tutorial can be downloaded from the GitHub repository (https://github.com/FridleyLab/spatialGEdev/tree/main/inst/extdata), but can be 
loaded into the STList using the following commands:
```{r}

```




