---
title: "spatialGE: Reading GeoMx-DSP outputs"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{spatialGE_tutorial_geomx}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

The package `spatialGE` provides a set of tools for the visualization of spatially-resolved 
transcriptomics. In this vignette we provide an example on how to read GeoMx-DSP 
data for downstream analysis with `spatialGE`. The GeoMx platform produces data for
a series of region of interest (ROIs). In order to read GeoMx data in `spatialGE`, three
components are required: 
  1. Count data for each ROI in the form of `.dcc` files (one per ROI)
  2. Spatial information within the sample annotation file (usually needs to be exported from the GeoMx analysis suite)
  3. A probe annotation file (`.pkc`) to obtain gene names.

For this example, we will use the Nanostring annotated kidney data set containing 
4 diabetic kidney disease samples and 3 normal kidney samples. Each tissue slice 
contains a variable number of ROIs.

## Data acquisition
The data set is publicly available and can be downloaded from the Nanostring Resources 
[website](https://www.nanostring.com/resources/annotated-kidney-dataset/). Users
need to register to access the data at this point. Then, click the "DOWNLOAD FILE"
button, and proceed to obtain the "GeoMxTools Input Files: Kidney_Dataset_for_GeomxTools.zip (17MB)"
file. This file contains the `.dcc` and `.pkc` files.

Next, click the "Kidney Dataset" link and download the "Sample Annotations: Kidney_Sample_Annotations.txt"
file (users will probably need to download by right-cliking the link). This file 
contains sample metadata, as well as the spatial information (x, y coordinates). 
Optionally, download the images in "Slide images: ROI_reports.zip" (~2.5Gb).

Place all the files in a folder in your desktop. In this vignette, we have named that
folder as "geomx_input". Proceed to decompress the zipped files (including those
within the "ROI reports" folder, in case that file was downloaded).

## spatialGE installation
The `spatialGE` repository is available at GitHub and can be installed via `devtools`.
To install `devtools`, in case is not installed in your R console, please run the 
following code:
```{r install_devtools}
if("devtools" %in% rownames(installed.packages()) == FALSE) {
  install.packages("devtools")
}
```

After making sure `devtools` is installed, proceed to install `spatialGE`:
```{r install_spatialGE}
# devtools::install_github("fridleylab/spatialGE")
```

To use `spatialGE`, load the package using the command:
```{r load_spatialGE}
library('spatialGE')
```

## Creating an STList (Spatial Transcriptomics List)
Raw data and results from analyses are stored in an STList (S4 class object). The STList
can be created with the function `STList()`, and in the case of GeoMx-DSP data, 
`.dcc` files are passed to the `rnacounts` argument. The `.pkc` file is passed to
`gmx_pkc`. 

The annotation data is passed to `samples`, however, additional arguments
are required. The `gmx_slide_col` indicated the column name within the annotation 
file with the slide ID (each slide contains several ROIs and is considered a sample
in `spatialGE`). The `gmx_roi_col` indicates the column with ROI IDs, matching the
ROI IDs in the `.dcc` files. `gmx_x_col` and `gmx_y_col` indicate the columns with
x and y coordinates. Additional columns from the annotation file can be imported 
via the `gmx_meta_cols` argument.

Assuming the user created folders using the instructions in this vignette 
(see *Data acquisition*), the output files are specified like so:
```{r file_paths}
dcc_files =  "~/Desktop/geomx_input/Kidney_Dataset/dccs"
pkc = '~/Desktop/geomx_input/Kidney_Dataset/pkcs/TAP_H_WTA_v1.0.pkc'
```

Annotation data is specified here:
```{r}
sample_annots = "~/Desktop/geomx_input/Kidney_Sample_Annotations.txt"
slide_col="SlideName"
id_col="Sample_ID"
x_col="ROICoordinateX"
y_col="ROICoordinateY"
meta_cols=c('disease_status',	'pathology', 'region')
```

Then, we use these arguments with the `STList` function.
```{r create_STList}
kdgeomx <- STList(rnacounts=dcc_files, gmx_pkc=pkc, samples=sample_annots,
                  gmx_slide_col=slide_col, 
                  gmx_roi_col=id_col, 
                  gmx_x_col=x_col, 
                  gmx_y_col=y_col, 
                  gmx_meta_cols=meta_cols)
```

The `kdgeomx` object is an STList.
```{r view_STList}
kdgeomx
```

## Loading images
If images were downloaded ("ROI reports"), users can load them into the STList using
the `load_images()` function. The function only requires two arguments: The STList to
load images to, and a folder with the images. The file names of the images must match
at least partially, the sample names in the STList (i.e. "disease3", "disease4",
"normal3", etc...)

First, we prepare a folder with the images to load. In your "geomx_input" folder, 
create a sub folder called "images". As previously mentioned, after decompressing
the "ROI reports" file, users should decompress the zipped files within. From each
of the resulting folder, copy the file ending in "_scan.png". For example, from 
the folder "disease1B_scan", users should copy "disease1B_scan.png". This file is the
image of the entire slide.
```{r}
images = '~/Desktop/geomx_input/images/'
kdgeomx = load_images(kdgeomx, images)
```

## Data normalization of spatially-resolved transcriptomic data
The suggested workflow for analysis of GeoMx includes quantile normalization of counts.
This type of normalization has not been implemented in `spatialGE`, and thus, we
will use the log transformation in the function `spatialTransform()`.
```{r log_transform}
kdgeomx <- spatialTransform(kdgeomx)
```

## Visualization of gene expression
We can plot now the gene expression levels of specific genes. Let's create plots 
for _TXNIP_:
```{r plot_gene_quilt}
qplot <- plot_gene_quilt(kdgeomx, image=T, genes='TXNIP', visium=F, color_pal='sunset', ptsize=2)
```

The `plot_gene_quilt` returns a list of plots. We can use `ggarrange()` to display
these plots in an organized manner. The tissue images in the right show a number of
areas that are colored different to the background, which correspond to the sampled
ROIs. The quilt plots on the left show the expression of _TXNIP_ in those ROIs. We have 
envisioned for future versions of `spatialGE`, plotting options allowing users to
visiualize gene expression directly over the images.
```{r, fig.width=8, fig.height=24}
ggpubr::ggarrange(plotlist=qplot, ncol=2, nrow=7)
```

<details>
  <summary>**Session Info**</summary>
```{r}
sessionInfo()
```
</details>

