---
title: "Spatially-informed differential gene expression"
date: '`r format(Sys.Date())`'
output: rmarkdown::html_vignette
bibliography: vignette_references.bib
vignette: >
  %\VignetteIndexEntry{spatialge_tutorial_stdiff}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r vig_settings, include=F}
knitr::opts_chunk$set(
  collapse=TRUE,
  comment="#>",
  fig.width=10, fig.height=7
)
```

Differential gene expression testing has been a fundamental part in studies using "bulk" RNAseq data, and continues to be critical in the analysis of single-cell and spatial transcriptomics. The `STdiff` function from the ***`spatialGE`*** package provides methods for traditional, non-spatial differential gene expression analysis such as Wilcoxon's rank test, T-test, and mixed models.

Spatial transcriptomics (ST) presents a new challenge for differential gene expression analysis. In ST, gene expression is measured at locations (spots or cells) that are separated by fixed or variable distance. Within each assayed tissue, cells or spots that are close are more likely to be transcriptomically similar compared to distant cells or spots. This correlation between (transcriptomic) similarity and spatial distance is known as spatial dependency. Neglecting the spatial dependency in ST may result in inflation of type I error and an excess of differentially expressed genes (false positives). 

To visually illustrate the presence of spatial dependency in ST data, a variogram can be used. A variogram is a plot that shows how a variable (gene expression in this case), varies with respect to distance among the locations (spots or cells) where the variable was measured.

## Spatial dependency in ST data (melanoma stage III tumor biopsies)

The spatial dependency is described quantitatively by means of the spatial covariance among spots/cells. The estimation of spatial covariance is computationally expensive, especially if the ST data that contains thousands of spots/cells. For convenience, in this tutorial the smaller ST data set generated by @thrane2018spatially will be used. This data set includes lymph node biopsies from four patients with stage III melanoma, with two tissue slices per biopsy. This technology is a predecessor of the Visium assay, and allowed RNA capture in up to 1,007 spots separated each other by 200μM. The spots are 100μM in diameter, which corresponds to 5-40 cells according to the authors. The data set is included in the `spatialGE` repository, but users can also find the data in the [authors' website](https://www.spatialresearch.org/resources-published-datasets/doi-10-1158-0008-5472-can-18-0747/).

### Creating an STlist (Spatial Transcriptomics List)

The `spatialGE` package does not have built-in methods to generate variogram plots, however an STlist object will be created to normalize the ST data, which is necessary for variogram analysis. The same STlist will be used later to conduct differential expression analysis with `STdiff`.

Load the `spatialGE` package:

```{r load_spatialge, message=F, warning=F}
library('spatialGE')
```

The `STlist` function takes data in several formats. The reader is encouraged to see the `STlist` reference (see [here](https://fridleylab.github.io/spatialGE/reference/STList.html) to learn about the different options to read ST data in `spatialGE`. In this guide, we provide the `STlist` function with comma-delimited files containing gene expression counts per spot and comma-delimited containing the spatial coordinates per spot. The example files in this tutorial can be downloaded from the [GitHub repository](https://github.com/FridleyLab/spatialGEdev/tree/main/inst/extdata). They can also be accessed directly from R like so:

```{r data_fpaths1}
# spatialGE contains an object containing the file path to a built-in folder with the Thrane et al. data set
data_files <- system.file("extdata/melanoma_thrane", package="spatialGE")
# data_files <- '../inst/extdata/melanoma_thrane/'

# Find expression and coordinates files within folder
count_files <- list.files(data_files, full.names=T, pattern='counts')
coord_files <- list.files(data_files, full.names=T, pattern='mapping')
#count_files <- list.files(thrane_files(), full.names=T, pattern='counts')
#coord_files <- list.files(thrane_files(), full.names=T, pattern='mapping')
```

This melanoma ST data set also contains sample-level meta data. We will also provide this file to specify sample names:

```{r data_fpaths2}
# Find clinical data within folder
clin_file <- list.files(data_files, full.names=T, pattern='clinical')
#clin_file <- list.files(thrane_files(), full.names=T, pattern='clinical')
```

We can load the data into an STlist object with:

```{r create_stlist, warning=F}
melanoma <- STlist(rnacounts=count_files, spotcoords=coord_files, samples=clin_file)
```

*Note*: If you receive a warning indicating that an input file has a incomplete final line, you can ignore it. Alternatively, modify the input files with a text editor by adding an empty line at the end of the files.

The `melanoma` object is an STlist that contains the count data, spot coordinates, and clinical meta data.

```{r call_stlist}
melanoma
```

To normalize the ST data, the `transform_data` is used:

```{r norm_chunk}
melanoma <- transform_data(melanoma)
```

### A variogram of gene expression in ST data

The expression of the gene _COL1A1_ will be used to illustrate spatial dependency in ST. The gene _COL1A1_ is one of the collagen genes, a key component of the extracellular matrix and the stroma compartment. The stroma is the region in a cancer tissue that surrounds and often supports the tumor cells. The variogram can show how its expression varies with respect to the distance between spots within the sample ST_mel1_rep1.

Load `tidtverse` for data manipulation:

```{r variogram_libraries, message=F, warning=F}
library('tidyverse')
```

Then, data frame is created with both gene _COL1A1_ gene expression and spatial coordinates from sample ST_mel1_rep1.

```{r variogram_chunk}
# Sample and gene to be used in example
samplename <- 'ST_mel1_rep1'
gene <- 'COL1A1'

# Prepare data with coordinate and expression data
df_vgm <- melanoma@spatial_meta[[samplename]] %>%
  select(libname, xpos, ypos) %>%
  left_join(., tibble(libname=colnames(melanoma@tr_counts[[samplename]]),
                      geneexpr=as.vector(melanoma@tr_counts[[samplename]][gene, ])), by='libname') %>%
  column_to_rownames(var='libname')
```

The data frame contains the spatial coordinates of each spot (`xpos`, `ypos`) and transformed expression of _MLANA_ (`geneexpr`).

```{r vgmdf_chunk}
head(df_vgm)
```

In a variogram, each point describes the variance in gene expression (semiovariance) with respect to the distance between spots. If spatial dependency is present trend should be observed wherein semiovariance increases as the distance between spots increases. In other words, variance between close spots is lower than variance between distant spots. To calculate the variogram, the `geoR` package is used:

```{r vgm_chunk, echo=F, message=F}
# Create geodata object
geodata_obj <- geoR::as.geodata(df_vgm, coords.col=c('xpos', 'ypos'), data.col='geneexpr')

# Calculate variogram
vgm <- geoR::variog(geodata_obj, uvec=100, trend='cte')
```

Now, the variogram can be plotted:

```{r vgm_plot, fig.asp=1, fig.width=5, fig.align='center'}
plot(vgm, pch=16, cex=0.5, ylab='Semiovariance', xlab='Distance', 
     main=paste0('Variogram of ', gene, ' expression'))
text(c(2, 5, 14), c(0.65, 2.1, 0.65), 
     labels=c('Nugget','Sill', 'Range'),
     col=c('purple', 'green', 'goldenrod1'))
clip(x1=-1, x2=12, y1=-1, y2=2.1)
abline(h=2, col='green', lwd=3, lty=2)
clip(x1=-1, x2=1, y1=-1, y2=2)
abline(h=0.75, col='purple', lwd=3, lty=2)
clip(x1=-1, x2=12.1, y1=-1, y2=2)
abline(v=12, col='goldenrod1', lwd=3, lty=2)
```

We can see that variance is higher at longer distances. Note, that this relationship is disrupted at longer spot-to-spot distances (> 15 in this case). This disruption occurs because as spot-to-spot distances increase, some spots may belong to distant tissue niches that hold similarity. Regardless, the positive relationship between _COL1A1_ expression variance and distance is clear a short to medium distances, first increasing and then forming a plateau. Since a relationship exists between _COL1A1_ expression variance and distance, spatial covariance models defined by the _nugget_, _sill_, and _range_ parameters can be fitted to account for spatial dependency.



<details>

<summary>**How is the information in variogram used by `STdiff` ?**</summary>

To account for spatial dependency, the `STdiff` calculates three parameters from the variogram that are necessary to describe the relationship between gene expression variance and distance between spots or cells. The three parameters are known as _nugget_, _sill_, and _range_).

The spatial covariance can be modeled using different covariance structures. The covariance structure that best describes the relationship varies from gene to gene, and from sample to sample. Three popular spatial covariance structures are the exponential, spherical, and gaussian covariance structures, all of them defined in terms of the _nugget_, _sill_, and _range_ parameters. 

Using the variogram calculated previously, the different covariance structures can be fitted and visualized with geoR like so:

```{r vgm_model_plot, message=F, fig.asp=1, fig.width=5, fig.align='center'}
# Exponential covariance structure:
exp_model <- geoR::variofit(vgm, fix.nugget=F, cov.model='exponential',
                            ini.cov.pars=c(0.2, 11), nugget=0.7, # Sill, range, and nugget specified here
                            max.dist=15) # Maximum distance to which model will be fitted

sph_model <- geoR::variofit(vgm, fix.nugget=F, cov.model='spherical',
                            ini.cov.pars=c(0.2, 11), nugget=0.7, 
                            max.dist=15)

gau_model <- geoR::variofit(vgm, fix.nugget=F, cov.model='gaussian',
                            ini.cov.pars=c(0.2, 11), nugget=0.7, 
                            max.dist=15)

plot(vgm, pch=16, cex=0.5, ylab='Semiovariance', xlab='Distance',
     main=paste0('Variogram of ', gene, ' expression with\nfitted covariance structures'))
text(c(20.5, 20.5, 20.5), c(0.8, 0.6, 0.4), labels=c('Exponential','Spherical', 'Gaussian'))
clip(x1=15, x2=17, y1=-1, y2=2.1)
abline(h=0.8, col=alpha('red', 0.7), lwd=3)
abline(h=0.6, col=alpha('blue', 0.7), lwd=3)
abline(h=0.4, col=alpha('orange', 0.7), lwd=3)
clip(x1=-1, x2=12, y1=-1, y2=2.1)
abline(h=2, col='green', lwd=3, lty=2)
clip(x1=-1, x2=1, y1=-1, y2=2)
abline(h=0.75, col='purple', lwd=3, lty=2)
clip(x1=-1, x2=12.1, y1=-1, y2=2)
abline(v=12, col='goldenrod1', lwd=3, lty=2)
lines(exp_model, col=alpha('red', 0.7), lwd=3)
lines(sph_model, col=alpha('blue', 0.7), lwd=3)
lines(gau_model, col=alpha('orange', 0.7), lwd=3)
```

In the case of _COL1A1_ for sample 'ST_mel1_rep1', all models (exponential, spherical, and gaussian) fit well the variogram. Due to computational constraints, the `geoR` package is not used in `spatialGE` to estimate the covariance structures. Rather, the `spaMM` package has been used and only the exponential covariance structure has been implemented. While this choice is not optimal for all genes, accounting for the spatial dependency even if the model fit is not optimal, is likely better than not accounting for the spatial dependency at all.

</details>



## Differential gene expression analysis with `STdiff`

In addition to Wilcoxon's rank test and T-test, the `STdiff` function can also use *non-spatial* mixed models to test for differentially expressed genes. In the *non-spatial* mixed models the the expression $y(s_i)$ of a gene in spot/cell $i$ is described as:

$$
y(s_i) = μ + ε(s_i)
$$ 

where $μ$ is the average expression of the gene in the sample, and $ε(s_i)$ is the random error at spot/cell $i$.

To reduce the number of false positives in differential expression analysis of ST data, the `STdiff` function features a novel implementation of *spatial* mixed models with that expands the non-spatial case to include spatial covariance structures. In the *spatial* mixed models, the expression $y(s_i)$ of a gene in spot/cell $i$ is described as:

$$
Y(s_i) = μ + w(s_i) + ε(s_i) 
$$

where $w(s_i)$ denotes the exponential spatial covariance of spot/cell $i$.

In order to detect differentially expressed genes, groups of spots/cells should be defined first. In the context of ST data, these groups of spots/cells often represent tissue domains. The `spatialGE` package includes a spatially-informed tissue domain detection method called `STclust`.

### Detection of tissue domains with `STclust`

The `STclust` function groups spots/cells into tissue domains using hierarchical clustering on a wighted similarity matrix. The similarity maxtrox is derived from Euclidean transcriptomics distances "shrunk" by the spatial Euclidean distances. Please, refer to the journal article describing the method for more details [@ospina_2022]. To detect tissue domains, a spatial weight of 0.025 (`w=0.025`) will be used. The number of domains will be automatically defined by applying dynamicTreeClusters (`ks='dtc'`).

```{r clustespots_chunk, warning=F}
melanoma <- STclust(melanoma, 
                    ks='dtc', 
                    ws=0.025)
```

Results of `STclust` domain detection can be plotted using the `STplot` function:

```{r plotclustspots_chunk, fig.align='center', fig.height=4, fig.width=3}
cluster_p <- STplot(melanoma, 
                    samples='ST_mel1_rep1', 
                    ws=0.025, 
                    deepSplit=F,
                    ptsize=2, color=c('#00BA38', '#619CFF', '#F8766D'))

print(cluster_p[[1]])
```

With the domains identified, the `STdiff` function can be used. The following command tests for differentially expressed genes on the 1000 genes with the highest variation (as defined by `Seurat`'s function `FindVariableFeatures`).

```{r stdiff_chunk}
deg <- STdiff(melanoma, 
              samples='ST_mel1_rep1', 
              w=0.025, k='dtc', deepSplit=F, 
              topgenes=1000,
              sp_topgenes=0.2) # Test the 20% of the top differtially expressed genes (based on adjusted p-value)
```

The output of `STdiff` is a list, where each element is a data frame containing the test results for each sample:

```{r stdiff_res1}
print(deg[['ST_mel1_rep1']])
```

The columns `p_val` and `adj_p_val` are respectively the nominal and adjusted p-values resulting from the differential expression tests using the _non-spatial_ models. The `exp_p_val` and `exp_adj_p_val` are the p-values resulting from the _spatial_ models.

### Comparing results from the _non-spatial_ and _spatial_ models

We can look at the -log10(p-values) to compare each spatial model to the non-spatial model:

```{r compare_pval, warning=F}
# First load ggrepel to add gene names to plot
library('ggrepel')

# Calculate -log10 p-values
log_pval <- deg[['ST_mel1_rep1']] %>%
  select(gene, cluster_1, mm_p_val, exp_p_val) %>%
  mutate(log_p_val= -log10(mm_p_val+1e-100)) %>% # Add a constant to avoid -log10(0)
  mutate(log_exp_p_val= -log10(exp_p_val+1e-100))

# Create plot
p1 <- ggplot(log_pval, aes(x=log_p_val, y=log_exp_p_val)) +
  geom_abline(intercept=0, slope=1, linetype="dashed") + # Diagonal line... All p-values over this line if equal
  geom_hline(yintercept=-log10(0.05), linetype="dashed", color='gray50') + # Line p-value <0.05
  geom_vline(xintercept=-log10(0.05), linetype="dashed", color='gray50') + # Line p-value <0.05
  geom_point(aes(color=cluster_1)) +
  xlim(0, 20) + ylim(0, 20) + # Restrict axes to ease visualization (very small p-values)
  geom_text_repel(aes(label=gene), size=2, show.legend=F) +
  theme_bw()
```

```{r compare_pval_p, warning=F, fig.width=7.5, fig.height=4, fig.align='center'}
p1
```

When comparing the -log10(p-values) from the _non-spatial_ and _spatial_ models, it is possible to see that the p-values from the _non-spatial_ model tended to provide smaller, more significant p-values (i.e., larger -log10(p-values), below the 1:1 dotted red line). Some genes such as _DCN_, _MYL9_, _IGF2_ are shown as differentially expressed by the _non-spatial_ models, however, are not differentially expressed in the _spatial_ models (p-values under the -log10(0.05) indicated with horizontal and vertical dashed lines). This pattern is suggestive of type I error inflation. 

Testing spatial models with `STdiff` is both computationally- and time-intensive. The analysis in this tutorial took approximately 18 minutes in a OS machine with a 6-cores Intel Core i7 and 16GB memory. Should the user decide to disable spatial testing, the user only needs to set `sp_topgenes=0`. When running _spatial_ tests with more genes, and samples assayed with more densely sampled technologies (e.g., Visium, CosMx-SMI), the user should consider running `STdiff` in a High-Performance Computing environment (HPC, or "cluster"), setting the number of cores to use via the `cores` parameter.



<details>

<summary>**Does `spatialGE` support other differential expression test types? **</summary>

It has been shown above that accounting for spatial dependence between spots/cells is important to control inflation of type-I error (i.e., too many differentially expressed genes). Nonetheless, the `STdiff` function also implements differential expression analysis using tests commonly used in non-spatial single cell analysis. Specifically, the argument `test_type` can be set to "wilcoxon" or "t_test" to perform either Wilcoxon's Rank tests or T-tests respectively:

```{r stdiff_wilcox}
deg_wilcox <- STdiff(melanoma, 
                     samples='ST_mel1_rep1', 
                     w=0.025, k='dtc', deepSplit=F, 
                     topgenes=1000,
                     test_type='wilcoxon')
```

No spatial tests are executed when Wilcoxon's Rank tests or T-tests are requested.

```{r stdiff_wilcox_res}
print(deg_wilcox[['ST_mel1_rep1']])
```

</details>



### Visualizng `STdiff` results

The `spatialGE` package has a built-in function (`STdiff_volcano`) to generate volcano plots. The function takes `STdiff` outputs and generates a volcano plot for each sample and cluster (or cluster pair if pairwise testing was requested).

```{r volc_plots1}
vp <- STdiff_volcano(deg)
```

Now, the `ggpubr` package can be used to arrange all plots:

```{r volc_plots2, warning=F, fig.width=7.5, fig.height=4, fig.align='center'}
ggpubr::ggarrange(plotlist=vp, ncol=3, common.legend=T)
```

## References

<div id="refs"></div>



<details>

<summary>**Session Info**</summary>

```{r session_info}
sessionInfo()
```

</details>
