<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Basic functionality: Visualization, domain detection, and spatial heterogeneity • spatialGE</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Basic functionality: Visualization, domain detection, and spatial heterogeneity">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basic_functions_vignette.html">Basic functionality: Visualization, domain detection, and spatial heterogeneity</a>
    </li>
    <li>
      <a href="../articles/spatial_differential_expression.html">Spatially-informed differential gene expression</a>
    </li>
    <li>
      <a href="../articles/spatial_enrichment_gradients_smi.html">Detection of gene set spatial patterns and expression gradients in CosMx-SMI</a>
    </li>
    <li>
      <a href="../articles/troubleshooting.html">Troubleshooting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><script src="basic_functions_vignette_files/kePrint-0.0.1/kePrint.js"></script><link href="basic_functions_vignette_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Basic functionality: Visualization, domain
detection, and spatial heterogeneity</h1>
            
            <h4 data-toc-skip class="date">2025-04-07</h4>
      

      <div class="hidden name"><code>basic_functions_vignette.Rmd</code></div>

    </div>

    
    
<p><img src="img%2Fdir_structure.png" width="385"><img src="img%2Fmetadata_file.png" width="315"></p>
<p>The package <strong><em><code>spatialGE</code></em></strong> provides
a collection of tools for the visualization of gene expression from
spatially-resolved transcriptomic experiments. The data input methods
have been designed so that any data can be analyzed as long as it
contains gene expression counts per spot or cell, and the spatial
coordinates of those spots or cells. Specialized data input options are
available to allow interoperability with workflows such as <a href="https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/what-is-space-ranger" class="external-link"><em>Space
Ranger</em></a>.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The <code>spatialGE</code> repository is available at GitHub and can
be installed via <code>devtools</code>. To install <code>devtools</code>
(in case it is not already installed in your R), please run the
following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="st">"devtools"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>After making sure <code>devtools</code> is installed, proceed to
install <code>spatialGE</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#devtools::install_github("fridleylab/spatialGE")</span></span></code></pre></div>
<p>To use <code>spatialGE</code>, load the package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">'spatialGE'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="spatially-resolved-expression-of-triple-negative-breast-cancer-tumor-biopsies">Spatially-resolved expression of triple negative breast cancer tumor
biopsies<a class="anchor" aria-label="anchor" href="#spatially-resolved-expression-of-triple-negative-breast-cancer-tumor-biopsies"></a>
</h2>
<p>To show the utility of some of the functions in
<code>spatialGE</code>, we use the spatial transcriptomics data set
generated with the platform Visium by <span class="citation">Bassiouni
et al. (2023)</span>. This data set includes triple negative breast
cancer biopsies from 22 patients, with two tissue slices per patient.
The Visium platform allows gene expression quantitation in approximately
5000 capture areas (i.e., “spots”) separated each other by 100μM. The
spots are 55μM in diameter, which corresponds to 1-10 cells
approximately <a href="https://kb.10xgenomics.com/hc/en-us/articles/360035487952-How-many-cells-are-captured-in-a-single-spot" class="external-link">according
to the manufacturer</a>.</p>
<p>Data for all the tissue slices are available at the Gene Expression
Omnibus (GEO). For the purpose of this tutorial, we will use eight
samples from four patients.</p>
<p>The GEO repositories can be accessed using the following links:</p>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6433599" class="external-link">sample_117d</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6433600" class="external-link">sample_117e</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6433611" class="external-link">sample_120d</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6433612" class="external-link">sample_120e</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6433585" class="external-link">sample_092a</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6433586" class="external-link">sample_092b</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6433593" class="external-link">sample_094c</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6433594" class="external-link">sample_094d</a></li>
</ul>
<p>For convenience, the data is also available at our <a href="https://github.com/FridleyLab/spatialGE_Data" class="external-link">spatialGE_Data
GitHub repository</a>. During this tutorial, we will pull the data from
this repository. Nonetheless, users are eoncouraged to download the data
and explore its contents in order to familiarize with the input formats
accepted by <code>spatialGE</code>. For each sample directory, the file
<code>filtered_feature_bc_matrix.h5</code> contains the gene expression
counts. The <code>spatial</code> sub-directory contains the
<code>tissue_positions_list.csv</code>,
<code>tissue_hires_image.png</code>, and
<code>scalefactors_json.json</code> files. This file structure,
corresponds roughly to the file structure in outputs generated by Space
Ranger, the software provided by 10X Genomics to process raw Visium
data.</p>
<p align="center">
<img src="img%2Fdir_structure.png" height="550" width="550"></p>
<p>In this tutorial, data will be deposited in a temporary directory.
However, the download path can be changed within the following code
block:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">visium_tmp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">visium_tmp</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">visium_tmp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">'https://github.com/FridleyLab/spatialGE_Data/raw/refs/heads/main/tnbc_bassiouni.zip?download='</span>, </span>
<span>              destfile<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">visium_tmp</span>, <span class="st">'/'</span>, <span class="st">'tnbc_bassiouni.zip'</span><span class="op">)</span>, mode<span class="op">=</span><span class="st">'wb'</span><span class="op">)</span></span>
<span><span class="va">zip_tmp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">visium_tmp</span>, pattern<span class="op">=</span><span class="st">'tnbc_bassiouni.zip$'</span>, full.names<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span>zipfile<span class="op">=</span><span class="va">zip_tmp</span>, exdir<span class="op">=</span><span class="va">visium_tmp</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="creating-an-stlist-spatial-transcriptomics-list">Creating an STList (Spatial Transcriptomics List)<a class="anchor" aria-label="anchor" href="#creating-an-stlist-spatial-transcriptomics-list"></a>
</h2>
<p>In <code>spatialGE</code>, raw and processed data are stored in an
<strong><em>STlist</em></strong> (S4 class object). As previously
mentioned, an STlist can be created with the function
<code>STlist</code>, from a variety of data formats (see <a href="https://fridleylab.github.io/spatialGE/reference/STlist.html" class="external-link">here</a>
for more info or type <code><a href="../reference/STlist.html">?STlist</a></code> in the R console). In this
tutorial we will provide the file paths to the folders downloaded in the
previous step.</p>
<p>Additionally, we will input meta data associated with each sample.
The meta data is provided in the form of a comma-delimited file. We have
extracted some of the clinical meta data for the eight samples from the
original publication and saved it as part of the <code>spatialGE</code>
package. The user is encouraged to look at the structure of this file by
downloading it from the <a href="https://github.com/FridleyLab/spatialGE_Data" class="external-link">spatialGE_Data
GitHub repository</a>. The most important aspect when constructing this
file is that the sample names are in the first column, and they match
the names of the folders containing the data:</p>
<p align="center">
<img src="img%2Fmetadata_file.png" height="250" width="550"></p>
<p>From the temporary directory, we can use R to generate the file paths
to be passed to the <code>STlist</code> function:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">visium_folders</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.dirs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">visium_tmp</span>, <span class="st">'/tnbc_bassiouni'</span><span class="op">)</span>, full.names<span class="op">=</span><span class="cn">TRUE</span>, recursive<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The meta data can be accessed directly from the
<code>spatialGE</code> package installed in the computer like so:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clin_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">visium_tmp</span>, <span class="st">'/tnbc_bassiouni'</span><span class="op">)</span>, full.names<span class="op">=</span><span class="cn">TRUE</span>, recursive<span class="op">=</span><span class="cn">FALSE</span>, pattern<span class="op">=</span><span class="st">'clinical'</span><span class="op">)</span></span></code></pre></div>
<p>We can load the files into an STlist using this command:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tnbc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STlist.html">STlist</a></span><span class="op">(</span>rnacounts<span class="op">=</span><span class="va">visium_folders</span>, samples<span class="op">=</span><span class="va">clin_file</span>, cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Found Visium data</span></span>
<span><span class="co">#&gt;  Found 8 Visium samples</span></span>
<span><span class="co">#&gt;  Data read completed</span></span>
<span><span class="co">#&gt; Matching gene expression and coordinate data...</span></span>
<span><span class="co">#&gt; Completed STlist!</span></span></code></pre></div>
<p>The <code>tnbc</code> object is an STlist that contains the count
data, spot coordinates, and clinical meta data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tnbc</span></span>
<span><span class="co">#&gt; Spatial Transcriptomics List (STlist).</span></span>
<span><span class="co">#&gt; 8 spatial array(s):</span></span>
<span><span class="co">#&gt;  sample_117d (1949 ROIs|spots|cells x 23755 genes)</span></span>
<span><span class="co">#&gt;  sample_117e (754 ROIs|spots|cells x 21662 genes)</span></span>
<span><span class="co">#&gt;  sample_120d (1359 ROIs|spots|cells x 21886 genes)</span></span>
<span><span class="co">#&gt;  sample_120e (844 ROIs|spots|cells x 20389 genes)</span></span>
<span><span class="co">#&gt;  sample_092a (1289 ROIs|spots|cells x 24227 genes)</span></span>
<span><span class="co">#&gt;  sample_092b (1376 ROIs|spots|cells x 23511 genes)</span></span>
<span><span class="co">#&gt;  sample_094c (1542 ROIs|spots|cells x 24581 genes)</span></span>
<span><span class="co">#&gt;  sample_094d (1391 ROIs|spots|cells x 24167 genes)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3 variables in sample-level data:</span></span>
<span><span class="co">#&gt;   patient_id, neoadjuvant, overall_survival_days</span></span></code></pre></div>
<p>As observed, by calling the <code>tnbc</code> object, information on
the number of spots and genes per sample is displayed. For count
statistics, the <code>summarize_STlist</code> function can be used:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarize_STlist.html">summarize_STlist</a></span><span class="op">(</span><span class="va">tnbc</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 8 × 9</span></span></span>
<span><span class="co">#&gt;   sample_name spotscells genes min_counts_per_spotcell mean_counts_per_spotcell</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sample_117d       <span style="text-decoration: underline;">1</span>949 <span style="text-decoration: underline;">23</span>755                     828                   <span style="text-decoration: underline;">35</span>264.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sample_117e        754 <span style="text-decoration: underline;">21</span>662                    <span style="text-decoration: underline;">1</span>022                   <span style="text-decoration: underline;">34</span>607.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sample_120d       <span style="text-decoration: underline;">1</span>359 <span style="text-decoration: underline;">21</span>886                     171                   <span style="text-decoration: underline;">13</span>831.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> sample_120e        844 <span style="text-decoration: underline;">20</span>389                      42                    <span style="text-decoration: underline;">9</span>997.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> sample_092a       <span style="text-decoration: underline;">1</span>289 <span style="text-decoration: underline;">24</span>227                     559                   <span style="text-decoration: underline;">43</span>225.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> sample_092b       <span style="text-decoration: underline;">1</span>376 <span style="text-decoration: underline;">23</span>511                    <span style="text-decoration: underline;">1</span>413                   <span style="text-decoration: underline;">34</span>809.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> sample_094c       <span style="text-decoration: underline;">1</span>542 <span style="text-decoration: underline;">24</span>581                      71                   <span style="text-decoration: underline;">45</span>805.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> sample_094d       <span style="text-decoration: underline;">1</span>391 <span style="text-decoration: underline;">24</span>167                     298                   <span style="text-decoration: underline;">37</span>157.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 4 more variables: max_counts_per_spotcell &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   min_genes_per_spotcell &lt;int&gt;, mean_genes_per_spotcell &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   max_genes_per_spotcell &lt;int&gt;</span></span></span></code></pre></div>
<p>The minimum number of counts per spot is 42, which seems low. We can
look at the distribution of counts and genes per spot using the
<code>distribution_plots</code> function:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/distribution_plots.html">distribution_plots</a></span><span class="op">(</span><span class="va">tnbc</span>, plot_type<span class="op">=</span><span class="st">'violin'</span>, plot_meta<span class="op">=</span><span class="st">'total_counts'</span><span class="op">)</span></span>
<span><span class="va">cp</span><span class="op">[[</span><span class="st">'total_counts'</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="basic_functions_vignette_files/figure-html/count_dstr-1.png" width="768"></p>
<p>Now, let us remove spots with low counts by keeping only those spots
with at least 5000 counts. We will also restrict the data set to spots
expressing at least 1000 genes. We also will remove a few spots that
have abnormally large number of counts, as in sample_094c. These
criteria are not a rule, and samples in each study have to be carefully
examined. For example, this criteria may be not enough to reduce the
differences in counts, especially for sample_120d and samples_120e.</p>
<p>We run this filter with the <code>filter_data</code> function:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tnbc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_data.html">filter_data</a></span><span class="op">(</span><span class="va">tnbc</span>, spot_minreads<span class="op">=</span><span class="fl">5000</span>, spot_mingenes<span class="op">=</span><span class="fl">1000</span>, spot_maxreads<span class="op">=</span><span class="fl">150000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/distribution_plots.html">distribution_plots</a></span><span class="op">(</span><span class="va">tnbc</span>, plot_type<span class="op">=</span><span class="st">'violin'</span>, plot_meta<span class="op">=</span><span class="st">'total_counts'</span><span class="op">)</span></span>
<span><span class="va">cp2</span><span class="op">[[</span><span class="st">'total_counts'</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="basic_functions_vignette_files/figure-html/filter_chunk-1.png" width="768"></p>
</div>
<div class="section level2">
<h2 id="exploring-variation-between-spatial-arrays">Exploring variation between spatial arrays<a class="anchor" aria-label="anchor" href="#exploring-variation-between-spatial-arrays"></a>
</h2>
<p>The functions <code>pseudobulk_samples</code> and
<code>pseudobulk_pca_plot</code> are the initial steps to obtain a quick
snapshot of the variation in gene expression among samples. The function
<code>pseudobulk_samples</code> creates (pseudo) “bulk” RNAseq data sets
by combining all counts from each sample. Then, it log transforms the
“pseudo bulk” RNAseq counts and runs a Principal Component Analysis
(PCA). Note that the spatial coordinate information is not considered
here, which is intended only as an exploratory analysis analysis. The
<code>max_var_genes</code> argument is used to specify the maximum
number of genes used for computation of PCA. The genes are selected
based on their standard deviation across samples.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tnbc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pseudobulk_samples.html">pseudobulk_samples</a></span><span class="op">(</span><span class="va">tnbc</span>, max_var_genes<span class="op">=</span><span class="fl">5000</span><span class="op">)</span></span></code></pre></div>
<p>In this case, we apply the function to look for agreement between
samples from the same patient: It is expected that tissue slices from
the same patient are more similar among them than tissue slices from
other patients. The <code>pseudobulk_pca</code> allows to map a
sample-level variable to the points in the PCA by including the name of
the column from the sample metadata (<code>patient_id</code> in this
example).</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pseudobulk_dim_plot.html">pseudobulk_dim_plot</a></span><span class="op">(</span><span class="va">tnbc</span>, plot_meta<span class="op">=</span><span class="st">'patient_id'</span><span class="op">)</span></span></code></pre></div>
<p><img src="basic_functions_vignette_files/figure-html/pca_chunk-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Users can also generate a heatmap from pseudobulk counts by calling
the <code>pseudobulk_heatmap</code> function, which also requires prior
use of the <code>pseudobulk_samples</code> function. The number of
variable genes to show can be controlled via the
<code>hm_display_genes</code> argument.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hm_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pseudobulk_heatmap.html">pseudobulk_heatmap</a></span><span class="op">(</span><span class="va">tnbc</span>, plot_meta<span class="op">=</span><span class="st">'patient_id'</span>, hm_display_genes<span class="op">=</span><span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p><img src="basic_functions_vignette_files/figure-html/heatmap_chunk-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="transformation-of-spatially-resolved-transcriptomics-data">Transformation of spatially-resolved transcriptomics data<a class="anchor" aria-label="anchor" href="#transformation-of-spatially-resolved-transcriptomics-data"></a>
</h2>
<p>Many transformation methods are available for RNAseq count data. In
<code>spatialGE</code>, the function <code>transform_data</code> applies
log-transformation to the counts, after library size normalization
performed on each sample separately. Similar to <a href="https://satijalab.org/seurat/reference/normalizedata" class="external-link">Seurat</a>,
it applies a scaling factor (<code>scale_f=10000</code> by default).</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tnbc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transform_data.html">transform_data</a></span><span class="op">(</span><span class="va">tnbc</span>, method<span class="op">=</span><span class="st">'log'</span>, cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>Users also can apply variance-stabilizing transformation (SCT; <span class="citation">Hafemeister and Satija (2019)</span>), which is another
method increasingly used in single-cell and spatial transcriptomics
studies. See <a href="https://fridleylab.github.io/spatialGE/reference/transform_data.html" class="external-link">here</a>
for details.</p>
</div>
<div class="section level2">
<h2 id="visualization-of-gene-expression-from-spatially-resolved-transcriptomics-data">Visualization of gene expression from spatially-resolved
transcriptomics data<a class="anchor" aria-label="anchor" href="#visualization-of-gene-expression-from-spatially-resolved-transcriptomics-data"></a>
</h2>
<p>After data transformation, expression of specific genes can be
visualized using “quilt” plots. The function <code>STplot</code> shows
the transformed expression of one or several genes. We have adopted the
color palettes from the packages <code>khroma</code> and
<code>RColorBrewer</code>. The name of a color palette can be passed
using the argument <code>color_pal</code>. The default behavior of the
function produces plots for all samples within the STlist, but we can
pass specific samples to be plotted using the argument
<code>samples</code>.</p>
<p>Let’s produce a quilt plot for the genes <em>NDRG1</em> and
<em>IGKC</em> (hypoxia and B-cell markers, respectively), for sample
number 1 of patient 14 (<code>samples=sample_094c</code>).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">quilts1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STplot.html">STplot</a></span><span class="op">(</span><span class="va">tnbc</span>, </span>
<span>                  genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'NDRG1'</span>, <span class="st">'IGKC'</span><span class="op">)</span>, </span>
<span>                  samples<span class="op">=</span><span class="st">'sample_094c'</span>, </span>
<span>                  color_pal<span class="op">=</span><span class="st">'BuRd'</span>, </span>
<span>                  ptsize<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<p>Because <code>spatialGE</code> functions output lists of ggplot
objects, we can plot the results side-by-side using functions such as
<code>ggarrange()</code>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">quilts1</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></span></code></pre></div>
<p><img src="basic_functions_vignette_files/figure-html/genequilt_chunk2-1.png" width="768" style="display: block; margin: auto;"></p>
<p>We can see that gene expression patterns of both genes are
non-overlapping: <em>IGKC</em> is expressed in the upper right portion
of the tissue, whereas <em>NDRG1</em> is expressed in the right bottom
portion (although relatively lower compared to <em>IGKC</em> as
indicated by the white-colored spots). The location of gene expression
of those two genes may be indicative of an immune-infiltrated area and a
tumor area. With the help of spatial interpolation, visualization of
these regions can be easier as will be showed next.</p>
<div class="section level3">
<h3 id="spatial-interpolation-of-gene-expression">Spatial interpolation of gene expression<a class="anchor" aria-label="anchor" href="#spatial-interpolation-of-gene-expression"></a>
</h3>
<p>We can predict a smooth gene expression surface for each sample. In
<code>spatialGE</code>, this prediction is achieved by using a spatial
interpolation method very popular in spatial statistics. The method
known as ‘kriging’ allows the estimation of gene expression values in
the un-sampled areas between spots, or cells/spots that were filtered
during data quality control. Estimating a transcriptomic surface via
kriging assumes that gene expression of two given points is correlated
to the spatial distance between them.</p>
<p>The function <code>gene_interpolation</code> performs kriging of gene
expression via the <code>fields</code> package. We specify that kriging
will be performed for two of the spatial samples
(<code>samples=c('sample_094c', 'sample_117e')</code>):</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tnbc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_interpolation.html">gene_interpolation</a></span><span class="op">(</span><span class="va">tnbc</span>, </span>
<span>                           genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'NDRG1'</span>, <span class="st">'IGKC'</span><span class="op">)</span>,</span>
<span>                           samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'sample_094c'</span>, <span class="st">'sample_117e'</span><span class="op">)</span>, cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Gene interpolation started.</span></span>
<span><span class="co">#&gt; Gene interpolation completed in 0.53 min.</span></span></code></pre></div>
<p>Generating gene expression surfaces can be time consuming. The finer
resolution to which the surface is to be predicted (<code>ngrid</code>
argument), the longer the time it takes. The execution time also depends
on the number of spots/cells. The surfaces can be visualized using the
<code><a href="../reference/STplot_interpolation.html">STplot_interpolation()</a></code> function:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kriges1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STplot_interpolation.html">STplot_interpolation</a></span><span class="op">(</span><span class="va">tnbc</span>,</span>
<span>                                genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'NDRG1'</span>, <span class="st">'IGKC'</span><span class="op">)</span>,</span>
<span>                                samples<span class="op">=</span><span class="st">'sample_094c'</span><span class="op">)</span></span>
<span><span class="co">#ggpubr::ggarrange(plotlist=kriges1, nrow=1, ncol=2, common.legend=TRUE, legend='bottom')</span></span></code></pre></div>
<p>By looking at the transcriptomic surfaces of the two genes, it is
easier to detect where “pockets” of high and low expression are located
within the tissue. It is now more evident that expression of
<em>NDRG1</em> is higher in the lower right region of the tumor slice,
as well in a smaller area to the left (potentially another hypoxic
region).</p>
<p>If tissue images were uploaded to the STlist
(<code>tissue_hires_image.png.gz</code>), it can be displayed for
comparison using the <code>plot_image</code> function:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ip</span> <span class="op">=</span> <span class="fu"><a href="../reference/plot_image.html">plot_image</a></span><span class="op">(</span><span class="va">tnbc</span>, samples<span class="op">=</span><span class="st">'sample_094c'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ip</span><span class="op">[[</span><span class="st">'image_sample_094c'</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="basic_functions_vignette_files/figure-html/plot_tissue-1.png" width="768"></p>
</div>
</div>
<div class="section level2">
<h2 id="unsupervised-spatially-informed-clustering-stclust">Unsupervised spatially-informed clustering (<em>STclust</em>)<a class="anchor" aria-label="anchor" href="#unsupervised-spatially-informed-clustering-stclust"></a>
</h2>
<p>Detecting tissue compartments or niches is an important part of the
study of the tissue architecture. We can do this by applying
<strong><em><code>STclust</code></em></strong>, a spatially-informed
clustering method implemented in <code>spatialGE</code>. The
<code>STclust</code> method uses weighted average matrices to capture
the transcriptomic differences among the cells/spots. As a first step in
<code>STclust</code>, top variable genes are identified via Seurat’s
<code>FindVariableFeatures</code>, and transcriptomic scaled distances
are calculated using only those genes. Next, scaled euclidean distances
are computed from the spatial coordinates of the spots/cells. The user
defines a weight (<code>ws</code>) from 0 to 1, to apply to the physical
distances. The higher the weight, the less biologically meaningful the
clustering solution is, given that the clusters would only reflect the
physical distances between the spots/cells and less information on the
transcriptomic profiles will be used. After many tests, we have found
that weights between 0.025 - 0.25 are enough to capture the tissue
heterogeneity. By default, <code>STclust</code> uses dynamic tree cuts
<span class="citation">(Langfelder, Zhang, and Horvath 2007)</span> to
define the number of clusters. But users can also test a series of k
values (<code>ks</code>). For a more detailed description of the method,
please refer to the paper describing <code>spatialGE</code> and
<code>STclust</code> <span class="citation">(Ospina et al.
2022)</span>.</p>
<p>We’ll try several weights to see it’s effect on the cluster
assignments:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tnbc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STclust.html">STclust</a></span><span class="op">(</span><span class="va">tnbc</span>, </span>
<span>                ks<span class="op">=</span><span class="st">'dtc'</span>, </span>
<span>                ws<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.025</span>, <span class="fl">0.05</span>, <span class="fl">0.2</span><span class="op">)</span>, cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; STclust started...</span></span>
<span><span class="co">#&gt; Updating STlist with results...</span></span>
<span><span class="co">#&gt; STclust completed in 0.43 min.</span></span></code></pre></div>
<p>Results of clustering can be plotted via the <code>STplot</code>
function:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STplot.html">STplot</a></span><span class="op">(</span><span class="va">tnbc</span>, </span>
<span>                    samples<span class="op">=</span><span class="st">'sample_094c'</span>, </span>
<span>                    ws<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.025</span>, <span class="fl">0.05</span>, <span class="fl">0.2</span><span class="op">)</span>,</span>
<span>                    color_pal<span class="op">=</span><span class="st">'highcontrast'</span><span class="op">)</span></span>
<span><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">cluster_p</span>, nrow<span class="op">=</span><span class="fl">2</span>, ncol<span class="op">=</span><span class="fl">2</span>, legend<span class="op">=</span><span class="st">'right'</span><span class="op">)</span></span></code></pre></div>
<p><img src="basic_functions_vignette_files/figure-html/plotclustspots_chunk-1.png" width="768" style="display: block; margin: auto;"></p>
<p>We can see that from <code>w=0</code> and <code>w=0.05</code>, we can
only detect two tissue niches. At <code>w=0.025</code>, we gain higher
resolution as one of the clusters is split and a third (‘yellow’)
cluster appears, potentially indicating that an different
transcriptional profile is present there. At <code>w=0.2</code>, the
clusters seem too compact, indicating that the weight of spatial
information is probably too high. We have used here the dynamic tree
cuts (<code>dtc</code>) to automatically select the number of clusters,
resulting in very coarse resolution tissue niches, however, users can
define their own range of k to be evaluated, allowing further detection
of tissue compartments.</p>
</div>
<div class="section level2">
<h2 id="association-between-spatial-heterogeneity-and-sample-level-variables">Association between spatial heterogeneity and sample-level
variables<a class="anchor" aria-label="anchor" href="#association-between-spatial-heterogeneity-and-sample-level-variables"></a>
</h2>
<p>To explore the relationship between a clinical (sample-level)
variable of interest and the level of gene expression spatial uniformity
within a sample, we can use the <code><a href="../reference/SThet.html">SThet()</a></code> function:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tnbc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SThet.html">SThet</a></span><span class="op">(</span><span class="va">tnbc</span>, </span>
<span>              genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'NDRG1'</span>, <span class="st">'IGKC'</span><span class="op">)</span>,</span>
<span>              method<span class="op">=</span><span class="st">'moran'</span>, cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; SThet started.</span></span>
<span><span class="co">#&gt;  Calculating spatial weights...</span></span>
<span><span class="co">#&gt; SThet completed in 0.04 min.</span></span></code></pre></div>
<p>The <code>SThet</code> function calculates the Moran’s I statsitic
(or Geary’s C) to measure the level of spatial heterogeneity in the
expression of the genes ( <em>NDRG1</em>, <em>IGKC</em>). The estimates
can be compared across samples using the function
<code><a href="../reference/compare_SThet.html">compare_SThet()</a></code></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compare_SThet.html">compare_SThet</a></span><span class="op">(</span><span class="va">tnbc</span>, </span>
<span>                   samplemeta<span class="op">=</span><span class="st">'overall_survival_days'</span>, </span>
<span>                   color_by<span class="op">=</span><span class="st">'patient_id'</span>,</span>
<span>                   gene<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'NDRG1'</span>, <span class="st">'IGKC'</span><span class="op">)</span>, </span>
<span>                   color_pal<span class="op">=</span><span class="st">'muted'</span>,</span>
<span>                   ptsize<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="basic_functions_vignette_files/figure-html/clinplot_chunk-1.png" width="768" style="display: block; margin: auto;"></p>
<p>The calculation of spatial statistics with <code>SThet</code> and and
multi-sample comparison with <code>compare_SThet</code> provides and
easy way to identify samples and genes exhibiting spatial patterns. The
previous figure shows that expression of <em>NDRG1</em> is more
spatially uniform (lower Moran’s I) across the tissues in samples from
patients 2 and 8 compared to patents 9 and 14. The samples with higher
spatial uniformity in the expression of <em>NDRG1</em> also tended to
have higher overall survival. Trends are less clear for <em>IGKC</em>,
however, it looks like samples where expression of <em>IGKC</em> was
spatially aggregated in “pockets” (higher Moran’s I) tended to have
lower survival (but note that patient 9 does not follow this trend). As
studies using spatial transcriptomics become larger, more samples will
provide more insightful patterns into the association of gene expression
spatial distribution and non-spatial traits associated with the
tissues.</p>
<p>The computed statistics are stored in the STlist for additional
analysis/plotting that the user may want to complete. The statistics
value can be accessed as a data frame using the
<code>get_gene_meta</code> function:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_gene_meta.html">get_gene_meta</a></span><span class="op">(</span><span class="va">tnbc</span>, sthet_only<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 16 × 10</span></span></span>
<span><span class="co">#&gt;    sample      patient_id neoadjuvant overall_survival_days gene  gene_mean</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> sample_117d          2 yes                          <span style="text-decoration: underline;">2</span>733 IGKC      3.07 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> sample_117d          2 yes                          <span style="text-decoration: underline;">2</span>733 NDRG1     2.13 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> sample_117e          2 yes                          <span style="text-decoration: underline;">2</span>733 IGKC      2.54 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> sample_117e          2 yes                          <span style="text-decoration: underline;">2</span>733 NDRG1     2.45 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> sample_120d          8 no                           <span style="text-decoration: underline;">1</span>453 IGKC      2.98 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> sample_120d          8 no                           <span style="text-decoration: underline;">1</span>453 NDRG1     0.672</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> sample_120e          8 no                           <span style="text-decoration: underline;">1</span>453 IGKC      2.77 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> sample_120e          8 no                           <span style="text-decoration: underline;">1</span>453 NDRG1     1.05 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> sample_092a          9 no                            308 IGKC      0.371</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> sample_092a          9 no                            308 NDRG1     1.11 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> sample_092b          9 no                            308 IGKC      0.105</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> sample_092b          9 no                            308 NDRG1     1.63 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span> sample_094c         14 yes                           718 IGKC      4.06 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span> sample_094c         14 yes                           718 NDRG1     1.38 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">15</span> sample_094d         14 yes                           718 IGKC      3.23 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">16</span> sample_094d         14 yes                           718 NDRG1     1.25 </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 4 more variables: gene_stdevs &lt;dbl&gt;, vst.variance.standardized &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   moran_i &lt;dbl&gt;, geary_c &lt;lgl&gt;</span></span></span></code></pre></div>
<details><summary><strong>How can the statistics generated by <code>SThet</code> can be
interpreted? </strong>
</summary><p>See the table below for a simplistic interpretation of the spatial
autocorrelation statistics calculated in <code>spatialGE</code>:</p>
<table class="table table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:center;">
Statistic
</th>
<th style="text-align:center;">
Clustered expression
</th>
<th style="text-align:center;">
No expression pattern
</th>
<th style="text-align:center;">
Uniform expression
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
Moran’s I
</td>
<td style="text-align:center;">
Closer to 1
</td>
<td style="text-align:center;">
Closer to 0
</td>
<td style="text-align:center;">
Closer to -1
</td>
</tr>
<tr>
<td style="text-align:center;">
Geary’s C
</td>
<td style="text-align:center;">
Closer to 0
</td>
<td style="text-align:center;">
Closer to 1
</td>
<td style="text-align:center;">
Closer to 2
</td>
</tr>
</tbody>
<tfoot><tr>
<td style="padding: 0; border:0;" colspan="100%">
<sup></sup> Note: The boundaries indicated for each statistic are
reached when the number of spots/cells is very large.
</td>
</tr></tfoot>
</table>
<p>To better understand how the Moran’s I and Geary’s C statistics
quantify spatial heterogeneity, tissue can be simulated using the
<code>scSpatialSIM</code> and <code>spatstat</code> packages. Also
<code>tidyverse</code> and <code>janitor</code> for some data
manipulation.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/bethatkinson/rpart" class="external-link">'rpart'</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://spatstat.org/" class="external-link">'spatstat'</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/FridleyLab/scSpatialSIM" class="external-link">'scSpatialSIM'</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://tidyverse.tidyverse.org" class="external-link">'tidyverse'</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/sfirke/janitor" class="external-link">'janitor'</a></span><span class="op">)</span></span></code></pre></div>
<p>The <code>sc.SpatialSIM</code> package uses spatial point processes
to simulate the locations of spots/cells within a tissue. To facilitate
interpretability of the Moran’s I and Geary’s C statistics, first tissue
will be simulated, and then gene expression values will be
simulated.</p>
<p>The first step is to create a <code>spatstat</code> observation
window:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wdw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spatstat.geom/man/owin.html" class="external-link">owin</a></span><span class="op">(</span>xrange<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span>, yrange<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sim_visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scSpatialSIM/man/CreateSimulationObject.html" class="external-link">CreateSimulationObject</a></span><span class="op">(</span>sims<span class="op">=</span><span class="fl">1</span>, cell_types<span class="op">=</span><span class="fl">1</span>, window<span class="op">=</span><span class="va">wdw</span><span class="op">)</span></span></code></pre></div>
<p>Next, the <code>sc.SpatialSIM</code> is used to generate the spatial
point process (positions of the spots) within the observation window.
Then, assignments of spots to tissue domains are simulated and
visualized:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate point process</span></span>
<span><span class="co"># Then, simulate tissue compartments</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">sim_visium</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scSpatialSIM/man/GenerateSpatialPattern.html" class="external-link">GenerateSpatialPattern</a></span><span class="op">(</span><span class="va">sim_visium</span>, gridded<span class="op">=</span><span class="cn">TRUE</span>, overwrite<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scSpatialSIM/man/GenerateTissue.html" class="external-link">GenerateTissue</a></span><span class="op">(</span>k<span class="op">=</span><span class="fl">1</span>, xmin<span class="op">=</span><span class="fl">1</span>, xmax<span class="op">=</span><span class="fl">2</span>, ymin<span class="op">=</span><span class="fl">1</span>, ymax<span class="op">=</span><span class="fl">2</span>, sdmin<span class="op">=</span><span class="fl">1</span>, sdmax<span class="op">=</span><span class="fl">2</span>, overwrite<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; x and y range inside window boundary</span></span>
<span><span class="co">#&gt; Computing tissue probability</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scSpatialSIM/man/PlotSimulation.html" class="external-link">PlotSimulation</a></span><span class="op">(</span><span class="va">sim_visium</span>, which<span class="op">=</span><span class="fl">1</span>, what<span class="op">=</span><span class="st">"tissue points"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_shape_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">19</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="basic_functions_vignette_files/figure-html/tissue_sim_2-1.png" width="576"></p>
<p>Now, the simulated tissue domain assignments are extracted from the
<code>SpatSimObj</code> object. Gene counts will be simulated in such a
way that:</p>
<ul>
<li>Expression of “gene_1” is exclusive to “Tissue_1”. If a spot was
assigned to “Tissue 1”, expression of “gene_1” is 1 (high expression).
If assigned to “Tissue 2”, expression of “gene_1” is 0.1 (low
expression). Given “Tissue 1” spots are aggregated towards the center of
the tissue, it is expected that Moran’s I &gt; 0 and Geary’s C &lt; 1
for “gene_1”.<br>
</li>
<li>Spots with high expression of “gene_2” (expression = 1) are equally
separated from spots with low expression (expression = 0.1). This
pattern is highly unlikely in a biological tissue. It is expected this
pattern yields a Moran’s I &lt; 0 and a Geary’s &gt; 1.</li>
<li>Expression of “gene_3” results from randomly assigning expression of
“gene_2” across the entire tissue. This pattern should result in Moran’s
I closer to 0 and Geary’s C closer to 1.</li>
</ul>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract tissue assignments from the `SpatSimObj` object</span></span>
<span><span class="co"># Simulate expression of 'gene_1'</span></span>
<span><span class="va">sim_visium_df</span> <span class="op">&lt;-</span> <span class="va">sim_visium</span><span class="op">@</span><span class="va">`Spatial Files`</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://sfirke.github.io/janitor/reference/clean_names.html" class="external-link">clean_names</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>gene_1<span class="op">=</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span><span class="va">tissue_assignment</span> <span class="op">==</span> <span class="st">'Tissue 1'</span> <span class="op">~</span> <span class="fl">1</span>, <span class="cn">TRUE</span> <span class="op">~</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate expression patter of "gene_2"</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sim_visium_df</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="va">i</span><span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span><span class="fl">2</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">sim_visium_df</span><span class="op">[</span><span class="va">i</span>, <span class="st">'gene_2'</span><span class="op">]</span> <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span><span class="op">{</span></span>
<span>    <span class="va">sim_visium_df</span><span class="op">[</span><span class="va">i</span>, <span class="st">'gene_2'</span><span class="op">]</span> <span class="op">=</span> <span class="fl">0.1</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Generate expression of "gene_3"</span></span>
<span><span class="co"># Set seed for resproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">sim_visium_df</span><span class="op">[[</span><span class="st">'gene_3'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">sim_visium_df</span><span class="op">[[</span><span class="st">'gene_2'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>To visualize the simulated expression and run <code>SThet</code>, an
STlist is created:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract simulated expression data</span></span>
<span><span class="va">sim_expr</span> <span class="op">&lt;-</span> <span class="va">sim_visium_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/add_column.html" class="external-link">add_column</a></span><span class="op">(</span>libname<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'spot'</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'libname'</span>, <span class="st">'gene_1'</span>, <span class="st">'gene_2'</span>, <span class="st">'gene_3'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="st">'libname'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="st">'genename'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract simulated spot locations</span></span>
<span><span class="va">sim_xy</span> <span class="op">&lt;-</span> <span class="va">sim_visium_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/add_column.html" class="external-link">add_column</a></span><span class="op">(</span>libname<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'spot'</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'libname'</span>, <span class="st">'y'</span>, <span class="st">'x'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The `STlist` function can take a list of data frames</span></span>
<span><span class="va">simulated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STlist.html">STlist</a></span><span class="op">(</span>rnacounts<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>sim_visium<span class="op">=</span><span class="va">sim_expr</span><span class="op">)</span>, </span>
<span>                    spotcoords<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>sim_visium<span class="op">=</span><span class="va">sim_xy</span><span class="op">)</span>, cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Found list of dataframes.</span></span>
<span><span class="co">#&gt; Matching gene expression and coordinate data...</span></span>
<span><span class="co">#&gt; Converting counts to sparse matrices</span></span>
<span><span class="co">#&gt; Completed STlist!</span></span>
<span></span>
<span><span class="co"># Plot expression</span></span>
<span><span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STplot.html">STplot</a></span><span class="op">(</span><span class="va">simulated</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'gene_1'</span>, <span class="st">'gene_2'</span>, <span class="st">'gene_3'</span><span class="op">)</span>, data_type<span class="op">=</span><span class="st">'raw'</span>, color_pal<span class="op">=</span><span class="st">'sunset'</span>, ptsize<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">ps</span>, ncol<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="basic_functions_vignette_files/figure-html/tissue_sim_4-1.png" width="768"></p>
<p>As expected, simulated expression of “gene_1” is aggregated (i.e.,
clustered, red spots concentrated towards the center of the tissue).
This result was intently obtained by setting the simulation parameters
to <code>xmin=2</code>, <code>xmax=3</code>, <code>ymin=2</code>,
<code>ymax=3</code> in <code>sc.SpatialSIM</code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The `SThet` function requires normalized data</span></span>
<span><span class="va">simulated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transform_data.html">transform_data</a></span><span class="op">(</span><span class="va">simulated</span>, cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run `SThet`</span></span>
<span><span class="va">simulated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SThet.html">SThet</a></span><span class="op">(</span><span class="va">simulated</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'gene_1'</span>, <span class="st">'gene_2'</span>, <span class="st">'gene_3'</span><span class="op">)</span>, method<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'moran'</span>, <span class="st">'geary'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; SThet started.</span></span>
<span><span class="co">#&gt;  Calculating spatial weights...</span></span>
<span><span class="co">#&gt; SThet completed in 0.01 min.</span></span>
<span></span>
<span><span class="co"># Extract results</span></span>
<span><span class="fu"><a href="../reference/get_gene_meta.html">get_gene_meta</a></span><span class="op">(</span><span class="va">simulated</span>, sthet_only<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 6</span></span></span>
<span><span class="co">#&gt;   sample     gene   gene_mean gene_stdevs  moran_i geary_c</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sim_visium gene_1      7.80       0.974  0.019<span style="text-decoration: underline;">3</span>    0.966</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sim_visium gene_2      7.68       0.957 -<span style="color: #BB0000;">0.015</span><span style="color: #BB0000; text-decoration: underline;">6</span>    1.01 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sim_visium gene_3      7.68       1.01   0.003<span style="text-decoration: underline;">71</span>   0.987</span></span></code></pre></div>
<p>The results for each of the metrics are as expected: “gene_1” shows
aggregation/clustering, indicative of “hot-spot” expression. “gene_2”
and “gene_3” show uniform and random expression respectively. Notice
that none of these values are very far from the “random” expectation
(Moran’s I = 0 and Geary’s C = 1). One of reasons for this result is the
effect of library-size normalization. In addition, obtaining extreme
values of I and C require extreme spatial patterns, which are unlikely
to be observed even in the way data has been simulated here.</p>
</details>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-bassiouni_2023" class="csl-entry">
Bassiouni, Rania, Michael O. Idowu, Lee D. Gibbs, Valentina Robila,
Pamela J. Grizzard, Michelle G. Webb, Jiarong Song, Ashley Noriega,
David W. Craig, and John D. Carpten. 2023. <span>“Spatial Transcriptomic
Analysis of a Diverse Patient Cohort Reveals a Conserved Architecture in
Triple-Negative Breast Cancer.”</span> <em>Cancer Research</em> 83:
34–48. <a href="https://doi.org/10.1158/0008-5472.CAN-22-2682" class="external-link">https://doi.org/10.1158/0008-5472.CAN-22-2682</a>.
</div>
<div id="ref-hafemeister_2019" class="csl-entry">
Hafemeister, Christoph, and Rahul Satija. 2019. <span>“Normalization and
Variance Stabilization of Single-Cell RNA-Seq Data Using Regularized
Negative Binomial Regression.”</span> <em>Genome Biology</em> 20: 296.
<a href="https://doi.org/10.1186/s13059-019-1874-1" class="external-link">https://doi.org/10.1186/s13059-019-1874-1</a>.
</div>
<div id="ref-langfelder_2008" class="csl-entry">
Langfelder, Peter, Bin Zhang, and Steve Horvath. 2007. <span>“<span class="nocase">Defining clusters from a hierarchical cluster tree: the
Dynamic Tree Cut package for R</span>.”</span> <em>Bioinformatics</em>
24: 719–20. <a href="https://doi.org/10.1093/bioinformatics/btm563" class="external-link">https://doi.org/10.1093/bioinformatics/btm563</a>.
</div>
<div id="ref-ospina_2022" class="csl-entry">
Ospina, Oscar E, Christopher M Wilson, Alex C Soupir, Anders Berglund,
Inna Smalley, Kenneth Y Tsai, and Brooke L Fridley. 2022. <span>“<span class="nocase">spatialGE: quantification and visualization of the tumor
microenvironment heterogeneity using spatial
transcriptomics</span>.”</span> <em>Bioinformatics</em> 38: 2645–47. <a href="https://doi.org/10.1093/bioinformatics/btac145" class="external-link">https://doi.org/10.1093/bioinformatics/btac145</a>.
</div>
</div>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Sequoia 15.3.2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/New_York</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] janitor_2.2.0          lubridate_1.9.3        forcats_1.0.0         </span></span>
<span><span class="co">#&gt;  [4] stringr_1.5.1          dplyr_1.1.4            purrr_1.0.2           </span></span>
<span><span class="co">#&gt;  [7] readr_2.1.5            tidyr_1.3.1            tibble_3.2.1          </span></span>
<span><span class="co">#&gt; [10] ggplot2_3.5.1          tidyverse_2.0.0        scSpatialSIM_0.1.3.4  </span></span>
<span><span class="co">#&gt; [13] spatstat_3.3-1         spatstat.linnet_3.2-5  spatstat.model_3.3-4  </span></span>
<span><span class="co">#&gt; [16] spatstat.explore_3.3-4 nlme_3.1-166           spatstat.random_3.3-2 </span></span>
<span><span class="co">#&gt; [19] spatstat.geom_3.3-5    spatstat.univar_3.1-1  spatstat.data_3.1-4   </span></span>
<span><span class="co">#&gt; [22] rpart_4.1.24           magrittr_2.0.3         spatialGE_1.2.0       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3    wk_0.9.2              rstudioapi_0.16.0    </span></span>
<span><span class="co">#&gt;   [4] jsonlite_1.8.8        shape_1.4.6.1         spatstat.utils_3.1-2 </span></span>
<span><span class="co">#&gt;   [7] magick_2.8.4          farver_2.1.2          rmarkdown_2.28       </span></span>
<span><span class="co">#&gt;  [10] GlobalOptions_0.1.2   fs_1.6.4              ragg_1.3.2           </span></span>
<span><span class="co">#&gt;  [13] vctrs_0.6.5           spdep_1.3-5           rstatix_0.7.2        </span></span>
<span><span class="co">#&gt;  [16] htmltools_0.5.8.1     curl_5.2.1            broom_1.0.6          </span></span>
<span><span class="co">#&gt;  [19] s2_1.1.7              spData_2.3.1          sass_0.4.9           </span></span>
<span><span class="co">#&gt;  [22] KernSmooth_2.23-24    bslib_0.8.0           htmlwidgets_1.6.4    </span></span>
<span><span class="co">#&gt;  [25] desc_1.4.3            cachem_1.1.0          ggpolypath_0.3.0     </span></span>
<span><span class="co">#&gt;  [28] lifecycle_1.0.4       iterators_1.0.14      pkgconfig_2.0.3      </span></span>
<span><span class="co">#&gt;  [31] Matrix_1.7-1          R6_2.5.1              fastmap_1.2.0        </span></span>
<span><span class="co">#&gt;  [34] snakecase_0.11.1      clue_0.3-65           digest_0.6.37        </span></span>
<span><span class="co">#&gt;  [37] colorspace_2.1-1      S4Vectors_0.42.1      tensor_1.5           </span></span>
<span><span class="co">#&gt;  [40] textshaping_0.4.0     ggpubr_0.6.0          labeling_0.4.3       </span></span>
<span><span class="co">#&gt;  [43] timechange_0.3.0      spatstat.sparse_3.1-0 fansi_1.0.6          </span></span>
<span><span class="co">#&gt;  [46] mgcv_1.9-1            polyclip_1.10-7       abind_1.4-5          </span></span>
<span><span class="co">#&gt;  [49] compiler_4.4.2        proxy_0.4-27          bit64_4.0.5          </span></span>
<span><span class="co">#&gt;  [52] withr_3.0.1           doParallel_1.0.17     backports_1.5.0      </span></span>
<span><span class="co">#&gt;  [55] carData_3.0-5         DBI_1.2.3             highr_0.11           </span></span>
<span><span class="co">#&gt;  [58] ggforce_0.4.2         ggsignif_0.6.4        MASS_7.3-61          </span></span>
<span><span class="co">#&gt;  [61] concaveman_1.1.0      rjson_0.2.22          classInt_0.4-10      </span></span>
<span><span class="co">#&gt;  [64] tools_4.4.2           units_0.8-5           goftest_1.2-3        </span></span>
<span><span class="co">#&gt;  [67] glue_1.7.0            grid_4.4.2            sf_1.0-19            </span></span>
<span><span class="co">#&gt;  [70] cluster_2.1.8         generics_0.1.3        hdf5r_1.3.11         </span></span>
<span><span class="co">#&gt;  [73] gtable_0.3.5          tzdb_0.4.0            class_7.3-22         </span></span>
<span><span class="co">#&gt;  [76] hms_1.1.3             sp_2.1-4              xml2_1.3.6           </span></span>
<span><span class="co">#&gt;  [79] car_3.1-2             utf8_1.2.4            BiocGenerics_0.50.0  </span></span>
<span><span class="co">#&gt;  [82] ggrepel_0.9.5         foreach_1.5.2         pillar_1.9.0         </span></span>
<span><span class="co">#&gt;  [85] vroom_1.6.5           splines_4.4.2         circlize_0.4.16      </span></span>
<span><span class="co">#&gt;  [88] tweenr_2.0.3          lattice_0.22-6        bit_4.0.5            </span></span>
<span><span class="co">#&gt;  [91] deldir_2.0-4          tidyselect_1.2.1      ComplexHeatmap_2.20.0</span></span>
<span><span class="co">#&gt;  [94] knitr_1.48            gridExtra_2.3         V8_5.0.0             </span></span>
<span><span class="co">#&gt;  [97] IRanges_2.38.1        svglite_2.1.3         stats4_4.4.2         </span></span>
<span><span class="co">#&gt; [100] xfun_0.47             matrixStats_1.3.0     stringi_1.8.4        </span></span>
<span><span class="co">#&gt; [103] boot_1.3-31           yaml_2.3.10           kableExtra_1.4.0     </span></span>
<span><span class="co">#&gt; [106] evaluate_0.24.0       codetools_0.2-20      cli_3.6.3            </span></span>
<span><span class="co">#&gt; [109] pbmcapply_1.5.1       systemfonts_1.1.0     munsell_0.5.1        </span></span>
<span><span class="co">#&gt; [112] jquerylib_0.1.4       Rcpp_1.0.13           png_0.1-8            </span></span>
<span><span class="co">#&gt; [115] parallel_4.4.2        pkgdown_2.1.1         viridisLite_0.4.2    </span></span>
<span><span class="co">#&gt; [118] scales_1.3.0          e1071_1.7-14          crayon_1.5.3         </span></span>
<span><span class="co">#&gt; [121] khroma_1.13.0         GetoptLong_1.0.5      rlang_1.1.4          </span></span>
<span><span class="co">#&gt; [124] cowplot_1.1.3</span></span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Oscar Ospina, Alex Soupir, Brooke Fridley.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
