<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="spatialGE">
<title>spatialGE: Guide to spatial heterogeneity analysis in melanoma (Thrane et al. 2018) • spatialGE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="spatialGE: Guide to spatial heterogeneity analysis in melanoma (Thrane et al. 2018)">
<meta property="og:description" content="spatialGE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">spatialGE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/melanoma_ST_vignette.html">spatialGE: Guide to spatial heterogeneity analysis in melanoma (Thrane et al. 2018)</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="melanoma_ST_vignette_files/kePrint-0.0.1/kePrint.js"></script><link href="melanoma_ST_vignette_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>spatialGE: Guide to spatial heterogeneity analysis in melanoma (Thrane et al. 2018)</h1>
            
      
      
      <div class="d-none name"><code>melanoma_ST_vignette.Rmd</code></div>
    </div>

    
    
<p>The package <code>spatialGE</code> provides a set of tools for the
visualization of gene expression from spatially-resolved transcriptomic
experiments, such as those generated by the 10X Visium platform.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The <code>spatialGE</code> repository is available at GitHub and can
be installed via <code>devtools</code>. To install
<code>devtools</code>, in case is not already installed in your R
console, please run the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="st">"devtools"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>After making sure <code>devtools</code> is installed, proceed to
install <code>spatialGE</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#devtools::install_github("fridleylab/spatialGE")</span></code></pre></div>
<p>To use <code>spatialGE</code>, load the package using:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://magrittr.tidyverse.org" class="external-link">'magrittr'</a></span><span class="op">)</span> <span class="co"># Use of pipe operator in this vignette</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">'spatialGE'</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="spatially-resolved-expression-of-melanoma-stage-iii-tumor-biopsies">Spatially-resolved expression of melanoma stage III tumor
biopsies<a class="anchor" aria-label="anchor" href="#spatially-resolved-expression-of-melanoma-stage-iii-tumor-biopsies"></a>
</h2>
<p>To show the utility of <code>spatialGE</code>, we use the Spatial
Transcriptomics (ST) data set generated by <a href="https:doi.org/10.1158/0008-5472.CAN-18-0747">Thrane et
al. (2018)</a>. This data set includes lymph node biopsies from four
patients with stage IIIc melanoma, with two tissue slices per biopsy.
The technology used in this study allows for RNA capture in up to 1,007
capture areas (i.e., “spots”) separated each other by 200μM. The spots
are 100μM in diameter, which corresponds to 5-40 cells according to the
authors.</p>
<p>The data set is included in the spatialGE repository, but users can
find the data in the <a href="https://www.spatialresearch.org/resources-published-datasets/doi-10-1158-0008-5472-can-18-0747/" class="external-link">authors’
website</a>.</p>
</div>
<div class="section level2">
<h2 id="creating-an-stlist-spatial-transcriptomics-list">Creating an STList (Spatial Transcriptomics List)<a class="anchor" aria-label="anchor" href="#creating-an-stlist-spatial-transcriptomics-list"></a>
</h2>
<p>In spatialGE, raw and processed data are stored in an STlist (S4
class object). The STlist can be created with the function
<code>STlist()</code>, which can take different inputs (see <a href="https://fridleylab.github.io/spatialGE/reference/STList.html" class="external-link">here</a>
for more info). For this example, we provide the STlist function with
the following inputs:</p>
<ol style="list-style-type: decimal">
<li><p>Raw RNA counts, one per sample. The raw count files have gene
names in the first column, and spots in subsequent columns. All files
must be either tab-delimited or comma-delimited.</p></li>
<li><p>Spatial y and x coordinates for each spot. The first column of
this file contains spot IDs matching the column names in the raw count
files. The second and third columns contain the y and x coordinates of
each spot in the raw count data.</p></li>
<li><p>A table with clinical data associated with each sample. This
table is optional. Alternatively users can provide only the names of
each sample (as a character vector). If the input is a table, sample
names are in the first column, and clinical or metadata variables are in
subsequent columns. The sample names should match partially the file
names of the RNA count and coordinate files (for example, the sample ID
“ST_mel4_rep2” matches the RNA count file “ST_mel4_rep2_counts.tsv” and
the coordinate file “ST_mel4_rep2_mapping.tsv”.</p></li>
</ol>
<p>The example files in this tutorial can be downloaded from the <a href="https://github.com/FridleyLab/spatialGEdev/tree/main/inst/extdata" class="external-link">GitHub
repository</a>. They are also stored in your local
<code>spatialGE</code> installation:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"spatialGE"</span><span class="op">)</span></code></pre></div>
<p>Then, we look for the necessary files within the
<code>spatialGE</code> data repository. These files contain the raw gene
counts (<code>count_files</code>), coordinates
(<code>coord_files</code>), and clinical (<code>clin_file</code>)
metadata.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">count_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">data_files</span>, full.names<span class="op">=</span><span class="cn">T</span>, pattern<span class="op">=</span><span class="st">'counts'</span><span class="op">)</span>
<span class="va">coord_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">data_files</span>, full.names<span class="op">=</span><span class="cn">T</span>, pattern<span class="op">=</span><span class="st">'mapping'</span><span class="op">)</span>
<span class="va">clin_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">data_files</span>, full.names<span class="op">=</span><span class="cn">T</span>, pattern<span class="op">=</span><span class="st">'clinical'</span><span class="op">)</span></code></pre></div>
<p>We can load the files into an STList object like so:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu">STlist</span><span class="op">(</span>rnacounts<span class="op">=</span><span class="va">count_files</span>, spotcoords<span class="op">=</span><span class="va">coord_files</span>, samples<span class="op">=</span><span class="va">clin_file</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #00BB00;">Found Matrix Data.</span></span>
<span class="co"><span style="color: #00BB00;">#&gt; Requested 8 Samples</span></span>
<span class="co"><span style="color: #00BB00;">#&gt; </span><span style="color: #BBBB00;">Cleaning Count and Coordinate Data Gene Names.</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; Converting Counts to Sparse Matrices</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; </span><span style="color: #00BB00; font-weight: bold;">Completed STlist!</span></span>
<span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; </span></span></code></pre></div>
<ul>
<li>Note: If you receive a warning indicating that an input file has a
incomplete final line, you can ignore it. Alternatively, modify the
input files with a text editor by adding an empty line at the end of the
files.</li>
</ul>
<p>The <code>melanoma</code> object is an STlist that contains the count
data, spot coordinates, and clinical data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span>
<span class="co">#&gt; Spatial Transcriptomics List (STlist).</span>
<span class="co">#&gt; 8 spatial array(s):</span>
<span class="co">#&gt;  ST_mel1_rep1</span>
<span class="co">#&gt;      ST_mel1_rep2</span>
<span class="co">#&gt;      ST_mel2_rep1</span>
<span class="co">#&gt;      ST_mel2_rep2</span>
<span class="co">#&gt;      ST_mel3_rep1</span>
<span class="co">#&gt;      ST_mel3_rep2</span>
<span class="co">#&gt;      ST_mel4_rep1</span>
<span class="co">#&gt;      ST_mel4_rep2</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 9 variables in sample data:</span>
<span class="co">#&gt;   patient, slice, gender, BRAF_status, NRAS_status, CDKN2A_status, survival, survival_months, RIN</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="exploring-variation-between-spatial-arrays">Exploring variation between spatial arrays<a class="anchor" aria-label="anchor" href="#exploring-variation-between-spatial-arrays"></a>
</h2>
<p>The function <code><a href="../reference/bulk_pca.html">bulk_pca()</a></code> allows for a quick snapshot of
the variation in gene expression among samples. The function creates
‘bulk’ RNAseq data sets by combining all counts from each sample. Then,
it log transforms the ‘pseudo bulk’ RNAseq counts and runs a Principal
Component Analysis (PCA). Note that the spatial (x, y) coordinate
information is not considered in this analysis, which is intended only
as an exploratory analysis. In this case, we apply the function to look
for agreement between samples from the same patient: It is expected that
tissue slices from the same patient are more similar among them than
tissue slices from other patients. The <code><a href="../reference/bulk_pca.html">bulk_pca()</a></code> allows to
map a clinical variable to the PCA (patient in this example), by
including the name of the column from the sample metadata file.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">pseudobulk_pca</span><span class="op">(</span><span class="va">melanoma</span>, plot_meta<span class="op">=</span><span class="st">'patient'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/pca_chunk-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="transformation-of-spatially-resolved-transcriptomics-data">Transformation of spatially-resolved transcriptomics data<a class="anchor" aria-label="anchor" href="#transformation-of-spatially-resolved-transcriptomics-data"></a>
</h2>
<p>Many transformation methods are available RNA-Seq count data. In
<code>spatialGE</code>, the function <code>transform_data()</code>
applies log-transformation to the data, after library size
normalization. Similar to Seurat, it applies a scaling factor
(<code>scale_f=10000</code> by default).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu">transform_data</span><span class="op">(</span><span class="va">melanoma</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization-of-gene-expression-from-spatially-resolved-transcriptomics-data">Visualization of gene expression from spatially-resolved
transcriptomics data<a class="anchor" aria-label="anchor" href="#visualization-of-gene-expression-from-spatially-resolved-transcriptomics-data"></a>
</h2>
<p>After data transformation, expression of specific genes can be
visualized using ‘quilt’ plots. The function
<code><a href="../reference/plot_gene_quilt.html">plot_gene_quilt()</a></code> shows the transformed expression of one
or several genes. We have adopted the color palettes from the packages
<code>khroma</code> and <code>RColorBrewer</code>. The name of a color
palette can be passed using the argument <code>color_pal</code>. The
default behavior of the function produces plots for all samples within
the STlist, but we can pass specific samples to be plotted using the
argument <code>samples</code>.</p>
<p>Let’s produce a quilt plot for the genes <em>MS4A1</em> and
<em>MLANA</em> (B-cell marker and melanoma antigen, respectively), for
sample number 2 of patient 1 (<code>samples=2</code>).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quilts1</span> <span class="op">&lt;-</span> <span class="fu">STplot</span><span class="op">(</span><span class="va">melanoma</span>, 
                  genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MS4A1'</span>, <span class="st">'MLANA'</span><span class="op">)</span>, 
                  samples<span class="op">=</span><span class="st">'ST_mel1_rep2'</span>, 
                  visium<span class="op">=</span><span class="cn">F</span>, 
                  color_pal<span class="op">=</span><span class="st">'BuRd'</span>, 
                  ptsize<span class="op">=</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
<p>Because <code>spatialGE</code> functions output lists of ggplot
objects, we can plot the results side-by-side using functions such as
<code>ggarrange()</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">quilts1</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/genequilt_chunk2-1.png" width="576" style="display: block; margin: auto;"></p>
<p>We can see that gene expression patterns of both genes are almost
opposite: <em>MS4A1</em> is expressed in the upper right portion of the
tissue, whereas <em>MLANA</em> is expressed in the left portion. The
left portion of the tissue corresponds to the melanoma compartment
detected by Thrane et al., and the upper right portion corresponds to
the a lymphoid tissue area (see Figure 3 of <a href="https:doi.org/10.1158/0008-5472.CAN-18-0747">their study</a>).</p>
<div class="section level3">
<h3 id="spatial-interpolation-of-gene-expression">Spatial interpolation of gene expression<a class="anchor" aria-label="anchor" href="#spatial-interpolation-of-gene-expression"></a>
</h3>
<p>We can predict a smooth gene expression surface for each sample. In
<code>spatialGE</code>, this prediction is achieved by a widely used
spatial interpolation method in spatial statistics. The method known as
‘kriging’ allows for the estimation of gene expression values in the
un-sampled areas between spots, or cells/spots for which the sample
processing failed to provide data. Estimating a transcriptomic surface
via kriging assumes that gene expression of two given points is
correlated to the spatial distance between them.</p>
<p>The function <code><a href="../reference/gene_krige.html">gene_krige()</a></code> performs kriging of gene
expression via the <code>geoR</code> package. We specify that kriging
will be performed for two of the spatial arrays
(<code>samples=c('ST_mel1_rep2', 'ST_mel2_rep1')</code>):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu">gene_interpolation</span><span class="op">(</span><span class="va">melanoma</span>, 
                               genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MS4A1'</span>, <span class="st">'MLANA'</span><span class="op">)</span>, 
                               samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'ST_mel1_rep2'</span>, <span class="st">'ST_mel2_rep1'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The surfaces can be visualized using the
<code>STplot_interpolation()</code> function:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges1</span> <span class="op">&lt;-</span> <span class="fu">STplot_interpolation</span><span class="op">(</span><span class="va">melanoma</span>,
                                genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MS4A1'</span>, <span class="st">'MLANA'</span><span class="op">)</span>,
                                samples<span class="op">=</span><span class="st">'ST_mel1_rep2'</span>,
                                visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: ggpolypath</span>
<span class="co">#&gt; Loading required package: ggplot2</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges1</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotkrige_chunk1-1.png" width="576" style="display: block; margin: auto;"></p>
<p>By looking at the transcriptomic surfaces of the two genes, it is
easier to detect where “pockets” of high and low expression are located
within the tissue. It is now more evident that expression of
<em>MS4A1</em> is lower on the left (melanoma) region of the tumor
slice, compared to the rest of the slice.</p>
<p>Just like with <code>STplot()</code>, we can control which tissue
slice to plot with the <code>samples</code> argument. These commands
plot sample 1 from patient 2 (“ST_mel2_rep1”).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges2</span> <span class="op">&lt;-</span> <span class="fu">STplot_interpolation</span><span class="op">(</span><span class="va">melanoma</span>,
                                genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MS4A1'</span>, <span class="st">'MLANA'</span><span class="op">)</span>,
                                samples<span class="op">=</span><span class="st">'ST_mel2_rep1'</span>,
                                visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges2</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotgenekrige_chunk2-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="unsupervised-spatially-informed-clustering-stclust">Unsupervised spatially-informed clustering (<em>STclust</em>)<a class="anchor" aria-label="anchor" href="#unsupervised-spatially-informed-clustering-stclust"></a>
</h2>
<p>Detecting tissue compartments or niches is an important part of the
study of the tissue architecture. We can do this by applying
<em>STclust</em>, a spatially-informed clustering method implemented in
<code>spatialGE</code>. STclust uses weighted average matrices to
capture the transcriptomic differences among the cells/spots. First, top
variable genes are identified via Seurat’s
<code>FindVariableFeatures</code>, and transcriptomic scaled distances
are calculated using only those genes. Next, scaled euclidean distances
are computed for the physical (x, y) distances among cells/spots. The
user defines a weight (<code>ws</code>) from 0 to 1, to apply to the
physical distances. The higher the weight, the less biologically
meaningful the clustering is given that the groups will only reflect the
physical distances between the cells/spots and less information on the
transcriptomic profiles will be used. After many tests, we have found
that weights between 0.025 - 0.25 are enough to capture the tissue
heterogeneity. The method uses <a href="https://doi.org/10.1093/bioinformatics/btm563" class="external-link">dynamic tree
cuts</a> to define the number of clusters. But users can also test a
series of k values (<code>ks</code>). For a more detailed description of
the method, please refer to the <a href="https://doi.org/10.1093/bioinformatics/btac145" class="external-link">paper describing
spatialGE</a>.</p>
<p>We’ll try several weights to see it’s effect on the cluster
assignments:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STclust.html">STclust</a></span><span class="op">(</span><span class="va">melanoma</span>, 
                    ks<span class="op">=</span><span class="st">'dtc'</span>, 
                    ws<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.025</span>, <span class="fl">0.05</span>, <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Registered S3 method overwritten by 'spatstat.geom':</span>
<span class="co">#&gt;   method     from</span>
<span class="co">#&gt;   print.boxx cli</span></code></pre></div>
<p>Results of clustering can be plotted via the <code>STplot()</code>
function:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cluster_p</span> <span class="op">&lt;-</span> <span class="fu">STplot</span><span class="op">(</span><span class="va">melanoma</span>, 
                    samples<span class="op">=</span><span class="st">'ST_mel1_rep2'</span>, 
                    ws<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.025</span>, <span class="fl">0.05</span>, <span class="fl">0.2</span><span class="op">)</span>,
                    visium<span class="op">=</span><span class="cn">F</span>, 
                    ptsize<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">cluster_p</span>, nrow<span class="op">=</span><span class="fl">2</span>, ncol<span class="op">=</span><span class="fl">2</span>, legend<span class="op">=</span><span class="st">'right'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotclustspots_chunk-1.png" width="672" style="display: block; margin: auto;"></p>
<p>We can see that at <code>w=0.025</code> and <code>w=0.05</code>,
tissue compartments are observed for the lymphoid tissue (upper right
portion of the tissue in yellow), with small pockets surrounding the
melanoma area (blue), possibly indicating immune activity around the
tumor spots. The color red likely corresponds to stromal spots. We have
used here the dynamic tree cuts (<code>dtc</code>) to automatically
select the number of clusters, but users can define their own range of k
to be evaluated, allowing further isolation of cell types.</p>
</div>
<div class="section level2">
<h2 id="association-between-spatial-heterogeneity-and-clinical-variables-sthet">Association between spatial heterogeneity and clinical variables
(<em>SThet</em>)<a class="anchor" aria-label="anchor" href="#association-between-spatial-heterogeneity-and-clinical-variables-sthet"></a>
</h2>
<p>To explore the relationship between a clinical variable of interest
and the level of gene expression spatial uniformity within a sample, we
can use the <code>SThet()</code> function:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu">SThet</span><span class="op">(</span><span class="va">melanoma</span>, 
                  genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MAPKAPK2'</span>, <span class="st">'TYR'</span>, <span class="st">'CD19'</span><span class="op">)</span>, 
                  method<span class="op">=</span><span class="st">'moran'</span><span class="op">)</span>
<span class="co">#&gt; CD19: Not present in the transformed counts for sample ST_mel3_rep1 .</span>
<span class="co">#&gt; TYR, CD19: Not present in the transformed counts for sample ST_mel3_rep2 .</span></code></pre></div>
<p>The <code>SThet</code> function calculates the Moran’s I statsitic
(or Geary’s C) to measure the level of spatial heterogeneity in the
expression of the genes (‘PDCD1’, ‘MS4A1’, ‘CD3E’). The estimates can be
compared across samples using the function
<code>compare_SThet()</code></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span> <span class="op">=</span> <span class="fu">compare_SThet</span><span class="op">(</span><span class="va">melanoma</span>, 
                  samplemeta<span class="op">=</span><span class="st">'survival_months'</span>, 
                  color_by<span class="op">=</span><span class="st">'patient'</span>,
                  gene<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MAPKAPK2'</span>, <span class="st">'TYR'</span>, <span class="st">'CD19'</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 3 rows containing missing values (geom_point).</span>
<span class="co">#&gt; Removed 3 rows containing missing values (geom_point).</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/clinplot_chunk-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The calculation of spatial statistics and multi-sample comparison
with <code>SThet</code> and <code>compare_SThet</code> provides and easy
way to identify samples and genes exhibiting spatial patterns. The
previous figure shows that the two samples from patient 1 are different
in the spatial expression patterns of <em>CD19</em>, <em>MAPKAPK2</em>,
and <em>TYR</em> (we have already explored this data set to pick these
genes). In the case of the gene <em>CD19</em> and <em>TYR</em>, the
differences likely correspond to the marked immune/B-cell and tumor
compartments respectively, which can be observed in the gene expression
surfaces of samples 1 and 2 from patient 1. Statistics like Moran’s I or
Geary’s C quantify the level to which the expression of a gene is
“concentrated” in an region of the sample (or conversely if it is evenly
distributed). It can also be seen in this plot that patient 1 also had
the highest overal survival within the data set. It is difficult to make
assertions on the effect of spatial heterogeinity on overal survival
with this small sample size, but we present this plot as an example of
the type of exploratory analysis that spatialGE can conduct.</p>
<p>See the table below for a simplistic interpretation of the spatial
autocorrelation statistics calculated in <code>spatialGE</code>:</p>
<table class="table table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:center;">
Low
</th>
<th style="text-align:center;">
Statistic
</th>
<th style="text-align:center;">
High
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
Dispersion
</td>
<td style="text-align:center;">
Moran’s I
</td>
<td style="text-align:center;">
Clustering
</td>
</tr>
<tr>
<td style="text-align:center;">
Clustering
</td>
<td style="text-align:center;">
Geary’s C
</td>
<td style="text-align:center;">
Dispersion
</td>
</tr>
</tbody>
</table>
<p>The computed statistics are stored in the STlist for additional
analysis/plotting that the user may want to complete. The statistics
value can be accessed as a data frame using a command like:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span><span class="op">@</span><span class="va">gene_meta</span><span class="op">[[</span><span class="st">'ST_mel1_rep1'</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MAPKAPK2'</span>, <span class="st">'TYR'</span>, <span class="st">'CD19'</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 6</span></span>
<span class="co">#&gt;   gene     gene_mean gene_stdevs vst.variance.standardized moran_i geary_c</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> MAPKAPK2     0.673       0.828                     0.781  0.031<span style="text-decoration: underline;">1</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> CD19         0.460       0.814                     3.67   0.251  <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> TYR          0.381       0.645                     1.29   0.133  <span style="color: #BB0000;">NA</span></span></code></pre></div>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Big Sur 10.16</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] ggpolypath_0.1.0  ggplot2_3.3.5     fields_14.1       viridis_0.6.2    </span>
<span class="co">#&gt; [5] viridisLite_0.4.0 spam_2.9-1        spatialGE_1.1.0   magrittr_2.0.2   </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] utf8_1.2.2            reticulate_1.24       tidyselect_1.1.1     </span>
<span class="co">#&gt;   [4] htmlwidgets_1.5.4     grid_4.1.2            Rtsne_0.15           </span>
<span class="co">#&gt;   [7] munsell_0.5.0         codetools_0.2-18      ragg_1.2.5           </span>
<span class="co">#&gt;  [10] ica_1.0-2             future_1.23.0         miniUI_0.1.1.1       </span>
<span class="co">#&gt;  [13] withr_2.4.3           colorspace_2.0-2      highr_0.9            </span>
<span class="co">#&gt;  [16] knitr_1.37            rstudioapi_0.13       Seurat_4.1.0         </span>
<span class="co">#&gt;  [19] ROCR_1.0-11           ggsignif_0.6.3        tensor_1.5           </span>
<span class="co">#&gt;  [22] listenv_0.8.0         labeling_0.4.2        polyclip_1.10-0      </span>
<span class="co">#&gt;  [25] bit64_4.0.5           farver_2.1.0          rprojroot_2.0.2      </span>
<span class="co">#&gt;  [28] parallelly_1.30.0     vctrs_0.3.8           generics_0.1.2       </span>
<span class="co">#&gt;  [31] xfun_0.31             R6_2.5.1              khroma_1.8.0         </span>
<span class="co">#&gt;  [34] concaveman_1.1.0      spatstat.utils_3.0-1  cachem_1.0.6         </span>
<span class="co">#&gt;  [37] reshape_0.8.9         assertthat_0.2.1      promises_1.2.0.1     </span>
<span class="co">#&gt;  [40] scales_1.2.0          vroom_1.5.7           rgeos_0.5-9          </span>
<span class="co">#&gt;  [43] gtable_0.3.0          globals_0.14.0        goftest_1.2-3        </span>
<span class="co">#&gt;  [46] rlang_1.0.0           systemfonts_1.0.4     splines_4.1.2        </span>
<span class="co">#&gt;  [49] rstatix_0.7.0         rgdal_1.6-4           lazyeval_0.2.2       </span>
<span class="co">#&gt;  [52] spatstat.geom_3.0-6   broom_0.7.12          yaml_2.2.2           </span>
<span class="co">#&gt;  [55] reshape2_1.4.4        abind_1.4-5           backports_1.4.1      </span>
<span class="co">#&gt;  [58] httpuv_1.6.5          tools_4.1.2           ellipsis_0.3.2       </span>
<span class="co">#&gt;  [61] spatstat.core_2.3-2   kableExtra_1.3.4      raster_3.5-15        </span>
<span class="co">#&gt;  [64] jquerylib_0.1.4       RColorBrewer_1.1-2    dynamicTreeCut_1.63-1</span>
<span class="co">#&gt;  [67] wordspace_0.2-6       ggridges_0.5.3        Rcpp_1.0.10          </span>
<span class="co">#&gt;  [70] plyr_1.8.6            purrr_0.3.4           ggpubr_0.4.0         </span>
<span class="co">#&gt;  [73] rpart_4.1-15          deldir_1.0-6          pbapply_1.5-0        </span>
<span class="co">#&gt;  [76] cowplot_1.1.1         zoo_1.8-9             SeuratObject_4.0.4   </span>
<span class="co">#&gt;  [79] ggrepel_0.9.1         cluster_2.1.2         fs_1.5.2             </span>
<span class="co">#&gt;  [82] data.table_1.14.2     scattermore_0.7       lmtest_0.9-39        </span>
<span class="co">#&gt;  [85] RANN_2.6.1            fitdistrplus_1.1-6    matrixStats_0.61.0   </span>
<span class="co">#&gt;  [88] hms_1.1.1             patchwork_1.1.1       mime_0.12            </span>
<span class="co">#&gt;  [91] evaluate_0.14         xtable_1.8-4          sparsesvd_0.2        </span>
<span class="co">#&gt;  [94] gridExtra_2.3         compiler_4.1.2        tibble_3.1.6         </span>
<span class="co">#&gt;  [97] maps_3.4.0            KernSmooth_2.23-20    V8_4.0.0             </span>
<span class="co">#&gt; [100] crayon_1.4.2          htmltools_0.5.2       mgcv_1.8-38          </span>
<span class="co">#&gt; [103] later_1.3.0           tzdb_0.2.0            tidyr_1.2.0          </span>
<span class="co">#&gt; [106] DBI_1.1.2             MASS_7.3-54           Matrix_1.4-1         </span>
<span class="co">#&gt; [109] car_3.0-12            readr_2.1.2           cli_3.1.1            </span>
<span class="co">#&gt; [112] parallel_4.1.2        dotCall64_1.0-1       igraph_1.2.11        </span>
<span class="co">#&gt; [115] pkgconfig_2.0.3       pkgdown_2.0.7         sp_1.4-6             </span>
<span class="co">#&gt; [118] terra_1.7-3           plotly_4.10.0         spatstat.sparse_3.0-0</span>
<span class="co">#&gt; [121] xml2_1.3.3            svglite_2.0.0         bslib_0.3.1          </span>
<span class="co">#&gt; [124] iotools_0.3-2         webshot_0.5.2         rvest_1.0.2          </span>
<span class="co">#&gt; [127] stringr_1.4.0         digest_0.6.29         sctransform_0.3.3    </span>
<span class="co">#&gt; [130] RcppAnnoy_0.0.19      spatstat.data_3.0-0   rmarkdown_2.14       </span>
<span class="co">#&gt; [133] leiden_0.3.9          uwot_0.1.11           curl_4.3.2           </span>
<span class="co">#&gt; [136] shiny_1.7.1           lifecycle_1.0.1       nlme_3.1-153         </span>
<span class="co">#&gt; [139] jsonlite_1.7.3        carData_3.0-5         desc_1.4.0           </span>
<span class="co">#&gt; [142] fansi_1.0.2           pillar_1.7.0          lattice_0.20-45      </span>
<span class="co">#&gt; [145] fastmap_1.1.0         httr_1.4.2            survival_3.4-0       </span>
<span class="co">#&gt; [148] glue_1.6.1            png_0.1-7             bit_4.0.4            </span>
<span class="co">#&gt; [151] stringi_1.7.6         sass_0.4.0            textshaping_0.3.6    </span>
<span class="co">#&gt; [154] memoise_2.0.1         dplyr_1.0.7           irlba_2.3.5          </span>
<span class="co">#&gt; [157] future.apply_1.8.1</span></code></pre></div>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Oscar Ospina, Brooke Fridley, Alex Soupir.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
