<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>spatialGE: Basic tutorial for visualization and spatial heterogeneity analysis • spatialGEdev</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="spatialGE: Basic tutorial for visualization and spatial heterogeneity analysis">
<meta property="og:description" content="spatialGEdev">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGEdev</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/melanoma_ST_vignette.html">spatialGE: Basic tutorial for visualization and spatial heterogeneity analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="melanoma_ST_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>spatialGE: Basic tutorial for visualization and spatial heterogeneity analysis</h1>
            
      
      
      <div class="hidden name"><code>melanoma_ST_vignette.Rmd</code></div>

    </div>

    
    
<p>The package <code>spatialGE</code> provides a set of tools for the visualization of gene expression from spatially-resolved arrays, such as those generated by the 10X Visium platform.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>Before installing <code>spatialGE</code>, please make sure that <code>xCell</code> is already installed.</p>
<p>The <code>spatialGE</code> repository is available at GitHub and can be installed via <code>devtools</code>. To install <code>devtools</code> in case is not installed in your R console, please run the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="st">"devtools"</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>After making sure <code>devtools</code> is installed, proceed to install <code>spatialGE</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#devtools::install_github("fridleylab/spatialGEdev")</span></code></pre></div>
<p>To use <code>spatialGE</code>, load the package using the command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">spatialGEdev</span><span class="op">)</span></code></pre></div>
</div>
<div id="spatially-resolved-expression-of-melanoma-stage-iii-tumor-biopsies" class="section level2">
<h2 class="hasAnchor">
<a href="#spatially-resolved-expression-of-melanoma-stage-iii-tumor-biopsies" class="anchor"></a>Spatially-resolved expression of melanoma stage III tumor biopsies</h2>
<p>To show the utility of <code>spatialGE</code>, we use the Spatial Transcriptomics data set generated by Thrane et al. (2018) that included lymph node biopsies from four stage IIIc melanoma patients, with two arrays per biopsy (DOI: 10.1158/0008-5472.CAN-18-0747). The spatial slides used in this melanoma comprised 1,007 ‘spots’ with 200μM of distance between their centers. The spots had a diameter of 100μM, covering 5-40 cells according to the authors.</p>
<p>In the data provided as example for <code>spatialGE</code>, we have removed probe names and duplicate gene names (resulting from different probes). The user can find the original raw data at the authors website (<a href="https://www.spatialresearch.org/resources-published-datasets/doi-10-1158-0008-5472-can-18-0747/" class="uri">https://www.spatialresearch.org/resources-published-datasets/doi-10-1158-0008-5472-can-18-0747/</a>).</p>
</div>
<div id="creating-an-stlist-spatial-transcriptomics-list" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-an-stlist-spatial-transcriptomics-list" class="anchor"></a>Creating an STList (Spatial Transcriptomics List)</h2>
<p>Raw and processed data are stored in an STList (an R S4 class object). The STList can be created with the function <code><a href="../reference/STList.html">STList()</a></code>, which takes minimally two files: 1. A text file containing file paths of gene raw count matrices, one per line. The gene count matrices must be tab-separated, with gene names in the first column, and spots in subsequent columns. Matrices of different for each tissue slice are allowed, but duplicated gene names are not. 2. A text file containing file paths of x,y coordinates for each spot, one per line. The order of the coordinate file paths must match that of the raw count matrices.</p>
<p>Optionally, the user can provide a table with clinical/phenotype data associated with each spatial array. This table is a comma-separated file with sample ID in the first column, and relevant variables in subsequent columns. The order of the rows in the clinical file must match that of the file paths.</p>
<p>The example files in this tutorial can be downloaded from the GitHub repository (<a href="https://github.com/FridleyLab/spatialGEdev/tree/main/inst/extdata" class="uri">https://github.com/FridleyLab/spatialGEdev/tree/main/inst/extdata</a>). They are also stored in your local <code>spatialGE</code> installation:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"spatialGEdev"</span><span class="op">)</span></code></pre></div>
<p>We then create two files (<code>fpaths_count_files.txt</code> and <code>fpaths_mapping_files.txt</code>), containing the file paths where the count and coordinate files are stored. Those two files will be the input to later create the Spatial Transcriptomics List (STList).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">count_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"genes"</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">data_files</span>, full.names<span class="op">=</span><span class="cn">T</span><span class="op">)</span>, value<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/write.html">write</a></span><span class="op">(</span><span class="va">count_files</span>, file<span class="op">=</span><span class="st">"fpaths_count_files.txt"</span><span class="op">)</span>
<span class="va">coord_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"mapping"</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">data_files</span>, full.names<span class="op">=</span><span class="cn">T</span><span class="op">)</span>, value<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/write.html">write</a></span><span class="op">(</span><span class="va">coord_files</span>, file<span class="op">=</span><span class="st">"fpaths_mapping_files.txt"</span><span class="op">)</span></code></pre></div>
<p>We also locate the file containing the clinical variables. This is a single file containing a comma-separated table, with one row per each of the count files. An additional first row contains the names of the variables. The first column is the sample ID of each spatial array.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clin_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"thrane_clinical.csv"</span>, package <span class="op">=</span> <span class="st">"spatialGEdev"</span><span class="op">)</span></code></pre></div>
<p>We can load the files into an STList object like so:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STList.html">STList</a></span><span class="op">(</span>countfiles <span class="op">=</span> <span class="st">"fpaths_count_files.txt"</span>, 
                   coordfiles <span class="op">=</span> <span class="st">"fpaths_mapping_files.txt"</span>,
                   clinical <span class="op">=</span> <span class="va">clin_file</span>
<span class="op">)</span>
<span class="co">#&gt; Creating STList...</span></code></pre></div>
<p>The <code>melanoma</code> object is an STList, and contains the count data, spot coordinates, and clinical data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span>
<span class="co">#&gt; Spatial Transcriptomics List (STList)</span>
<span class="co">#&gt; 4 spatial arrays.</span>
<span class="co">#&gt; 6 variables in clinical data.</span></code></pre></div>
</div>
<div id="exploring-variation-between-spatial-arrays" class="section level2">
<h2 class="hasAnchor">
<a href="#exploring-variation-between-spatial-arrays" class="anchor"></a>Exploring variation between spatial arrays</h2>
<p>The function <code>STbulk_pca()</code> allows for a quick snapshot of the variation in gene expression among samples. The function creates ‘bulk’ RNA-Seq data sets by combining all counts from a spatial array. Then, it applies limma-voom normalization to these ‘bulk’ RNA-Seq libraries and performs a PCA. Note that spatial (x,y) coordinate information is not included in this analysis, which is intended as a tool for data exploration. In this case, we apply the function to look for agreement between samples from the same patient: it is expected that tissue slices from the same patient are more similar between them, than slices from other patients. The <code>STbulk_pca()</code> allows to map a clinical variable to the PCA (gender in this example), by including the exact name of the column from the clinical/phenotype file.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bulk_pca.html">bulk_pca</a></span><span class="op">(</span><span class="va">melanoma</span>, clinvar<span class="op">=</span><span class="st">'Gender'</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: magrittr</span>
<span class="co">#&gt; Loading required package: ggplot2</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/pca_chunk-1.png" width="960"></p>
</div>
<div id="data-normalization-of-spatially-resolved-transcriptomic-data" class="section level2">
<h2 class="hasAnchor">
<a href="#data-normalization-of-spatially-resolved-transcriptomic-data" class="anchor"></a>Data normalization of spatially-resolved transcriptomic data</h2>
<p>Many normalization methods are available for both bulk and single-cell RNA-Seq data. In <code>spatialGE</code>, we apply limma-voom normalization to RNA-Seq libraries from each of the spots in order to obtain a gaussian-shape normal distribution required by most methods in this package. Many of the methods implemented in <code>spatialGE</code> have been borrowed from the field of geostatistics, which often require this kind of data normalization.</p>
<p>Normalization is achieved by applying the function <code><a href="../reference/voom_norm.html">voom_norm()</a></code> to the <code>melanoma</code> object.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/voom_norm.html">voom_norm</a></span><span class="op">(</span><span class="va">melanoma</span><span class="op">)</span>
<span class="co">#&gt; Normalizing spatial array #1...</span>
<span class="co">#&gt; Normalizing spatial array #2...</span>
<span class="co">#&gt; Normalizing spatial array #3...</span>
<span class="co">#&gt; Normalizing spatial array #4...</span></code></pre></div>
</div>
<div id="visualization-of-gene-expression-from-spatially-resolved-transcriptomics-data" class="section level2">
<h2 class="hasAnchor">
<a href="#visualization-of-gene-expression-from-spatially-resolved-transcriptomics-data" class="anchor"></a>Visualization of gene expression from spatially-resolved transcriptomics data</h2>
<p>After normalization, expression of specific genes can be visualized using ‘quilt’ plots. The function <code><a href="../reference/plot_gene_quilt.html">plot_gene_quilt()</a></code> re-creates the shape of the spatial array and shows the scaled normalized expression of a gene. We have adopted the color palettes from the package <code>khroma</code>. The color palette can be passed using the argument <code>color_pal</code>. The default behavior of the function produces plots for all spatial arrays in the STList, but we can define who should be plotted using the argument <code>plot_who</code>.</p>
<p>Let’s produce a quilt plot for the genes IGLL5 and SOX10, an immunoglobulin gene and a melanoma marker.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quilts1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'IGLL5'</span>, <span class="st">'SOX10'</span><span class="op">)</span>, color_pal<span class="op">=</span><span class="st">"YlOrBr"</span>, 
                           plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">quilts1</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/genequilt_chunk-1.png" width="960"></p>
<p>We can see that gene expression patterns of both genes are different, with a more dispersed pattern in SOX10, compared to IGLL5 The large area of lower IGLL5 expression corresponds to the melanoma pocket detected by Thrane et al. (see Figure 3 of their study).</p>
<div id="spatial-interpolation-kriging-of-gene-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#spatial-interpolation-kriging-of-gene-expression" class="anchor"></a>Spatial interpolation (‘kriging’) of gene expression</h3>
<p>We can predict a smooth surface of the gene expression in a spatial array. In <code>spatialGE</code>, this prediction is achieved by a type of spatial interpolation widely used on geographic data. The method named ‘kriging’ allows for the estimation of the normalized gene expression value of the unsampled areas between spots, or spots for which a RNA-Seq library failed to provide data. Estimating a transcriptomic surface via kriging assumes that gene expression of two given points is correlated with the physical distance between them.</p>
<p>The function <code><a href="../reference/gene_krige.html">gene_krige()</a></code> performs kriging of gene expression via the <code>geoR</code> package:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_krige.html">gene_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'IGLL5'</span>, <span class="st">'SOX10'</span><span class="op">)</span>, who<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Performing spatial interpolation ('kriging') of gene IGLL5 for subject 2...</span>
<span class="co">#&gt; Performing spatial interpolation ('kriging') of gene SOX10 for subject 2...</span>
<span class="co">#&gt; Performing spatial interpolation ('kriging') of gene IGLL5 for subject 3...</span>
<span class="co">#&gt; Performing spatial interpolation ('kriging') of gene SOX10 for subject 3...</span></code></pre></div>
<p>The surface can be visualized using the <code><a href="../reference/plot_gene_krige.html">plot_gene_krige()</a></code> function:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_krige.html">plot_gene_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'IGLL5'</span>, <span class="st">'SOX10'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="co">#&gt; Loading required namespace: rgeos</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges1</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotgenekrige_chunk1-1.png" width="960"></p>
<p>By looking at the transcriptomic surfaces of the tho genes, it is easier to detect pockets of higher (or lower) expression within the tissue. It is now more evident that expression of IGLL5 is lower on the left (melanoma) portion of the tumor slice, compared to the rest of the slice.</p>
<p>In the transcriptomic surface plots, <code>spatialGE</code> outputs three spatial heterogeneity statistics: Moran’s I, Geary’s C, and Getis-Ord Gi.</p>
<table class="table">
<caption>Heterogeneity statitstics in <code>spatialGE</code>
</caption>
<thead><tr class="header">
<th align="center">Low</th>
<th align="center">Statistic</th>
<th align="center">High</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">Dispersion</td>
<td align="center">Moran’s I</td>
<td align="center">Clustering</td>
</tr>
<tr class="even">
<td align="center">Clustering</td>
<td align="center">Geary’s C</td>
<td align="center">Dispersion</td>
</tr>
<tr class="odd">
<td align="center">“Cold spots”</td>
<td align="center">Getis-Ord Gi</td>
<td align="center">“Hot spots”</td>
</tr>
</tbody>
</table>
<p>A higher Moran’s I for IGLL5 indicates that there are clusters of high expression. In other words, gene expression of IGLL5 is not uniform across the tissue. Geary’s C was significantly different than zero in IGLL5, but not SOX10 (indicated by asterisk), further validating the higher gene expression clustering in IGLL5. Statistical significance (<em>p</em>&lt;0.05) in <code>spatialGE</code> is obtained via permutation (performed with <code>spdep</code>).</p>
<p>We can control which tissue slice to plot with the <code>plot_who</code> argument. This argument is also included in the quilt plot functions.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_krige.html">plot_gene_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'IGLL5'</span>, <span class="st">'SOX10'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">3</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges2</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotgenekrige_chunk2-1.png" width="960"></p>
</div>
</div>
<div id="gene-expression-deconvolution-in-spatial-arrays" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-expression-deconvolution-in-spatial-arrays" class="anchor"></a>Gene expression deconvolution in spatial arrays</h2>
<p>Many gene expression deconvolution methods are available, and their use will depend largely on the question and biological system to be studied. In <code>spatialGE</code>, we have implemented xCell to provide a qualitatuve snapshot of the cell composition within a spatial array. Decovolution in xCell focuses on immunological cell types, which is of crucial importance in the study of tumor microenvironment.</p>
<p>Let’s apply xCell to the spatial arrays:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatial_deconv.html">spatial_deconv</a></span><span class="op">(</span><span class="va">melanoma</span>, method<span class="op">=</span><span class="st">'xcell'</span><span class="op">)</span>
<span class="co">#&gt; This will take some time...</span>
<span class="co">#&gt; Loading required package: estimate</span>
<span class="co">#&gt; Estimating 'purity' scores for spatial array #1...</span>
<span class="co">#&gt; [1] "Merged dataset includes 8576 genes (1836 mismatched)."</span>
<span class="co">#&gt; [1] "1 gene set: StromalSignature  overlap= 134"</span>
<span class="co">#&gt; [1] "2 gene set: ImmuneSignature  overlap= 139"</span>
<span class="co">#&gt; Estimating 'purity' scores for spatial array #2...</span>
<span class="co">#&gt; [1] "Merged dataset includes 8713 genes (1699 mismatched)."</span>
<span class="co">#&gt; [1] "1 gene set: StromalSignature  overlap= 132"</span>
<span class="co">#&gt; [1] "2 gene set: ImmuneSignature  overlap= 139"</span>
<span class="co">#&gt; Estimating 'purity' scores for spatial array #3...</span>
<span class="co">#&gt; [1] "Merged dataset includes 8775 genes (1637 mismatched)."</span>
<span class="co">#&gt; [1] "1 gene set: StromalSignature  overlap= 133"</span>
<span class="co">#&gt; [1] "2 gene set: ImmuneSignature  overlap= 138"</span>
<span class="co">#&gt; Estimating 'purity' scores for spatial array #4...</span>
<span class="co">#&gt; [1] "Merged dataset includes 8708 genes (1704 mismatched)."</span>
<span class="co">#&gt; [1] "1 gene set: StromalSignature  overlap= 133"</span>
<span class="co">#&gt; [1] "2 gene set: ImmuneSignature  overlap= 137"</span>
<span class="co">#&gt; Loading required package: xCell</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to spatial array #1...</span>
<span class="co">#&gt; Warning in .gsva(expr, mapped.gset.idx.list, method, kcdf, rnaseq,</span>
<span class="co">#&gt; abs.ranking, : Some gene sets have size one. Consider setting 'min.sz &gt; 1'.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Calculating p-values for spatial array #1...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to spatial array #2...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Calculating p-values for spatial array #2...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to spatial array #3...</span>
<span class="co">#&gt; Warning in .gsva(expr, mapped.gset.idx.list, method, kcdf, rnaseq,</span>
<span class="co">#&gt; abs.ranking, : Some gene sets have size one. Consider setting 'min.sz &gt; 1'.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Calculating p-values for spatial array #3...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to spatial array #4...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Calculating p-values for spatial array #4...</span></code></pre></div>
<p>With the function <code><a href="../reference/plot_deconv_quilt.html">plot_deconv_quilt()</a></code>, we can visualize the cell type scores inferred by xCell. For example, B cells and dendritic cells.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quilts2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_quilt.html">plot_deconv_quilt</a></span><span class="op">(</span><span class="va">melanoma</span>, cells<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'b_cells'</span>, <span class="st">'i_dc'</span><span class="op">)</span>, 
                           color_pal<span class="op">=</span><span class="st">"YlOrBr"</span>, plot_who<span class="op">=</span><span class="fl">2</span>, purity<span class="op">=</span><span class="cn">T</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">quilts2</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">3</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/deconvquilts_chunk-1.png" width="960"></p>
<p>By plotting the tumor/stroma classification of the spots, the differences in cell scores can quiclky be detected. Melanoma areas (squares) show lower B cell and dendritic cell scores, as expected. Nonetheless, B cells were abundant in the lymphoid pocket (upper right part of the tissue).</p>
<p>Similar plots can be generated for gene expression, as opposed to deconvolution scores, with the <code><a href="../reference/plot_gene_quilt.html">plot_gene_quilt()</a></code> function:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quilts3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'IGLL5'</span>, <span class="st">'SOX10'</span><span class="op">)</span>, color_pal<span class="op">=</span><span class="st">"YlOrBr"</span>, 
                           plot_who<span class="op">=</span><span class="fl">2</span>, purity<span class="op">=</span><span class="cn">T</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>

<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">quilts3</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">3</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/genepurity_chunk-1.png" width="960"></p>
<p>Spatial interpolation of cell scores is also possible with the <code><a href="../reference/deconv_krige.html">deconv_krige()</a></code> function:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconv_krige.html">deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, cells<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'b_cells'</span>, <span class="st">'i_dc'</span><span class="op">)</span>, who<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; Performing spatial interpolation ('kriging') of b_cells for subject 2...</span>
<span class="co">#&gt; Performing spatial interpolation ('kriging') of i_dc for subject 2...</span></code></pre></div>
<p>We plot the surfaces using the <code><a href="../reference/plot_deconv_krige.html">plot_deconv_krige()</a></code> function.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_krige.html">plot_deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>,  cells<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'b_cells'</span>, <span class="st">'i_dc'</span><span class="op">)</span>,
                           color_pal<span class="op">=</span><span class="st">"YlOrBr"</span>, plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges3</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotdeconvkrige_chunk-1.png" width="960"></p>
<p>Cell scores in xCell are accompanied by p-values to support the significance of the scores. The p-values are calculated automatically during deconvolution in <code>spatialGE</code>, and we can plot them using the <code>pvalue</code> option:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_krige.html">plot_deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>,  cells<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'b_cells'</span>, <span class="st">'i_dc'</span><span class="op">)</span>,
                           color_pal<span class="op">=</span><span class="st">"YlOrBr"</span>, plot_who<span class="op">=</span><span class="fl">2</span>, pvalues<span class="op">=</span><span class="cn">T</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges4</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotkrigepvals_chunk-1.png" width="960"></p>
<p>B cells are present throughout the tissue, but less abundant in the melanoma pocket. The estimates were significantly different from zero as sugggested by xCell’s permutation method. The xCell scores for the dendritic cells were only significantly different from zero only for a few spots (gray hollow circles).</p>
<p>The <code><a href="../reference/plot_deconv_krige.html">plot_deconv_krige()</a></code> also allows annotation of tumor spots:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_krige.html">plot_deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>,  cells<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'b_cells'</span>, <span class="st">'i_dc'</span><span class="op">)</span>,
                           color_pal<span class="op">=</span><span class="st">"YlOrBr"</span>, plot_who<span class="op">=</span><span class="fl">2</span>, purity<span class="op">=</span><span class="cn">T</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges5</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotkrigepurity_chunk-1.png" width="960"></p>
</div>
<div id="spatially-informed-clustering-of-spots" class="section level2">
<h2 class="hasAnchor">
<a href="#spatially-informed-clustering-of-spots" class="anchor"></a>Spatially-informed clustering of spots</h2>
<p>We can now perform hierarchical clustering to detect differences in gene expression across the tissue. In <code>spatialGE</code>, we use weighted average matrices to capture the the transcriptomic differences between the spots. The transcriptomic (euclidean) distance matrix is calculated and scaled. Following, scaled euclidean distances are computed for the spatial distances between spots. The user defines a weight, from 0 to 1, to apply to the spatial distances. The higher the weight, the less biologically meaningful the clustering is given that the groups will only reflect the physical distances between the spots and less information on the transcriptomic profiles will be used. After many tests, we have found that weights between 0.1 - 0.25 should be enough to capture the tissue heterogeneity. This approach to dspatial clustering can be performed using the <code><a href="../reference/cluster_STspot.html">cluster_STspot()</a></code> function:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cluster_STspot.html">cluster_STspot</a></span><span class="op">(</span><span class="va">melanoma</span>, ks<span class="op">=</span><span class="st">'dtc'</span>, weights<span class="op">=</span><span class="fl">0.05</span><span class="op">)</span></code></pre></div>
<p>Results of clustering can be plotted via the <code><a href="../reference/plot_STclusters.html">plot_STclusters()</a></code> function:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cluster_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_STclusters.html">plot_STclusters</a></span><span class="op">(</span><span class="va">melanoma</span>, purity<span class="op">=</span><span class="cn">T</span>, plot_who<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>

<span class="co"># In case specific k values are selected:</span>
<span class="co"># for(i in 1:2){</span>
<span class="co">#  print(ggpubr::ggarrange(plotlist=cluster_p[[i]], nrow=2, ncol=2, common.legend=T, legend='bottom'))</span>
<span class="co"># }</span>

<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">cluster_p</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotclustspots_chunk-1.png" width="960"></p>
<p>We have used here the dynamicTreeCut (<code>dtc</code> option) to select the number of clusters, but users can define their own range of k to evaluate. The two tissues plotted here, show compartmentalization, but in the sample on the left, the lymphoid tissue on the right portion of the tissue is shown in yellow.</p>
</div>
<div id="association-between-spatial-heterogeneity-and-clinical-variables" class="section level2">
<h2 class="hasAnchor">
<a href="#association-between-spatial-heterogeneity-and-clinical-variables" class="anchor"></a>Association between spatial heterogeneity and clinical variables</h2>
<p>To explore the relationship between a clinical variable of interest and the level of gene expression uniformity within a tissue section, we can use the function <code><a href="../reference/plot_phenovar.html">plot_phenovar()</a></code>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pheno_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_phenovar.html">plot_phenovar</a></span><span class="op">(</span><span class="va">melanoma</span>, phenovar<span class="op">=</span><span class="st">'Survival'</span>, gene<span class="op">=</span><span class="st">'SPP1'</span><span class="op">)</span>

<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">pheno_p</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">3</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/unnamed-chunk-1-1.png" width="576"></p>
<p>In the case of the phosphoprotein SPP1 gene, samples with a more uniform expression came from subjects with a higher survival time (low Moran’s I; high Geary’s C). A tendency for fewer “hotspots” of high expression was also observed in those samples (lower Getis-Ord Gi compared to other samples).</p>
</div>
<div id="summary" class="section level2">
<h2 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h2>
<p><code>spatialGE</code> provides a suite of tools for the study of tissue heterogeneity via spatial transcriptomics. These tools are designed for exploratory analysis, and can be specially useful to understand the immune landscape of tumors.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Oscar Ospina, Brooke Fridley.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
