<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>spatialGE: Guide Spatial heterogeneity analysis in melanoma (Thrane et al. 2018) • spatialGE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="spatialGE: Guide Spatial heterogeneity analysis in melanoma (Thrane et al. 2018)">
<meta property="og:description" content="spatialGE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0.0000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/melanoma_ST_vignette.html">spatialGE: Guide Spatial heterogeneity analysis in melanoma (Thrane et al. 2018)</a>
    </li>
    <li>
      <a href="../articles/prostatecancer_Visium_vignette.html">spatialGE: Using PyKrige to study Visium data sets</a>
    </li>
    <li>
      <a href="../articles/read_GeoMx_output.html">spatialGE: Reading GeoMx-DSP outputs</a>
    </li>
    <li>
      <a href="../articles/spatialGE_troubleshoot.html">Troubleshooting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="melanoma_ST_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="melanoma_ST_vignette_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="melanoma_ST_vignette_files/plotly-binding-4.9.3/plotly.js"></script><script src="melanoma_ST_vignette_files/typedarray-0.1/typedarray.min.js"></script><script src="melanoma_ST_vignette_files/jquery-3.5.1/jquery.min.js"></script><link href="melanoma_ST_vignette_files/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="melanoma_ST_vignette_files/crosstalk-1.1.1/js/crosstalk.min.js"></script><link href="melanoma_ST_vignette_files/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="melanoma_ST_vignette_files/plotly-main-1.57.1/plotly-latest.min.js"></script><script src="melanoma_ST_vignette_files/kePrint-0.0.1/kePrint.js"></script><link href="melanoma_ST_vignette_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>spatialGE: Guide Spatial heterogeneity analysis in melanoma (Thrane et al. 2018)</h1>
            
      
      
      <div class="hidden name"><code>melanoma_ST_vignette.Rmd</code></div>

    </div>

    
    
<p>The package <code>spatialGE</code> provides a set of tools for the visualization of gene expression from spatially-resolved transcriptomic experiments, such as those generated by the 10X Visium platform.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>The <code>spatialGE</code> repository is available at GitHub and can be installed via <code>devtools</code>. To install <code>devtools</code>, in case is not already installed in your R console, please run the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="st">"devtools"</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>After making sure <code>devtools</code> is installed, proceed to install <code>spatialGE</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># devtools::install_github("fridleylab/spatialGE")</span></code></pre></div>
<p>To use <code>spatialGE</code>, load the package using:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">spatialGE</span><span class="op">)</span></code></pre></div>
</div>
<div id="spatially-resolved-expression-of-melanoma-stage-iii-tumor-biopsies" class="section level2">
<h2 class="hasAnchor">
<a href="#spatially-resolved-expression-of-melanoma-stage-iii-tumor-biopsies" class="anchor"></a>Spatially-resolved expression of melanoma stage III tumor biopsies</h2>
<p>To show the utility of <code>spatialGE</code>, we use the Spatial Transcriptomics (ST) data set generated by Thrane et al. (2018). This data set includes lymph node biopsies from four patients with stage IIIc melanoma, two spatial arrays (samples) per biopsy (<a href="https:doi.org/10.1158/0008-5472.CAN-18-0747" class="uri">https:doi.org/10.1158/0008-5472.CAN-18-0747</a>). The technology used in this study (ST) allows for the collection of up to 1,007 “spots” with 200μM of distance between their centers. Data for each sample, however, has a variable number of spots due to differences in size of the tissue slices. The spots had a diameter of 100μM, covering 5-40 cells according to the authors.</p>
<p>The data set is included in the spatialGE repository, but users can find the data in the authors’ website (<a href="https://www.spatialresearch.org/resources-published-datasets/doi-10-1158-0008-5472-can-18-0747/" class="uri">https://www.spatialresearch.org/resources-published-datasets/doi-10-1158-0008-5472-can-18-0747/</a>).</p>
</div>
<div id="creating-an-stlist-spatial-transcriptomics-list" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-an-stlist-spatial-transcriptomics-list" class="anchor"></a>Creating an STList (Spatial Transcriptomics List)</h2>
<p>Raw and processed data are stored in an STList (S4 class object). The STList can be created with the function <code><a href="../reference/STList.html">STList()</a></code>, which can take different inputs (see <a href="https://fridleylab.github.io/spatialGE/reference/STList.html">here</a> for more info). For this example, we provide to the STList function the following inputs:</p>
<ol style="list-style-type: decimal">
<li><p>Files with raw RNA counts, one per sample. The raw count files have gene names in the first column, and spots in subsequent columns.</p></li>
<li><p>Files with y and x coordinates for each spot. The first column of this file contains spot IDs matching column names in the raw count files. The second and third columns contain the Y and X coordinates of each spot respectively.</p></li>
<li><p>Names of the samples. These names should match, at least partially, the names of the count and coordinate files. In this example, sample names are included in the first column of the metadata file.</p></li>
</ol>
<p>We also input a table with clinical data associated with each sample. This table is optional, and must be a comma- or tab-delimited file with IDs for each sample in the first column, and relevant variables in subsequent columns. The array IDs should match, at least partially, the file names of the RNA count and coordinate files (for example, the sample ID “ST_mel4_rep2” matches the RNA count file “ST_mel4_rep2_counts.tsv” and coordinate file “ST_mel4_rep2_mapping.tsv”.</p>
<p>The example files in this tutorial can be downloaded from the GitHub repository (<a href="https://github.com/FridleyLab/spatialGEdev/tree/main/inst/extdata" class="uri">https://github.com/FridleyLab/spatialGEdev/tree/main/inst/extdata</a>). They are also stored in your local <code>spatialGE</code> installation:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"spatialGE"</span><span class="op">)</span></code></pre></div>
<p>Then, we look for the find the necessary files within the data repository in <code>spatialGE</code>. These files contain the raw gene counts (<code>count_files</code>), coordinates (<code>coord_files</code>), and <code>clinical</code> metadata.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">count_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">data_files</span>, full.names<span class="op">=</span><span class="cn">T</span>, pattern<span class="op">=</span><span class="st">'counts'</span><span class="op">)</span>
<span class="va">coord_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">data_files</span>, full.names<span class="op">=</span><span class="cn">T</span>, pattern<span class="op">=</span><span class="st">'mapping'</span><span class="op">)</span>
<span class="va">clin_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">data_files</span>, full.names<span class="op">=</span><span class="cn">T</span>, pattern<span class="op">=</span><span class="st">'clinical'</span><span class="op">)</span></code></pre></div>
<p>We can load the files into an STList object like so:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STList.html">STList</a></span><span class="op">(</span>rnacounts<span class="op">=</span><span class="va">count_files</span>, spotcoords<span class="op">=</span><span class="va">coord_files</span>, samples<span class="op">=</span><span class="va">clin_file</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: magrittr</span>
<span class="co">#&gt; Found Matrix Data.</span>
<span class="co">#&gt; Requested 8 Samples</span>
<span class="co">#&gt; Cleaning Count and Coordinate Data Gene Names.</span>
<span class="co">#&gt; Converting Counts to Sparse Matrices</span>
<span class="co">#&gt; Completed STList!</span></code></pre></div>
<ul>
<li>Note: If you receive a warning indicating that an input file has a incomplete final line, you can ignore it. Alternatively, modify the input files with a text editor by adding an empty line at the end of the files.</li>
</ul>
<p>The <code>melanoma</code> object is an STList, and contains the count data, spot coordinates, and clinical data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span>
<span class="co">#&gt; Spatial Transcriptomics List (STList).</span>
<span class="co">#&gt; 8 spatial array(s):</span>
<span class="co">#&gt;  ST_mel1_rep1</span>
<span class="co">#&gt;      ST_mel1_rep2</span>
<span class="co">#&gt;      ST_mel2_rep1</span>
<span class="co">#&gt;      ST_mel2_rep2</span>
<span class="co">#&gt;      ST_mel3_rep1</span>
<span class="co">#&gt;      ST_mel3_rep2</span>
<span class="co">#&gt;      ST_mel4_rep1</span>
<span class="co">#&gt;      ST_mel4_rep2</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 9 variables in sample data:</span>
<span class="co">#&gt;   patient, slice, gender, BRAF_status, NRAS_statuds, CDKN2A_status, survival, survival_months, RIN</span></code></pre></div>
</div>
<div id="exploring-variation-between-spatial-arrays" class="section level2">
<h2 class="hasAnchor">
<a href="#exploring-variation-between-spatial-arrays" class="anchor"></a>Exploring variation between spatial arrays</h2>
<p>The function <code><a href="../reference/bulk_pca.html">bulk_pca()</a></code> allows for a quick snapshot of the variation in gene expression among samples. The function creates ‘bulk’ RNAseq data sets by combining all counts from a spatial array. Then, it log transform the ‘bulk’ RNAseq counts and performs a PCA. Note that spatial (x,y) coordinate information is not considered in this analysis, which is intended only as a tool for exploratory analysis. In this case, we apply the function to look for agreement between samples from the same patient: It is expected that tissue slices from the same patient are more similar between them, than tissue slices from other patients. The <code><a href="../reference/bulk_pca.html">bulk_pca()</a></code> allows to map a clinical variable to the PCA (gender in this example), by including the exact, case-sensitive name of the column from the sample metadata file.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bulk_pca.html">bulk_pca</a></span><span class="op">(</span><span class="va">melanoma</span>, clinvar<span class="op">=</span><span class="st">'gender'</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: Matrix</span>
<span class="co">#&gt; Loading required package: ggplot2</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/pca_chunk-1.png" width="960"></p>
</div>
<div id="transformation-of-spatially-resolved-transcriptomic-data" class="section level2">
<h2 class="hasAnchor">
<a href="#transformation-of-spatially-resolved-transcriptomic-data" class="anchor"></a>Transformation of spatially-resolved transcriptomic data</h2>
<p>Many transformation methods are available for both bulk and single-cell RNA-Seq count data. In <code>spatialGE</code>, the function <code><a href="../reference/spatialTransform.html">spatialTransform()</a></code> applies log-transformation to the data, after library size normalization. Similar to Seurat, it applies a scaling factor (<code>scale_f=10000</code> by default). <code>spatialGE</code> borrows methods from the field of spatial statistics, often requiring gaussian-shaped data distributions. Hence, users also can apply voom transformation, which could in certain situations, provide more gaussian-shaped distributions than log-transformed data.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialTransform.html">spatialTransform</a></span><span class="op">(</span><span class="va">melanoma</span><span class="op">)</span></code></pre></div>
</div>
<div id="visualization-of-gene-expression-from-spatially-resolved-transcriptomics-data" class="section level2">
<h2 class="hasAnchor">
<a href="#visualization-of-gene-expression-from-spatially-resolved-transcriptomics-data" class="anchor"></a>Visualization of gene expression from spatially-resolved transcriptomics data</h2>
<p>After data transformation, expression of specific genes can be visualized using ‘quilt’ plots. The function <code><a href="../reference/plot_gene_quilt.html">plot_gene_quilt()</a></code> shows the transformed expression of one or several genes. We have adopted the color palettes from the packages <code>khroma</code> and <code>RColorBrewer</code>. The name of a color palette can be passed using the argument <code>color_pal</code>. The default behavior of the function produces plots for all samples within the STList, but we can define specific samples to be plotted using the argument <code>plot_who</code>.</p>
<p>Let’s produce a quilt plot for the genes <em>CD19</em> and <em>MLANA</em> (an immune gene and a melanoma marker), for sample number 2 of patient 1 (<code>plot_who=2</code>).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quilts1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'CD19'</span>, <span class="st">'MLANA'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span>, 
                           color_pal<span class="op">=</span><span class="st">'BuRd'</span>, ptsize<span class="op">=</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
<p>Because <code>spatialGE</code> functions output lists of ggplot objects, we can plot the results side-by-side using functions such as <code>ggarrange</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">quilts1</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/unnamed-chunk-1-1.png" width="960"></p>
<p>We can see that gene expression patterns of both genes are almost opposite: <em>CD19</em> is expressed in the upper right portion of the tissue, whereas <em>MLANA</em> is expressed in the left portion. The left portion of the tissue corresponds to the melanoma compartment detected by Thrane et al., and the upper right portion corresponds to the a lymphoid tissue area (see Figure 3 of their study).</p>
<div id="spatial-interpolation-kriging-of-gene-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#spatial-interpolation-kriging-of-gene-expression" class="anchor"></a>Spatial interpolation (“kriging”) of gene expression</h3>
<p>We can predict a smooth gene expression surface for each sample. In <code>spatialGE</code>, this prediction is achieved by a widely used spatial interpolation method in spatial statistics. The method known as ‘kriging’ allows for the estimation of gene expression values in the unsampled areas between spots, or spots for which a the sample preparation failed to provide data. Estimating a transcriptomic surface via kriging assumes that gene expression of two given points is correlated with the physical distance between them.</p>
<p>The function <code><a href="../reference/gene_krige.html">gene_krige()</a></code> performs kriging of gene expression via the <code>geoR</code> package. The argument <code>res=0.2</code>, sets the resolution of kriging, the smaller the value of <code>res</code>, the more resolution and computing time rquired. We specify that kriging will be performed for two of the spatial arrays (<code>who=c(2, 3)</code>):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_krige.html">gene_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'CD19'</span>, <span class="st">'MLANA'</span><span class="op">)</span>, who<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, python<span class="op">=</span><span class="cn">F</span>, 
                       res<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span></code></pre></div>
<p>The surfaces can be visualized using the <code><a href="../reference/plot_gene_krige.html">plot_gene_krige()</a></code> function:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_krige.html">plot_gene_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'CD19'</span>, <span class="st">'MLANA'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="co">#&gt; Loading required namespace: rgeos</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges1</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotgenekrige_chunk1-1.png" width="960"></p>
<p>By looking at the transcriptomic surfaces of the tho genes, it is easier to detect “pockets” of high (or low) expression within the tissue. It is now more evident that expression of CD19 is lower on the left (melanoma) region of the tumor slice, compared to the rest of the slice.</p>
<p>Just like in <code><a href="../reference/plot_gene_quilt.html">plot_gene_quilt()</a></code>, we can control which tissue slice to plot with the <code>plot_who</code> argument.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_krige.html">plot_gene_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'CD19'</span>, <span class="st">'MLANA'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">3</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges2</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotgenekrige_chunk2-1.png" width="960"></p>
</div>
</div>
<div id="gene-expression-deconvolution-in-spatial-arrays" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-expression-deconvolution-in-spatial-arrays" class="anchor"></a>Gene expression deconvolution in spatial arrays</h2>
<p>In <code>spatialGE</code>, we have implemented xCell to provide a qualitative snapshot of a tissue’s cell composition. Decovolution in xCell focuses on immunological cell types, which is of crucial importance in the study of the tumor microenvironment. Other deconvolution methods will be implemented in future versions of <code>spatialGE</code>. The user may also want to look at other methods such as <a href="https://academic.oup.com/nar/article/49/9/e50/6129341">SPOTlight</a>.</p>
<p>In addition to xCell, the <code>spatialDeconv</code> function automatically applies ESTIMATE to obtain tumor purity scores. The purity scores are internally clustered using model-based clustering (<code>mclust</code>) to detect tumor/stroma compartments.</p>
<p>Let’s apply xCell to the samples:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialDeconv.html">spatialDeconv</a></span><span class="op">(</span><span class="va">melanoma</span>, method<span class="op">=</span><span class="st">'xcell'</span><span class="op">)</span>
<span class="co">#&gt; This will take some time...</span>
<span class="co">#&gt; Loading required package: estimate</span>
<span class="co">#&gt; Loading required package: xCell</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel1_rep1...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel1_rep2...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel2_rep1...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel2_rep2...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel3_rep1...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel3_rep2...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel4_rep1...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel4_rep2...</span></code></pre></div>
<p>With the function <code><a href="../reference/plot_deconv_quilt.html">plot_deconv_quilt()</a></code>, we can visualize the cell type scores inferred by xCell. For example, B cells.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quilts2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_quilt.html">plot_deconv_quilt</a></span><span class="op">(</span><span class="va">melanoma</span>, cells<span class="op">=</span><span class="st">'b_cells'</span>, plot_who<span class="op">=</span><span class="fl">2</span>, purity<span class="op">=</span><span class="cn">T</span>, 
                             color_pal<span class="op">=</span><span class="st">'BuRd'</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">quilts2</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/deconvquilts_chunk-1.png" width="960"></p>
<p>By plotting the tumor/stroma classification of the spots, the differences in cell scores can quickly be detected. Melanoma areas (squares) show lower B cell as expected. Nonetheless, B cells were abundant in the lymphoid pocket (upper right part of the tissue).</p>
<p>Similar plots can be generated for gene expression. These plots can also be made interactive by making <code>inter=T</code>, which converts plots to Plotly objects. For example, let us make an interactive plot for gene <em>CD74</em>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quilts3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="st">'CD74'</span>, plot_who<span class="op">=</span><span class="fl">2</span>, purity<span class="op">=</span><span class="cn">T</span>, visium<span class="op">=</span><span class="cn">F</span>,
                           inter<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="va">quilts3</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<div id="htmlwidget-36691df21fefe9ca0c5c" style="width:960px;height:672px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-36691df21fefe9ca0c5c">{"x":{"data":[{"x":[15,16,17,18,13,14,15,16,17,18,19,13,14,15,16,17,18,13,15,16,17,18,12,13,14,16,17,12,13,16,17,11,14,16,17,18,9,13,14,16,17,18,5,6,7,8,9,12,16,17,18,3,4,8,16,17,18,19,4,7,8,15,16,17,18,19,20,21,4,5,6,15,16,17,18,19,20,21,22,23,24,4,5,15,16,17,18,19,20,21,22,23,24,25,26,27,28,5,6,8,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,11,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,8,12,13,14,15,16,17,18,19,20,21,22,23,24,25,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,11,12,13,14,15,16,17,18,19,20,21,22,23,14,15,17,18,19,20,21,22],"y":[7,7,7,7,8,8,8,8,8,8,8,9,9,9,9,9,9,10,10,10,10,10,11,11,11,11,11,12,12,12,12,13,13,13,13,13,14,14,14,14,14,14,15,15,15,15,15,15,15,15,15,16,16,16,16,16,16,16,17,17,17,17,17,17,17,17,17,17,18,18,18,18,18,18,18,18,18,18,18,18,18,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,21,21,21,21,21,21,21,21,21,21,21,21,21,21,21,21,22,22,22,22,22,22,22,22,22,22,22,22,22,22,22,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,24,24,24,24,24,24,24,24,24,24,24,24,24,24,24,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,26,26,26,26,26,26,26,26,26,26,26,26,26,27,27,27,27,27,27,27,27],"text":["X: 15<br />Y:  7<br />Tumor_Stroma: stroma<br />Value: 3.4698913","X: 16<br />Y:  7<br />Tumor_Stroma: stroma<br />Value: 1.8521417","X: 17<br />Y:  7<br />Tumor_Stroma: stroma<br />Value: 1.7892638","X: 18<br />Y:  7<br />Tumor_Stroma: stroma<br />Value: 2.8153241","X: 13<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 2.6339190","X: 14<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 3.8956918","X: 15<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 3.4381994","X: 16<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 3.1477717","X: 17<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 4.2310590","X: 18<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 2.9806352","X: 19<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 1.8267669","X: 13<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 3.0886211","X: 14<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 3.5270017","X: 15<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 3.5436595","X: 16<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 3.5297854","X: 17<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 4.2461740","X: 18<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 3.1649939","X: 13<br />Y: 10<br />Tumor_Stroma: stroma<br />Value: 3.4009944","X: 15<br />Y: 10<br />Tumor_Stroma: stroma<br />Value: 3.0243401","X: 16<br />Y: 10<br />Tumor_Stroma: stroma<br />Value: 3.8350437","X: 17<br />Y: 10<br />Tumor_Stroma: stroma<br />Value: 3.3945702","X: 18<br />Y: 10<br />Tumor_Stroma: stroma<br />Value: 3.1121117","X: 12<br />Y: 11<br />Tumor_Stroma: stroma<br />Value: 3.7247877","X: 13<br />Y: 11<br />Tumor_Stroma: stroma<br />Value: 3.9326308","X: 14<br />Y: 11<br />Tumor_Stroma: stroma<br />Value: 2.6837989","X: 16<br />Y: 11<br />Tumor_Stroma: stroma<br />Value: 2.7490922","X: 17<br />Y: 11<br />Tumor_Stroma: stroma<br />Value: 2.8685405","X: 12<br />Y: 12<br />Tumor_Stroma: stroma<br />Value: 2.5195777","X: 13<br />Y: 12<br />Tumor_Stroma: stroma<br />Value: 2.5789460","X: 16<br />Y: 12<br />Tumor_Stroma: stroma<br />Value: 4.0211494","X: 17<br />Y: 12<br />Tumor_Stroma: stroma<br />Value: 3.4199636","X: 11<br />Y: 13<br />Tumor_Stroma: stroma<br />Value: 2.2239772","X: 14<br />Y: 13<br />Tumor_Stroma: stroma<br />Value: 3.0943368","X: 16<br />Y: 13<br />Tumor_Stroma: stroma<br />Value: 3.4666744","X: 17<br />Y: 13<br />Tumor_Stroma: stroma<br />Value: 3.3036698","X: 18<br />Y: 13<br />Tumor_Stroma: stroma<br />Value: 2.4586885","X:  9<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 3.0849853","X: 13<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 2.3469448","X: 14<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 2.9433975","X: 16<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 3.6181276","X: 17<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 3.6650018","X: 18<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 2.7376651","X:  5<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 2.4756382","X:  6<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 2.6405443","X:  7<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 3.1898577","X:  8<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 2.6393359","X:  9<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 2.5622330","X: 12<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 2.4346942","X: 16<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 2.0618528","X: 17<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 3.6071831","X: 18<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 3.2106565","X:  3<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 3.4297388","X:  4<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 3.1310413","X:  8<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 3.0311608","X: 16<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 3.0347969","X: 17<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 3.1348010","X: 18<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 3.1133433","X: 19<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 3.2074271","X:  4<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 3.4687716","X:  7<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 1.9112001","X:  8<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 2.6670644","X: 15<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 2.6819013","X: 16<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 3.4629790","X: 17<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 3.1234936","X: 18<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 2.8748298","X: 19<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 1.9724295","X: 20<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 3.2244080","X: 21<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 3.0951083","X:  4<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  5<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 2.4081858","X:  6<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 1.5719573","X: 15<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 3.7719064","X: 16<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 3.5200727","X: 17<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 3.1504415","X: 18<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 2.4763033","X: 19<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 3.3193720","X: 20<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 3.1855400","X: 21<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 2.4774686","X: 22<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 1.0758100","X: 23<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 3.0683231","X: 24<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 1.4735581","X:  4<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 2.7329474","X:  5<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 15<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 2.7991539","X: 16<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 3.8104910","X: 17<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 2.4966756","X: 18<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 2.9498362","X: 19<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 3.5931823","X: 20<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 3.2986337","X: 21<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 22<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 2.4767193","X: 23<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 3.0324130","X: 24<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 3.0684535","X: 25<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 2.2122989","X: 26<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 2.9763556","X: 27<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 3.7306345","X: 28<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 2.8257157","X:  5<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  6<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.1100924","X:  8<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 1.0148737","X: 13<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 3.6230602","X: 14<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.3668053","X: 15<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.9042964","X: 16<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.7615782","X: 17<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.2162942","X: 18<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.9301530","X: 19<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 3.5426115","X: 20<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.9096879","X: 21<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 3.5469663","X: 22<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 3.5721798","X: 23<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.7024680","X: 24<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 3.1837640","X: 25<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 3.8478601","X: 26<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 1.9703016","X: 27<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.7683800","X: 28<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 3.2355372","X: 11<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 2.1749255","X: 14<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 2.8722846","X: 15<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 3.8941280","X: 16<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 2.9352023","X: 17<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 2.9064128","X: 18<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 19<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 2.3332100","X: 20<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 1.9581813","X: 21<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 3.6460376","X: 22<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 3.8676940","X: 23<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 3.9571248","X: 24<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 3.9557659","X: 25<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 3.6991366","X: 26<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 3.7604190","X: 27<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 3.6692543","X: 28<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 3.6037451","X: 13<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 2.8781671","X: 14<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.3158540","X: 15<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.5772507","X: 16<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.2231867","X: 17<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.0759149","X: 18<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.8650812","X: 19<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 2.8416431","X: 20<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.5379039","X: 21<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.5510834","X: 22<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.4112499","X: 23<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.8122934","X: 24<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 2.8006871","X: 25<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 4.1884146","X: 26<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 4.3855754","X: 27<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 4.6785998","X: 12<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 2.8012680","X: 13<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 2.5648755","X: 14<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 3.0524938","X: 15<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 3.0287085","X: 16<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 3.3742736","X: 17<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 2.1924978","X: 18<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 3.2358724","X: 19<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 3.7705246","X: 20<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 3.7516757","X: 21<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 4.0249284","X: 22<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 3.8361643","X: 23<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 3.4909001","X: 24<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 3.4102186","X: 25<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 2.8123773","X: 26<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 3.4570094","X:  8<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 1.8331382","X: 12<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 3.1372176","X: 13<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 2.6235414","X: 14<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 2.6162996","X: 15<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 2.8975214","X: 16<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 3.4132506","X: 17<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 2.8021400","X: 18<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 3.1691164","X: 19<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 20<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 3.5617813","X: 21<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 4.0952628","X: 22<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 3.9826916","X: 23<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 3.0030360","X: 24<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 3.2824523","X: 25<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 3.5503826","X:  9<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 2.7269187","X: 10<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 11<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 12<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 1.9169909","X: 13<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 3.1700524","X: 14<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 3.4460646","X: 15<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 3.0120926","X: 16<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 3.2645291","X: 17<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 3.2857487","X: 18<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 3.1892936","X: 19<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 2.8646320","X: 20<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 3.5139566","X: 21<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 4.3085595","X: 22<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 3.9113648","X: 23<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 2.9473093","X: 24<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 3.8717976","X: 11<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 12<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 2.3540713","X: 13<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 3.0214837","X: 14<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 3.1901440","X: 15<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 1.3924761","X: 16<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 3.3649815","X: 17<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 3.3340385","X: 18<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 1.9815405","X: 19<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 2.1543114","X: 20<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 2.8035953","X: 21<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 4.2874899","X: 22<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 4.0005754","X: 23<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 3.2849235","X: 14<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 3.1292978","X: 15<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 17<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 3.0988072","X: 18<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 3.6078095","X: 19<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 2.4611358","X: 20<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 3.1158115","X: 21<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 4.1482049","X: 22<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 3.8816862"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":["rgba(206,79,3,1)","rgba(255,185,87,1)","rgba(255,189,92,1)","rgba(232,123,26,1)","rgba(239,135,32,1)","rgba(169,63,5,1)","rgba(207,81,4,1)","rgba(219,101,14,1)","rgba(139,51,7,1)","rgba(226,112,20,1)","rgba(255,186,89,1)","rgba(221,105,16,1)","rgba(202,75,2,1)","rgba(201,75,2,1)","rgba(202,75,2,1)","rgba(138,51,7,1)","rgba(218,100,14,1)","rgba(208,84,5,1)","rgba(224,109,19,1)","rgba(174,65,5,1)","rgba(209,84,5,1)","rgba(220,103,16,1)","rgba(184,69,4,1)","rgba(166,61,6,1)","rgba(237,132,30,1)","rgba(235,127,28,1)","rgba(230,120,24,1)","rgba(244,142,35,1)","rgba(242,138,33,1)","rgba(158,58,6,1)","rgba(208,82,5,1)","rgba(252,161,53,1)","rgba(221,105,16,1)","rgba(206,79,3,1)","rgba(212,90,8,1)","rgba(246,146,37,1)","rgba(221,105,17,1)","rgba(251,154,41,1)","rgba(227,115,21,1)","rgba(194,72,3,1)","rgba(190,71,4,1)","rgba(235,128,28,1)","rgba(246,145,37,1)","rgba(239,134,31,1)","rgba(217,98,13,1)","rgba(239,135,31,1)","rgba(242,140,34,1)","rgba(247,148,38,1)","rgba(254,172,68,1)","rgba(195,73,3,1)","rgba(216,97,12,1)","rgba(207,82,4,1)","rgba(220,102,15,1)","rgba(224,109,18,1)","rgba(223,109,18,1)","rgba(219,102,15,1)","rgba(220,103,16,1)","rgba(216,97,12,1)","rgba(206,79,3,1)","rgba(255,181,82,1)","rgba(238,133,30,1)","rgba(238,132,30,1)","rgba(206,79,3,1)","rgba(220,103,15,1)","rgba(230,119,24,1)","rgba(254,177,76,1)","rgba(216,96,11,1)","rgba(221,105,16,1)","rgba(255,255,229,1)","rgba(248,150,39,1)","rgba(255,202,111,1)","rgba(180,67,5,1)","rgba(203,76,2,1)","rgba(219,101,14,1)","rgba(246,145,37,1)","rgba(212,89,8,1)","rgba(217,99,13,1)","rgba(246,145,37,1)","rgba(254,229,152,1)","rgba(222,106,17,1)","rgba(255,208,119,1)","rgba(236,128,28,1)","rgba(255,255,229,1)","rgba(233,124,26,1)","rgba(177,66,5,1)","rgba(245,144,36,1)","rgba(227,114,21,1)","rgba(196,73,3,1)","rgba(213,91,9,1)","rgba(255,255,229,1)","rgba(246,145,37,1)","rgba(224,109,18,1)","rgba(222,106,17,1)","rgba(252,162,54,1)","rgba(226,112,20,1)","rgba(184,68,4,1)","rgba(232,122,25,1)","rgba(255,255,229,1)","rgba(253,169,64,1)","rgba(255,231,156,1)","rgba(194,72,3,1)","rgba(250,152,40,1)","rgba(229,117,23,1)","rgba(234,127,27,1)","rgba(252,162,54,1)","rgba(228,116,22,1)","rgba(201,75,2,1)","rgba(228,117,23,1)","rgba(200,75,2,1)","rgba(198,74,3,1)","rgba(237,130,29,1)","rgba(217,99,13,1)","rgba(173,64,5,1)","rgba(254,177,77,1)","rgba(234,126,27,1)","rgba(215,95,11,1)","rgba(253,164,58,1)","rgba(230,119,24,1)","rgba(169,63,5,1)","rgba(227,115,22,1)","rgba(229,117,23,1)","rgba(255,255,229,1)","rgba(251,154,42,1)","rgba(254,178,78,1)","rgba(191,71,3,1)","rgba(171,64,5,1)","rgba(163,61,6,1)","rgba(164,61,6,1)","rgba(187,69,4,1)","rgba(181,67,4,1)","rgba(189,70,4,1)","rgba(195,73,3,1)","rgba(230,119,24,1)","rgba(212,90,8,1)","rgba(198,74,3,1)","rgba(216,96,12,1)","rgba(222,106,17,1)","rgba(172,64,5,1)","rgba(231,121,25,1)","rgba(201,75,2,1)","rgba(200,75,2,1)","rgba(208,83,5,1)","rgba(176,65,5,1)","rgba(233,124,26,1)","rgba(143,53,7,1)","rgba(126,46,7,1)","rgba(102,37,6,1)","rgba(233,124,26,1)","rgba(242,139,34,1)","rgba(223,107,18,1)","rgba(224,109,19,1)","rgba(210,86,6,1)","rgba(252,163,56,1)","rgba(215,95,11,1)","rgba(180,67,4,1)","rgba(182,68,4,1)","rgba(157,58,6,1)","rgba(174,65,5,1)","rgba(205,77,2,1)","rgba(208,83,5,1)","rgba(232,123,26,1)","rgba(206,80,3,1)","rgba(255,186,89,1)","rgba(219,102,15,1)","rgba(240,136,32,1)","rgba(240,136,32,1)","rgba(229,118,23,1)","rgba(208,83,5,1)","rgba(233,124,26,1)","rgba(218,100,14,1)","rgba(255,255,229,1)","rgba(199,74,3,1)","rgba(151,56,6,1)","rgba(161,60,6,1)","rgba(225,111,19,1)","rgba(213,92,9,1)","rgba(200,75,2,1)","rgba(236,129,29,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(254,181,81,1)","rgba(218,100,14,1)","rgba(207,81,4,1)","rgba(224,110,19,1)","rgba(214,93,10,1)","rgba(213,92,9,1)","rgba(217,98,13,1)","rgba(230,120,24,1)","rgba(204,76,2,1)","rgba(133,49,7,1)","rgba(167,62,5,1)","rgba(227,114,21,1)","rgba(171,63,5,1)","rgba(255,255,229,1)","rgba(250,153,41,1)","rgba(224,110,19,1)","rgba(217,98,13,1)","rgba(255,213,126,1)","rgba(210,86,6,1)","rgba(211,88,7,1)","rgba(254,177,76,1)","rgba(253,166,60,1)","rgba(233,124,26,1)","rgba(135,50,7,1)","rgba(160,59,6,1)","rgba(213,92,9,1)","rgba(220,102,15,1)","rgba(255,255,229,1)","rgba(221,104,16,1)","rgba(195,73,3,1)","rgba(246,146,37,1)","rgba(220,103,15,1)","rgba(147,54,7,1)","rgba(170,63,5,1)"],"opacity":1,"size":1.88976377952756,"symbol":"cross-thin-open","line":{"width":3.77952755905512,"color":["rgba(206,79,3,1)","rgba(255,185,87,1)","rgba(255,189,92,1)","rgba(232,123,26,1)","rgba(239,135,32,1)","rgba(169,63,5,1)","rgba(207,81,4,1)","rgba(219,101,14,1)","rgba(139,51,7,1)","rgba(226,112,20,1)","rgba(255,186,89,1)","rgba(221,105,16,1)","rgba(202,75,2,1)","rgba(201,75,2,1)","rgba(202,75,2,1)","rgba(138,51,7,1)","rgba(218,100,14,1)","rgba(208,84,5,1)","rgba(224,109,19,1)","rgba(174,65,5,1)","rgba(209,84,5,1)","rgba(220,103,16,1)","rgba(184,69,4,1)","rgba(166,61,6,1)","rgba(237,132,30,1)","rgba(235,127,28,1)","rgba(230,120,24,1)","rgba(244,142,35,1)","rgba(242,138,33,1)","rgba(158,58,6,1)","rgba(208,82,5,1)","rgba(252,161,53,1)","rgba(221,105,16,1)","rgba(206,79,3,1)","rgba(212,90,8,1)","rgba(246,146,37,1)","rgba(221,105,17,1)","rgba(251,154,41,1)","rgba(227,115,21,1)","rgba(194,72,3,1)","rgba(190,71,4,1)","rgba(235,128,28,1)","rgba(246,145,37,1)","rgba(239,134,31,1)","rgba(217,98,13,1)","rgba(239,135,31,1)","rgba(242,140,34,1)","rgba(247,148,38,1)","rgba(254,172,68,1)","rgba(195,73,3,1)","rgba(216,97,12,1)","rgba(207,82,4,1)","rgba(220,102,15,1)","rgba(224,109,18,1)","rgba(223,109,18,1)","rgba(219,102,15,1)","rgba(220,103,16,1)","rgba(216,97,12,1)","rgba(206,79,3,1)","rgba(255,181,82,1)","rgba(238,133,30,1)","rgba(238,132,30,1)","rgba(206,79,3,1)","rgba(220,103,15,1)","rgba(230,119,24,1)","rgba(254,177,76,1)","rgba(216,96,11,1)","rgba(221,105,16,1)","rgba(255,255,229,1)","rgba(248,150,39,1)","rgba(255,202,111,1)","rgba(180,67,5,1)","rgba(203,76,2,1)","rgba(219,101,14,1)","rgba(246,145,37,1)","rgba(212,89,8,1)","rgba(217,99,13,1)","rgba(246,145,37,1)","rgba(254,229,152,1)","rgba(222,106,17,1)","rgba(255,208,119,1)","rgba(236,128,28,1)","rgba(255,255,229,1)","rgba(233,124,26,1)","rgba(177,66,5,1)","rgba(245,144,36,1)","rgba(227,114,21,1)","rgba(196,73,3,1)","rgba(213,91,9,1)","rgba(255,255,229,1)","rgba(246,145,37,1)","rgba(224,109,18,1)","rgba(222,106,17,1)","rgba(252,162,54,1)","rgba(226,112,20,1)","rgba(184,68,4,1)","rgba(232,122,25,1)","rgba(255,255,229,1)","rgba(253,169,64,1)","rgba(255,231,156,1)","rgba(194,72,3,1)","rgba(250,152,40,1)","rgba(229,117,23,1)","rgba(234,127,27,1)","rgba(252,162,54,1)","rgba(228,116,22,1)","rgba(201,75,2,1)","rgba(228,117,23,1)","rgba(200,75,2,1)","rgba(198,74,3,1)","rgba(237,130,29,1)","rgba(217,99,13,1)","rgba(173,64,5,1)","rgba(254,177,77,1)","rgba(234,126,27,1)","rgba(215,95,11,1)","rgba(253,164,58,1)","rgba(230,119,24,1)","rgba(169,63,5,1)","rgba(227,115,22,1)","rgba(229,117,23,1)","rgba(255,255,229,1)","rgba(251,154,42,1)","rgba(254,178,78,1)","rgba(191,71,3,1)","rgba(171,64,5,1)","rgba(163,61,6,1)","rgba(164,61,6,1)","rgba(187,69,4,1)","rgba(181,67,4,1)","rgba(189,70,4,1)","rgba(195,73,3,1)","rgba(230,119,24,1)","rgba(212,90,8,1)","rgba(198,74,3,1)","rgba(216,96,12,1)","rgba(222,106,17,1)","rgba(172,64,5,1)","rgba(231,121,25,1)","rgba(201,75,2,1)","rgba(200,75,2,1)","rgba(208,83,5,1)","rgba(176,65,5,1)","rgba(233,124,26,1)","rgba(143,53,7,1)","rgba(126,46,7,1)","rgba(102,37,6,1)","rgba(233,124,26,1)","rgba(242,139,34,1)","rgba(223,107,18,1)","rgba(224,109,19,1)","rgba(210,86,6,1)","rgba(252,163,56,1)","rgba(215,95,11,1)","rgba(180,67,4,1)","rgba(182,68,4,1)","rgba(157,58,6,1)","rgba(174,65,5,1)","rgba(205,77,2,1)","rgba(208,83,5,1)","rgba(232,123,26,1)","rgba(206,80,3,1)","rgba(255,186,89,1)","rgba(219,102,15,1)","rgba(240,136,32,1)","rgba(240,136,32,1)","rgba(229,118,23,1)","rgba(208,83,5,1)","rgba(233,124,26,1)","rgba(218,100,14,1)","rgba(255,255,229,1)","rgba(199,74,3,1)","rgba(151,56,6,1)","rgba(161,60,6,1)","rgba(225,111,19,1)","rgba(213,92,9,1)","rgba(200,75,2,1)","rgba(236,129,29,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(254,181,81,1)","rgba(218,100,14,1)","rgba(207,81,4,1)","rgba(224,110,19,1)","rgba(214,93,10,1)","rgba(213,92,9,1)","rgba(217,98,13,1)","rgba(230,120,24,1)","rgba(204,76,2,1)","rgba(133,49,7,1)","rgba(167,62,5,1)","rgba(227,114,21,1)","rgba(171,63,5,1)","rgba(255,255,229,1)","rgba(250,153,41,1)","rgba(224,110,19,1)","rgba(217,98,13,1)","rgba(255,213,126,1)","rgba(210,86,6,1)","rgba(211,88,7,1)","rgba(254,177,76,1)","rgba(253,166,60,1)","rgba(233,124,26,1)","rgba(135,50,7,1)","rgba(160,59,6,1)","rgba(213,92,9,1)","rgba(220,102,15,1)","rgba(255,255,229,1)","rgba(221,104,16,1)","rgba(195,73,3,1)","rgba(246,146,37,1)","rgba(220,103,15,1)","rgba(147,54,7,1)","rgba(170,63,5,1)"]}},"hoveron":"points","name":"stroma","legendgroup":"stroma","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[15,14,15,12,13,15,7,8,10,11,12,15,10,11,13,14,15,5,6,7,9,10,11,12,13,14,15,5,6,9,10,11,12,13,14,7,8,9,10,11,12,13,14,6,7,8,9,10,11,12,13,14,7,9,10,11,12,6,7,8,9,10,12,13,6,7,8,9,10,11,12,7,8,9,10,11,9,10,11],"y":[11,12,12,13,13,13,14,14,14,14,14,14,15,15,15,15,15,16,16,16,16,16,16,16,16,16,16,17,17,17,17,17,17,17,17,18,18,18,18,18,18,18,18,19,19,19,19,19,19,19,19,19,20,20,20,20,20,21,21,21,21,21,21,21,22,22,22,22,22,22,22,23,23,23,23,23,24,24,24],"text":["X: 15<br />Y: 11<br />Tumor_Stroma: tumor<br />Value: 1.7065680","X: 14<br />Y: 12<br />Tumor_Stroma: tumor<br />Value: 2.5022386","X: 15<br />Y: 12<br />Tumor_Stroma: tumor<br />Value: 2.7889901","X: 12<br />Y: 13<br />Tumor_Stroma: tumor<br />Value: 1.3300558","X: 13<br />Y: 13<br />Tumor_Stroma: tumor<br />Value: 1.8827494","X: 15<br />Y: 13<br />Tumor_Stroma: tumor<br />Value: 2.7828264","X:  7<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 0.8134384","X:  8<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 2.2988138","X: 10<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 0.8763564","X: 11<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 1.0883192","X: 12<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 2.1283676","X: 15<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 2.9399569","X: 10<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 0.4263893","X: 11<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 2.3242207","X: 13<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 2.1300274","X: 14<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 2.2079629","X: 15<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 2.2286119","X:  5<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  6<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 2.1028007","X:  7<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 1.9283952","X:  9<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 0.6560307","X: 10<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 1.2031102","X: 11<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 1.5418215","X: 12<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 0.8851240","X: 13<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 2.2249335","X: 14<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 1.2277143","X: 15<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 1.4694216","X:  5<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 2.7159426","X:  6<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 2.3526712","X:  9<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 1.5264164","X: 10<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 1.4102389","X: 11<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 1.6044736","X: 12<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 1.7098560","X: 13<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 1.2652836","X: 14<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 0.7402078","X:  7<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 0.8728521","X:  8<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 1.6189128","X:  9<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 1.0839631","X: 10<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 1.3077329","X: 11<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 1.4045690","X: 12<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 1.7669870","X: 13<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 1.2397910","X: 14<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 2.0325944","X:  6<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 1.3762171","X:  7<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  8<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  9<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 10<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 1.2728988","X: 11<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 1.3278438","X: 12<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 1.7643762","X: 13<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 1.2979270","X: 14<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 1.8329806","X:  7<br />Y: 20<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  9<br />Y: 20<br />Tumor_Stroma: tumor<br />Value: 1.1739939","X: 10<br />Y: 20<br />Tumor_Stroma: tumor<br />Value: 0.5219478","X: 11<br />Y: 20<br />Tumor_Stroma: tumor<br />Value: 1.8969783","X: 12<br />Y: 20<br />Tumor_Stroma: tumor<br />Value: 0.8475265","X:  6<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 1.8238563","X:  7<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  8<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 0.9945759","X:  9<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 1.6452587","X: 10<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 1.2684275","X: 12<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 1.8556840","X: 13<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 1.7583215","X:  6<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 0.9642614","X:  7<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 0.9713458","X:  8<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 1.5379356","X:  9<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 0.7137133","X: 10<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 1.1663298","X: 11<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 12<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 2.1894404","X:  7<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 2.4974793","X:  8<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 1.9197170","X:  9<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 1.5046610","X: 10<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 11<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  9<br />Y: 24<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 10<br />Y: 24<br />Tumor_Stroma: tumor<br />Value: 1.1892661","X: 11<br />Y: 24<br />Tumor_Stroma: tumor<br />Value: 0.0000000"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":["rgba(255,194,100,1)","rgba(245,143,36,1)","rgba(233,125,27,1)","rgba(255,217,131,1)","rgba(255,183,84,1)","rgba(234,125,27,1)","rgba(255,235,171,1)","rgba(251,157,45,1)","rgba(255,234,166,1)","rgba(254,229,151,1)","rgba(253,167,62,1)","rgba(227,115,22,1)","rgba(255,245,198,1)","rgba(251,155,43,1)","rgba(253,167,62,1)","rgba(252,162,55,1)","rgba(252,161,53,1)","rgba(255,255,229,1)","rgba(253,169,65,1)","rgba(254,180,80,1)","rgba(255,239,182,1)","rgba(254,225,142,1)","rgba(255,204,114,1)","rgba(255,234,165,1)","rgba(252,161,53,1)","rgba(254,223,140,1)","rgba(255,208,120,1)","rgba(236,130,29,1)","rgba(250,153,41,1)","rgba(255,205,115,1)","rgba(255,212,125,1)","rgba(255,200,108,1)","rgba(255,194,99,1)","rgba(255,221,137,1)","rgba(255,237,176,1)","rgba(255,234,166,1)","rgba(255,199,107,1)","rgba(254,229,151,1)","rgba(255,218,133,1)","rgba(255,212,125,1)","rgba(255,190,94,1)","rgba(254,223,139,1)","rgba(254,173,71,1)","rgba(255,214,128,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,221,136,1)","rgba(255,217,132,1)","rgba(255,190,95,1)","rgba(255,219,134,1)","rgba(255,186,89,1)","rgba(255,255,229,1)","rgba(254,227,145,1)","rgba(255,242,192,1)","rgba(255,182,83,1)","rgba(255,235,168,1)","rgba(255,186,89,1)","rgba(255,255,229,1)","rgba(255,231,158,1)","rgba(255,198,105,1)","rgba(255,221,137,1)","rgba(255,184,87,1)","rgba(255,191,95,1)","rgba(255,232,160,1)","rgba(255,232,159,1)","rgba(255,204,114,1)","rgba(255,238,178,1)","rgba(254,227,145,1)","rgba(255,255,229,1)","rgba(253,164,56,1)","rgba(245,144,36,1)","rgba(254,180,81,1)","rgba(255,206,117,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(254,226,143,1)","rgba(255,255,229,1)"],"opacity":1,"size":1.88976377952756,"symbol":"square","line":{"width":3.77952755905512,"color":["rgba(255,194,100,1)","rgba(245,143,36,1)","rgba(233,125,27,1)","rgba(255,217,131,1)","rgba(255,183,84,1)","rgba(234,125,27,1)","rgba(255,235,171,1)","rgba(251,157,45,1)","rgba(255,234,166,1)","rgba(254,229,151,1)","rgba(253,167,62,1)","rgba(227,115,22,1)","rgba(255,245,198,1)","rgba(251,155,43,1)","rgba(253,167,62,1)","rgba(252,162,55,1)","rgba(252,161,53,1)","rgba(255,255,229,1)","rgba(253,169,65,1)","rgba(254,180,80,1)","rgba(255,239,182,1)","rgba(254,225,142,1)","rgba(255,204,114,1)","rgba(255,234,165,1)","rgba(252,161,53,1)","rgba(254,223,140,1)","rgba(255,208,120,1)","rgba(236,130,29,1)","rgba(250,153,41,1)","rgba(255,205,115,1)","rgba(255,212,125,1)","rgba(255,200,108,1)","rgba(255,194,99,1)","rgba(255,221,137,1)","rgba(255,237,176,1)","rgba(255,234,166,1)","rgba(255,199,107,1)","rgba(254,229,151,1)","rgba(255,218,133,1)","rgba(255,212,125,1)","rgba(255,190,94,1)","rgba(254,223,139,1)","rgba(254,173,71,1)","rgba(255,214,128,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,221,136,1)","rgba(255,217,132,1)","rgba(255,190,95,1)","rgba(255,219,134,1)","rgba(255,186,89,1)","rgba(255,255,229,1)","rgba(254,227,145,1)","rgba(255,242,192,1)","rgba(255,182,83,1)","rgba(255,235,168,1)","rgba(255,186,89,1)","rgba(255,255,229,1)","rgba(255,231,158,1)","rgba(255,198,105,1)","rgba(255,221,137,1)","rgba(255,184,87,1)","rgba(255,191,95,1)","rgba(255,232,160,1)","rgba(255,232,159,1)","rgba(255,204,114,1)","rgba(255,238,178,1)","rgba(254,227,145,1)","rgba(255,255,229,1)","rgba(253,164,56,1)","rgba(245,144,36,1)","rgba(254,180,81,1)","rgba(255,206,117,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(254,226,143,1)","rgba(255,255,229,1)"]}},"hoveron":"points","name":"tumor","legendgroup":"tumor","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[10],"y":[10],"name":"99_6bb8586e0ff8917f79385464a6e2aae6","type":"scatter","mode":"markers","opacity":0,"hoverinfo":"skip","showlegend":false,"marker":{"color":[0,1],"colorscale":[[0,"#FFFFE5"],[0.00334448160535117,"#FFFFE4"],[0.00668896321070234,"#FFFEE3"],[0.0100334448160535,"#FFFEE2"],[0.0133779264214047,"#FFFDE0"],[0.0167224080267559,"#FFFDDF"],[0.020066889632107,"#FFFDDE"],[0.0234113712374582,"#FFFCDD"],[0.0267558528428094,"#FFFCDC"],[0.0301003344481605,"#FFFCDB"],[0.0334448160535117,"#FFFBDA"],[0.0367892976588629,"#FFFBD9"],[0.040133779264214,"#FFFAD7"],[0.0434782608695652,"#FFFAD6"],[0.0468227424749164,"#FFFAD5"],[0.0501672240802676,"#FFF9D4"],[0.0535117056856187,"#FFF9D3"],[0.0568561872909699,"#FFF9D2"],[0.0602006688963211,"#FFF8D1"],[0.0635451505016722,"#FFF8D0"],[0.0668896321070234,"#FFF7CE"],[0.0702341137123746,"#FFF7CD"],[0.0735785953177258,"#FFF7CC"],[0.0769230769230769,"#FFF6CB"],[0.0802675585284281,"#FFF6CA"],[0.0836120401337793,"#FFF5C9"],[0.0869565217391304,"#FFF5C8"],[0.0903010033444816,"#FFF5C7"],[0.0936454849498328,"#FFF4C6"],[0.096989966555184,"#FFF4C4"],[0.100334448160535,"#FFF4C3"],[0.103678929765886,"#FFF3C2"],[0.107023411371237,"#FFF3C1"],[0.110367892976589,"#FFF2C0"],[0.11371237458194,"#FFF2BF"],[0.117056856187291,"#FFF2BE"],[0.120401337792642,"#FFF1BD"],[0.123745819397993,"#FFF1BB"],[0.127090301003344,"#FFF1BA"],[0.130434782608696,"#FFF0B9"],[0.133779264214047,"#FFF0B8"],[0.137123745819398,"#FFEFB7"],[0.140468227424749,"#FFEFB6"],[0.1438127090301,"#FFEFB5"],[0.147157190635452,"#FFEEB4"],[0.150501672240803,"#FFEEB2"],[0.153846153846154,"#FFEEB1"],[0.157190635451505,"#FFEDB0"],[0.160535117056856,"#FFEDAF"],[0.163879598662207,"#FFEDAE"],[0.167224080267559,"#FFECAD"],[0.17056856187291,"#FFECAC"],[0.173913043478261,"#FFEBAB"],[0.177257525083612,"#FFEBAA"],[0.180602006688963,"#FFEBA8"],[0.183946488294314,"#FFEAA7"],[0.187290969899666,"#FFEAA6"],[0.190635451505017,"#FFEAA5"],[0.193979933110368,"#FFE9A4"],[0.197324414715719,"#FFE9A3"],[0.20066889632107,"#FFE8A2"],[0.204013377926421,"#FFE8A1"],[0.207357859531773,"#FFE89F"],[0.210702341137124,"#FFE79E"],[0.214046822742475,"#FFE79D"],[0.217391304347826,"#FFE79C"],[0.220735785953177,"#FFE69B"],[0.224080267558528,"#FFE69A"],[0.22742474916388,"#FFE599"],[0.230769230769231,"#FEE598"],[0.234113712374582,"#FEE596"],[0.237458193979933,"#FEE495"],[0.240802675585284,"#FEE494"],[0.244147157190635,"#FEE493"],[0.247491638795987,"#FEE392"],[0.250836120401338,"#FEE391"],[0.254180602006689,"#FEE28F"],[0.25752508361204,"#FEE18E"],[0.260869565217391,"#FEE08D"],[0.264214046822742,"#FEDF8B"],[0.267558528428094,"#FFDE8A"],[0.270903010033445,"#FFDD89"],[0.274247491638796,"#FFDC87"],[0.277591973244147,"#FFDB86"],[0.280936454849498,"#FFDA85"],[0.28428093645485,"#FFD983"],[0.287625418060201,"#FFD882"],[0.290969899665552,"#FFD781"],[0.294314381270903,"#FFD67F"],[0.297658862876254,"#FFD57E"],[0.301003344481605,"#FFD47D"],[0.304347826086957,"#FFD37C"],[0.307692307692308,"#FFD27A"],[0.311036789297659,"#FFD179"],[0.31438127090301,"#FFD078"],[0.317725752508361,"#FFCF76"],[0.321070234113712,"#FFCE75"],[0.324414715719064,"#FFCD74"],[0.327759197324415,"#FFCC72"],[0.331103678929766,"#FFCB71"],[0.334448160535117,"#FFCB70"],[0.337792642140468,"#FFCA6E"],[0.341137123745819,"#FFC96D"],[0.344481605351171,"#FFC86C"],[0.347826086956522,"#FFC76A"],[0.351170568561873,"#FFC669"],[0.354515050167224,"#FFC568"],[0.357859531772575,"#FFC466"],[0.361204013377926,"#FFC365"],[0.364548494983278,"#FFC264"],[0.367892976588629,"#FFC162"],[0.37123745819398,"#FFC061"],[0.374581939799331,"#FFBF60"],[0.377926421404682,"#FFBE5E"],[0.381270903010033,"#FFBD5D"],[0.384615384615385,"#FFBC5C"],[0.387959866220736,"#FFBB5A"],[0.391304347826087,"#FFBA59"],[0.394648829431438,"#FFB957"],[0.397993311036789,"#FFB856"],[0.40133779264214,"#FFB755"],[0.404682274247492,"#FFB653"],[0.408026755852843,"#FFB552"],[0.411371237458194,"#FEB451"],[0.414715719063545,"#FEB34F"],[0.418060200668896,"#FEB24E"],[0.421404682274247,"#FEB14D"],[0.424749163879599,"#FEB04B"],[0.42809364548495,"#FEAF4A"],[0.431438127090301,"#FEAE48"],[0.434782608695652,"#FEAD47"],[0.438127090301003,"#FEAC45"],[0.441471571906355,"#FEAB44"],[0.444816053511706,"#FDAA43"],[0.448160535117057,"#FDA941"],[0.451505016722408,"#FDA840"],[0.454849498327759,"#FDA73E"],[0.45819397993311,"#FDA63D"],[0.461538461538462,"#FDA53B"],[0.464882943143813,"#FDA43A"],[0.468227424749164,"#FDA338"],[0.471571906354515,"#FCA237"],[0.474916387959866,"#FCA135"],[0.478260869565217,"#FCA034"],[0.481605351170569,"#FC9F32"],[0.48494983277592,"#FC9E31"],[0.488294314381271,"#FC9D2F"],[0.491638795986622,"#FB9C2D"],[0.494983277591973,"#FB9B2C"],[0.498327759197324,"#FB9A2A"],[0.501672240802676,"#FB9929"],[0.505016722408027,"#FA9828"],[0.508361204013378,"#F99728"],[0.511705685618729,"#F99627"],[0.51505016722408,"#F89527"],[0.518394648829432,"#F89426"],[0.521739130434783,"#F79326"],[0.525083612040134,"#F69225"],[0.528428093645485,"#F69125"],[0.531772575250836,"#F59024"],[0.535117056856187,"#F58F24"],[0.538461538461539,"#F48E23"],[0.54180602006689,"#F38D23"],[0.545150501672241,"#F38C22"],[0.548494983277592,"#F28B22"],[0.551839464882943,"#F28A21"],[0.555183946488294,"#F18921"],[0.558528428093646,"#F08820"],[0.561872909698997,"#F08720"],[0.565217391304348,"#EF861F"],[0.568561872909699,"#EE851F"],[0.57190635451505,"#EE841E"],[0.575250836120401,"#ED831E"],[0.578595317725752,"#ED821D"],[0.581939799331104,"#EC811D"],[0.585284280936455,"#EB801C"],[0.588628762541806,"#EB7F1C"],[0.591973244147157,"#EA7E1B"],[0.595317725752508,"#E97D1B"],[0.59866220735786,"#E97C1A"],[0.602006688963211,"#E87B1A"],[0.605351170568562,"#E87A19"],[0.608695652173913,"#E77919"],[0.612040133779264,"#E67818"],[0.615384615384615,"#E67718"],[0.618729096989967,"#E57617"],[0.622073578595318,"#E47517"],[0.625418060200669,"#E47416"],[0.62876254180602,"#E37315"],[0.632107023411371,"#E37215"],[0.635451505016722,"#E27114"],[0.638795986622074,"#E17014"],[0.642140468227425,"#E16F13"],[0.645484949832776,"#E06E13"],[0.648829431438127,"#DF6D12"],[0.652173913043478,"#DF6C12"],[0.65551839464883,"#DE6A11"],[0.658862876254181,"#DE6911"],[0.662207357859532,"#DD6810"],[0.665551839464883,"#DC6710"],[0.668896321070234,"#DC660F"],[0.672240802675585,"#DB650E"],[0.675585284280937,"#DA640E"],[0.678929765886288,"#DA630D"],[0.682274247491639,"#D9620D"],[0.68561872909699,"#D8610C"],[0.688963210702341,"#D8600C"],[0.692307692307692,"#D75F0B"],[0.695652173913044,"#D65E0A"],[0.698996655518395,"#D65D0A"],[0.702341137123746,"#D55C09"],[0.705685618729097,"#D55B09"],[0.709030100334448,"#D45A08"],[0.712374581939799,"#D35807"],[0.71571906354515,"#D35707"],[0.719063545150502,"#D25606"],[0.722408026755853,"#D15506"],[0.725752508361204,"#D15405"],[0.729096989966555,"#D05305"],[0.732441471571906,"#CF5204"],[0.735785953177257,"#CF5104"],[0.739130434782609,"#CE5003"],[0.74247491638796,"#CD4F03"],[0.745819397993311,"#CD4D03"],[0.749163879598662,"#CC4C02"],[0.752508361204013,"#CB4C02"],[0.755852842809365,"#C94B02"],[0.759197324414716,"#C84A02"],[0.762541806020067,"#C74A03"],[0.765886287625418,"#C54903"],[0.769230769230769,"#C44903"],[0.77257525083612,"#C24803"],[0.775919732441472,"#C14803"],[0.779264214046823,"#BF4703"],[0.782608695652174,"#BE4704"],[0.785953177257525,"#BD4604"],[0.789297658862876,"#BB4604"],[0.792642140468227,"#BA4504"],[0.795986622073579,"#B84504"],[0.79933110367893,"#B74404"],[0.802675585284281,"#B54304"],[0.806020066889632,"#B44305"],[0.809364548494983,"#B34205"],[0.812709030100335,"#B14205"],[0.816053511705686,"#B04105"],[0.819397993311037,"#AE4105"],[0.822742474916388,"#AD4005"],[0.826086956521739,"#AC4005"],[0.82943143812709,"#AA3F05"],[0.832775919732442,"#A93F05"],[0.836120401337793,"#A73E05"],[0.839464882943144,"#A63E06"],[0.842809364548495,"#A53D06"],[0.846153846153846,"#A33D06"],[0.849498327759197,"#A23C06"],[0.852842809364548,"#A13B06"],[0.8561872909699,"#9F3B06"],[0.859531772575251,"#9E3A06"],[0.862876254180602,"#9C3A06"],[0.866220735785953,"#9B3906"],[0.869565217391304,"#9A3906"],[0.872909698996656,"#983806"],[0.876254180602007,"#973806"],[0.879598662207358,"#963706"],[0.882943143812709,"#943707"],[0.88628762541806,"#933607"],[0.889632107023411,"#913607"],[0.892976588628763,"#903507"],[0.896321070234114,"#8F3507"],[0.899665551839465,"#8D3407"],[0.903010033444816,"#8C3407"],[0.906354515050167,"#8B3307"],[0.909698996655518,"#893307"],[0.91304347826087,"#883207"],[0.916387959866221,"#873207"],[0.919732441471572,"#853107"],[0.923076923076923,"#843107"],[0.926421404682274,"#833007"],[0.929765886287626,"#813007"],[0.933110367892977,"#802F07"],[0.936454849498328,"#7F2F07"],[0.939799331103679,"#7D2E07"],[0.94314381270903,"#7C2E07"],[0.946488294314381,"#7B2D07"],[0.949832775919732,"#792C07"],[0.953177257525084,"#782C07"],[0.956521739130435,"#772B07"],[0.959866220735786,"#752B07"],[0.963210702341137,"#742A07"],[0.966555183946488,"#732A07"],[0.96989966555184,"#722907"],[0.973244147157191,"#702907"],[0.976588628762542,"#6F2807"],[0.979933110367893,"#6E2807"],[0.983277591973244,"#6C2707"],[0.986622073578595,"#6B2706"],[0.989966555183947,"#6A2606"],[0.993311036789298,"#692606"],[0.996655518394649,"#672506"],[1,"#662506"]],"colorbar":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"thickness":23.04,"title":"log_expr","titlefont":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"tickmode":"array","ticktext":["0","1","2","3","4"],"tickvals":[0,0.213739161107169,0.427478322214338,0.641217483321508,0.854956644428677],"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"ticklen":2,"len":0.5,"yanchor":"top","y":1}},"xaxis":"x","yaxis":"y","frame":null}],"layout":{"margin":{"t":43.9712981082844,"r":7.30593607305936,"b":40.3913894324853,"l":37.2602739726027},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"title":{"text":"CD74<br />sample: ST_mel1_rep2","font":{"color":"rgba(0,0,0,1)","family":"","size":17.5342465753425},"x":0,"xref":"paper"},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[1.75,29.25],"tickmode":"array","ticktext":["10","20"],"tickvals":[10,20],"categoryorder":"array","categoryarray":["10","20"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"X Position","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"scaleanchor":"y","scaleratio":1,"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[6,28],"tickmode":"array","ticktext":["10","15","20","25"],"tickvals":[10,15,20,25],"categoryorder":"array","categoryarray":["10","15","20","25"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"Y Position","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"scaleanchor":"x","scaleratio":1,"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"y":0.5,"yanchor":"top"},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"12a3b3323e88e":{"x":{},"y":{},"fill":{},"shape":{},"colour":{},"type":"scatter"}},"cur_data":"12a3b3323e88e","visdat":{"12a3b3323e88e":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script><p>Spatial interpolation of cell scores is also possible with the <code><a href="../reference/deconv_krige.html">deconv_krige()</a></code> function:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconv_krige.html">deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, cells<span class="op">=</span><span class="st">'b_cells'</span>, who<span class="op">=</span><span class="fl">2</span>, python<span class="op">=</span><span class="cn">F</span>, res<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span></code></pre></div>
<p>We plot the surfaces using the <code><a href="../reference/plot_deconv_krige.html">plot_deconv_krige()</a></code> function.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_krige.html">plot_deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>,  cells<span class="op">=</span><span class="st">'b_cells'</span>, plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges3</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">1</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotdeconvkrige_chunk-1.png" width="960"></p>
<p>Cell scores in xCell are accompanied by p-values (test for cell scores different to zero). The p-values are calculated automatically during deconvolution in <code>spatialGE</code>, and we can plot them using the <code>pvalue</code> option:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_krige.html">plot_deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>,  cells<span class="op">=</span><span class="st">'b_cells'</span>, plot_who<span class="op">=</span><span class="fl">2</span>, pvalues<span class="op">=</span><span class="cn">T</span>, 
                             visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges4</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">1</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotkrigepvals_chunk-1.png" width="960"></p>
<p>As observed in the surfaces, B cells are present throughout the tissue, but less abundant in the melanoma pocket. Most of the spots had B cell scores significantly different to zero, as suggested by xCell’s permutation method (gray hollow circles). Most of the non-sognificant scores were observed in the melanoma area, reinforcing the observation of lower infiltration in that area.</p>
<p>The <code><a href="../reference/plot_deconv_krige.html">plot_deconv_krige()</a></code> also allows annotation of tumor spots (hollow squares):</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_krige.html">plot_deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>,  cells<span class="op">=</span><span class="st">'b_cells'</span>, plot_who<span class="op">=</span><span class="fl">2</span>, purity<span class="op">=</span><span class="cn">T</span>, 
                             visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges5</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotkrigepurity_chunk-1.png" width="960"></p>
</div>
<div id="unsupervised-spatially-informed-clustering-stclust" class="section level2">
<h2 class="hasAnchor">
<a href="#unsupervised-spatially-informed-clustering-stclust" class="anchor"></a>Unsupervised spatially-informed clustering (<em>STclust</em>)</h2>
<p>We can also perform <em>STclust</em>, the spatially informed clustering method in <code>spatialGE</code> to detect compartments in the tissue. In STclust, we use weighted average matrices to capture the transcriptomic differences among the spots. A transcriptomic (Manhattan) distance matrix is calculated and scaled from PCA. Following, scaled euclidean distances are computed for the spatial distances among spots. The user defines a weight, from 0 to 1, to apply to the spatial distances. The higher the weight, the less biologically meaningful the clustering is given that the groups will only reflect the physical distances between the spots and less information on the transcriptomic profiles will be used. After many tests, we have found that weights between 0.025 - 0.25 is enough to capture the tissue heterogeneity. This approach to spatial clustering can be performed using the <code><a href="../reference/STclust.html">STclust()</a></code> function:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STclust.html">STclust</a></span><span class="op">(</span><span class="va">melanoma</span>, ks<span class="op">=</span><span class="st">'dtc'</span>, weights<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.025</span>, <span class="fl">0.05</span>, <span class="fl">0.075</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Using vperc = 0.8</span></code></pre></div>
<p>Results of clustering can be plotted via the <code><a href="../reference/plot_STclusters.html">plot_STclusters()</a></code> function:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cluster_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_STclusters.html">plot_STclusters</a></span><span class="op">(</span><span class="va">melanoma</span>, plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">cluster_p</span>, nrow<span class="op">=</span><span class="fl">2</span>, ncol<span class="op">=</span><span class="fl">2</span>, legend<span class="op">=</span><span class="st">'right'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotclustspots_chunk-1.png" width="960"></p>
<p>We can see that at <code>w=0.025</code>, tissue compartments are observed for the lymphoid tissue (upper right portion of the tissue in red), which also sureounds the melnoma area (yellow), possibly indicating immune activity around the tumor spots. Blue likely corresponds to stromal spots. We have used here the dynamicTreeCut (<code>dtc</code> option) to select the number of clusters, but users can define their own range of k to be evaluated, allowing to further isolate cell types.</p>
</div>
<div id="association-between-spatial-heterogeneity-and-clinical-variables-sthet" class="section level2">
<h2 class="hasAnchor">
<a href="#association-between-spatial-heterogeneity-and-clinical-variables-sthet" class="anchor"></a>Association between spatial heterogeneity and clinical variables (<em>SThet</em>)</h2>
<p>To explore the relationship between a clinical variable of interest and the level of gene expression uniformity within a tissue section, we can use the function <code>plot_phenovar()</code>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatial_stat_plot.html">spatial_stat_plot</a></span><span class="op">(</span><span class="va">melanoma</span>, samplevar<span class="op">=</span><span class="st">'survival_months'</span>, gene<span class="op">=</span><span class="st">'CD74'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/clinplot_chunk-1.png" width="960"></p>
<p>In the case of the gene CD74, an immune gene, samples with a less uniform expression (i.e., clustered gene expression) also were from a patient with a higher survival time (high Moran’s I; low Geary’s C). The survival of the patients did not show an association with the tendency of CD74 to generate expression “hot spots” (Getis-Ord Gi was intermediate in patient 1). See the table below for a simplistic interpretation of the spatial autocorrelation statistics calculated in <code>spatialGE</code>:</p>
<table class="table table">
<caption>
Heterogeneity statitstics in <code>spatialGE</code>
</caption>
<thead><tr>
<th style="text-align:center;">
Low
</th>
<th style="text-align:center;">
Statistic
</th>
<th style="text-align:center;">
High
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
Dispersion
</td>
<td style="text-align:center;">
Moran’s I
</td>
<td style="text-align:center;">
Clustering
</td>
</tr>
<tr>
<td style="text-align:center;">
Clustering
</td>
<td style="text-align:center;">
Geary’s C
</td>
<td style="text-align:center;">
Dispersion
</td>
</tr>
<tr>
<td style="text-align:center;">
“Cold spots”
</td>
<td style="text-align:center;">
Getis-Ord Gi
</td>
<td style="text-align:center;">
“Hot spots”
</td>
</tr>
</tbody>
</table>
<p>Estimating these statistics can be computationally expensive. Hence, plots can be shown again by accessing the appropriate STList slot (<code>melanoma@pheno_plots[['CD74']]</code>).</p>
<details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.0.3 (2020-10-10)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.7</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] xCell_1.1.0          mclust_5.4.7         estimate_1.0.13     </span>
<span class="co">#&gt; [4] ggplot2_3.3.5        Matrix_1.2-18        magrittr_2.0.1      </span>
<span class="co">#&gt; [7] spatialGE_1.0.0.0000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] readxl_1.3.1                backports_1.2.1            </span>
<span class="co">#&gt;   [3] systemfonts_1.0.2           lazyeval_0.2.2             </span>
<span class="co">#&gt;   [5] sp_1.4-5                    GSEABase_1.52.1            </span>
<span class="co">#&gt;   [7] BiocParallel_1.24.1         crosstalk_1.1.1            </span>
<span class="co">#&gt;   [9] GenomeInfoDb_1.26.4         digest_0.6.27              </span>
<span class="co">#&gt;  [11] htmltools_0.5.1.1           fansi_0.4.2                </span>
<span class="co">#&gt;  [13] memoise_2.0.0               RandomFieldsUtils_0.5.3    </span>
<span class="co">#&gt;  [15] cluster_2.1.0               openxlsx_4.2.3             </span>
<span class="co">#&gt;  [17] readr_1.4.0                 annotate_1.68.0            </span>
<span class="co">#&gt;  [19] matrixStats_0.58.0          svglite_2.0.0              </span>
<span class="co">#&gt;  [21] pkgdown_1.6.1               colorspace_2.0-0           </span>
<span class="co">#&gt;  [23] rvest_1.0.0                 blob_1.2.1                 </span>
<span class="co">#&gt;  [25] ggrepel_0.9.1               textshaping_0.3.4          </span>
<span class="co">#&gt;  [27] haven_2.3.1                 xfun_0.22                  </span>
<span class="co">#&gt;  [29] dplyr_1.0.5                 sparsesvd_0.2              </span>
<span class="co">#&gt;  [31] tcltk_4.0.3                 crayon_1.4.1               </span>
<span class="co">#&gt;  [33] RCurl_1.98-1.3              jsonlite_1.7.2             </span>
<span class="co">#&gt;  [35] graph_1.68.0                glue_1.4.2                 </span>
<span class="co">#&gt;  [37] kableExtra_1.3.4.9000       gtable_0.3.0               </span>
<span class="co">#&gt;  [39] zlibbioc_1.36.0             XVector_0.30.0             </span>
<span class="co">#&gt;  [41] webshot_0.5.2               DelayedArray_0.16.3        </span>
<span class="co">#&gt;  [43] V8_3.4.0                    car_3.0-10                 </span>
<span class="co">#&gt;  [45] khroma_1.6.0                BiocGenerics_0.36.0        </span>
<span class="co">#&gt;  [47] abind_1.4-5                 scales_1.1.1               </span>
<span class="co">#&gt;  [49] DBI_1.1.1                   rstatix_0.7.0              </span>
<span class="co">#&gt;  [51] Rcpp_1.0.6                  RandomFields_3.3.8         </span>
<span class="co">#&gt;  [53] viridisLite_0.4.0           xtable_1.8-4               </span>
<span class="co">#&gt;  [55] foreign_0.8-80              bit_4.0.4                  </span>
<span class="co">#&gt;  [57] stats4_4.0.3                GSVA_1.38.2                </span>
<span class="co">#&gt;  [59] htmlwidgets_1.5.3           httr_1.4.2                 </span>
<span class="co">#&gt;  [61] ellipsis_0.3.1              pkgconfig_2.0.3            </span>
<span class="co">#&gt;  [63] XML_3.99-0.6                farver_2.1.0               </span>
<span class="co">#&gt;  [65] sass_0.3.1                  dynamicTreeCut_1.63-1      </span>
<span class="co">#&gt;  [67] utf8_1.2.1                  janitor_2.1.0              </span>
<span class="co">#&gt;  [69] tidyselect_1.1.0            labeling_0.4.2             </span>
<span class="co">#&gt;  [71] rlang_0.4.10                AnnotationDbi_1.52.0       </span>
<span class="co">#&gt;  [73] munsell_0.5.0               cellranger_1.1.0           </span>
<span class="co">#&gt;  [75] tools_4.0.3                 cachem_1.0.4               </span>
<span class="co">#&gt;  [77] generics_0.1.0              RSQLite_2.2.5              </span>
<span class="co">#&gt;  [79] splancs_2.01-40             broom_0.7.5                </span>
<span class="co">#&gt;  [81] geoR_1.8-1                  evaluate_0.14              </span>
<span class="co">#&gt;  [83] stringr_1.4.0               fastmap_1.1.0              </span>
<span class="co">#&gt;  [85] yaml_2.2.1                  ragg_1.1.2                 </span>
<span class="co">#&gt;  [87] knitr_1.33                  bit64_4.0.5                </span>
<span class="co">#&gt;  [89] fs_1.5.0                    zip_2.1.1                  </span>
<span class="co">#&gt;  [91] purrr_0.3.4                 wordspace_0.2-6            </span>
<span class="co">#&gt;  [93] xml2_1.3.2                  pracma_2.3.3               </span>
<span class="co">#&gt;  [95] rstudioapi_0.13             concaveman_1.1.0           </span>
<span class="co">#&gt;  [97] compiler_4.0.3              plotly_4.9.3               </span>
<span class="co">#&gt;  [99] curl_4.3                    ggsignif_0.6.1             </span>
<span class="co">#&gt; [101] tibble_3.1.0                bslib_0.2.4                </span>
<span class="co">#&gt; [103] stringi_1.5.3               highr_0.8                  </span>
<span class="co">#&gt; [105] desc_1.3.0                  forcats_0.5.1              </span>
<span class="co">#&gt; [107] rgeos_0.5-5                 lattice_0.20-41            </span>
<span class="co">#&gt; [109] vctrs_0.3.6                 pillar_1.5.1               </span>
<span class="co">#&gt; [111] lifecycle_1.0.0             jquerylib_0.1.3            </span>
<span class="co">#&gt; [113] data.table_1.14.0           cowplot_1.1.1              </span>
<span class="co">#&gt; [115] bitops_1.0-6                raster_3.4-5               </span>
<span class="co">#&gt; [117] GenomicRanges_1.42.0        R6_2.5.0                   </span>
<span class="co">#&gt; [119] gridExtra_2.3               rio_0.5.26                 </span>
<span class="co">#&gt; [121] IRanges_2.24.1              codetools_0.2-16           </span>
<span class="co">#&gt; [123] ggpolypath_0.1.0            MASS_7.3-53                </span>
<span class="co">#&gt; [125] assertthat_0.2.1            SummarizedExperiment_1.20.0</span>
<span class="co">#&gt; [127] rprojroot_2.0.2             withr_2.4.1                </span>
<span class="co">#&gt; [129] S4Vectors_0.28.1            GenomeInfoDbData_1.2.4     </span>
<span class="co">#&gt; [131] iotools_0.3-1               parallel_4.0.3             </span>
<span class="co">#&gt; [133] hms_1.0.0                   quadprog_1.5-8             </span>
<span class="co">#&gt; [135] grid_4.0.3                  tidyr_1.1.3                </span>
<span class="co">#&gt; [137] rmarkdown_2.10              snakecase_0.11.0           </span>
<span class="co">#&gt; [139] MatrixGenerics_1.2.1        carData_3.0-4              </span>
<span class="co">#&gt; [141] ggpubr_0.4.0                Biobase_2.50.0             </span>
<span class="co">#&gt; [143] lubridate_1.7.10</span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Oscar Ospina, Brooke Fridley, Alex Soupir.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
