<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>spatialGE: Guide to spatial heterogeneity analysis in melanoma (Thrane et al. 2018) • spatialGE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="spatialGE: Guide to spatial heterogeneity analysis in melanoma (Thrane et al. 2018)">
<meta property="og:description" content="spatialGE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0.0000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/melanoma_ST_vignette.html">spatialGE: Guide to spatial heterogeneity analysis in melanoma (Thrane et al. 2018)</a>
    </li>
    <li>
      <a href="../articles/prostatecancer_Visium_vignette.html">spatialGE: Kriging for Visium data sets</a>
    </li>
    <li>
      <a href="../articles/read_GeoMx_output.html">spatialGE: Reading GeoMx-DSP outputs</a>
    </li>
    <li>
      <a href="../articles/spatialGE_troubleshoot.html">Troubleshooting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="melanoma_ST_vignette_files/htmlwidgets-1.5.4/htmlwidgets.js"></script><script src="melanoma_ST_vignette_files/plotly-binding-4.10.0/plotly.js"></script><script src="melanoma_ST_vignette_files/typedarray-0.1/typedarray.min.js"></script><link href="melanoma_ST_vignette_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="melanoma_ST_vignette_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><link href="melanoma_ST_vignette_files/plotly-htmlwidgets-css-2.5.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="melanoma_ST_vignette_files/plotly-main-2.5.1/plotly-latest.min.js"></script><script src="melanoma_ST_vignette_files/kePrint-0.0.1/kePrint.js"></script><link href="melanoma_ST_vignette_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>spatialGE: Guide to spatial heterogeneity
analysis in melanoma (Thrane et al. 2018)</h1>
            
      
      
      <div class="hidden name"><code>melanoma_ST_vignette.Rmd</code></div>

    </div>

    
    
<p>The package <code>spatialGE</code> provides a set of tools for the
visualization of gene expression from spatially-resolved transcriptomic
experiments, such as those generated by the 10X Visium platform.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The <code>spatialGE</code> repository is available at GitHub and can
be installed via <code>devtools</code>. To install
<code>devtools</code>, in case is not already installed in your R
console, please run the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="st">"devtools"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>After making sure <code>devtools</code> is installed, proceed to
install <code>spatialGE</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"fridleylab/spatialGE"</span><span class="op">)</span></code></pre></div>
<p>To use <code>spatialGE</code>, load the package using:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">spatialGE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="spatially-resolved-expression-of-melanoma-stage-iii-tumor-biopsies">Spatially-resolved expression of melanoma stage III tumor
biopsies<a class="anchor" aria-label="anchor" href="#spatially-resolved-expression-of-melanoma-stage-iii-tumor-biopsies"></a>
</h2>
<p>To show the utility of <code>spatialGE</code>, we use the Spatial
Transcriptomics (ST) data set generated by <a href="https:doi.org/10.1158/0008-5472.CAN-18-0747">Thrane et
al. (2018)</a>. This data set includes lymph node biopsies from four
patients with stage IIIc melanoma, two spatial arrays (samples) per
biopsy. The technology used in this study allows for RNA collection in
up to 1,007 capture areas (i.e., “spots”) with 200μM of distance between
their centers. Data for each sample, however, has a variable number of
spots due to differences in area of the tissue slices. The spots had a
diameter of 100μM, covering 5-40 cells according to the authors.</p>
<p>The data set is included in the spatialGE repository, but users can
find the data in the <a href="https://www.spatialresearch.org/resources-published-datasets/doi-10-1158-0008-5472-can-18-0747/" class="external-link">authors’
website</a>.</p>
</div>
<div class="section level2">
<h2 id="creating-an-stlist-spatial-transcriptomics-list">Creating an STList (Spatial Transcriptomics List)<a class="anchor" aria-label="anchor" href="#creating-an-stlist-spatial-transcriptomics-list"></a>
</h2>
<p>Raw and processed data are stored in an STList (S4 class object). The
STList can be created with the function <code><a href="../reference/STList.html">STList()</a></code>, which can
take different inputs (see <a href="https://fridleylab.github.io/spatialGE/reference/STList.html" class="external-link">here</a>
for more info). For this example, we provide to the STList function the
following inputs:</p>
<ol style="list-style-type: decimal">
<li><p>Tab-delimited raw RNA counts, one per sample. The raw count files
have gene names in the first column, and spots in subsequent
columns.</p></li>
<li><p>Tab-delimited y and x coordinates for each spot. The first column
of this file contains spot IDs matching column names in the raw count
files. The second and third columns contain the Y and X coordinates of
each spot respectively.</p></li>
<li><p>A comma-delimited table with clinical data associated with each
sample. This table is optional and can by a list of user-provided names
for each sample (count file). If a table, sample names are in the first
column, and clinical variables in subsequent columns. The sample names
should match, at least partially, the file names of the RNA count and
coordinate files (for example, the sample ID “ST_mel4_rep2” matches the
RNA count file “ST_mel4_rep2_counts.tsv” and the coordinate file
“ST_mel4_rep2_mapping.tsv”.</p></li>
</ol>
<p>The example files in this tutorial can be downloaded from the <a href="https://github.com/FridleyLab/spatialGEdev/tree/main/inst/extdata" class="external-link">GitHub
repository</a>. They are also stored in your local
<code>spatialGE</code> installation:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"spatialGE"</span><span class="op">)</span></code></pre></div>
<p>Then, we look for the find the necessary files within the data
repository in <code>spatialGE</code>. These files contain the raw gene
counts (<code>count_files</code>), coordinates
(<code>coord_files</code>), and <code>clinical</code> metadata.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">count_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">data_files</span>, full.names<span class="op">=</span><span class="cn">T</span>, pattern<span class="op">=</span><span class="st">'counts'</span><span class="op">)</span>
<span class="va">coord_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">data_files</span>, full.names<span class="op">=</span><span class="cn">T</span>, pattern<span class="op">=</span><span class="st">'mapping'</span><span class="op">)</span>
<span class="va">clin_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">data_files</span>, full.names<span class="op">=</span><span class="cn">T</span>, pattern<span class="op">=</span><span class="st">'clinical'</span><span class="op">)</span></code></pre></div>
<p>We can load the files into an STList object like so:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STList.html">STList</a></span><span class="op">(</span>rnacounts<span class="op">=</span><span class="va">count_files</span>, spotcoords<span class="op">=</span><span class="va">coord_files</span>, samples<span class="op">=</span><span class="va">clin_file</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #BBBB00;">Found Matrix Data.</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; Requested 8 Samples</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; Cleaning Count and Coordinate Data Gene Names.</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; Converting Counts to Sparse Matrices</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; </span><span style="color: #00BB00; font-weight: bold;">Completed STList!</span></span>
<span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; </span></span></code></pre></div>
<ul>
<li>Note: If you receive a warning indicating that an input file has a
incomplete final line, you can ignore it. Alternatively, modify the
input files with a text editor by adding an empty line at the end of the
files.</li>
</ul>
<p>The <code>melanoma</code> object is an STList, and contains the count
data, spot coordinates, and clinical data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span>
<span class="co">#&gt; Spatial Transcriptomics List (STList).</span>
<span class="co">#&gt; 8 spatial array(s):</span>
<span class="co">#&gt;  ST_mel1_rep1</span>
<span class="co">#&gt;      ST_mel1_rep2</span>
<span class="co">#&gt;      ST_mel2_rep1</span>
<span class="co">#&gt;      ST_mel2_rep2</span>
<span class="co">#&gt;      ST_mel3_rep1</span>
<span class="co">#&gt;      ST_mel3_rep2</span>
<span class="co">#&gt;      ST_mel4_rep1</span>
<span class="co">#&gt;      ST_mel4_rep2</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 9 variables in sample data:</span>
<span class="co">#&gt;   patient, slice, gender, BRAF_status, NRAS_statuds, CDKN2A_status, survival, survival_months, RIN</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="exploring-variation-between-spatial-arrays">Exploring variation between spatial arrays<a class="anchor" aria-label="anchor" href="#exploring-variation-between-spatial-arrays"></a>
</h2>
<p>The function <code><a href="../reference/bulk_pca.html">bulk_pca()</a></code> allows for a quick snapshot of
the variation in gene expression among samples. The function creates
‘bulk’ RNAseq data sets by combining all counts from each sample. Then,
it log transforms the ‘bulk’ RNAseq counts and performs a PCA. Note that
spatial (x, y) coordinate information is not considered in this
analysis, which is intended only as a tool for exploratory analysis. In
this case, we apply the function to look for agreement between samples
from the same patient: It is expected that tissue slices from the same
patient are more similar between them than tissue slices from other
patients. The <code><a href="../reference/bulk_pca.html">bulk_pca()</a></code> allows to map a clinical variable
to the PCA (gender in this example), by including the exact,
case-sensitive name of the column from the sample metadata file.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bulk_pca.html">bulk_pca</a></span><span class="op">(</span><span class="va">melanoma</span>, clinvar<span class="op">=</span><span class="st">'gender'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/pca_chunk-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="transformation-of-spatially-resolved-transcriptomic-data">Transformation of spatially-resolved transcriptomic data<a class="anchor" aria-label="anchor" href="#transformation-of-spatially-resolved-transcriptomic-data"></a>
</h2>
<p>Many transformation methods are available for both bulk and
single-cell RNA-Seq count data. In <code>spatialGE</code>, the function
<code><a href="../reference/spatialTransform.html">spatialTransform()</a></code> applies log-transformation to the data,
after library size normalization. Similar to Seurat, it applies a
scaling factor (<code>scale_f=10000</code> by default).
<code>spatialGE</code> borrows methods from the field of spatial
statistics, often requiring gaussian-shaped data distributions. Hence,
users also can apply voom transformation which could, in certain
situations, provide more gaussian-shaped distributions than
log-transformed data.</p>
<p>We will use log-transformation for this example:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialTransform.html">spatialTransform</a></span><span class="op">(</span><span class="va">melanoma</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization-of-gene-expression-from-spatially-resolved-transcriptomics-data">Visualization of gene expression from spatially-resolved
transcriptomics data<a class="anchor" aria-label="anchor" href="#visualization-of-gene-expression-from-spatially-resolved-transcriptomics-data"></a>
</h2>
<p>After data transformation, expression of specific genes can be
visualized using ‘quilt’ plots. The function
<code><a href="../reference/plot_gene_quilt.html">plot_gene_quilt()</a></code> shows the transformed expression of one
or several genes. We have adopted the color palettes from the packages
<code>khroma</code> and <code>RColorBrewer</code>. The name of a color
palette can be passed using the argument <code>color_pal</code>. The
default behavior of the function produces plots for all samples within
the STList, but we can define specific samples to be plotted using the
argument <code>plot_who</code>.</p>
<p>Let’s produce a quilt plot for the genes <em>MS4A1</em> and
<em>MLANA</em> (B-cell marker and melanoma antigen, respectively), for
sample number 2 of patient 1 (<code>plot_who=2</code>).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quilts1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MS4A1'</span>, <span class="st">'MLANA'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span>, 
                           color_pal<span class="op">=</span><span class="st">'BuRd'</span>, ptsize<span class="op">=</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
<p>Because <code>spatialGE</code> functions output lists of ggplot
objects, we can plot the results side-by-side using functions such as
<code>ggarrange</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">quilts1</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/unnamed-chunk-1-1.png" width="960"></p>
<p>We can see that gene expression patterns of both genes are almost
opposite: <em>MS4A1</em> is expressed in the upper right portion of the
tissue, whereas <em>MLANA</em> is expressed in the left portion. The
left portion of the tissue corresponds to the melanoma compartment
detected by Thrane et al., and the upper right portion corresponds to
the a lymphoid tissue area (see Figure 3 of their study).</p>
<div class="section level3">
<h3 id="spatial-interpolation-kriging-of-gene-expression">Spatial interpolation (“kriging”) of gene expression<a class="anchor" aria-label="anchor" href="#spatial-interpolation-kriging-of-gene-expression"></a>
</h3>
<p>We can predict a smooth gene expression surface for each sample. In
<code>spatialGE</code>, this prediction is achieved by a widely used
spatial interpolation method in spatial statistics. The method known as
‘kriging’ allows for the estimation of gene expression values in the
unsampled areas between spots, or spots for which the sample processing
failed to provide data. Estimating a transcriptomic surface via kriging
assumes that gene expression of two given points is correlated with the
physical distance between them.</p>
<p>The function <code><a href="../reference/gene_krige.html">gene_krige()</a></code> performs kriging of gene
expression via the <code>geoR</code> package. The argument
<code>res=0.2</code>, sets the resolution of kriging, the smaller the
value of <code>res</code>, the more resolution and computing time
required. We specify that kriging will be performed for two of the
spatial arrays (<code>who=c(2, 3)</code>):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_krige.html">gene_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MS4A1'</span>, <span class="st">'MLANA'</span><span class="op">)</span>, who<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, python<span class="op">=</span><span class="cn">F</span>, res<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: magrittr</span></code></pre></div>
<p>The surfaces can be visualized using the
<code><a href="../reference/plot_gene_krige.html">plot_gene_krige()</a></code> function:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_krige.html">plot_gene_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MS4A1'</span>, <span class="st">'MLANA'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="co">#&gt; Loading required namespace: rgeos</span>
<span class="co">#&gt; Loading required package: ggplot2</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges1</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotgenekrige_chunk1-1.png" width="960"></p>
<p>By looking at the transcriptomic surfaces of the tho genes, it is
easier to detect “pockets” of high (or low) expression within the
tissue. It is now more evident that expression of <em>MS4A1</em> is
lower on the left (melanoma) region of the tumor slice, compared to the
rest of the slice.</p>
<p>Just like in <code><a href="../reference/plot_gene_quilt.html">plot_gene_quilt()</a></code>, we can control which
tissue slice to plot with the <code>plot_who</code> argument. These
commands plot sample 1 from patient 2.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_krige.html">plot_gene_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MS4A1'</span>, <span class="st">'MLANA'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">3</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges2</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotgenekrige_chunk2-1.png" width="960"></p>
</div>
</div>
<div class="section level2">
<h2 id="gene-expression-deconvolution-in-spatial-arrays">Gene expression deconvolution in spatial arrays<a class="anchor" aria-label="anchor" href="#gene-expression-deconvolution-in-spatial-arrays"></a>
</h2>
<p>In <code>spatialGE</code>, we have implemented xCell to provide a
qualitative snapshot of a tissue’s cell composition. Deconvolution in
xCell focuses on immunological cell types, which is of crucial
importance in the study of the tumor microenvironment. The
implementation of additional is being evaluated for future versions of
<code>spatialGE</code>. The user may also want to look at other methods
such as <a href="https://academic.oup.com/nar/article/49/9/e50/6129341" class="external-link">SPOTlight</a>.</p>
<p>In addition to xCell, the <code>spatialDeconv</code> function
automatically applies ESTIMATE to obtain tumor purity scores. The purity
scores are internally clustered using model-based clustering
(<code>mclust</code>) to approximate tumor/stroma compartments.</p>
<p>Let’s apply xCell to the samples:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialDeconv.html">spatialDeconv</a></span><span class="op">(</span><span class="va">melanoma</span>, method<span class="op">=</span><span class="st">'xcell'</span><span class="op">)</span>
<span class="co">#&gt; This will take some time...</span>
<span class="co">#&gt; Loading required package: estimate</span>
<span class="co">#&gt; Loading required package: xCell</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel1_rep1...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel1_rep2...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel2_rep1...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel2_rep2...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel3_rep1...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel3_rep2...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel4_rep1...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Applying xCell to sample: ST_mel4_rep2...</span></code></pre></div>
<p>With the function <code><a href="../reference/plot_deconv_quilt.html">plot_deconv_quilt()</a></code>, we can visualize
the cell type scores inferred by xCell. For example, B-cells:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quilts2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_quilt.html">plot_deconv_quilt</a></span><span class="op">(</span><span class="va">melanoma</span>, cells<span class="op">=</span><span class="st">'b_cells'</span>, plot_who<span class="op">=</span><span class="fl">2</span>, purity<span class="op">=</span><span class="cn">T</span>, 
                             color_pal<span class="op">=</span><span class="st">'BuRd'</span>, visium<span class="op">=</span><span class="cn">F</span>, ptsize<span class="op">=</span><span class="fl">1</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">quilts2</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/deconvquilts_chunk-1.png" width="960"></p>
<p>By plotting the tumor/stroma classification of the spots, the
differences in cell scores can quickly be detected. Melanoma areas
(squares) show lower B-cells. Nonetheless, B-cells were abundant in the
lymphoid pocket (upper right part of the tissue).</p>
<p>Similar plots can be generated for gene expression. These plots can
also be made interactive by making <code>inter=T</code>, which converts
plots to Plotly objects. For example, let us make an interactive plot
for gene <em>MS4A1</em>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quilts3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="st">'MS4A1'</span>, plot_who<span class="op">=</span><span class="fl">2</span>, purity<span class="op">=</span><span class="cn">T</span>, visium<span class="op">=</span><span class="cn">F</span>, inter<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="va">quilts3</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<div id="htmlwidget-cd4a7afb8fdaf08cd8f2" style="width:960px;height:672px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-cd4a7afb8fdaf08cd8f2">{"x":{"data":[{"x":[15,16,17,18,13,14,15,16,17,18,19,13,14,15,16,17,18,13,15,16,17,18,12,13,14,16,17,12,13,16,17,11,14,16,17,18,9,13,14,16,17,18,5,6,7,8,9,12,16,17,18,3,4,8,16,17,18,19,4,7,8,15,16,17,18,19,20,21,4,5,6,15,16,17,18,19,20,21,22,23,24,4,5,15,16,17,18,19,20,21,22,23,24,25,26,27,28,5,6,8,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,11,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,8,12,13,14,15,16,17,18,19,20,21,22,23,24,25,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,11,12,13,14,15,16,17,18,19,20,21,22,23,14,15,17,18,19,20,21,22],"y":[7,7,7,7,8,8,8,8,8,8,8,9,9,9,9,9,9,10,10,10,10,10,11,11,11,11,11,12,12,12,12,13,13,13,13,13,14,14,14,14,14,14,15,15,15,15,15,15,15,15,15,16,16,16,16,16,16,16,17,17,17,17,17,17,17,17,17,17,18,18,18,18,18,18,18,18,18,18,18,18,18,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,20,21,21,21,21,21,21,21,21,21,21,21,21,21,21,21,21,22,22,22,22,22,22,22,22,22,22,22,22,22,22,22,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,24,24,24,24,24,24,24,24,24,24,24,24,24,24,24,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,26,26,26,26,26,26,26,26,26,26,26,26,26,27,27,27,27,27,27,27,27],"text":["X: 15<br />Y:  7<br />Tumor_Stroma: stroma<br />Value: 2.4316597","X: 16<br />Y:  7<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 17<br />Y:  7<br />Tumor_Stroma: stroma<br />Value: 2.3951725","X: 18<br />Y:  7<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 13<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 2.0101081","X: 14<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 1.7016154","X: 15<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 0.6210503","X: 16<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 1.8826707","X: 17<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 2.7425443","X: 18<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 19<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 13<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 1.6466102","X: 14<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 1.7891600","X: 15<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 2.2405310","X: 16<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 1.7759975","X: 17<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 3.3133143","X: 18<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 13<br />Y: 10<br />Tumor_Stroma: stroma<br />Value: 1.3608204","X: 15<br />Y: 10<br />Tumor_Stroma: stroma<br />Value: 0.8748858","X: 16<br />Y: 10<br />Tumor_Stroma: stroma<br />Value: 1.9708330","X: 17<br />Y: 10<br />Tumor_Stroma: stroma<br />Value: 2.7344274","X: 18<br />Y: 10<br />Tumor_Stroma: stroma<br />Value: 1.6665152","X: 12<br />Y: 11<br />Tumor_Stroma: stroma<br />Value: 1.9140283","X: 13<br />Y: 11<br />Tumor_Stroma: stroma<br />Value: 2.9791986","X: 14<br />Y: 11<br />Tumor_Stroma: stroma<br />Value: 2.0097350","X: 16<br />Y: 11<br />Tumor_Stroma: stroma<br />Value: 0.9013167","X: 17<br />Y: 11<br />Tumor_Stroma: stroma<br />Value: 2.5996088","X: 12<br />Y: 12<br />Tumor_Stroma: stroma<br />Value: 1.7172765","X: 13<br />Y: 12<br />Tumor_Stroma: stroma<br />Value: 1.7704157","X: 16<br />Y: 12<br />Tumor_Stroma: stroma<br />Value: 2.2954701","X: 17<br />Y: 12<br />Tumor_Stroma: stroma<br />Value: 1.5978592","X: 11<br />Y: 13<br />Tumor_Stroma: stroma<br />Value: 1.6335472","X: 14<br />Y: 13<br />Tumor_Stroma: stroma<br />Value: 1.6514469","X: 16<br />Y: 13<br />Tumor_Stroma: stroma<br />Value: 1.1849402","X: 17<br />Y: 13<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 18<br />Y: 13<br />Tumor_Stroma: stroma<br />Value: 1.8476254","X:  9<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 13<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 14<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 1.0978795","X: 16<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 2.3254989","X: 17<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 2.2805646","X: 18<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  5<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 1.5325059","X:  6<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  7<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 0.6395227","X:  8<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  9<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 12<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 16<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 0.9490278","X: 17<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 0.8668107","X: 18<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 2.5570474","X:  3<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 1.9421482","X:  4<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  8<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 16<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 1.2918647","X: 17<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 1.8711890","X: 18<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 1.6675608","X: 19<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 1.5999065","X:  4<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  7<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  8<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 15<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 16<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 1.3376745","X: 17<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 1.4118798","X: 18<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 2.0397860","X: 19<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 1.9724295","X: 20<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 2.0259456","X: 21<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 1.5073916","X:  4<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  5<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 1.2608573","X:  6<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 15<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 1.6575201","X: 16<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 1.9667684","X: 17<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 18<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 19<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 20<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 2.9726824","X: 21<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 2.4774686","X: 22<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 1.5820121","X: 23<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 2.4206263","X: 24<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  4<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  5<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 15<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 0.3652528","X: 16<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 1.4811256","X: 17<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 18<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 2.1091374","X: 19<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 20<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 2.6417544","X: 21<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 22<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 2.2166584","X: 23<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 2.8212472","X: 24<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 2.5881525","X: 25<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 26<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 1.4116451","X: 27<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 28<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  5<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  6<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X:  8<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 13<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.8242734","X: 14<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 1.6974436","X: 15<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 0.4940727","X: 16<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 1.8548650","X: 17<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 1.6266168","X: 18<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 19<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 20<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.2696005","X: 21<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.1389640","X: 22<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.3248447","X: 23<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 3.2131337","X: 24<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 1.7276630","X: 25<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 2.9630564","X: 26<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 1.9703016","X: 27<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 1.7880193","X: 28<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 1.9607996","X: 11<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 14<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 2.4215220","X: 15<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 1.5630976","X: 16<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 1.3789891","X: 17<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 18<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 19<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 20<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 21<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 2.9365075","X: 22<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 3.1371697","X: 23<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 3.2208748","X: 24<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 2.7769118","X: 25<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 2.6056710","X: 26<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 3.0437609","X: 27<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 1.7540283","X: 28<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 1.9395940","X: 13<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 2.1351733","X: 14<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 1.4165927","X: 15<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 16<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 1.2269470","X: 17<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 18<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 19<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 2.0692421","X: 20<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 2.2039576","X: 21<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 2.1933230","X: 22<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.6692233","X: 23<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.4783479","X: 24<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 2.8006871","X: 25<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.0807086","X: 26<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.9078223","X: 27<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 3.8160560","X: 12<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 1.9721632","X: 13<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 2.3607816","X: 14<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 2.0441626","X: 15<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 1.2409260","X: 16<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 17<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 18<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 1.8301286","X: 19<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 1.6563754","X: 20<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 2.8699966","X: 21<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 3.6613719","X: 22<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 3.9471786","X: 23<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 2.9389610","X: 24<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 25<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 26<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 2.7948994","X:  8<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 12<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 1.0998137","X: 13<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 2.1599463","X: 14<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 15<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 16<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 17<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 1.8180701","X: 18<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 1.4481437","X: 19<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 20<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 2.4039170","X: 21<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 3.5355285","X: 22<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 3.4369556","X: 23<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 1.8672471","X: 24<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 2.2562279","X: 25<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 2.2467192","X:  9<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 10<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 11<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 12<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 13<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 2.0171379","X: 14<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 15<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 16<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 17<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 18<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 19<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 20<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 2.0771408","X: 21<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 3.6287755","X: 22<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 2.3205558","X: 23<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 2.6769696","X: 24<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 3.0076110","X: 11<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 12<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 2.0975589","X: 13<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 2.3759154","X: 14<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 15<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 16<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 17<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 18<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 1.4175397","X: 19<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 2.7877176","X: 20<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 1.4111760","X: 21<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 3.6801585","X: 22<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 3.1773912","X: 23<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 2.6285358","X: 14<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 15<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 17<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 2.0865615","X: 18<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 19<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 0.0000000","X: 20<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 1.4058072","X: 21<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 2.8779912","X: 22<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 2.9346399"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":["rgba(230,119,23,1)","rgba(255,255,229,1)","rgba(231,122,25,1)","rgba(255,255,229,1)","rgba(249,151,40,1)","rgba(254,174,72,1)","rgba(255,237,176,1)","rgba(252,161,52,1)","rgba(215,94,10,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(254,178,78,1)","rgba(253,168,63,1)","rgba(239,133,31,1)","rgba(253,169,64,1)","rgba(166,62,6,1)","rgba(255,255,229,1)","rgba(255,200,108,1)","rgba(255,230,155,1)","rgba(251,154,41,1)","rgba(215,95,11,1)","rgba(254,177,76,1)","rgba(252,158,49,1)","rgba(202,75,2,1)","rgba(249,151,40,1)","rgba(255,229,152,1)","rgba(222,106,17,1)","rgba(254,173,71,1)","rgba(253,169,65,1)","rgba(236,129,29,1)","rgba(255,182,83,1)","rgba(254,179,80,1)","rgba(254,178,78,1)","rgba(255,212,125,1)","rgba(255,255,229,1)","rgba(253,163,56,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,219,134,1)","rgba(235,127,28,1)","rgba(237,130,29,1)","rgba(255,255,229,1)","rgba(255,187,90,1)","rgba(255,255,229,1)","rgba(255,237,175,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(254,228,148,1)","rgba(255,230,155,1)","rgba(224,109,18,1)","rgba(251,156,45,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,205,114,1)","rgba(252,162,54,1)","rgba(254,177,76,1)","rgba(255,182,83,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,201,110,1)","rgba(255,196,102,1)","rgba(248,149,38,1)","rgba(251,154,41,1)","rgba(249,150,39,1)","rgba(255,189,93,1)","rgba(255,255,229,1)","rgba(255,207,118,1)","rgba(255,255,229,1)","rgba(254,178,77,1)","rgba(251,155,42,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(203,75,2,1)","rgba(227,115,22,1)","rgba(255,183,85,1)","rgba(230,120,24,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,244,198,1)","rgba(255,191,95,1)","rgba(255,255,229,1)","rgba(245,144,36,1)","rgba(255,255,229,1)","rgba(220,102,15,1)","rgba(255,255,229,1)","rgba(240,135,32,1)","rgba(211,88,7,1)","rgba(222,106,17,1)","rgba(255,255,229,1)","rgba(255,196,102,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(211,87,7,1)","rgba(254,175,73,1)","rgba(255,241,187,1)","rgba(252,163,56,1)","rgba(254,180,80,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(237,131,30,1)","rgba(243,141,35,1)","rgba(235,127,28,1)","rgba(177,66,5,1)","rgba(254,172,70,1)","rgba(204,76,2,1)","rgba(251,154,41,1)","rgba(253,168,63,1)","rgba(251,155,43,1)","rgba(255,255,229,1)","rgba(230,119,24,1)","rgba(255,185,87,1)","rgba(255,198,106,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(205,78,3,1)","rgba(185,69,4,1)","rgba(176,65,5,1)","rgba(213,91,9,1)","rgba(221,105,16,1)","rgba(195,73,3,1)","rgba(253,170,67,1)","rgba(251,157,45,1)","rgba(244,142,35,1)","rgba(255,195,102,1)","rgba(255,255,229,1)","rgba(255,209,121,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(247,147,37,1)","rgba(240,136,32,1)","rgba(241,137,33,1)","rgba(129,48,7,1)","rgba(149,55,7,1)","rgba(212,89,8,1)","rgba(191,71,3,1)","rgba(106,38,6,1)","rgba(115,42,7,1)","rgba(251,154,41,1)","rgba(233,124,26,1)","rgba(248,149,38,1)","rgba(255,208,120,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(253,165,58,1)","rgba(254,178,77,1)","rgba(208,84,5,1)","rgba(130,48,7,1)","rgba(102,37,6,1)","rgba(205,78,3,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(212,90,8,1)","rgba(255,255,229,1)","rgba(255,219,134,1)","rgba(242,140,34,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(253,166,60,1)","rgba(255,193,99,1)","rgba(255,255,229,1)","rgba(231,121,25,1)","rgba(143,53,7,1)","rgba(153,57,6,1)","rgba(252,162,54,1)","rgba(238,132,30,1)","rgba(238,133,31,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(249,151,39,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(246,146,37,1)","rgba(133,49,7,1)","rgba(235,127,28,1)","rgba(218,99,13,1)","rgba(199,74,3,1)","rgba(255,255,229,1)","rgba(245,144,36,1)","rgba(232,123,26,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,195,102,1)","rgba(212,90,8,1)","rgba(255,196,102,1)","rgba(128,47,7,1)","rgba(181,67,4,1)","rgba(220,103,15,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(246,145,37,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,196,103,1)","rgba(208,83,5,1)","rgba(205,78,3,1)"],"opacity":1,"size":1.88976377952756,"symbol":"cross-thin-open","line":{"width":3.77952755905512,"color":["rgba(230,119,23,1)","rgba(255,255,229,1)","rgba(231,122,25,1)","rgba(255,255,229,1)","rgba(249,151,40,1)","rgba(254,174,72,1)","rgba(255,237,176,1)","rgba(252,161,52,1)","rgba(215,94,10,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(254,178,78,1)","rgba(253,168,63,1)","rgba(239,133,31,1)","rgba(253,169,64,1)","rgba(166,62,6,1)","rgba(255,255,229,1)","rgba(255,200,108,1)","rgba(255,230,155,1)","rgba(251,154,41,1)","rgba(215,95,11,1)","rgba(254,177,76,1)","rgba(252,158,49,1)","rgba(202,75,2,1)","rgba(249,151,40,1)","rgba(255,229,152,1)","rgba(222,106,17,1)","rgba(254,173,71,1)","rgba(253,169,65,1)","rgba(236,129,29,1)","rgba(255,182,83,1)","rgba(254,179,80,1)","rgba(254,178,78,1)","rgba(255,212,125,1)","rgba(255,255,229,1)","rgba(253,163,56,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,219,134,1)","rgba(235,127,28,1)","rgba(237,130,29,1)","rgba(255,255,229,1)","rgba(255,187,90,1)","rgba(255,255,229,1)","rgba(255,237,175,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(254,228,148,1)","rgba(255,230,155,1)","rgba(224,109,18,1)","rgba(251,156,45,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,205,114,1)","rgba(252,162,54,1)","rgba(254,177,76,1)","rgba(255,182,83,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,201,110,1)","rgba(255,196,102,1)","rgba(248,149,38,1)","rgba(251,154,41,1)","rgba(249,150,39,1)","rgba(255,189,93,1)","rgba(255,255,229,1)","rgba(255,207,118,1)","rgba(255,255,229,1)","rgba(254,178,77,1)","rgba(251,155,42,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(203,75,2,1)","rgba(227,115,22,1)","rgba(255,183,85,1)","rgba(230,120,24,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,244,198,1)","rgba(255,191,95,1)","rgba(255,255,229,1)","rgba(245,144,36,1)","rgba(255,255,229,1)","rgba(220,102,15,1)","rgba(255,255,229,1)","rgba(240,135,32,1)","rgba(211,88,7,1)","rgba(222,106,17,1)","rgba(255,255,229,1)","rgba(255,196,102,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(211,87,7,1)","rgba(254,175,73,1)","rgba(255,241,187,1)","rgba(252,163,56,1)","rgba(254,180,80,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(237,131,30,1)","rgba(243,141,35,1)","rgba(235,127,28,1)","rgba(177,66,5,1)","rgba(254,172,70,1)","rgba(204,76,2,1)","rgba(251,154,41,1)","rgba(253,168,63,1)","rgba(251,155,43,1)","rgba(255,255,229,1)","rgba(230,119,24,1)","rgba(255,185,87,1)","rgba(255,198,106,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(205,78,3,1)","rgba(185,69,4,1)","rgba(176,65,5,1)","rgba(213,91,9,1)","rgba(221,105,16,1)","rgba(195,73,3,1)","rgba(253,170,67,1)","rgba(251,157,45,1)","rgba(244,142,35,1)","rgba(255,195,102,1)","rgba(255,255,229,1)","rgba(255,209,121,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(247,147,37,1)","rgba(240,136,32,1)","rgba(241,137,33,1)","rgba(129,48,7,1)","rgba(149,55,7,1)","rgba(212,89,8,1)","rgba(191,71,3,1)","rgba(106,38,6,1)","rgba(115,42,7,1)","rgba(251,154,41,1)","rgba(233,124,26,1)","rgba(248,149,38,1)","rgba(255,208,120,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(253,165,58,1)","rgba(254,178,77,1)","rgba(208,84,5,1)","rgba(130,48,7,1)","rgba(102,37,6,1)","rgba(205,78,3,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(212,90,8,1)","rgba(255,255,229,1)","rgba(255,219,134,1)","rgba(242,140,34,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(253,166,60,1)","rgba(255,193,99,1)","rgba(255,255,229,1)","rgba(231,121,25,1)","rgba(143,53,7,1)","rgba(153,57,6,1)","rgba(252,162,54,1)","rgba(238,132,30,1)","rgba(238,133,31,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(249,151,39,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(246,146,37,1)","rgba(133,49,7,1)","rgba(235,127,28,1)","rgba(218,99,13,1)","rgba(199,74,3,1)","rgba(255,255,229,1)","rgba(245,144,36,1)","rgba(232,123,26,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,195,102,1)","rgba(212,90,8,1)","rgba(255,196,102,1)","rgba(128,47,7,1)","rgba(181,67,4,1)","rgba(220,103,15,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(246,145,37,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,196,103,1)","rgba(208,83,5,1)","rgba(205,78,3,1)"]}},"hoveron":"points","name":"stroma","legendgroup":"stroma","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[15,14,15,12,13,15,7,8,10,11,12,15,10,11,13,14,15,5,6,7,9,10,11,12,13,14,15,5,6,9,10,11,12,13,14,7,8,9,10,11,12,13,14,6,7,8,9,10,11,12,13,14,7,9,10,11,12,6,7,8,9,10,12,13,6,7,8,9,10,11,12,7,8,9,10,11,9,10,11],"y":[11,12,12,13,13,13,14,14,14,14,14,14,15,15,15,15,15,16,16,16,16,16,16,16,16,16,16,17,17,17,17,17,17,17,17,18,18,18,18,18,18,18,18,19,19,19,19,19,19,19,19,19,20,20,20,20,20,21,21,21,21,21,21,21,22,22,22,22,22,22,22,23,23,23,23,23,24,24,24],"text":["X: 15<br />Y: 11<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 14<br />Y: 12<br />Tumor_Stroma: tumor<br />Value: 1.4927171","X: 15<br />Y: 12<br />Tumor_Stroma: tumor<br />Value: 1.2652836","X: 12<br />Y: 13<br />Tumor_Stroma: tumor<br />Value: 0.8715557","X: 13<br />Y: 13<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 15<br />Y: 13<br />Tumor_Stroma: tumor<br />Value: 1.1525883","X:  7<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  8<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 10<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 0.8763564","X: 11<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 12<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 1.1361046","X: 15<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 10<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 11<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 1.1953392","X: 13<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 0.5547524","X: 14<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 1.0293945","X: 15<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 0.7810904","X:  5<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 0.7690936","X:  6<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  7<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  9<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 10<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 11<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 12<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 13<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 0.6017371","X: 14<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 0.7915024","X: 15<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 0.6710068","X:  5<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 1.2717430","X:  6<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  9<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 10<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 11<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 12<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 1.0336349","X: 13<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 14<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 1.3193336","X:  7<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  8<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  9<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 10<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 11<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 12<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 13<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 14<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  6<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  7<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 0.7433755","X:  8<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  9<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 10<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 0.8266263","X: 11<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 12<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 13<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 0.5100801","X: 14<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 1.3866507","X:  7<br />Y: 20<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  9<br />Y: 20<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 10<br />Y: 20<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 11<br />Y: 20<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 12<br />Y: 20<br />Tumor_Stroma: tumor<br />Value: 0.5109857","X:  6<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  7<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  8<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  9<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 10<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 12<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 13<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 0.9558562","X:  6<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  7<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  8<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  9<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 0.7137133","X: 10<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 11<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 0.9445227","X: 12<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  7<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 0.9528609","X:  8<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  9<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 1.0120784","X: 10<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 11<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X:  9<br />Y: 24<br />Tumor_Stroma: tumor<br />Value: 1.2308803","X: 10<br />Y: 24<br />Tumor_Stroma: tumor<br />Value: 0.0000000","X: 11<br />Y: 24<br />Tumor_Stroma: tumor<br />Value: 0.0000000"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":["rgba(255,255,229,1)","rgba(255,190,94,1)","rgba(255,207,117,1)","rgba(255,230,155,1)","rgba(255,255,229,1)","rgba(255,215,128,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,230,154,1)","rgba(255,255,229,1)","rgba(255,216,130,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,212,124,1)","rgba(255,239,182,1)","rgba(254,224,141,1)","rgba(255,233,163,1)","rgba(255,233,164,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,238,178,1)","rgba(255,232,162,1)","rgba(255,236,172,1)","rgba(255,206,116,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(254,224,140,1)","rgba(255,255,229,1)","rgba(255,203,112,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,234,166,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,231,159,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,240,186,1)","rgba(255,198,105,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,240,186,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(254,228,148,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,235,168,1)","rgba(255,255,229,1)","rgba(254,228,149,1)","rgba(255,255,229,1)","rgba(254,228,148,1)","rgba(255,255,229,1)","rgba(254,225,142,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,209,121,1)","rgba(255,255,229,1)","rgba(255,255,229,1)"],"opacity":1,"size":1.88976377952756,"symbol":"square","line":{"width":3.77952755905512,"color":["rgba(255,255,229,1)","rgba(255,190,94,1)","rgba(255,207,117,1)","rgba(255,230,155,1)","rgba(255,255,229,1)","rgba(255,215,128,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,230,154,1)","rgba(255,255,229,1)","rgba(255,216,130,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,212,124,1)","rgba(255,239,182,1)","rgba(254,224,141,1)","rgba(255,233,163,1)","rgba(255,233,164,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,238,178,1)","rgba(255,232,162,1)","rgba(255,236,172,1)","rgba(255,206,116,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(254,224,140,1)","rgba(255,255,229,1)","rgba(255,203,112,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,234,166,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,231,159,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,240,186,1)","rgba(255,198,105,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,240,186,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(254,228,148,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,235,168,1)","rgba(255,255,229,1)","rgba(254,228,149,1)","rgba(255,255,229,1)","rgba(254,228,148,1)","rgba(255,255,229,1)","rgba(254,225,142,1)","rgba(255,255,229,1)","rgba(255,255,229,1)","rgba(255,209,121,1)","rgba(255,255,229,1)","rgba(255,255,229,1)"]}},"hoveron":"points","name":"tumor","legendgroup":"tumor","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[10],"y":[10],"name":"99_cabc574b4876c2da18faf6e434b3d77b","type":"scatter","mode":"markers","opacity":0,"hoverinfo":"skip","showlegend":false,"marker":{"color":[0,1],"colorscale":[[0,"#FFFFE5"],[0.00334448160535117,"#FFFFE4"],[0.00668896321070234,"#FFFEE3"],[0.0100334448160535,"#FFFEE2"],[0.0133779264214047,"#FFFDE0"],[0.0167224080267559,"#FFFDDF"],[0.020066889632107,"#FFFDDE"],[0.0234113712374582,"#FFFCDD"],[0.0267558528428094,"#FFFCDC"],[0.0301003344481605,"#FFFCDB"],[0.0334448160535117,"#FFFBDA"],[0.0367892976588629,"#FFFBD9"],[0.040133779264214,"#FFFAD7"],[0.0434782608695652,"#FFFAD6"],[0.0468227424749164,"#FFFAD5"],[0.0501672240802676,"#FFF9D4"],[0.0535117056856187,"#FFF9D3"],[0.0568561872909699,"#FFF9D2"],[0.0602006688963211,"#FFF8D1"],[0.0635451505016722,"#FFF8D0"],[0.0668896321070234,"#FFF7CE"],[0.0702341137123746,"#FFF7CD"],[0.0735785953177257,"#FFF7CC"],[0.0769230769230769,"#FFF6CB"],[0.0802675585284281,"#FFF6CA"],[0.0836120401337793,"#FFF5C9"],[0.0869565217391304,"#FFF5C8"],[0.0903010033444816,"#FFF5C7"],[0.0936454849498328,"#FFF4C6"],[0.0969899665551839,"#FFF4C4"],[0.100334448160535,"#FFF4C3"],[0.103678929765886,"#FFF3C2"],[0.107023411371237,"#FFF3C1"],[0.110367892976589,"#FFF2C0"],[0.11371237458194,"#FFF2BF"],[0.117056856187291,"#FFF2BE"],[0.120401337792642,"#FFF1BD"],[0.123745819397993,"#FFF1BB"],[0.127090301003344,"#FFF1BA"],[0.130434782608696,"#FFF0B9"],[0.133779264214047,"#FFF0B8"],[0.137123745819398,"#FFEFB7"],[0.140468227424749,"#FFEFB6"],[0.1438127090301,"#FFEFB5"],[0.147157190635451,"#FFEEB4"],[0.150501672240803,"#FFEEB2"],[0.153846153846154,"#FFEEB1"],[0.157190635451505,"#FFEDB0"],[0.160535117056856,"#FFEDAF"],[0.163879598662207,"#FFEDAE"],[0.167224080267559,"#FFECAD"],[0.17056856187291,"#FFECAC"],[0.173913043478261,"#FFEBAB"],[0.177257525083612,"#FFEBAA"],[0.180602006688963,"#FFEBA8"],[0.183946488294314,"#FFEAA7"],[0.187290969899666,"#FFEAA6"],[0.190635451505017,"#FFEAA5"],[0.193979933110368,"#FFE9A4"],[0.197324414715719,"#FFE9A3"],[0.20066889632107,"#FFE8A2"],[0.204013377926421,"#FFE8A1"],[0.207357859531773,"#FFE89F"],[0.210702341137124,"#FFE79E"],[0.214046822742475,"#FFE79D"],[0.217391304347826,"#FFE79C"],[0.220735785953177,"#FFE69B"],[0.224080267558528,"#FFE69A"],[0.22742474916388,"#FFE599"],[0.230769230769231,"#FEE598"],[0.234113712374582,"#FEE596"],[0.237458193979933,"#FEE495"],[0.240802675585284,"#FEE494"],[0.244147157190635,"#FEE493"],[0.247491638795987,"#FEE392"],[0.250836120401338,"#FEE391"],[0.254180602006689,"#FEE28F"],[0.25752508361204,"#FEE18E"],[0.260869565217391,"#FEE08D"],[0.264214046822742,"#FEDF8B"],[0.267558528428094,"#FFDE8A"],[0.270903010033445,"#FFDD89"],[0.274247491638796,"#FFDC87"],[0.277591973244147,"#FFDB86"],[0.280936454849498,"#FFDA85"],[0.284280936454849,"#FFD983"],[0.287625418060201,"#FFD882"],[0.290969899665552,"#FFD781"],[0.294314381270903,"#FFD67F"],[0.297658862876254,"#FFD57E"],[0.301003344481605,"#FFD47D"],[0.304347826086956,"#FFD37C"],[0.307692307692308,"#FFD27A"],[0.311036789297659,"#FFD179"],[0.31438127090301,"#FFD078"],[0.317725752508361,"#FFCF76"],[0.321070234113712,"#FFCE75"],[0.324414715719063,"#FFCD74"],[0.327759197324415,"#FFCC72"],[0.331103678929766,"#FFCB71"],[0.334448160535117,"#FFCB70"],[0.337792642140468,"#FFCA6E"],[0.341137123745819,"#FFC96D"],[0.344481605351171,"#FFC86C"],[0.347826086956522,"#FFC76A"],[0.351170568561873,"#FFC669"],[0.354515050167224,"#FFC568"],[0.357859531772575,"#FFC466"],[0.361204013377926,"#FFC365"],[0.364548494983278,"#FFC264"],[0.367892976588629,"#FFC162"],[0.37123745819398,"#FFC061"],[0.374581939799331,"#FFBF60"],[0.377926421404682,"#FFBE5E"],[0.381270903010033,"#FFBD5D"],[0.384615384615385,"#FFBC5C"],[0.387959866220736,"#FFBB5A"],[0.391304347826087,"#FFBA59"],[0.394648829431438,"#FFB957"],[0.397993311036789,"#FFB856"],[0.40133779264214,"#FFB755"],[0.404682274247492,"#FFB653"],[0.408026755852843,"#FFB552"],[0.411371237458194,"#FEB451"],[0.414715719063545,"#FEB34F"],[0.418060200668896,"#FEB24E"],[0.421404682274247,"#FEB14D"],[0.424749163879599,"#FEB04B"],[0.42809364548495,"#FEAF4A"],[0.431438127090301,"#FEAE48"],[0.434782608695652,"#FEAD47"],[0.438127090301003,"#FEAC45"],[0.441471571906354,"#FEAB44"],[0.444816053511706,"#FDAA43"],[0.448160535117057,"#FDA941"],[0.451505016722408,"#FDA840"],[0.454849498327759,"#FDA73E"],[0.45819397993311,"#FDA63D"],[0.461538461538462,"#FDA53B"],[0.464882943143813,"#FDA43A"],[0.468227424749164,"#FDA338"],[0.471571906354515,"#FCA237"],[0.474916387959866,"#FCA135"],[0.478260869565217,"#FCA034"],[0.481605351170569,"#FC9F32"],[0.48494983277592,"#FC9E31"],[0.488294314381271,"#FC9D2F"],[0.491638795986622,"#FB9C2D"],[0.494983277591973,"#FB9B2C"],[0.498327759197324,"#FB9A2A"],[0.501672240802676,"#FB9929"],[0.505016722408027,"#FA9828"],[0.508361204013378,"#F99728"],[0.511705685618729,"#F99627"],[0.51505016722408,"#F89527"],[0.518394648829431,"#F89426"],[0.521739130434783,"#F79326"],[0.525083612040134,"#F69225"],[0.528428093645485,"#F69125"],[0.531772575250836,"#F59024"],[0.535117056856187,"#F58F24"],[0.538461538461538,"#F48E23"],[0.54180602006689,"#F38D23"],[0.545150501672241,"#F38C22"],[0.548494983277592,"#F28B22"],[0.551839464882943,"#F28A21"],[0.555183946488294,"#F18921"],[0.558528428093645,"#F08820"],[0.561872909698997,"#F08720"],[0.565217391304348,"#EF861F"],[0.568561872909699,"#EE851F"],[0.57190635451505,"#EE841E"],[0.575250836120401,"#ED831E"],[0.578595317725752,"#ED821D"],[0.581939799331104,"#EC811D"],[0.585284280936455,"#EB801C"],[0.588628762541806,"#EB7F1C"],[0.591973244147157,"#EA7E1B"],[0.595317725752508,"#E97D1B"],[0.59866220735786,"#E97C1A"],[0.602006688963211,"#E87B1A"],[0.605351170568562,"#E87A19"],[0.608695652173913,"#E77919"],[0.612040133779264,"#E67818"],[0.615384615384615,"#E67718"],[0.618729096989967,"#E57617"],[0.622073578595318,"#E47517"],[0.625418060200669,"#E47416"],[0.62876254180602,"#E37315"],[0.632107023411371,"#E37215"],[0.635451505016722,"#E27114"],[0.638795986622074,"#E17014"],[0.642140468227425,"#E16F13"],[0.645484949832776,"#E06E13"],[0.648829431438127,"#DF6D12"],[0.652173913043478,"#DF6C12"],[0.655518394648829,"#DE6A11"],[0.658862876254181,"#DE6911"],[0.662207357859532,"#DD6810"],[0.665551839464883,"#DC6710"],[0.668896321070234,"#DC660F"],[0.672240802675585,"#DB650E"],[0.675585284280936,"#DA640E"],[0.678929765886288,"#DA630D"],[0.682274247491639,"#D9620D"],[0.68561872909699,"#D8610C"],[0.688963210702341,"#D8600C"],[0.692307692307692,"#D75F0B"],[0.695652173913043,"#D65E0A"],[0.698996655518395,"#D65D0A"],[0.702341137123746,"#D55C09"],[0.705685618729097,"#D55B09"],[0.709030100334448,"#D45A08"],[0.712374581939799,"#D35807"],[0.71571906354515,"#D35707"],[0.719063545150502,"#D25606"],[0.722408026755853,"#D15506"],[0.725752508361204,"#D15405"],[0.729096989966555,"#D05305"],[0.732441471571906,"#CF5204"],[0.735785953177257,"#CF5104"],[0.739130434782609,"#CE5003"],[0.74247491638796,"#CD4F03"],[0.745819397993311,"#CD4D03"],[0.749163879598662,"#CC4C02"],[0.752508361204013,"#CB4C02"],[0.755852842809364,"#C94B02"],[0.759197324414716,"#C84A02"],[0.762541806020067,"#C74A03"],[0.765886287625418,"#C54903"],[0.769230769230769,"#C44903"],[0.77257525083612,"#C24803"],[0.775919732441471,"#C14803"],[0.779264214046823,"#BF4703"],[0.782608695652174,"#BE4704"],[0.785953177257525,"#BD4604"],[0.789297658862876,"#BB4604"],[0.792642140468227,"#BA4504"],[0.795986622073579,"#B84504"],[0.79933110367893,"#B74404"],[0.802675585284281,"#B54304"],[0.806020066889632,"#B44305"],[0.809364548494983,"#B34205"],[0.812709030100334,"#B14205"],[0.816053511705686,"#B04105"],[0.819397993311037,"#AE4105"],[0.822742474916388,"#AD4005"],[0.826086956521739,"#AC4005"],[0.82943143812709,"#AA3F05"],[0.832775919732441,"#A93F05"],[0.836120401337793,"#A73E05"],[0.839464882943144,"#A63E06"],[0.842809364548495,"#A53D06"],[0.846153846153846,"#A33D06"],[0.849498327759197,"#A23C06"],[0.852842809364548,"#A13B06"],[0.8561872909699,"#9F3B06"],[0.859531772575251,"#9E3A06"],[0.862876254180602,"#9C3A06"],[0.866220735785953,"#9B3906"],[0.869565217391304,"#9A3906"],[0.872909698996655,"#983806"],[0.876254180602007,"#973806"],[0.879598662207358,"#963706"],[0.882943143812709,"#943707"],[0.88628762541806,"#933607"],[0.889632107023411,"#913607"],[0.892976588628763,"#903507"],[0.896321070234114,"#8F3507"],[0.899665551839465,"#8D3407"],[0.903010033444816,"#8C3407"],[0.906354515050167,"#8B3307"],[0.909698996655518,"#893307"],[0.91304347826087,"#883207"],[0.916387959866221,"#873207"],[0.919732441471572,"#853107"],[0.923076923076923,"#843107"],[0.926421404682274,"#833007"],[0.929765886287625,"#813007"],[0.933110367892977,"#802F07"],[0.936454849498328,"#7F2F07"],[0.939799331103679,"#7D2E07"],[0.94314381270903,"#7C2E07"],[0.946488294314381,"#7B2D07"],[0.949832775919732,"#792C07"],[0.953177257525084,"#782C07"],[0.956521739130435,"#772B07"],[0.959866220735786,"#752B07"],[0.963210702341137,"#742A07"],[0.966555183946488,"#732A07"],[0.969899665551839,"#722907"],[0.973244147157191,"#702907"],[0.976588628762542,"#6F2807"],[0.979933110367893,"#6E2807"],[0.983277591973244,"#6C2707"],[0.986622073578595,"#6B2706"],[0.989966555183946,"#6A2606"],[0.993311036789298,"#692606"],[0.996655518394649,"#672506"],[1,"#662506"]],"colorbar":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"thickness":23.04,"title":"log_expr","titlefont":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"tickmode":"array","ticktext":["0","1","2","3"],"tickvals":[0,0.2533455147618,0.506691029523601,0.760036544285401],"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"ticklen":2,"len":0.5,"yanchor":"top","y":1}},"xaxis":"x","yaxis":"y","frame":null}],"layout":{"margin":{"t":43.9712981082844,"r":7.30593607305936,"b":40.3913894324853,"l":37.2602739726027},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"title":{"text":"MS4A1<br />sample: ST_mel1_rep2","font":{"color":"rgba(0,0,0,1)","family":"","size":17.5342465753425},"x":0,"xref":"paper"},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[1.75,29.25],"tickmode":"array","ticktext":["10","20"],"tickvals":[10,20],"categoryorder":"array","categoryarray":["10","20"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"X Position","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"scaleanchor":"y","scaleratio":1,"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[6,28],"tickmode":"array","ticktext":["10","15","20","25"],"tickvals":[10,15,20,25],"categoryorder":"array","categoryarray":["10","15","20","25"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"Y Position","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"scaleanchor":"x","scaleratio":1,"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"y":0.5,"yanchor":"top","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"121ec3cf76117":{"x":{},"y":{},"fill":{},"shape":{},"colour":{},"type":"scatter"}},"cur_data":"121ec3cf76117","visdat":{"121ec3cf76117":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script><p>Spatial interpolation of cell scores is also possible with the
<code><a href="../reference/deconv_krige.html">deconv_krige()</a></code> function:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconv_krige.html">deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, cells<span class="op">=</span><span class="st">'b_cells'</span>, who<span class="op">=</span><span class="fl">2</span>, python<span class="op">=</span><span class="cn">F</span>, res<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span></code></pre></div>
<p>We plot the surfaces using the <code><a href="../reference/plot_deconv_krige.html">plot_deconv_krige()</a></code>
function.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_krige.html">plot_deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>,  cells<span class="op">=</span><span class="st">'b_cells'</span>, plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges3</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">1</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotdeconvkrige_chunk-1.png" width="960"></p>
<p>The <code><a href="../reference/plot_deconv_krige.html">plot_deconv_krige()</a></code> also allows annotation of tumor
spots (solid squares):</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_krige.html">plot_deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>,  cells<span class="op">=</span><span class="st">'b_cells'</span>, plot_who<span class="op">=</span><span class="fl">2</span>, purity<span class="op">=</span><span class="cn">T</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges4</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotkrigepurity_chunk-1.png" width="960"></p>
<p>Cell scores in xCell are accompanied by p-values (test for cell
scores different to zero). The p-values are calculated automatically
during deconvolution in <code>spatialGE</code>, and we can plot them
using the <code>pvalue</code> option:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_krige.html">plot_deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>,  cells<span class="op">=</span><span class="st">'b_cells'</span>, plot_who<span class="op">=</span><span class="fl">2</span>, pvalues<span class="op">=</span><span class="cn">T</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges5</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">1</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotkrigepvals_chunk-1.png" width="960"></p>
<p>As observed in the surfaces, B-cells are mostly present in the
lymphoid tissue pocket. Most of the spots had B-cell scores
significantly different to zero (gray hollow circles), as suggested by
xCell’s permutation method . Most of the non-significant scores were
observed in the melanoma area, reinforcing the observation of lower
infiltration in that part of the tissue.</p>
</div>
<div class="section level2">
<h2 id="unsupervised-spatially-informed-clustering-stclust">Unsupervised spatially-informed clustering (<em>STclust</em>)<a class="anchor" aria-label="anchor" href="#unsupervised-spatially-informed-clustering-stclust"></a>
</h2>
<p>We can also perform <em>STclust</em>, the spatially informed
clustering method in <code>spatialGE</code> to detect compartments in
the tissue. In STclust, we use weighted average matrices to capture the
transcriptomic differences among the spots. First, top variable genes
are identified via Seurat’s <code>FindVariableFeatures</code>, and
transcriptomic scaled distances are calculated using only those genes.
Next, scaled euclidean distances are computed for the physical (x, y)
distances among spots. The user defines a weight, from 0 to 1, to apply
to the physical distances. The higher the weight, the less biologically
meaningful the clustering is given that the groups will only reflect the
physical distances between the spots and less information on the
transcriptomic profiles will be used. After many tests, we have found
that weights between 0.025 - 0.25 are enough to capture the tissue
heterogeneity. This approach to spatial clustering can be performed
using the <code><a href="../reference/STclust.html">STclust()</a></code> function:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STclust.html">STclust</a></span><span class="op">(</span><span class="va">melanoma</span>, ks<span class="op">=</span><span class="st">'dtc'</span>, weights<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Registered S3 method overwritten by 'spatstat.geom':</span>
<span class="co">#&gt;   method     from</span>
<span class="co">#&gt;   print.boxx cli</span></code></pre></div>
<p>Results of clustering can be plotted via the
<code><a href="../reference/plot_STclusters.html">plot_STclusters()</a></code> function:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cluster_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_STclusters.html">plot_STclusters</a></span><span class="op">(</span><span class="va">melanoma</span>, plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span>, ptsize<span class="op">=</span><span class="fl">1</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">cluster_p</span>, nrow<span class="op">=</span><span class="fl">2</span>, ncol<span class="op">=</span><span class="fl">2</span>, legend<span class="op">=</span><span class="st">'right'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotclustspots_chunk-1.png" width="960"></p>
<p>We can see that at <code>w=0.05</code>, tissue compartments are
observed for the lymphoid tissue (upper right portion of the tissue in
red), with small pockets surrounding the melanoma area (yellow),
possibly indicating immune activity around the tumor spots. In that same
plot (<code>w=0.05</code>), red likely corresponds to stromal spots. We
have used here the dynamicTreeCut (<code>dtc</code> option) to
automatically select the number of clusters, but users can define their
own range of k to be evaluated, allowing to further isolate cell
types.</p>
</div>
<div class="section level2">
<h2 id="association-between-spatial-heterogeneity-and-clinical-variables-sthet">Association between spatial heterogeneity and clinical variables
(<em>SThet</em>)<a class="anchor" aria-label="anchor" href="#association-between-spatial-heterogeneity-and-clinical-variables-sthet"></a>
</h2>
<p>To explore the relationship between a clinical variable of interest
and the level of gene expression spatial uniformity within a sample, we
can use the function <code><a href="../reference/spatial_stat_plot.html">spatial_stat_plot()</a></code>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatial_stat_plot.html">spatial_stat_plot</a></span><span class="op">(</span><span class="va">melanoma</span>, var<span class="op">=</span><span class="st">'survival_months'</span>, gene<span class="op">=</span><span class="st">'CD74'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/clinplot_chunk-1.png" width="960"></p>
<p>In the case of the gene <em>CD74</em>, an immune gene important for
<a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=CD74" class="external-link">antigen
presentation</a>, samples with a less uniform expression (i.e.,
clustered gene expression) came from a patient with a longer survival
time (high Moran’s I; low Geary’s C). The survival of the patients did
not show an association with the tendency of <em>CD74</em> to generate
expression “hot spots” (Getis-Ord Gi was intermediate in patient 1). See
the table below for a simplistic interpretation of the spatial
autocorrelation statistics calculated in <code>spatialGE</code>:</p>
<table class="table table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
Heterogeneity statitstics in <code>spatialGE</code>
</caption>
<thead><tr>
<th style="text-align:center;">
Low
</th>
<th style="text-align:center;">
Statistic
</th>
<th style="text-align:center;">
High
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
Dispersion
</td>
<td style="text-align:center;">
Moran’s I
</td>
<td style="text-align:center;">
Clustering
</td>
</tr>
<tr>
<td style="text-align:center;">
Clustering
</td>
<td style="text-align:center;">
Geary’s C
</td>
<td style="text-align:center;">
Dispersion
</td>
</tr>
<tr>
<td style="text-align:center;">
“Cold spots”
</td>
<td style="text-align:center;">
Getis-Ord Gi
</td>
<td style="text-align:center;">
“Hot spots”
</td>
</tr>
</tbody>
</table>
<p>Estimating these statistics is computationally expensive, which is
important to consider specially for newer ST technologies such as 10X
Visium. The computation time of spatial statistics for a given gene,
increases with the number of spots (for example, 10X Visium contains
almost 5X spots compared to this data set). Hence, computed plots are
stored in the STList and can be shown again by accessing the appropriate
STList slot (<code>melanoma@spstats_plots[['CD74']]</code> in this
case).</p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.7</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] xCell_1.1.0          estimate_1.0.13      ggplot2_3.3.5       </span>
<span class="co">#&gt; [4] magrittr_2.0.2       spatialGE_1.0.0.0000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] scattermore_0.7             SeuratObject_4.0.4         </span>
<span class="co">#&gt;   [3] ragg_1.2.2                  tidyr_1.2.0                </span>
<span class="co">#&gt;   [5] bit64_4.0.5                 knitr_1.37                 </span>
<span class="co">#&gt;   [7] irlba_2.3.5                 DelayedArray_0.20.0        </span>
<span class="co">#&gt;   [9] data.table_1.14.2           rpart_4.1-15               </span>
<span class="co">#&gt;  [11] KEGGREST_1.34.0             RCurl_1.98-1.5             </span>
<span class="co">#&gt;  [13] generics_0.1.2              BiocGenerics_0.40.0        </span>
<span class="co">#&gt;  [15] ScaledMatrix_1.2.0          terra_1.5-17               </span>
<span class="co">#&gt;  [17] cowplot_1.1.1               RSQLite_2.2.9              </span>
<span class="co">#&gt;  [19] RANN_2.6.1                  future_1.23.0              </span>
<span class="co">#&gt;  [21] wordspace_0.2-6             bit_4.0.4                  </span>
<span class="co">#&gt;  [23] tzdb_0.2.0                  webshot_0.5.2              </span>
<span class="co">#&gt;  [25] xml2_1.3.3                  spatstat.data_2.1-2        </span>
<span class="co">#&gt;  [27] lubridate_1.8.0             httpuv_1.6.5               </span>
<span class="co">#&gt;  [29] SummarizedExperiment_1.24.0 assertthat_0.2.1           </span>
<span class="co">#&gt;  [31] xfun_0.29                   hms_1.1.1                  </span>
<span class="co">#&gt;  [33] jquerylib_0.1.4             evaluate_0.14              </span>
<span class="co">#&gt;  [35] promises_1.2.0.1            fansi_1.0.2                </span>
<span class="co">#&gt;  [37] igraph_1.2.11               DBI_1.1.2                  </span>
<span class="co">#&gt;  [39] htmlwidgets_1.5.4           sparsesvd_0.2              </span>
<span class="co">#&gt;  [41] spatstat.geom_2.3-1         stats4_4.1.2               </span>
<span class="co">#&gt;  [43] purrr_0.3.4                 ellipsis_0.3.2             </span>
<span class="co">#&gt;  [45] crosstalk_1.2.0             dplyr_1.0.7                </span>
<span class="co">#&gt;  [47] ggpubr_0.4.0                backports_1.4.1            </span>
<span class="co">#&gt;  [49] V8_4.0.0                    annotate_1.72.0            </span>
<span class="co">#&gt;  [51] deldir_1.0-6                sparseMatrixStats_1.6.0    </span>
<span class="co">#&gt;  [53] MatrixGenerics_1.6.0        vctrs_0.3.8                </span>
<span class="co">#&gt;  [55] SingleCellExperiment_1.16.0 Biobase_2.54.0             </span>
<span class="co">#&gt;  [57] ROCR_1.0-11                 abind_1.4-5                </span>
<span class="co">#&gt;  [59] cachem_1.0.6                withr_2.4.3                </span>
<span class="co">#&gt;  [61] vroom_1.5.7                 sctransform_0.3.3          </span>
<span class="co">#&gt;  [63] mclust_5.4.9                goftest_1.2-3              </span>
<span class="co">#&gt;  [65] svglite_2.0.0               cluster_2.1.2              </span>
<span class="co">#&gt;  [67] splancs_2.01-42             lazyeval_0.2.2             </span>
<span class="co">#&gt;  [69] crayon_1.4.2                geoR_1.8-1                 </span>
<span class="co">#&gt;  [71] pkgconfig_2.0.3             labeling_0.4.2             </span>
<span class="co">#&gt;  [73] GenomeInfoDb_1.30.1         nlme_3.1-153               </span>
<span class="co">#&gt;  [75] rlang_1.0.0                 globals_0.14.0             </span>
<span class="co">#&gt;  [77] lifecycle_1.0.1             miniUI_0.1.1.1             </span>
<span class="co">#&gt;  [79] rsvd_1.0.5                  tcltk_4.1.2                </span>
<span class="co">#&gt;  [81] rprojroot_2.0.2             polyclip_1.10-0            </span>
<span class="co">#&gt;  [83] GSVA_1.42.0                 matrixStats_0.61.0         </span>
<span class="co">#&gt;  [85] lmtest_0.9-39               graph_1.72.0               </span>
<span class="co">#&gt;  [87] Matrix_1.3-4                raster_3.5-15              </span>
<span class="co">#&gt;  [89] carData_3.0-5               Rhdf5lib_1.16.0            </span>
<span class="co">#&gt;  [91] zoo_1.8-9                   ggridges_0.5.3             </span>
<span class="co">#&gt;  [93] png_0.1-7                   viridisLite_0.4.0          </span>
<span class="co">#&gt;  [95] bitops_1.0-7                KernSmooth_2.23-20         </span>
<span class="co">#&gt;  [97] rhdf5filters_1.6.0          Biostrings_2.62.0          </span>
<span class="co">#&gt;  [99] blob_1.2.2                  DelayedMatrixStats_1.16.0  </span>
<span class="co">#&gt; [101] stringr_1.4.0               parallelly_1.30.0          </span>
<span class="co">#&gt; [103] iotools_0.3-2               readr_2.1.2                </span>
<span class="co">#&gt; [105] rstatix_0.7.0               S4Vectors_0.32.3           </span>
<span class="co">#&gt; [107] ggsignif_0.6.3              beachmat_2.10.0            </span>
<span class="co">#&gt; [109] scales_1.1.1                memoise_2.0.1              </span>
<span class="co">#&gt; [111] GSEABase_1.56.0             plyr_1.8.6                 </span>
<span class="co">#&gt; [113] ica_1.0-2                   zlibbioc_1.40.0            </span>
<span class="co">#&gt; [115] compiler_4.1.2              kableExtra_1.3.4           </span>
<span class="co">#&gt; [117] concaveman_1.1.0            RColorBrewer_1.1-2         </span>
<span class="co">#&gt; [119] fitdistrplus_1.1-6          snakecase_0.11.0           </span>
<span class="co">#&gt; [121] cli_3.1.1                   XVector_0.34.0             </span>
<span class="co">#&gt; [123] listenv_0.8.0               patchwork_1.1.1            </span>
<span class="co">#&gt; [125] pbapply_1.5-0               MASS_7.3-54                </span>
<span class="co">#&gt; [127] mgcv_1.8-38                 tidyselect_1.1.1           </span>
<span class="co">#&gt; [129] stringi_1.7.6               textshaping_0.3.6          </span>
<span class="co">#&gt; [131] highr_0.9                   yaml_2.2.2                 </span>
<span class="co">#&gt; [133] BiocSingular_1.10.0         RandomFields_3.3.14        </span>
<span class="co">#&gt; [135] ggrepel_0.9.1               grid_4.1.2                 </span>
<span class="co">#&gt; [137] sass_0.4.0                  tools_4.1.2                </span>
<span class="co">#&gt; [139] future.apply_1.8.1          parallel_4.1.2             </span>
<span class="co">#&gt; [141] rstudioapi_0.13             khroma_1.8.0               </span>
<span class="co">#&gt; [143] janitor_2.1.0               gridExtra_2.3              </span>
<span class="co">#&gt; [145] farver_2.1.0                Rtsne_0.15                 </span>
<span class="co">#&gt; [147] digest_0.6.29               rgeos_0.5-9                </span>
<span class="co">#&gt; [149] shiny_1.7.1                 pracma_2.3.6               </span>
<span class="co">#&gt; [151] quadprog_1.5-8              Rcpp_1.0.8                 </span>
<span class="co">#&gt; [153] GenomicRanges_1.46.1        car_3.0-12                 </span>
<span class="co">#&gt; [155] broom_0.7.12                later_1.3.0                </span>
<span class="co">#&gt; [157] RcppAnnoy_0.0.19            httr_1.4.2                 </span>
<span class="co">#&gt; [159] AnnotationDbi_1.56.2        colorspace_2.0-2           </span>
<span class="co">#&gt; [161] rvest_1.0.2                 XML_3.99-0.8               </span>
<span class="co">#&gt; [163] fs_1.5.2                    tensor_1.5                 </span>
<span class="co">#&gt; [165] reticulate_1.24             IRanges_2.28.0             </span>
<span class="co">#&gt; [167] splines_4.1.2               ggpolypath_0.1.0           </span>
<span class="co">#&gt; [169] uwot_0.1.11                 spatstat.utils_2.3-0       </span>
<span class="co">#&gt; [171] pkgdown_2.0.2               sp_1.4-6                   </span>
<span class="co">#&gt; [173] plotly_4.10.0               systemfonts_1.0.4          </span>
<span class="co">#&gt; [175] xtable_1.8-4                jsonlite_1.7.3             </span>
<span class="co">#&gt; [177] dynamicTreeCut_1.63-1       R6_2.5.1                   </span>
<span class="co">#&gt; [179] pillar_1.7.0                htmltools_0.5.2            </span>
<span class="co">#&gt; [181] mime_0.12                   glue_1.6.1                 </span>
<span class="co">#&gt; [183] fastmap_1.1.0               BiocParallel_1.28.3        </span>
<span class="co">#&gt; [185] codetools_0.2-18            utf8_1.2.2                 </span>
<span class="co">#&gt; [187] lattice_0.20-45             bslib_0.3.1                </span>
<span class="co">#&gt; [189] spatstat.sparse_2.1-0       tibble_3.1.6               </span>
<span class="co">#&gt; [191] curl_4.3.2                  RandomFieldsUtils_1.1.0    </span>
<span class="co">#&gt; [193] leiden_0.3.9                survival_3.2-13            </span>
<span class="co">#&gt; [195] rmarkdown_2.11              desc_1.4.0                 </span>
<span class="co">#&gt; [197] munsell_0.5.0               rhdf5_2.38.0               </span>
<span class="co">#&gt; [199] GenomeInfoDbData_1.2.7      HDF5Array_1.22.1           </span>
<span class="co">#&gt; [201] reshape2_1.4.4              gtable_0.3.0               </span>
<span class="co">#&gt; [203] spatstat.core_2.3-2         Seurat_4.1.0</span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Oscar Ospina, Brooke Fridley, Alex Soupir.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
