<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>spatialGE: Visualization and spatial heterogeneity analysis of melanoma • spatialGE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="spatialGE: Visualization and spatial heterogeneity analysis of melanoma">
<meta property="og:description" content="spatialGE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/melanoma_ST_vignette.html">spatialGE: Visualization and spatial heterogeneity analysis of melanoma</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="melanoma_ST_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="melanoma_ST_vignette_files/kePrint-0.0.1/kePrint.js"></script><link href="melanoma_ST_vignette_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<script src="melanoma_ST_vignette_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="melanoma_ST_vignette_files/plotly-binding-4.9.4.1/plotly.js"></script><script src="melanoma_ST_vignette_files/typedarray-0.1/typedarray.min.js"></script><script src="melanoma_ST_vignette_files/jquery-3.5.1/jquery.min.js"></script><link href="melanoma_ST_vignette_files/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="melanoma_ST_vignette_files/crosstalk-1.1.1/js/crosstalk.min.js"></script><link href="melanoma_ST_vignette_files/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="melanoma_ST_vignette_files/plotly-main-1.57.1/plotly-latest.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>spatialGE: Visualization and spatial heterogeneity analysis of melanoma</h1>
            
      
      
      <div class="hidden name"><code>melanoma_ST_vignette.Rmd</code></div>

    </div>

    
    
<p>The package <code>spatialGE</code> provides a set of tools for the visualization of gene expression from spatially-resolved arrays, such as those generated by the 10X Visium platform.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>The <code>spatialGE</code> repository is available at GitHub and can be installed via <code>devtools</code>. To install <code>devtools</code>, in case is not installed in your R console, please run the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="st">"devtools"</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>After making sure <code>devtools</code> is installed, proceed to install <code>spatialGE</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># devtools::install_github("fridleylab/spatialGE")</span></code></pre></div>
<p>To use <code>spatialGE</code>, load the package using the command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">spatialGE</span><span class="op">)</span></code></pre></div>
</div>
<div id="spatially-resolved-expression-of-melanoma-stage-iii-tumor-biopsies" class="section level2">
<h2 class="hasAnchor">
<a href="#spatially-resolved-expression-of-melanoma-stage-iii-tumor-biopsies" class="anchor"></a>Spatially-resolved expression of melanoma stage III tumor biopsies</h2>
<p>To show the utility of <code>spatialGE</code>, we use the Spatial Transcriptomics data set generated by Thrane et al. (2018). This data set includes lymph node biopsies from four stage IIIc melanoma patients, with two spatial arrays per biopsy (<a href="https:doi.org/10.1158/0008-5472.CAN-18-0747" class="uri">https:doi.org/10.1158/0008-5472.CAN-18-0747</a>). The spatial arrays comprised 1,007 ‘spots’ with 200μM of distance between their centers. The spots had a diameter of 100μM, covering 5-40 cells according to the authors.</p>
<p>In the version of this data set provided along <code>spatialGE</code>, we have removed probe names and duplicate gene names (resulting from different probes). The user can find the original raw data at the authors website (<a href="https://www.spatialresearch.org/resources-published-datasets/doi-10-1158-0008-5472-can-18-0747/" class="uri">https://www.spatialresearch.org/resources-published-datasets/doi-10-1158-0008-5472-can-18-0747/</a>).</p>
</div>
<div id="creating-an-stlist-spatial-transcriptomics-list" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-an-stlist-spatial-transcriptomics-list" class="anchor"></a>Creating an STList (Spatial Transcriptomics List)</h2>
<p>Raw and processed data are stored in an STList (an R S4 class object). The STList can be created with the function <code><a href="../reference/STList.html">STList()</a></code>, which can take minimally inputs in different ways. Minimally, raw RNA counts and spot coordinates must be provided. For this example, we provide to the STList function the following inputs:</p>
<ol style="list-style-type: decimal">
<li><p>Comma- or tab-deimited files with raw RNA counts, one per spatial array. The raw count files must have gene names in the first column, and spots in subsequent columns. Duplicated gene names are not allowed.</p></li>
<li><p>Comma- or tab-deimited files with y and x coordinates for each spot. The first column of this file contains spot IDs matching column names in the raw count files. The seond and third columns must match the Y and X coordinates of each spot respectively.</p></li>
</ol>
<p>We also input a table with clinical data associated with each spatial array. This table is optional, and must be a comma- or tab-delimited file with IDs for each spatial array in the first column, and relevant variables in subsequent columns. The array IDs should match the file paths of the RNA count and coordinate files (for example, the sample ID “ST_mel4_rep2” matches the RNA count file “./inst/extdata/ST_mel4_rep2_counts_dedup_genes.tsv” and coordinate file “./inst/extdata/ST_mel4_rep2_counts_mapping.txt”.</p>
<p>The example files in this tutorial can be downloaded from the GitHub repository (<a href="https://github.com/FridleyLab/spatialGEdev/tree/main/inst/extdata" class="uri">https://github.com/FridleyLab/spatialGEdev/tree/main/inst/extdata</a>). They are also stored in your local <code>spatialGE</code> installation:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"spatialGE"</span><span class="op">)</span>
<span class="va">data_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">data_files</span>, full.names<span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<p>We then find the necessary files within the data repository in <code>spatialGE</code>. Those two files will be the input to later create the Spatial Transcriptomics List (STList).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">count_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"genes"</span>, <span class="va">data_files</span>, value<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="va">coord_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"mapping"</span>, <span class="va">data_files</span>, value<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="va">clin_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"clinical"</span>, <span class="va">data_files</span>, value<span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<p>We can load the files into an STList object like so:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STList.html">STList</a></span><span class="op">(</span>rnacounts<span class="op">=</span><span class="va">count_files</span>, spotcoords<span class="op">=</span><span class="va">coord_files</span>, samples<span class="op">=</span><span class="va">clin_file</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; Creating STList...</span></code></pre></div>
<ul>
<li>Note: If you receive a warning indicating that an input file has a incomplete final line, you can ingnore it. Alternatively, modify the input files with a text editor by adding an empty line at the end of the files.</li>
</ul>
<p>The <code>melanoma</code> object is an STList, and contains the count data, spot coordinates, and clinical data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span>
<span class="co">#&gt; Spatial Transcriptomics List (STList).</span>
<span class="co">#&gt; 8 spatial array(s):</span>
<span class="co">#&gt;  ST_mel1_rep1</span>
<span class="co">#&gt;      ST_mel1_rep2</span>
<span class="co">#&gt;      ST_mel2_rep1</span>
<span class="co">#&gt;      ST_mel2_rep2</span>
<span class="co">#&gt;      ST_mel3_rep1</span>
<span class="co">#&gt;      ST_mel3_rep2</span>
<span class="co">#&gt;      ST_mel4_rep1</span>
<span class="co">#&gt;      ST_mel4_rep2</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 9 variables in sample data:</span>
<span class="co">#&gt;   patient, slice, gender, BRAF_status, NRAS_statuds, CDKN2A_status, survival, survival_months, RIN</span></code></pre></div>
</div>
<div id="exploring-variation-between-spatial-arrays" class="section level2">
<h2 class="hasAnchor">
<a href="#exploring-variation-between-spatial-arrays" class="anchor"></a>Exploring variation between spatial arrays</h2>
<p>The function <code><a href="../reference/bulk_pca.html">bulk_pca()</a></code> allows for a quick snapshot of the variation in gene expression among samples. The function creates ‘bulk’ RNA-Seq data sets by combining all counts from a spatial array. Then, it applies voom transformation to the ‘bulk’ RNA-Seq libraries and performs a PCA. Note that spatial (x,y) coordinate information is not included in this analysis, which is intended only as a tool for data exploration. In this case, we apply the function to look for agreement between samples from the same patient: it is expected that tissue slices from the same patient are more similar between them, than tissue slices from other patients. The <code><a href="../reference/bulk_pca.html">bulk_pca()</a></code> allows to map a clinical variable to the PCA (gender in this example), by including the exact, case-sensitive name of the column from the clinical/sample file.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bulk_pca.html">bulk_pca</a></span><span class="op">(</span><span class="va">melanoma</span>, clinvar<span class="op">=</span><span class="st">'gender'</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: magrittr</span>
<span class="co">#&gt; Loading required package: ggplot2</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/pca_chunk-1.png" width="960"></p>
</div>
<div id="data-normalization-of-spatially-resolved-transcriptomic-data" class="section level2">
<h2 class="hasAnchor">
<a href="#data-normalization-of-spatially-resolved-transcriptomic-data" class="anchor"></a>Data normalization of spatially-resolved transcriptomic data</h2>
<p>Many transformation methods are available for both bulk and single-cell RNA-Seq count data. In <code>spatialGE</code>, we apply voom transformation to RNA-Seq libraries from each of the spots in order to obtain a gaussian-shape normal distribution required by most methods in this package. Many of the methods implemented in <code>spatialGE</code> have been borrowed from the field of geostatistics, which often require this kind of data distribution.</p>
<p>Normalization is achieved by applying the function <code><a href="../reference/voom_norm.html">voom_norm()</a></code> to the <code>melanoma</code> object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/voom_norm.html">voom_norm</a></span><span class="op">(</span><span class="va">melanoma</span><span class="op">)</span></code></pre></div>
</div>
<div id="visualization-of-gene-expression-from-spatially-resolved-transcriptomics-data" class="section level2">
<h2 class="hasAnchor">
<a href="#visualization-of-gene-expression-from-spatially-resolved-transcriptomics-data" class="anchor"></a>Visualization of gene expression from spatially-resolved transcriptomics data</h2>
<p>After data transformation, expression of specific genes can be visualized using ‘quilt’ plots. The function <code><a href="../reference/plot_gene_quilt.html">plot_gene_quilt()</a></code> shows the transformed expression of one or several genes. We have adopted the color palettes from the package <code>khroma</code>. The color palette can be passed using the argument <code>color_pal</code>. Other palettes from the <code>RColorBrewer</code> package can also be specified. The default behavior of the function produces plots for allspatial arrays in the STList, but we can define which arrays will be plotted using the argument <code>plot_who</code>.</p>
<p>Let’s produce a quilt plot for the genes CD74 and SOX10, an immune gene and a melanoma marker.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quilts1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'CD74'</span>, <span class="st">'SOX10'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span></code></pre></div>
<p>Because <code>spatialGE</code> functions output ggplot objects, we can plot the results side-by-side using functions such as <code>ggarrange</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">quilts1</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/unnamed-chunk-1-1.png" width="960"></p>
<p>We can see that gene expression patterns of both genes are different, with a more dispersed pattern in SOX10, compared to CD74. The “pocket” of lower CD74 expression on the left side of the tissue corresponds to the melanoma pocket detected by Thrane et al. (see Figure 3 of their study).</p>
<div id="spatial-interpolation-kriging-of-gene-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#spatial-interpolation-kriging-of-gene-expression" class="anchor"></a>Spatial interpolation (‘kriging’) of gene expression</h3>
<p>We can predict a smooth transcriptomic surface of the gene expression in an spatial array. In <code>spatialGE</code>, this prediction is achieved by a widely used spatial interpolation method in geostatistics. The method known as ‘kriging’ allows for the estimation of the transfromed gene expression values in the unsampled areas between spots, or spots for which a RNA-Seq library failed to provide data. Estimating a transcriptomic surface via kriging assumes that gene expression of two given points is correlated with the physical distance between them.</p>
<p>The function <code><a href="../reference/gene_krige.html">gene_krige()</a></code> performs kriging of gene expression via the <code>geoR</code> package. We specify that kriging will be performed for two os the spatial arrays:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_krige.html">gene_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'CD74'</span>, <span class="st">'SOX10'</span><span class="op">)</span>, who<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, python<span class="op">=</span><span class="cn">F</span><span class="op">)</span></code></pre></div>
<p>The surfaces can be visualized using the <code><a href="../reference/plot_gene_krige.html">plot_gene_krige()</a></code> function:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_krige.html">plot_gene_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'CD74'</span>, <span class="st">'SOX10'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="co">#&gt; Loading required namespace: rgeos</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges1</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotgenekrige_chunk1-1.png" width="960"></p>
<p>By looking at the transcriptomic surfaces of the tho genes, it is easier to detect pockets of higher (or lower) expression within the tissue. It is now more evident that expression of CD74 is lower on the left (melanoma) region of the tumor slice, compared to the rest of the slice.</p>
<p>In the transcriptomic surface plots, <code>spatialGE</code> outputs three spatial heterogeneity statistics: Moran’s I, Geary’s C, and Getis-Ord Gi.</p>
<table class="table table">
<caption>
Heterogeneity statitstics in <code>spatialGE</code>
</caption>
<thead><tr>
<th style="text-align:center;">
Low
</th>
<th style="text-align:center;">
Statistic
</th>
<th style="text-align:center;">
High
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:center;">
Dispersion
</td>
<td style="text-align:center;">
Moran’s I
</td>
<td style="text-align:center;">
Clustering
</td>
</tr>
<tr>
<td style="text-align:center;">
Clustering
</td>
<td style="text-align:center;">
Geary’s C
</td>
<td style="text-align:center;">
Dispersion
</td>
</tr>
<tr>
<td style="text-align:center;">
“Cold spots”
</td>
<td style="text-align:center;">
Getis-Ord Gi
</td>
<td style="text-align:center;">
“Hot spots”
</td>
</tr>
</tbody>
</table>
<p>A higher Moran’s I for CD74 indicates that there are clusters of high expression. In other words, gene expression of CD74 is not uniform across the tissue. Geary’s C was significantly different than zero in CD74, but not SOX10 (indicated by asterisk), further validating the higher gene expression clustering in IGLL5. Statistical significance (<em>p</em>&lt;0.05) in <code>spatialGE</code> is obtained via permutation (performed with <code>spdep</code>).</p>
<p>We can control which tissue slice to plot with the <code>plot_who</code> argument. This argument is also included in the quilt plot functions.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_krige.html">plot_gene_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'CD74'</span>, <span class="st">'SOX10'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">3</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges2</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotgenekrige_chunk2-1.png" width="960"></p>
</div>
</div>
<div id="gene-expression-deconvolution-in-spatial-arrays" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-expression-deconvolution-in-spatial-arrays" class="anchor"></a>Gene expression deconvolution in spatial arrays</h2>
<p>Many gene expression deconvolution methods are available, and their use will depend largely on the question and biological system to be studied. In <code>spatialGE</code>, we have implemented xCell to provide a qualitative snapshot of a tissue’s cell composition. Decovolution in xCell focuses on immunological cell types, which is of crucial importance in the study of the tumor microenvironment.</p>
<p>In addition to xCell, the <code>spatial_deconv</code> function automatically applies ESTIMATE to obtain tumor purity scores. The purity scores are internally clustered using model-based clustering (<code>mclust</code>) to detect tumor/stroma compartments.</p>
<p>Let’s apply xCell to the spatial arrays:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatial_deconv.html">spatial_deconv</a></span><span class="op">(</span><span class="va">melanoma</span>, method<span class="op">=</span><span class="st">'xcell'</span><span class="op">)</span>
<span class="co">#&gt; This will take some time...</span>
<span class="co">#&gt; Loading required package: estimate</span>
<span class="co">#&gt; Loading required package: xCell</span></code></pre></div>
<p>With the function <code><a href="../reference/plot_deconv_quilt.html">plot_deconv_quilt()</a></code>, we can visualize the cell type scores inferred by xCell. For example, B cells and dendritic cells.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quilts2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_quilt.html">plot_deconv_quilt</a></span><span class="op">(</span><span class="va">melanoma</span>, cells<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'b_cells'</span>, <span class="st">'i_dc'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">2</span>, 
                             purity<span class="op">=</span><span class="cn">T</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">quilts2</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">3</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/deconvquilts_chunk-1.png" width="960"></p>
<p>By plotting the tumor/stroma classification of the spots, the differences in cell scores can quiclky be detected. Melanoma areas (squares) show lower B cell and dendritic cell scores, as expected. Nonetheless, B cells were abundant in the lymphoid pocket (upper right part of the tissue).</p>
<p>Similar plots can be generated for gene expression. These plots can also be made interactive by making <code>inter=T</code>, which converts plots to Plotly objects.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">quilts3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">melanoma</span>, genes<span class="op">=</span><span class="st">'CD74'</span>, plot_who<span class="op">=</span><span class="fl">2</span>, purity<span class="op">=</span><span class="cn">T</span>, visium<span class="op">=</span><span class="cn">F</span>,
                           inter<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="va">quilts3</span><span class="op">$</span><span class="va">CD74_2</span></code></pre></div>
<div id="htmlwidget-474b1f4b896660068c83" style="width:960px;height:672px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-474b1f4b896660068c83">{"x":{"data":[{"x":[3,4,4,4,4,5,5,5,5,6,9,10,11,11,11,12,12,12,13,13,13,13,13,13,14,14,14,14,14,14,15,15,15,15,15,15,16,16,16,16,16,16,16,16,16,16,16,16,16,16,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,19,19,19,19,19,19,19,19,19,19,19,19,20,20,20,20,20,20,20,20,20,20,20,21,21,21,21,21,22,22,22,22,23,23,23,23,23,23,24,24,24,24,24,24,24,25,25,25,25,25,25,26,26,26,27,27,27,27,28,28,28],"y":[16,16,17,18,19,15,18,19,20,15,25,25,13,25,26,11,25,26,8,9,23,24,25,26,22,23,24,25,26,27,22,23,24,25,26,27,7,8,10,11,17,18,19,20,21,22,23,24,25,26,7,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,7,8,9,10,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,16,17,18,19,20,21,22,23,24,25,26,27,17,18,19,20,21,22,23,24,25,26,27,17,18,19,21,22,18,19,20,21,18,19,20,21,22,26,18,19,20,21,22,23,24,19,20,21,22,23,24,19,20,23,19,20,21,22,19,20,21],"text":["X:  3<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 11.782639","X:  4<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 11.243923","X:  4<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 11.744181","X:  4<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 11.702237","X:  4<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 11.054270","X:  5<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 10.355004","X:  5<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 10.150448","X:  5<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 11.286144","X:  5<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 10.525833","X:  6<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 10.454319","X:  9<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 11.963775","X: 10<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 11.304576","X: 11<br />Y: 13<br />Tumor_Stroma: stroma<br />Value: 10.098113","X: 11<br />Y: 25<br />Tumor_Stroma: stroma<br />Value:  9.466686","X: 11<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 10.690319","X: 12<br />Y: 11<br />Tumor_Stroma: stroma<br />Value: 12.151743","X: 12<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 10.083743","X: 12<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 10.072141","X: 13<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 10.856972","X: 13<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 11.101420","X: 13<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 10.372201","X: 13<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 10.462854","X: 13<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 11.280931","X: 13<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 11.206467","X: 14<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 11.406783","X: 14<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 11.124798","X: 14<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 10.609352","X: 14<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 11.804119","X: 14<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 11.564892","X: 14<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 11.724157","X: 15<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 11.807728","X: 15<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 11.029596","X: 15<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 10.906136","X: 15<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 11.049858","X: 15<br />Y: 26<br />Tumor_Stroma: stroma<br />Value:  8.922529","X: 15<br />Y: 27<br />Tumor_Stroma: stroma<br />Value:  8.562773","X: 16<br />Y:  7<br />Tumor_Stroma: stroma<br />Value:  9.918810","X: 16<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 11.391982","X: 16<br />Y: 10<br />Tumor_Stroma: stroma<br />Value: 12.176373","X: 16<br />Y: 11<br />Tumor_Stroma: stroma<br />Value: 10.570746","X: 16<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 11.617542","X: 16<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 11.720194","X: 16<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 12.132051","X: 16<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 10.581763","X: 16<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 10.926710","X: 16<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 11.311198","X: 16<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 11.589545","X: 16<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 11.613494","X: 16<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 11.693559","X: 16<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 11.613707","X: 17<br />Y:  7<br />Tumor_Stroma: stroma<br />Value:  9.762714","X: 17<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 12.782051","X: 17<br />Y: 10<br />Tumor_Stroma: stroma<br />Value: 11.570455","X: 17<br />Y: 11<br />Tumor_Stroma: stroma<br />Value: 10.780059","X: 17<br />Y: 12<br />Tumor_Stroma: stroma<br />Value: 11.572501","X: 17<br />Y: 13<br />Tumor_Stroma: stroma<br />Value: 11.461006","X: 17<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 11.953901","X: 17<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 11.828707","X: 17<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 11.197396","X: 17<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 11.200372","X: 17<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 11.221881","X: 17<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 10.625926","X: 17<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 10.094217","X: 17<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 11.114456","X: 17<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 11.199750","X: 17<br />Y: 23<br />Tumor_Stroma: stroma<br />Value:  9.863957","X: 17<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 10.929935","X: 17<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 11.724311","X: 17<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 11.561007","X: 17<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 11.460113","X: 18<br />Y:  7<br />Tumor_Stroma: stroma<br />Value: 10.835145","X: 18<br />Y:  8<br />Tumor_Stroma: stroma<br />Value: 11.101946","X: 18<br />Y:  9<br />Tumor_Stroma: stroma<br />Value: 11.239630","X: 18<br />Y: 10<br />Tumor_Stroma: stroma<br />Value: 11.256187","X: 18<br />Y: 13<br />Tumor_Stroma: stroma<br />Value: 10.541242","X: 18<br />Y: 14<br />Tumor_Stroma: stroma<br />Value: 10.706874","X: 18<br />Y: 15<br />Tumor_Stroma: stroma<br />Value: 11.496373","X: 18<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 11.256180","X: 18<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 10.870392","X: 18<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 10.370364","X: 18<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 10.996976","X: 18<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 11.026583","X: 18<br />Y: 21<br />Tumor_Stroma: stroma<br />Value:  8.996592","X: 18<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 12.291952","X: 18<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 11.298657","X: 18<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 11.275584","X: 18<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 11.303198","X: 18<br />Y: 26<br />Tumor_Stroma: stroma<br />Value:  9.691568","X: 18<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 11.974014","X: 19<br />Y: 16<br />Tumor_Stroma: stroma<br />Value: 11.363908","X: 19<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 10.220747","X: 19<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 11.783176","X: 19<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 11.891993","X: 19<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 11.795930","X: 19<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 10.327839","X: 19<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 10.757415","X: 19<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 12.066550","X: 19<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 10.018268","X: 19<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 11.259976","X: 19<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 10.498933","X: 19<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 10.556583","X: 20<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 11.298243","X: 20<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 11.249691","X: 20<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 11.455512","X: 20<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 10.887857","X: 20<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 10.148382","X: 20<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 11.721331","X: 20<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 12.036231","X: 20<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 11.781614","X: 20<br />Y: 25<br />Tumor_Stroma: stroma<br />Value: 11.721917","X: 20<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 10.765497","X: 20<br />Y: 27<br />Tumor_Stroma: stroma<br />Value: 11.190958","X: 21<br />Y: 17<br />Tumor_Stroma: stroma<br />Value: 11.177853","X: 21<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 10.281581","X: 21<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 18.875968","X: 21<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 11.875696","X: 21<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 11.763295","X: 22<br />Y: 18<br />Tumor_Stroma: stroma<br />Value:  8.146981","X: 22<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 10.277991","X: 22<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 11.806304","X: 22<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 12.209434","X: 23<br />Y: 18<br />Tumor_Stroma: stroma<br />Value: 11.149049","X: 23<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 11.135572","X: 23<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 10.538775","X: 23<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 12.343518","X: 23<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 12.130171","X: 23<br />Y: 26<br />Tumor_Stroma: stroma<br />Value: 12.190167","X: 24<br />Y: 18<br />Tumor_Stroma: stroma<br />Value:  9.098624","X: 24<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 11.186793","X: 24<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 11.376119","X: 24<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 12.375081","X: 24<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 10.960292","X: 24<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 11.644352","X: 24<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 11.490847","X: 25<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 10.740638","X: 25<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 12.267837","X: 25<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 11.973487","X: 25<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 12.698792","X: 25<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 12.021718","X: 25<br />Y: 24<br />Tumor_Stroma: stroma<br />Value: 11.862117","X: 26<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 11.004240","X: 26<br />Y: 20<br />Tumor_Stroma: stroma<br />Value:  9.659446","X: 26<br />Y: 23<br />Tumor_Stroma: stroma<br />Value: 12.427180","X: 27<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 12.504720","X: 27<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 10.883269","X: 27<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 12.038403","X: 27<br />Y: 22<br />Tumor_Stroma: stroma<br />Value: 13.431199","X: 28<br />Y: 19<br />Tumor_Stroma: stroma<br />Value: 11.272739","X: 28<br />Y: 20<br />Tumor_Stroma: stroma<br />Value: 11.556422","X: 28<br />Y: 21<br />Tumor_Stroma: stroma<br />Value: 12.008087"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":["rgba(252,159,49,1)","rgba(253,170,66,1)","rgba(252,160,50,1)","rgba(252,160,52,1)","rgba(254,174,72,1)","rgba(255,189,93,1)","rgba(255,193,99,1)","rgba(253,169,65,1)","rgba(255,186,88,1)","rgba(255,187,90,1)","rgba(251,155,42,1)","rgba(253,169,65,1)","rgba(255,195,101,1)","rgba(255,208,119,1)","rgba(255,182,83,1)","rgba(249,151,39,1)","rgba(255,195,101,1)","rgba(255,195,101,1)","rgba(254,178,78,1)","rgba(254,173,71,1)","rgba(255,189,93,1)","rgba(255,187,90,1)","rgba(253,169,65,1)","rgba(254,171,68,1)","rgba(253,167,61,1)","rgba(254,173,70,1)","rgba(255,184,86,1)","rgba(252,158,48,1)","rgba(252,163,56,1)","rgba(252,160,51,1)","rgba(252,158,48,1)","rgba(254,175,73,1)","rgba(254,177,77,1)","rgba(254,174,72,1)","rgba(255,219,135,1)","rgba(254,227,145,1)","rgba(255,198,106,1)","rgba(253,167,62,1)","rgba(249,150,39,1)","rgba(255,185,87,1)","rgba(252,162,54,1)","rgba(252,160,51,1)","rgba(249,151,40,1)","rgba(255,184,87,1)","rgba(254,177,76,1)","rgba(253,169,64,1)","rgba(252,163,55,1)","rgba(252,162,55,1)","rgba(252,161,52,1)","rgba(252,162,55,1)","rgba(255,202,110,1)","rgba(241,137,32,1)","rgba(252,163,56,1)","rgba(254,180,81,1)","rgba(252,163,56,1)","rgba(253,166,60,1)","rgba(251,155,43,1)","rgba(252,158,47,1)","rgba(254,171,68,1)","rgba(254,171,68,1)","rgba(253,171,67,1)","rgba(255,183,85,1)","rgba(255,195,101,1)","rgba(254,173,71,1)","rgba(254,171,68,1)","rgba(255,200,107,1)","rgba(254,177,76,1)","rgba(252,160,51,1)","rgba(253,163,56,1)","rgba(253,166,60,1)","rgba(254,179,79,1)","rgba(254,173,71,1)","rgba(253,170,67,1)","rgba(253,170,66,1)","rgba(255,185,88,1)","rgba(255,182,83,1)","rgba(253,165,58,1)","rgba(253,170,66,1)","rgba(254,178,78,1)","rgba(255,189,93,1)","rgba(254,176,74,1)","rgba(254,175,73,1)","rgba(255,218,132,1)","rgba(247,148,38,1)","rgba(253,169,65,1)","rgba(253,170,65,1)","rgba(253,169,65,1)","rgba(255,203,112,1)","rgba(251,155,42,1)","rgba(253,168,63,1)","rgba(255,192,97,1)","rgba(252,159,49,1)","rgba(251,156,45,1)","rgba(252,158,48,1)","rgba(255,190,94,1)","rgba(254,181,81,1)","rgba(250,153,40,1)","rgba(255,196,103,1)","rgba(253,170,66,1)","rgba(255,186,89,1)","rgba(255,185,87,1)","rgba(253,169,65,1)","rgba(253,170,66,1)","rgba(253,166,60,1)","rgba(254,178,77,1)","rgba(255,194,99,1)","rgba(252,160,51,1)","rgba(251,153,41,1)","rgba(252,159,49,1)","rgba(252,160,51,1)","rgba(254,180,81,1)","rgba(254,171,68,1)","rgba(254,172,69,1)","rgba(255,191,95,1)","rgba(102,37,6,1)","rgba(251,157,46,1)","rgba(252,159,49,1)","rgba(255,230,155,1)","rgba(255,191,95,1)","rgba(252,158,48,1)","rgba(248,149,39,1)","rgba(254,172,69,1)","rgba(254,173,70,1)","rgba(255,185,88,1)","rgba(246,146,37,1)","rgba(249,151,40,1)","rgba(248,150,39,1)","rgba(255,216,130,1)","rgba(254,171,68,1)","rgba(253,167,62,1)","rgba(246,146,37,1)","rgba(254,176,75,1)","rgba(252,162,54,1)","rgba(253,165,59,1)","rgba(255,181,82,1)","rgba(247,148,38,1)","rgba(251,155,42,1)","rgba(242,139,33,1)","rgba(251,153,41,1)","rgba(251,157,46,1)","rgba(254,175,74,1)","rgba(255,204,113,1)","rgba(245,145,36,1)","rgba(244,143,35,1)","rgba(254,178,78,1)","rgba(250,153,41,1)","rgba(232,122,25,1)","rgba(253,170,66,1)","rgba(253,164,56,1)","rgba(251,154,41,1)"],"opacity":1,"size":1.88976377952756,"symbol":"cross-thin-open","line":{"width":3.77952755905512,"color":["rgba(252,159,49,1)","rgba(253,170,66,1)","rgba(252,160,50,1)","rgba(252,160,52,1)","rgba(254,174,72,1)","rgba(255,189,93,1)","rgba(255,193,99,1)","rgba(253,169,65,1)","rgba(255,186,88,1)","rgba(255,187,90,1)","rgba(251,155,42,1)","rgba(253,169,65,1)","rgba(255,195,101,1)","rgba(255,208,119,1)","rgba(255,182,83,1)","rgba(249,151,39,1)","rgba(255,195,101,1)","rgba(255,195,101,1)","rgba(254,178,78,1)","rgba(254,173,71,1)","rgba(255,189,93,1)","rgba(255,187,90,1)","rgba(253,169,65,1)","rgba(254,171,68,1)","rgba(253,167,61,1)","rgba(254,173,70,1)","rgba(255,184,86,1)","rgba(252,158,48,1)","rgba(252,163,56,1)","rgba(252,160,51,1)","rgba(252,158,48,1)","rgba(254,175,73,1)","rgba(254,177,77,1)","rgba(254,174,72,1)","rgba(255,219,135,1)","rgba(254,227,145,1)","rgba(255,198,106,1)","rgba(253,167,62,1)","rgba(249,150,39,1)","rgba(255,185,87,1)","rgba(252,162,54,1)","rgba(252,160,51,1)","rgba(249,151,40,1)","rgba(255,184,87,1)","rgba(254,177,76,1)","rgba(253,169,64,1)","rgba(252,163,55,1)","rgba(252,162,55,1)","rgba(252,161,52,1)","rgba(252,162,55,1)","rgba(255,202,110,1)","rgba(241,137,32,1)","rgba(252,163,56,1)","rgba(254,180,81,1)","rgba(252,163,56,1)","rgba(253,166,60,1)","rgba(251,155,43,1)","rgba(252,158,47,1)","rgba(254,171,68,1)","rgba(254,171,68,1)","rgba(253,171,67,1)","rgba(255,183,85,1)","rgba(255,195,101,1)","rgba(254,173,71,1)","rgba(254,171,68,1)","rgba(255,200,107,1)","rgba(254,177,76,1)","rgba(252,160,51,1)","rgba(253,163,56,1)","rgba(253,166,60,1)","rgba(254,179,79,1)","rgba(254,173,71,1)","rgba(253,170,67,1)","rgba(253,170,66,1)","rgba(255,185,88,1)","rgba(255,182,83,1)","rgba(253,165,58,1)","rgba(253,170,66,1)","rgba(254,178,78,1)","rgba(255,189,93,1)","rgba(254,176,74,1)","rgba(254,175,73,1)","rgba(255,218,132,1)","rgba(247,148,38,1)","rgba(253,169,65,1)","rgba(253,170,65,1)","rgba(253,169,65,1)","rgba(255,203,112,1)","rgba(251,155,42,1)","rgba(253,168,63,1)","rgba(255,192,97,1)","rgba(252,159,49,1)","rgba(251,156,45,1)","rgba(252,158,48,1)","rgba(255,190,94,1)","rgba(254,181,81,1)","rgba(250,153,40,1)","rgba(255,196,103,1)","rgba(253,170,66,1)","rgba(255,186,89,1)","rgba(255,185,87,1)","rgba(253,169,65,1)","rgba(253,170,66,1)","rgba(253,166,60,1)","rgba(254,178,77,1)","rgba(255,194,99,1)","rgba(252,160,51,1)","rgba(251,153,41,1)","rgba(252,159,49,1)","rgba(252,160,51,1)","rgba(254,180,81,1)","rgba(254,171,68,1)","rgba(254,172,69,1)","rgba(255,191,95,1)","rgba(102,37,6,1)","rgba(251,157,46,1)","rgba(252,159,49,1)","rgba(255,230,155,1)","rgba(255,191,95,1)","rgba(252,158,48,1)","rgba(248,149,39,1)","rgba(254,172,69,1)","rgba(254,173,70,1)","rgba(255,185,88,1)","rgba(246,146,37,1)","rgba(249,151,40,1)","rgba(248,150,39,1)","rgba(255,216,130,1)","rgba(254,171,68,1)","rgba(253,167,62,1)","rgba(246,146,37,1)","rgba(254,176,75,1)","rgba(252,162,54,1)","rgba(253,165,59,1)","rgba(255,181,82,1)","rgba(247,148,38,1)","rgba(251,155,42,1)","rgba(242,139,33,1)","rgba(251,153,41,1)","rgba(251,157,46,1)","rgba(254,175,74,1)","rgba(255,204,113,1)","rgba(245,145,36,1)","rgba(244,143,35,1)","rgba(254,178,78,1)","rgba(250,153,41,1)","rgba(232,122,25,1)","rgba(253,170,66,1)","rgba(253,164,56,1)","rgba(251,154,41,1)"]}},"hoveron":"points","name":"stroma","legendgroup":"stroma","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[5,5,6,6,6,6,6,6,6,7,7,7,7,7,7,7,7,7,7,8,8,8,8,8,8,8,8,8,8,8,9,9,9,9,9,9,9,9,9,9,9,10,10,10,10,10,10,10,10,10,10,10,11,11,11,11,11,11,11,11,11,11,11,12,12,12,12,12,12,12,12,12,12,12,12,12,13,13,13,13,13,13,13,13,13,13,13,13,13,14,14,14,14,14,14,14,14,14,14,14,14,14,15,15,15,15,15,15,15,15,15,15,15,15,15,15,15,16,16,16,16,16,16,17,19,21,21,21,21,21,21,22,22,22,22,22,22,23,23,23,24,26,26],"y":[16,17,16,17,18,19,20,21,22,14,15,16,17,18,19,20,21,22,23,14,15,16,17,18,19,20,21,22,23,24,14,15,16,17,18,19,20,21,22,23,24,14,15,16,17,18,19,20,21,22,23,24,14,15,16,17,18,19,20,21,22,23,24,12,13,14,15,16,17,18,19,20,21,22,23,24,10,11,12,13,14,15,16,17,18,19,20,21,22,8,9,11,12,13,14,15,16,17,18,19,20,21,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,9,12,13,14,15,16,8,8,20,23,24,25,26,27,22,23,24,25,26,27,23,24,25,25,21,22],"text":["X:  5<br />Y: 16<br />Tumor_Stroma: tumor<br />Value:  5.238202","X:  5<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 10.508769","X:  6<br />Y: 16<br />Tumor_Stroma: tumor<br />Value:  9.591976","X:  6<br />Y: 17<br />Tumor_Stroma: tumor<br />Value:  9.952540","X:  6<br />Y: 18<br />Tumor_Stroma: tumor<br />Value:  8.862237","X:  6<br />Y: 19<br />Tumor_Stroma: tumor<br />Value:  8.454123","X:  6<br />Y: 20<br />Tumor_Stroma: tumor<br />Value:  9.881974","X:  6<br />Y: 21<br />Tumor_Stroma: tumor<br />Value:  9.212455","X:  6<br />Y: 22<br />Tumor_Stroma: tumor<br />Value:  7.814173","X:  7<br />Y: 14<br />Tumor_Stroma: tumor<br />Value:  7.427137","X:  7<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 11.201172","X:  7<br />Y: 16<br />Tumor_Stroma: tumor<br />Value:  9.248973","X:  7<br />Y: 17<br />Tumor_Stroma: tumor<br />Value:  9.292169","X:  7<br />Y: 18<br />Tumor_Stroma: tumor<br />Value:  7.575703","X:  7<br />Y: 19<br />Tumor_Stroma: tumor<br />Value:  5.121320","X:  7<br />Y: 20<br />Tumor_Stroma: tumor<br />Value:  5.304605","X:  7<br />Y: 21<br />Tumor_Stroma: tumor<br />Value:  5.552776","X:  7<br />Y: 22<br />Tumor_Stroma: tumor<br />Value:  7.852201","X:  7<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 10.201815","X:  8<br />Y: 14<br />Tumor_Stroma: tumor<br />Value:  9.856270","X:  8<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 10.403213","X:  8<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 10.980336","X:  8<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 10.524515","X:  8<br />Y: 18<br />Tumor_Stroma: tumor<br />Value:  8.746477","X:  8<br />Y: 19<br />Tumor_Stroma: tumor<br />Value:  5.971442","X:  8<br />Y: 20<br />Tumor_Stroma: tumor<br />Value:  7.983038","X:  8<br />Y: 21<br />Tumor_Stroma: tumor<br />Value:  7.923205","X:  8<br />Y: 22<br />Tumor_Stroma: tumor<br />Value:  8.797972","X:  8<br />Y: 23<br />Tumor_Stroma: tumor<br />Value:  9.311517","X:  8<br />Y: 24<br />Tumor_Stroma: tumor<br />Value:  9.377714","X:  9<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 11.110437","X:  9<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 10.245363","X:  9<br />Y: 16<br />Tumor_Stroma: tumor<br />Value:  6.902274","X:  9<br />Y: 17<br />Tumor_Stroma: tumor<br />Value:  8.607317","X:  9<br />Y: 18<br />Tumor_Stroma: tumor<br />Value:  7.817379","X:  9<br />Y: 19<br />Tumor_Stroma: tumor<br />Value:  5.901918","X:  9<br />Y: 20<br />Tumor_Stroma: tumor<br />Value:  8.013068","X:  9<br />Y: 21<br />Tumor_Stroma: tumor<br />Value:  9.012618","X:  9<br />Y: 22<br />Tumor_Stroma: tumor<br />Value:  7.080242","X:  9<br />Y: 23<br />Tumor_Stroma: tumor<br />Value:  8.735501","X:  9<br />Y: 24<br />Tumor_Stroma: tumor<br />Value:  6.903373","X: 10<br />Y: 14<br />Tumor_Stroma: tumor<br />Value:  7.598476","X: 10<br />Y: 15<br />Tumor_Stroma: tumor<br />Value:  6.038771","X: 10<br />Y: 16<br />Tumor_Stroma: tumor<br />Value:  7.996497","X: 10<br />Y: 17<br />Tumor_Stroma: tumor<br />Value:  8.376327","X: 10<br />Y: 18<br />Tumor_Stroma: tumor<br />Value:  8.176844","X: 10<br />Y: 19<br />Tumor_Stroma: tumor<br />Value:  8.099044","X: 10<br />Y: 20<br />Tumor_Stroma: tumor<br />Value:  6.460538","X: 10<br />Y: 21<br />Tumor_Stroma: tumor<br />Value:  8.239476","X: 10<br />Y: 22<br />Tumor_Stroma: tumor<br />Value:  7.994702","X: 10<br />Y: 23<br />Tumor_Stroma: tumor<br />Value:  6.048524","X: 10<br />Y: 24<br />Tumor_Stroma: tumor<br />Value:  8.426927","X: 11<br />Y: 14<br />Tumor_Stroma: tumor<br />Value:  8.148795","X: 11<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 10.018566","X: 11<br />Y: 16<br />Tumor_Stroma: tumor<br />Value:  8.579459","X: 11<br />Y: 17<br />Tumor_Stroma: tumor<br />Value:  8.682035","X: 11<br />Y: 18<br />Tumor_Stroma: tumor<br />Value:  8.408397","X: 11<br />Y: 19<br />Tumor_Stroma: tumor<br />Value:  8.252239","X: 11<br />Y: 20<br />Tumor_Stroma: tumor<br />Value:  9.343143","X: 11<br />Y: 21<br />Tumor_Stroma: tumor<br />Value:  9.713450","X: 11<br />Y: 22<br />Tumor_Stroma: tumor<br />Value:  5.925127","X: 11<br />Y: 23<br />Tumor_Stroma: tumor<br />Value:  5.140845","X: 11<br />Y: 24<br />Tumor_Stroma: tumor<br />Value:  6.480180","X: 12<br />Y: 12<br />Tumor_Stroma: tumor<br />Value: 10.287124","X: 12<br />Y: 13<br />Tumor_Stroma: tumor<br />Value:  8.374698","X: 12<br />Y: 14<br />Tumor_Stroma: tumor<br />Value:  9.595113","X: 12<br />Y: 15<br />Tumor_Stroma: tumor<br />Value: 10.151799","X: 12<br />Y: 16<br />Tumor_Stroma: tumor<br />Value:  7.350420","X: 12<br />Y: 17<br />Tumor_Stroma: tumor<br />Value:  8.900944","X: 12<br />Y: 18<br />Tumor_Stroma: tumor<br />Value:  9.256809","X: 12<br />Y: 19<br />Tumor_Stroma: tumor<br />Value:  9.000363","X: 12<br />Y: 20<br />Tumor_Stroma: tumor<br />Value:  7.231244","X: 12<br />Y: 21<br />Tumor_Stroma: tumor<br />Value:  9.163864","X: 12<br />Y: 22<br />Tumor_Stroma: tumor<br />Value:  9.683781","X: 12<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 10.649088","X: 12<br />Y: 24<br />Tumor_Stroma: tumor<br />Value: 11.126505","X: 13<br />Y: 10<br />Tumor_Stroma: tumor<br />Value: 11.583415","X: 13<br />Y: 11<br />Tumor_Stroma: tumor<br />Value: 12.295965","X: 13<br />Y: 12<br />Tumor_Stroma: tumor<br />Value: 10.296769","X: 13<br />Y: 13<br />Tumor_Stroma: tumor<br />Value:  9.190521","X: 13<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 10.006612","X: 13<br />Y: 15<br />Tumor_Stroma: tumor<br />Value:  9.574232","X: 13<br />Y: 16<br />Tumor_Stroma: tumor<br />Value:  9.726519","X: 13<br />Y: 17<br />Tumor_Stroma: tumor<br />Value:  8.137322","X: 13<br />Y: 18<br />Tumor_Stroma: tumor<br />Value:  8.165378","X: 13<br />Y: 19<br />Tumor_Stroma: tumor<br />Value:  8.143065","X: 13<br />Y: 20<br />Tumor_Stroma: tumor<br />Value: 11.837035","X: 13<br />Y: 21<br />Tumor_Stroma: tumor<br />Value:  8.975865","X: 13<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 10.783942","X: 14<br />Y:  8<br />Tumor_Stroma: tumor<br />Value: 12.245473","X: 14<br />Y:  9<br />Tumor_Stroma: tumor<br />Value: 11.696562","X: 14<br />Y: 11<br />Tumor_Stroma: tumor<br />Value: 10.436309","X: 14<br />Y: 12<br />Tumor_Stroma: tumor<br />Value: 10.161365","X: 14<br />Y: 13<br />Tumor_Stroma: tumor<br />Value: 11.063473","X: 14<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 10.836232","X: 14<br />Y: 15<br />Tumor_Stroma: tumor<br />Value:  9.706659","X: 14<br />Y: 16<br />Tumor_Stroma: tumor<br />Value:  8.137277","X: 14<br />Y: 17<br />Tumor_Stroma: tumor<br />Value:  6.939025","X: 14<br />Y: 18<br />Tumor_Stroma: tumor<br />Value:  9.401826","X: 14<br />Y: 19<br />Tumor_Stroma: tumor<br />Value:  9.092942","X: 14<br />Y: 20<br />Tumor_Stroma: tumor<br />Value:  9.947589","X: 14<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 10.742314","X: 15<br />Y:  7<br />Tumor_Stroma: tumor<br />Value: 11.648922","X: 15<br />Y:  8<br />Tumor_Stroma: tumor<br />Value: 11.570958","X: 15<br />Y:  9<br />Tumor_Stroma: tumor<br />Value: 11.720970","X: 15<br />Y: 10<br />Tumor_Stroma: tumor<br />Value: 10.974311","X: 15<br />Y: 11<br />Tumor_Stroma: tumor<br />Value:  8.871065","X: 15<br />Y: 12<br />Tumor_Stroma: tumor<br />Value: 10.601542","X: 15<br />Y: 13<br />Tumor_Stroma: tumor<br />Value: 10.598200","X: 15<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 10.826014","X: 15<br />Y: 15<br />Tumor_Stroma: tumor<br />Value:  9.761089","X: 15<br />Y: 16<br />Tumor_Stroma: tumor<br />Value:  8.432734","X: 15<br />Y: 17<br />Tumor_Stroma: tumor<br />Value: 10.556061","X: 15<br />Y: 18<br />Tumor_Stroma: tumor<br />Value: 12.084102","X: 15<br />Y: 19<br />Tumor_Stroma: tumor<br />Value: 10.601288","X: 15<br />Y: 20<br />Tumor_Stroma: tumor<br />Value: 10.767198","X: 15<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 12.243091","X: 16<br />Y:  9<br />Tumor_Stroma: tumor<br />Value: 11.711978","X: 16<br />Y: 12<br />Tumor_Stroma: tumor<br />Value: 12.423145","X: 16<br />Y: 13<br />Tumor_Stroma: tumor<br />Value: 11.610119","X: 16<br />Y: 14<br />Tumor_Stroma: tumor<br />Value: 11.832388","X: 16<br />Y: 15<br />Tumor_Stroma: tumor<br />Value:  9.448343","X: 16<br />Y: 16<br />Tumor_Stroma: tumor<br />Value: 10.985136","X: 17<br />Y:  8<br />Tumor_Stroma: tumor<br />Value: 12.734377","X: 19<br />Y:  8<br />Tumor_Stroma: tumor<br />Value:  9.882438","X: 21<br />Y: 20<br />Tumor_Stroma: tumor<br />Value: 11.825666","X: 21<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 12.433821","X: 21<br />Y: 24<br />Tumor_Stroma: tumor<br />Value: 12.544547","X: 21<br />Y: 25<br />Tumor_Stroma: tumor<br />Value: 12.850103","X: 21<br />Y: 26<br />Tumor_Stroma: tumor<br />Value: 12.820743","X: 21<br />Y: 27<br />Tumor_Stroma: tumor<br />Value: 12.635909","X: 22<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 11.539051","X: 22<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 12.173422","X: 22<br />Y: 24<br />Tumor_Stroma: tumor<br />Value: 12.386835","X: 22<br />Y: 25<br />Tumor_Stroma: tumor<br />Value: 12.276927","X: 22<br />Y: 26<br />Tumor_Stroma: tumor<br />Value: 12.413212","X: 22<br />Y: 27<br />Tumor_Stroma: tumor<br />Value: 12.271169","X: 23<br />Y: 23<br />Tumor_Stroma: tumor<br />Value: 11.679471","X: 23<br />Y: 24<br />Tumor_Stroma: tumor<br />Value: 11.021364","X: 23<br />Y: 25<br />Tumor_Stroma: tumor<br />Value: 10.910759","X: 24<br />Y: 25<br />Tumor_Stroma: tumor<br />Value: 12.234592","X: 26<br />Y: 21<br />Tumor_Stroma: tumor<br />Value: 12.067849","X: 26<br />Y: 22<br />Tumor_Stroma: tumor<br />Value: 12.970996"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":["rgba(255,254,226,1)","rgba(255,186,89,1)","rgba(255,205,115,1)","rgba(255,198,105,1)","rgba(255,221,136,1)","rgba(254,228,148,1)","rgba(255,199,107,1)","rgba(255,213,126,1)","rgba(255,233,163,1)","rgba(255,236,173,1)","rgba(254,171,68,1)","rgba(255,212,125,1)","rgba(255,212,124,1)","rgba(255,235,169,1)","rgba(255,255,229,1)","rgba(255,253,225,1)","rgba(255,251,218,1)","rgba(255,233,162,1)","rgba(255,192,98,1)","rgba(255,200,108,1)","rgba(255,188,92,1)","rgba(254,176,75,1)","rgba(255,186,88,1)","rgba(254,223,140,1)","rgba(255,248,208,1)","rgba(255,232,159,1)","rgba(255,232,161,1)","rgba(255,222,138,1)","rgba(255,211,123,1)","rgba(255,210,121,1)","rgba(254,173,71,1)","rgba(255,191,96,1)","rgba(255,240,186,1)","rgba(254,226,144,1)","rgba(255,233,163,1)","rgba(255,249,210,1)","rgba(255,231,158,1)","rgba(255,217,132,1)","rgba(255,239,181,1)","rgba(254,223,140,1)","rgba(255,240,185,1)","rgba(255,235,169,1)","rgba(255,247,207,1)","rgba(255,232,159,1)","rgba(254,228,150,1)","rgba(255,230,154,1)","rgba(255,231,156,1)","rgba(255,244,196,1)","rgba(255,230,153,1)","rgba(255,232,159,1)","rgba(255,247,206,1)","rgba(254,228,148,1)","rgba(255,230,155,1)","rgba(255,196,103,1)","rgba(254,227,144,1)","rgba(254,224,141,1)","rgba(254,228,149,1)","rgba(255,229,153,1)","rgba(255,211,122,1)","rgba(255,203,112,1)","rgba(255,248,209,1)","rgba(255,255,229,1)","rgba(255,244,196,1)","rgba(255,191,95,1)","rgba(254,228,150,1)","rgba(255,205,115,1)","rgba(255,193,99,1)","rgba(255,237,175,1)","rgba(255,220,135,1)","rgba(255,212,125,1)","rgba(255,218,132,1)","rgba(255,238,178,1)","rgba(255,214,128,1)","rgba(255,203,113,1)","rgba(255,183,85,1)","rgba(254,173,70,1)","rgba(252,163,56,1)","rgba(247,147,38,1)","rgba(255,190,95,1)","rgba(255,214,127,1)","rgba(255,197,103,1)","rgba(255,206,116,1)","rgba(255,202,111,1)","rgba(255,230,155,1)","rgba(255,230,155,1)","rgba(255,230,155,1)","rgba(252,157,47,1)","rgba(255,218,133,1)","rgba(254,180,81,1)","rgba(248,149,38,1)","rgba(252,161,52,1)","rgba(255,187,91,1)","rgba(255,193,99,1)","rgba(254,174,72,1)","rgba(254,179,79,1)","rgba(255,203,112,1)","rgba(255,230,155,1)","rgba(255,240,185,1)","rgba(255,209,121,1)","rgba(255,216,130,1)","rgba(255,198,105,1)","rgba(255,181,82,1)","rgba(252,162,53,1)","rgba(252,163,56,1)","rgba(252,160,51,1)","rgba(254,176,75,1)","rgba(255,220,136,1)","rgba(255,184,86,1)","rgba(255,184,86,1)","rgba(254,179,79,1)","rgba(255,202,110,1)","rgba(254,228,148,1)","rgba(255,185,87,1)","rgba(250,152,40,1)","rgba(255,184,86,1)","rgba(254,180,81,1)","rgba(248,149,38,1)","rgba(252,160,51,1)","rgba(245,145,36,1)","rgba(252,162,55,1)","rgba(252,158,47,1)","rgba(255,208,119,1)","rgba(254,176,74,1)","rgba(241,138,33,1)","rgba(255,199,107,1)","rgba(252,158,47,1)","rgba(245,144,36,1)","rgba(244,142,35,1)","rgba(240,135,32,1)","rgba(240,136,32,1)","rgba(243,140,34,1)","rgba(253,164,57,1)","rgba(249,150,39,1)","rgba(246,145,37,1)","rgba(247,148,38,1)","rgba(246,145,36,1)","rgba(247,148,38,1)","rgba(252,161,52,1)","rgba(254,175,73,1)","rgba(254,177,77,1)","rgba(248,149,38,1)","rgba(250,152,40,1)","rgba(238,133,30,1)"],"opacity":1,"size":1.88976377952756,"symbol":"square","line":{"width":3.77952755905512,"color":["rgba(255,254,226,1)","rgba(255,186,89,1)","rgba(255,205,115,1)","rgba(255,198,105,1)","rgba(255,221,136,1)","rgba(254,228,148,1)","rgba(255,199,107,1)","rgba(255,213,126,1)","rgba(255,233,163,1)","rgba(255,236,173,1)","rgba(254,171,68,1)","rgba(255,212,125,1)","rgba(255,212,124,1)","rgba(255,235,169,1)","rgba(255,255,229,1)","rgba(255,253,225,1)","rgba(255,251,218,1)","rgba(255,233,162,1)","rgba(255,192,98,1)","rgba(255,200,108,1)","rgba(255,188,92,1)","rgba(254,176,75,1)","rgba(255,186,88,1)","rgba(254,223,140,1)","rgba(255,248,208,1)","rgba(255,232,159,1)","rgba(255,232,161,1)","rgba(255,222,138,1)","rgba(255,211,123,1)","rgba(255,210,121,1)","rgba(254,173,71,1)","rgba(255,191,96,1)","rgba(255,240,186,1)","rgba(254,226,144,1)","rgba(255,233,163,1)","rgba(255,249,210,1)","rgba(255,231,158,1)","rgba(255,217,132,1)","rgba(255,239,181,1)","rgba(254,223,140,1)","rgba(255,240,185,1)","rgba(255,235,169,1)","rgba(255,247,207,1)","rgba(255,232,159,1)","rgba(254,228,150,1)","rgba(255,230,154,1)","rgba(255,231,156,1)","rgba(255,244,196,1)","rgba(255,230,153,1)","rgba(255,232,159,1)","rgba(255,247,206,1)","rgba(254,228,148,1)","rgba(255,230,155,1)","rgba(255,196,103,1)","rgba(254,227,144,1)","rgba(254,224,141,1)","rgba(254,228,149,1)","rgba(255,229,153,1)","rgba(255,211,122,1)","rgba(255,203,112,1)","rgba(255,248,209,1)","rgba(255,255,229,1)","rgba(255,244,196,1)","rgba(255,191,95,1)","rgba(254,228,150,1)","rgba(255,205,115,1)","rgba(255,193,99,1)","rgba(255,237,175,1)","rgba(255,220,135,1)","rgba(255,212,125,1)","rgba(255,218,132,1)","rgba(255,238,178,1)","rgba(255,214,128,1)","rgba(255,203,113,1)","rgba(255,183,85,1)","rgba(254,173,70,1)","rgba(252,163,56,1)","rgba(247,147,38,1)","rgba(255,190,95,1)","rgba(255,214,127,1)","rgba(255,197,103,1)","rgba(255,206,116,1)","rgba(255,202,111,1)","rgba(255,230,155,1)","rgba(255,230,155,1)","rgba(255,230,155,1)","rgba(252,157,47,1)","rgba(255,218,133,1)","rgba(254,180,81,1)","rgba(248,149,38,1)","rgba(252,161,52,1)","rgba(255,187,91,1)","rgba(255,193,99,1)","rgba(254,174,72,1)","rgba(254,179,79,1)","rgba(255,203,112,1)","rgba(255,230,155,1)","rgba(255,240,185,1)","rgba(255,209,121,1)","rgba(255,216,130,1)","rgba(255,198,105,1)","rgba(255,181,82,1)","rgba(252,162,53,1)","rgba(252,163,56,1)","rgba(252,160,51,1)","rgba(254,176,75,1)","rgba(255,220,136,1)","rgba(255,184,86,1)","rgba(255,184,86,1)","rgba(254,179,79,1)","rgba(255,202,110,1)","rgba(254,228,148,1)","rgba(255,185,87,1)","rgba(250,152,40,1)","rgba(255,184,86,1)","rgba(254,180,81,1)","rgba(248,149,38,1)","rgba(252,160,51,1)","rgba(245,145,36,1)","rgba(252,162,55,1)","rgba(252,158,47,1)","rgba(255,208,119,1)","rgba(254,176,74,1)","rgba(241,138,33,1)","rgba(255,199,107,1)","rgba(252,158,47,1)","rgba(245,144,36,1)","rgba(244,142,35,1)","rgba(240,135,32,1)","rgba(240,136,32,1)","rgba(243,140,34,1)","rgba(253,164,57,1)","rgba(249,150,39,1)","rgba(246,145,37,1)","rgba(247,148,38,1)","rgba(246,145,36,1)","rgba(247,148,38,1)","rgba(252,161,52,1)","rgba(254,175,73,1)","rgba(254,177,77,1)","rgba(248,149,38,1)","rgba(250,152,40,1)","rgba(238,133,30,1)"]}},"hoveron":"points","name":"tumor","legendgroup":"tumor","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[10],"y":[10],"name":"99_c4a0d47a654c879c50e37a2b18ccead3","type":"scatter","mode":"markers","opacity":0,"hoverinfo":"skip","showlegend":false,"marker":{"color":[0,1],"colorscale":[[0,"#FFFFE5"],[0.0033444816053512,"#FFFFE4"],[0.00668896321070233,"#FFFEE3"],[0.0100334448160535,"#FFFEE2"],[0.0133779264214047,"#FFFDE0"],[0.0167224080267559,"#FFFDDF"],[0.020066889632107,"#FFFDDE"],[0.0234113712374582,"#FFFCDD"],[0.0267558528428093,"#FFFCDC"],[0.0301003344481605,"#FFFCDB"],[0.0334448160535117,"#FFFBDA"],[0.0367892976588629,"#FFFBD9"],[0.0401337792642141,"#FFFAD7"],[0.0434782608695652,"#FFFAD6"],[0.0468227424749164,"#FFFAD5"],[0.0501672240802675,"#FFF9D4"],[0.0535117056856187,"#FFF9D3"],[0.0568561872909699,"#FFF9D2"],[0.0602006688963211,"#FFF8D1"],[0.0635451505016723,"#FFF8D0"],[0.0668896321070234,"#FFF7CE"],[0.0702341137123746,"#FFF7CD"],[0.0735785953177257,"#FFF7CC"],[0.0769230769230769,"#FFF6CB"],[0.0802675585284281,"#FFF6CA"],[0.0836120401337793,"#FFF5C9"],[0.0869565217391304,"#FFF5C8"],[0.0903010033444816,"#FFF5C7"],[0.0936454849498328,"#FFF4C6"],[0.0969899665551839,"#FFF4C4"],[0.100334448160535,"#FFF4C3"],[0.103678929765886,"#FFF3C2"],[0.107023411371237,"#FFF3C1"],[0.110367892976589,"#FFF2C0"],[0.11371237458194,"#FFF2BF"],[0.117056856187291,"#FFF2BE"],[0.120401337792642,"#FFF1BD"],[0.123745819397993,"#FFF1BB"],[0.127090301003344,"#FFF1BA"],[0.130434782608696,"#FFF0B9"],[0.133779264214047,"#FFF0B8"],[0.137123745819398,"#FFEFB7"],[0.140468227424749,"#FFEFB6"],[0.1438127090301,"#FFEFB5"],[0.147157190635451,"#FFEEB4"],[0.150501672240803,"#FFEEB2"],[0.153846153846154,"#FFEEB1"],[0.157190635451505,"#FFEDB0"],[0.160535117056856,"#FFEDAF"],[0.163879598662207,"#FFEDAE"],[0.167224080267559,"#FFECAD"],[0.17056856187291,"#FFECAC"],[0.173913043478261,"#FFEBAB"],[0.177257525083612,"#FFEBAA"],[0.180602006688963,"#FFEBA8"],[0.183946488294314,"#FFEAA7"],[0.187290969899666,"#FFEAA6"],[0.190635451505017,"#FFEAA5"],[0.193979933110368,"#FFE9A4"],[0.197324414715719,"#FFE9A3"],[0.20066889632107,"#FFE8A2"],[0.204013377926421,"#FFE8A1"],[0.207357859531773,"#FFE89F"],[0.210702341137124,"#FFE79E"],[0.214046822742475,"#FFE79D"],[0.217391304347826,"#FFE79C"],[0.220735785953177,"#FFE69B"],[0.224080267558528,"#FFE69A"],[0.22742474916388,"#FFE599"],[0.230769230769231,"#FEE598"],[0.234113712374582,"#FEE596"],[0.237458193979933,"#FEE495"],[0.240802675585284,"#FEE494"],[0.244147157190635,"#FEE493"],[0.247491638795987,"#FEE392"],[0.250836120401338,"#FEE391"],[0.254180602006689,"#FEE28F"],[0.25752508361204,"#FEE18E"],[0.260869565217391,"#FEE08D"],[0.264214046822742,"#FEDF8B"],[0.267558528428094,"#FFDE8A"],[0.270903010033445,"#FFDD89"],[0.274247491638796,"#FFDC87"],[0.277591973244147,"#FFDB86"],[0.280936454849498,"#FFDA85"],[0.284280936454849,"#FFD983"],[0.287625418060201,"#FFD882"],[0.290969899665552,"#FFD781"],[0.294314381270903,"#FFD67F"],[0.297658862876254,"#FFD57E"],[0.301003344481605,"#FFD47D"],[0.304347826086957,"#FFD37C"],[0.307692307692308,"#FFD27A"],[0.311036789297659,"#FFD179"],[0.31438127090301,"#FFD078"],[0.317725752508361,"#FFCF76"],[0.321070234113712,"#FFCE75"],[0.324414715719063,"#FFCD74"],[0.327759197324415,"#FFCC72"],[0.331103678929766,"#FFCB71"],[0.334448160535117,"#FFCB70"],[0.337792642140468,"#FFCA6E"],[0.341137123745819,"#FFC96D"],[0.344481605351171,"#FFC86C"],[0.347826086956522,"#FFC76A"],[0.351170568561873,"#FFC669"],[0.354515050167224,"#FFC568"],[0.357859531772575,"#FFC466"],[0.361204013377926,"#FFC365"],[0.364548494983278,"#FFC264"],[0.367892976588629,"#FFC162"],[0.37123745819398,"#FFC061"],[0.374581939799331,"#FFBF60"],[0.377926421404682,"#FFBE5E"],[0.381270903010033,"#FFBD5D"],[0.384615384615385,"#FFBC5C"],[0.387959866220736,"#FFBB5A"],[0.391304347826087,"#FFBA59"],[0.394648829431438,"#FFB957"],[0.397993311036789,"#FFB856"],[0.401337792642141,"#FFB755"],[0.404682274247492,"#FFB653"],[0.408026755852843,"#FFB552"],[0.411371237458194,"#FEB451"],[0.414715719063545,"#FEB34F"],[0.418060200668896,"#FEB24E"],[0.421404682274247,"#FEB14D"],[0.424749163879599,"#FEB04B"],[0.42809364548495,"#FEAF4A"],[0.431438127090301,"#FEAE48"],[0.434782608695652,"#FEAD47"],[0.438127090301003,"#FEAC45"],[0.441471571906354,"#FEAB44"],[0.444816053511706,"#FDAA43"],[0.448160535117057,"#FDA941"],[0.451505016722408,"#FDA840"],[0.454849498327759,"#FDA73E"],[0.45819397993311,"#FDA63D"],[0.461538461538462,"#FDA53B"],[0.464882943143813,"#FDA43A"],[0.468227424749164,"#FDA338"],[0.471571906354515,"#FCA237"],[0.474916387959866,"#FCA135"],[0.478260869565217,"#FCA034"],[0.481605351170569,"#FC9F32"],[0.48494983277592,"#FC9E31"],[0.488294314381271,"#FC9D2F"],[0.491638795986622,"#FB9C2D"],[0.494983277591973,"#FB9B2C"],[0.498327759197324,"#FB9A2A"],[0.501672240802675,"#FB9929"],[0.505016722408027,"#FA9828"],[0.508361204013378,"#F99728"],[0.511705685618729,"#F99627"],[0.51505016722408,"#F89527"],[0.518394648829431,"#F89426"],[0.521739130434783,"#F79326"],[0.525083612040134,"#F69225"],[0.528428093645485,"#F69125"],[0.531772575250836,"#F59024"],[0.535117056856187,"#F58F24"],[0.538461538461538,"#F48E23"],[0.54180602006689,"#F38D23"],[0.545150501672241,"#F38C22"],[0.548494983277592,"#F28B22"],[0.551839464882943,"#F28A21"],[0.555183946488294,"#F18921"],[0.558528428093645,"#F08820"],[0.561872909698997,"#F08720"],[0.565217391304348,"#EF861F"],[0.568561872909699,"#EE851F"],[0.57190635451505,"#EE841E"],[0.575250836120401,"#ED831E"],[0.578595317725753,"#ED821D"],[0.581939799331104,"#EC811D"],[0.585284280936455,"#EB801C"],[0.588628762541806,"#EB7F1C"],[0.591973244147157,"#EA7E1B"],[0.595317725752508,"#E97D1B"],[0.598662207357859,"#E97C1A"],[0.602006688963211,"#E87B1A"],[0.605351170568562,"#E87A19"],[0.608695652173913,"#E77919"],[0.612040133779264,"#E67818"],[0.615384615384615,"#E67718"],[0.618729096989967,"#E57617"],[0.622073578595318,"#E47517"],[0.625418060200669,"#E47416"],[0.62876254180602,"#E37315"],[0.632107023411371,"#E37215"],[0.635451505016722,"#E27114"],[0.638795986622073,"#E17014"],[0.642140468227425,"#E16F13"],[0.645484949832776,"#E06E13"],[0.648829431438127,"#DF6D12"],[0.652173913043478,"#DF6C12"],[0.655518394648829,"#DE6A11"],[0.65886287625418,"#DE6911"],[0.662207357859532,"#DD6810"],[0.665551839464883,"#DC6710"],[0.668896321070234,"#DC660F"],[0.672240802675585,"#DB650E"],[0.675585284280936,"#DA640E"],[0.678929765886288,"#DA630D"],[0.682274247491639,"#D9620D"],[0.68561872909699,"#D8610C"],[0.688963210702341,"#D8600C"],[0.692307692307692,"#D75F0B"],[0.695652173913043,"#D65E0A"],[0.698996655518395,"#D65D0A"],[0.702341137123746,"#D55C09"],[0.705685618729097,"#D55B09"],[0.709030100334448,"#D45A08"],[0.712374581939799,"#D35807"],[0.71571906354515,"#D35707"],[0.719063545150502,"#D25606"],[0.722408026755853,"#D15506"],[0.725752508361204,"#D15405"],[0.729096989966555,"#D05305"],[0.732441471571906,"#CF5204"],[0.735785953177257,"#CF5104"],[0.739130434782609,"#CE5003"],[0.74247491638796,"#CD4F03"],[0.745819397993311,"#CD4D03"],[0.749163879598662,"#CC4C02"],[0.752508361204013,"#CB4C02"],[0.755852842809364,"#C94B02"],[0.759197324414716,"#C84A02"],[0.762541806020067,"#C74A03"],[0.765886287625418,"#C54903"],[0.769230769230769,"#C44903"],[0.77257525083612,"#C24803"],[0.775919732441471,"#C14803"],[0.779264214046823,"#BF4703"],[0.782608695652174,"#BE4704"],[0.785953177257525,"#BD4604"],[0.789297658862876,"#BB4604"],[0.792642140468227,"#BA4504"],[0.795986622073579,"#B84504"],[0.79933110367893,"#B74404"],[0.802675585284281,"#B54304"],[0.806020066889632,"#B44305"],[0.809364548494983,"#B34205"],[0.812709030100334,"#B14205"],[0.816053511705686,"#B04105"],[0.819397993311037,"#AE4105"],[0.822742474916388,"#AD4005"],[0.826086956521739,"#AC4005"],[0.82943143812709,"#AA3F05"],[0.832775919732442,"#A93F05"],[0.836120401337793,"#A73E05"],[0.839464882943144,"#A63E06"],[0.842809364548495,"#A53D06"],[0.846153846153846,"#A33D06"],[0.849498327759197,"#A23C06"],[0.852842809364548,"#A13B06"],[0.8561872909699,"#9F3B06"],[0.859531772575251,"#9E3A06"],[0.862876254180602,"#9C3A06"],[0.866220735785953,"#9B3906"],[0.869565217391304,"#9A3906"],[0.872909698996655,"#983806"],[0.876254180602007,"#973806"],[0.879598662207358,"#963706"],[0.882943143812709,"#943707"],[0.88628762541806,"#933607"],[0.889632107023411,"#913607"],[0.892976588628763,"#903507"],[0.896321070234114,"#8F3507"],[0.899665551839465,"#8D3407"],[0.903010033444816,"#8C3407"],[0.906354515050167,"#8B3307"],[0.909698996655518,"#893307"],[0.913043478260869,"#883207"],[0.916387959866221,"#873207"],[0.919732441471572,"#853107"],[0.923076923076923,"#843107"],[0.926421404682274,"#833007"],[0.929765886287625,"#813007"],[0.933110367892977,"#802F07"],[0.936454849498328,"#7F2F07"],[0.939799331103679,"#7D2E07"],[0.94314381270903,"#7C2E07"],[0.946488294314381,"#7B2D07"],[0.949832775919732,"#792C07"],[0.953177257525084,"#782C07"],[0.956521739130435,"#772B07"],[0.959866220735786,"#752B07"],[0.963210702341137,"#742A07"],[0.966555183946488,"#732A07"],[0.969899665551839,"#722907"],[0.97324414715719,"#702907"],[0.976588628762542,"#6F2807"],[0.979933110367893,"#6E2807"],[0.983277591973244,"#6C2707"],[0.986622073578595,"#6B2706"],[0.989966555183946,"#6A2606"],[0.993311036789298,"#692606"],[0.996655518394649,"#672506"],[1,"#662506"]],"colorbar":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"thickness":23.04,"title":"norm_expr","titlefont":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"tickmode":"array","ticktext":["10","15"],"tickvals":[0.354693183659784,0.718206670005888],"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"ticklen":2,"len":0.5,"yanchor":"top","y":1}},"xaxis":"x","yaxis":"y","frame":null}],"layout":{"margin":{"t":43.9712981082844,"r":7.30593607305936,"b":40.3913894324853,"l":37.2602739726027},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"title":{"text":"CD74 - subj 2","font":{"color":"rgba(0,0,0,1)","family":"","size":17.5342465753425},"x":0,"xref":"paper"},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[1.75,29.25],"tickmode":"array","ticktext":["10","20"],"tickvals":[10,20],"categoryorder":"array","categoryarray":["10","20"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"X Position","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"scaleanchor":"y","scaleratio":1,"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[6,28],"tickmode":"array","ticktext":["10","15","20","25"],"tickvals":[10,15,20,25],"categoryorder":"array","categoryarray":["10","15","20","25"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"Y Position","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"scaleanchor":"x","scaleratio":1,"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"y":0.476799775028121,"yanchor":"top"},"annotations":[{"text":"tumor/stroma","x":1.02,"y":0.5,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"left","yanchor":"bottom","legendTitle":true}],"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"d0852d1485b0":{"x":{},"y":{},"fill":{},"shape":{},"colour":{},"type":"scatter"}},"cur_data":"d0852d1485b0","visdat":{"d0852d1485b0":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script><p>Spatial interpolation of cell scores is also possible with the <code><a href="../reference/deconv_krige.html">deconv_krige()</a></code> function:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconv_krige.html">deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>, cells<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'b_cells'</span>, <span class="st">'i_dc'</span><span class="op">)</span>, who<span class="op">=</span><span class="fl">2</span>, python<span class="op">=</span><span class="cn">F</span><span class="op">)</span></code></pre></div>
<p>We plot the surfaces using the <code><a href="../reference/plot_deconv_krige.html">plot_deconv_krige()</a></code> function.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_krige.html">plot_deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>,  cells<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'b_cells'</span>, <span class="st">'i_dc'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">2</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges3</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotdeconvkrige_chunk-1.png" width="960"></p>
<p>Cell scores in xCell are accompanied by p-values (cell scores not equal to zero). The p-values are calculated automatically during deconvolution in <code>spatialGE</code>, and we can plot them using the <code>pvalue</code> option:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_krige.html">plot_deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>,  cells<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'b_cells'</span>, <span class="st">'i_dc'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">2</span>, 
                             pvalues<span class="op">=</span><span class="cn">T</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges4</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotkrigepvals_chunk-1.png" width="960"></p>
<p>As observed in the surfaces, B cells are present throughout the tissue, but less abundant in the melanoma pocket. The estimates were significantly different from zero as sugggested by xCell’s permutation method. The xCell scores for the dendritic cells were only significantly different from zero only for a few spots (gray hollow circles).</p>
<p>The <code><a href="../reference/plot_deconv_krige.html">plot_deconv_krige()</a></code> also allows annotation of tumor spots:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kriges5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_deconv_krige.html">plot_deconv_krige</a></span><span class="op">(</span><span class="va">melanoma</span>,  cells<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'b_cells'</span>, <span class="st">'i_dc'</span><span class="op">)</span>, plot_who<span class="op">=</span><span class="fl">2</span>, 
                             purity<span class="op">=</span><span class="cn">T</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kriges5</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">3</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotkrigepurity_chunk-1.png" width="960"></p>
</div>
<div id="unsupervised-spatially-informed-clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#unsupervised-spatially-informed-clustering" class="anchor"></a>Unsupervised spatially-informed clustering</h2>
<p>We can now perform unsupervised spatially-informed clustering to detect compartments in the tissue. In <code>spatialGE</code>, we use weighted average matrices to capture the transcriptomic differences among the spots. A transcriptomic (euclidean) distance matrix is calculated and scaled from PCA. Following, scaled euclidean distances are computed for the spatial distances among spots. The user defines a weight, from 0 to 1, to apply to the spatial distances. The higher the weight, the less biologically meaningful the clustering is given that the groups will only reflect the physical distances between the spots and less information on the transcriptomic profiles will be used. After many tests, we have found that weights between 0.1 - 0.25 should be enough to capture the tissue heterogeneity. This approach to spatial clustering can be performed using the <code><a href="../reference/cluster_STspot.html">cluster_STspot()</a></code> function:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cluster_STspot.html">cluster_STspot</a></span><span class="op">(</span><span class="va">melanoma</span>, ks<span class="op">=</span><span class="st">'dtc'</span>, weights<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span>
<span class="co">#&gt; Using vperc = 0.8</span></code></pre></div>
<p>Results of clustering can be plotted via the <code><a href="../reference/plot_STclusters.html">plot_STclusters()</a></code> function:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cluster_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_STclusters.html">plot_STclusters</a></span><span class="op">(</span><span class="va">melanoma</span>, purity<span class="op">=</span><span class="cn">T</span>, plot_who<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>, visium<span class="op">=</span><span class="cn">F</span><span class="op">)</span>

<span class="co"># In case specific k values are selected:</span>
<span class="co"># for(i in 1:2){</span>
<span class="co">#  print(ggpubr::ggarrange(plotlist=cluster_p[[i]], nrow=2, ncol=2, common.legend=T, legend='bottom'))</span>
<span class="co"># }</span>

<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">cluster_p</span>, nrow<span class="op">=</span><span class="fl">1</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/plotclustspots_chunk-1.png" width="960"></p>
<p>We have used here the dynamicTreeCut (<code>dtc</code> option) to select the number of clusters, but users can define their own range of k to evaluate. The two tissues plotted here, show compartmentalization, but in the sample on the left, the lymphoid tissue on the right portion of the tissue is shown in blue possibly indicating spots with immune activity, which also surround the melanoma pocket.</p>
</div>
<div id="association-between-spatial-heterogeneity-and-clinical-variables" class="section level2">
<h2 class="hasAnchor">
<a href="#association-between-spatial-heterogeneity-and-clinical-variables" class="anchor"></a>Association between spatial heterogeneity and clinical variables</h2>
<p>To explore the relationship between a clinical variable of interest and the level of gene expression uniformity within a tissue section, we can use the function <code>plot_phenovar()</code>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatial_stat_plot.html">spatial_stat_plot</a></span><span class="op">(</span><span class="va">melanoma</span>, samplevar<span class="op">=</span><span class="st">'survival_months'</span>, gene<span class="op">=</span><span class="st">'CD74'</span><span class="op">)</span></code></pre></div>
<p><img src="melanoma_ST_vignette_files/figure-html/clinplot_chunk-1.png" width="960"></p>
<p>In the case of the gene CD74, an immune gene, samples with a less uniform expression (i.e., clustered gene expression) came from subjects with a higher survival time (high Moran’s I; low Geary’s C). The survival of the subjects did not show an association with the tendency of CD74 to generate expression “hotspots” (Getis-Ord Gi was intermediate in subject 1)</p>
</div>
<div id="summary" class="section level2">
<h2 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h2>
<p><code>spatialGE</code> provides a suite of tools for the study of tissue heterogeneity via spatial transcriptomics. These tools are designed for exploratory analysis, and can be specially useful to understand the immune landscape of tumors.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Oscar Ospina, Brooke Fridley.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
