<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>spatialGE: Cancerous and normal prostate (Visium and Python kriging) • spatialGE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="spatialGE: Cancerous and normal prostate (Visium and Python kriging)">
<meta property="og:description" content="spatialGE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/melanoma_ST_vignette.html">spatialGE: Visualization and spatial heterogeneity analysis of melanoma</a>
    </li>
    <li>
      <a href="../articles/prostatecancer_Visium_vignette.html">spatialGE: Cancerous and normal prostate (Visium and Python kriging)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="prostatecancer_Visium_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>spatialGE: Cancerous and normal prostate (Visium and Python kriging)</h1>
            
      
      
      <div class="hidden name"><code>prostatecancer_Visium_vignette.Rmd</code></div>

    </div>

    
    
<p>The package <code>spatialGE</code> provides a set of tools for the visualization of gene expression from spatially-resolved arrays. In this vignette we explore gene expression in two Visium data sets: One of normal prostate tissue and other with prostate adenocarcinoma. These data sets were generated with the Visium FFPE technology, but the instructions in this vignette can be used with regular Visium data sets. In this vugnette, we introduce the Python implementation of kriging in <code>spatialGE</code>. We use <code>PyKrige</code> to generate transcriptomic surfaces in order to reduce run time of the interpolation. Nonetheless, kriging is a computationally expensive procedure, and even though <code>PyKrige</code> results in ~3X increases in speed, the interpolation of a Visium data set (~3,000 - 4,000 spots) takes some time.</p>
<div id="setting-up-a-python-environment" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-a-python-environment" class="anchor"></a>Setting up a Python environment</h2>
<p><code>spatialGE</code> uses <code>reticulate</code> to bridge with the Python module <code>PyKrige</code>. To do this, a Python environment needs to be set up. We recommend installing <a href="https://www.anaconda.com/products/individual">Anaconda</a> and create an environment (<code>r-reticulate</code> in this example), then install <code>PyKrige</code> within that environment. Next, the path to the Python environment needs to be set up for R to find it. The set up is done in the <code>.Renviron</code> file. On Mac OS, this is done by typing in the Terminal:</p>
<pre><code>echo "RETICULATE_PYTHON=\"$HOME/opt/anaconda3/envs/r-reticulate/bin/python3\"" &gt;&gt; ~/.Renviron</code></pre>
</div>
<div id="data-acquisition" class="section level2">
<h2 class="hasAnchor">
<a href="#data-acquisition" class="anchor"></a>Data acquisition</h2>
<p>The data sets are freely available and can be downloaded from the 10X Resources website. You may be asked toregister your information before accessing the data. The normal prostate (FFPE) can be downloaded <a href="https://www.10xgenomics.com/resources/datasets/normal-human-prostate-ffpe-1-standard">here</a> and the adenocarcinoma sample <a href="https://www.10xgenomics.com/resources/datasets/human-prostate-cancer-adenocarcinoma-with-invasive-carcinoma-ffpe-1-standard">here</a>. For this vignette, we assume the data will be downloaded in two folders on the computer’s Desktop. At the very least, the user will need to download ‘Spatial imaging data’ and the ‘Feature / cell matrix (raw)’ from each link.</p>
<p>The links to ‘Spatial imaging data’ and the ‘Feature / cell matrix (raw)’ contain compressed <code>.tar.gz</code> files. You can double click the files to decompress them.</p>
</div>
<div id="spatialge-installation" class="section level2">
<h2 class="hasAnchor">
<a href="#spatialge-installation" class="anchor"></a>spatialGE installation</h2>
<p>The <code>spatialGE</code> repository is available at GitHub and can be installed via <code>devtools</code>. To install <code>devtools</code>, in case is not installed in your R console, please run the following code:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="st">"devtools"</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>After making sure <code>devtools</code> is installed, proceed to install <code>spatialGE</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># devtools::install_github("fridleylab/spatialGE")</span></code></pre></div>
<p>To use <code>spatialGE</code>, load the package using the command:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">spatialGE</span><span class="op">)</span></code></pre></div>
</div>
<div id="creating-an-stlist-spatial-transcriptomics-list" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-an-stlist-spatial-transcriptomics-list" class="anchor"></a>Creating an STList (Spatial Transcriptomics List)</h2>
<p>Raw and processed data are stored in an STList (an R S4 class object). The STList can be created with the function <code><a href="../reference/STList.html">STList()</a></code>, and in the case of Visium data, only the file paths to the Visium folders need to be provided. Each folder must contain at least two subfolders: <code>filtered_feature_bc_matrix</code> and <code>spatial</code>. The <code>filtered_feature_bc_matrix</code> contains the files <code>features.tsv.gz</code>, <code>barcodes.tsv.gz</code>, and <code>matrix.mtx.gz</code>. The <code>spatial</code> folder contains <code>tissue_positions_list.csv</code>.</p>
<p>The user also provides array IDs partially matching names of the folders. Optionally, the Visium folders can be provided in the second column of the sample meta data table along with clinical/phenotype data associated with each spatial array. The first column contains array IDs.</p>
<p>Assuming the folders are in the computer’s Desktop, we specify the file paths.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">visiumfp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'~/Desktop/Human_Prostate_Cancer_Adenocarcinoma_Invasive_Carcinoma_FFPE/'</span>,
              <span class="st">'~/Desktop/Normal_Human_Prostate_FFPE/'</span><span class="op">)</span></code></pre></div>
<p>Then, we load the file paths to the <code>STList</code> function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STList.html">STList</a></span><span class="op">(</span>rnacounts<span class="op">=</span><span class="va">visiumfp</span>, samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Adenocarcinoma'</span>, <span class="st">'Normal'</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; Creating STList...</span></code></pre></div>
<p>The <code>prvisium</code> object is an STList.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span>
<span class="co">#&gt; Spatial Transcriptomics List (STList).</span>
<span class="co">#&gt; 2 spatial array(s):</span>
<span class="co">#&gt;  Adenocarcinoma</span>
<span class="co">#&gt;      Normal</span></code></pre></div>
</div>
<div id="data-normalization-of-spatially-resolved-transcriptomic-data" class="section level2">
<h2 class="hasAnchor">
<a href="#data-normalization-of-spatially-resolved-transcriptomic-data" class="anchor"></a>Data normalization of spatially-resolved transcriptomic data</h2>
<p>In <code>spatialGE</code>, we apply voom transformation to the RNA-Seq libraries fromeach of the spots in order to obtain a gaussian-shape normal distribution required by most methods in this package. Many of the methods implemented in <code>spatialGE</code> have been borrowed from the field of geostatistics, which often require this kind of data transformation</p>
<p>Transformation is achieved by applying the function <code><a href="../reference/voom_norm.html">voom_norm()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/voom_norm.html">voom_norm</a></span><span class="op">(</span><span class="va">prvisium</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: magrittr</span></code></pre></div>
<p>We can plot now the gene expression levels of specific genes. Let’s see <em>KLK3</em>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">prvisium</span>, genes<span class="op">=</span><span class="st">'KLK3'</span>, color_pal<span class="op">=</span><span class="st">'sunset'</span>, plot_who<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: ggplot2</span>
<span class="va">qplot</span>
<span class="co">#&gt; $KLK3_2</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>We can also generate transcriptomic surfaces for the samples using the <code>PyKrige</code> implementation:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_krige.html">gene_krige</a></span><span class="op">(</span><span class="va">prvisium</span>, genes<span class="op">=</span><span class="st">'KLK3'</span>, univ<span class="op">=</span><span class="cn">F</span>, python<span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<p>We can now plot the transcriptomic surfaces.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_krige.html">plot_gene_krige</a></span><span class="op">(</span><span class="va">prvisium</span>, genes<span class="op">=</span><span class="st">'KLK3'</span>, color_pal<span class="op">=</span><span class="st">'sunset'</span><span class="op">)</span>
<span class="co">#&gt; Loading required namespace: rgeos</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kplot</span>, ncol<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>As expected, the expression of <em>KLK3</em>, known as prostatic antigen, is higher across the entire cancerous sample (‘subj 1’), compared to the normal sample (‘subj 2’).</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Oscar Ospina, Brooke Fridley.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
