<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>spatialGE: Using PyKrige to study Visium data sets • spatialGE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="spatialGE: Using PyKrige to study Visium data sets">
<meta property="og:description" content="spatialGE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/melanoma_ST_vignette.html">spatialGE: Guide Spatial heterogeneity analysis in melanoma (Thrane et al. 2018)</a>
    </li>
    <li>
      <a href="../articles/prostatecancer_Visium_vignette.html">spatialGE: Using PyKrige to study Visium data sets</a>
    </li>
    <li>
      <a href="../articles/spatialGE_troubleshoot.html">Troubleshooting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="prostatecancer_Visium_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>spatialGE: Using PyKrige to study Visium data sets</h1>
            
      
      
      <div class="hidden name"><code>prostatecancer_Visium_vignette.Rmd</code></div>

    </div>

    
    
<p>The package <code>spatialGE</code> provides a set of tools for the visualization of spatially-resolved transcriptomics. In this vignette we explore gene expression in two 10X Visium samples: A normal prostate tissue and a prostate adenocarcinoma sample. These data sets were generated with the Visium FFPE technology, but the instructions in this vignette can be used with other Visium data sets. Here we introduce the Python implementation of kriging in <code>spatialGE</code>. We use <code>PyKrige</code> to generate transcriptomic surfaces in order to reduce run time of the spatial interpolation. Nonetheless, kriging is a computationally expensive procedure, and even though <code>PyKrige</code> results in ~3X increases in speed, the interpolation of a Visium data set (~3,000-4,000 spots) takes some time.</p>
<div id="setting-up-a-python-environment" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-a-python-environment" class="anchor"></a>Setting up a Python environment</h2>
<p><code>spatialGE</code> uses <code>reticulate</code> to bridge with the Python module <code>PyKrige</code>. To do this, a Python environment needs to be set up. We recommend installing <a href="https://www.anaconda.com/products/individual">Anaconda</a> and create a dedicated environment (<code>r-reticulate</code> in this example), then install <code>PyKrige</code> within that environment. Next, the path to the Python environment needs to be set up for R to find it. The set up is done in the <code>.Renviron</code> file. On Mac OS, this is done by typing in the Terminal:</p>
<pre><code>echo "RETICULATE_PYTHON=\"$HOME/opt/anaconda3/envs/r-reticulate/bin/python3\"" &gt;&gt; ~/.Renviron</code></pre>
</div>
<div id="data-acquisition" class="section level2">
<h2 class="hasAnchor">
<a href="#data-acquisition" class="anchor"></a>Data acquisition</h2>
<p>The data sets are publicly available and can be downloaded from the <a href="https://www.10xgenomics.com/resources/datasets/">10X Resources website</a>. Do a search for “prostate” and click on “Spatial Gene Expression”, and the “Visium Spatial for FFPE Demonstration (v1 Chemistry)”. For this vignette, we will use the normal, adenocarcinoma/invasive carcinoma, and acinar cell carcinoma samples. At this point, users may be asked to register to download the data. Before downloading the data, create three folders on your computer’s desktop. We have named these folders: <code>Visium_FFPE_Human_Normal_Prostate</code>, <code>Visium_FFPE_Human_Prostate_Acinar_Cell_Carcinoma</code>, and <code>Visium_FFPE_Human_Prostate_Cancer</code>. Please proceed to download the files listed in the output section of the 10X website, and place them in their respective folders in your computer’s desktop.</p>
<p>Once the data download is completed, decompress the <code>*_filtered_feature_bc_matrix.tar.gz</code> files from each sample. In MacOS, double clicking each file decompresses the files and produces the folder <code>filtered_feature_bc_matrix</code>. Repeat the procedure for the <code>*_spatial.tar.gz</code> files (which will produce the <code>spatial</code> folder).</p>
</div>
<div id="spatialge-installation" class="section level2">
<h2 class="hasAnchor">
<a href="#spatialge-installation" class="anchor"></a>spatialGE installation</h2>
<p>The <code>spatialGE</code> repository is available at GitHub and can be installed via <code>devtools</code>. To install <code>devtools</code>, in case is not installed in your R console, please run the following code:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="st">"devtools"</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>After making sure <code>devtools</code> is installed, proceed to install <code>spatialGE</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># devtools::install_github("fridleylab/spatialGE")</span></code></pre></div>
<p>To use <code>spatialGE</code>, load the package using the command:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">'spatialGE'</span><span class="op">)</span></code></pre></div>
</div>
<div id="creating-an-stlist-spatial-transcriptomics-list" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-an-stlist-spatial-transcriptomics-list" class="anchor"></a>Creating an STList (Spatial Transcriptomics List)</h2>
<p>Raw data and results from analyses are stored in an STList (S4 class object). The STList can be created with the function <code><a href="../reference/STList.html">STList()</a></code>, and in the case of Visium data, only the file paths to the Visium folders need to be provided. Each folder must contain at least two subfolders: <code>filtered_feature_bc_matrix</code> and <code>spatial</code>. The <code>filtered_feature_bc_matrix</code> contains the files <code>features.tsv.gz</code>, <code>barcodes.tsv.gz</code>, and <code>matrix.mtx.gz</code>. The <code>spatial</code> folder contains <code>tissue_positions_list.csv</code> and stained tissue images (e.g., H&amp;E images).</p>
<p>The user also provides sample IDs partially matching names of the folders. Optionally, the Visium folders can be provided in the second column of the sample metadata file along with clinical/phenotype data associated with each spatial array. We do not have metadata available for the Visium data used in this vignette, thus, we will provide only sample IDs.</p>
<p>Assuming the user created folders with using the instructions in this vignette (see <em>Data acquisition</em>), the output files are specified like so:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">visiumfp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'~/Desktop/Visium_FFPE_Human_Normal_Prostate/'</span>,
              <span class="st">'~/Desktop/Visium_FFPE_Human_Prostate_Cancer/'</span>,
              <span class="st">'~/Desktop/Visium_FFPE_Human_Prostate_Acinar_Cell_Carcinoma/'</span><span class="op">)</span></code></pre></div>
<p>Appropriate sample IDs could be:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">visiumIDs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Normal_Prostate'</span>, 
               <span class="st">'Prostate_Cancer'</span>,
               <span class="st">'Prostate_Acinar_Cell'</span><span class="op">)</span> </code></pre></div>
<p>Then, we load the file paths and sample IDs to the <code>STList</code> function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STList.html">STList</a></span><span class="op">(</span>rnacounts<span class="op">=</span><span class="va">visiumfp</span>, samples<span class="op">=</span><span class="va">visiumIDs</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: magrittr</span>
<span class="co">#&gt; Found Visium Data.</span>
<span class="co">#&gt; Found 3 Visium Samples</span>
<span class="co">#&gt;  Completed!</span>
<span class="co">#&gt; Requested 3 Samples</span>
<span class="co">#&gt; Cleaning Count and Coordinate Data Gene Names.</span>
<span class="co">#&gt; Converting Counts to Sparse Matrices</span>
<span class="co">#&gt; Completed STList!</span></code></pre></div>
<p>The <code>prvisium</code> object is an STList.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span>
<span class="co">#&gt; Spatial Transcriptomics List (STList).</span>
<span class="co">#&gt; 3 spatial array(s):</span>
<span class="co">#&gt;  Normal_Prostate</span>
<span class="co">#&gt;      Prostate_Cancer</span>
<span class="co">#&gt;      Prostate_Acinar_Cell</span></code></pre></div>
</div>
<div id="gene--and-spot-wise-filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#gene--and-spot-wise-filtering" class="anchor"></a>Gene- and spot-wise filtering</h2>
<p>As with other omics data, examination of its quality is crucial to extract a meaningful biological signal. Similar to scRNAseq data, we can examine the number of total counts.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qcplots1</span> <span class="op">=</span> <span class="fu"><a href="../reference/plot_QC_stats.html">plot_QC_stats</a></span><span class="op">(</span><span class="va">prvisium</span>, qc<span class="op">=</span><span class="st">'total_reads'</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: ggplot2</span>
<span class="co">#&gt; Loading required package: Matrix</span>
<span class="co">#&gt; Warning: Removed 1 rows containing missing values (geom_point).</span>
<span class="va">qcplots1</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/toral_reads-1.png" width="864"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qcplots2</span> <span class="op">=</span> <span class="fu"><a href="../reference/plot_QC_stats.html">plot_QC_stats</a></span><span class="op">(</span><span class="va">prvisium</span>, qc<span class="op">=</span><span class="st">'total_genes'</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 1 rows containing missing values (geom_point).</span>
<span class="va">qcplots2</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/total_genes-1.png" width="864"></p>
<p>We can also plot these numbers spatially, and to generate an organized layout, we use <code>ggarrange()</code>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qcplots3</span> <span class="op">=</span> <span class="fu"><a href="../reference/plot_QC_quilt.html">plot_QC_quilt</a></span><span class="op">(</span><span class="va">prvisium</span>, nreads<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">qcplots3</span>, ncol<span class="op">=</span><span class="fl">3</span>, nrow<span class="op">=</span><span class="fl">1</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/quilt_ngenes-1.png" width="864"></p>
<p>We can filter out spots with low counts (&lt;200). We also see a few spots with unusually high count in the acinar cell cancer sample. We can filter spots using the <code><a href="../reference/filter_data.html">filter_data()</a></code> function.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_data.html">filter_data</a></span><span class="op">(</span><span class="va">prvisium</span>, spot_maxreads<span class="op">=</span><span class="fl">50000</span><span class="op">)</span></code></pre></div>
</div>
<div id="data-normalization-of-spatially-resolved-transcriptomic-data" class="section level2">
<h2 class="hasAnchor">
<a href="#data-normalization-of-spatially-resolved-transcriptomic-data" class="anchor"></a>Data normalization of spatially-resolved transcriptomic data</h2>
<p>Many transformation methods are available for both bulk and single-cell RNA-Seq count data. In <code>spatialGE</code>, the function <code><a href="../reference/spatialTransform.html">spatialTransform()</a></code> applies log-transformation to the data, after library size normalization. Similar to Seurat, it applies a scaling factor (<code>scale_f=10000</code> by default). <code>spatialGE</code> borrows methods from the field of spatial statistics, often requiring gaussian-shaped data distributions. Hence, users also can apply voom transformation, which could in certain situations, provide more gaussian-shaped distributions than log-transformed data.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialTransform.html">spatialTransform</a></span><span class="op">(</span><span class="va">prvisium</span><span class="op">)</span></code></pre></div>
</div>
<div id="visualization-of-gene-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#visualization-of-gene-expression" class="anchor"></a>Visualization of gene expression</h2>
<p>We can plot now the gene expression levels of specific genes. When created from Visium outputs or Seurat objects, STLists contain image data as well. We can plot the images along with quilt plots by setting <code>image=T</code>. Let’s create plots for <em>KLK3</em>:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">prvisium</span>, genes<span class="op">=</span><span class="st">'KLK3'</span>, color_pal<span class="op">=</span><span class="st">'sunset'</span>, image<span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<p>The <code>plot_gene_quilt</code> returns a list of plots. We can use <code>ggarrange()</code> to diplay these plots in an organized manner:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">qplot</span>, ncol<span class="op">=</span><span class="fl">2</span>, nrow<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/unnamed-chunk-3-1.png" width="960"></p>
</div>
<div id="spatial-interpolation-of-gene-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#spatial-interpolation-of-gene-expression" class="anchor"></a>Spatial interpolation of gene expression</h2>
<p>We can also generate transcriptomic surfaces for the samples using the <code><a href="../reference/gene_krige.html">gene_krige()</a></code> and setting <code>python=T</code> to use the <code>PyKrige</code> implementation.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_krige.html">gene_krige</a></span><span class="op">(</span><span class="va">prvisium</span>, genes<span class="op">=</span><span class="st">'KLK3'</span>, univ<span class="op">=</span><span class="cn">F</span>, python<span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<p>As mentioned at the beginning of this vignette, we have leveraged the <code>reticulate</code> package to benefit from the speed that python can bring to computationally expensive functions. For Kriging, we pass in our STList, the same gene name that we were interested in before ( <em>KLK3</em> ), <code>univ</code> is a parameter letting us decide between universal kriging or ordinary kriging (universal takes spatial trends into account, but takes much longer to do. Here we have it perform ordinary with <code>univ=F</code>.</p>
<p>We can now plot the transcriptomic surfaces.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_krige.html">plot_gene_krige</a></span><span class="op">(</span><span class="va">prvisium</span>, genes<span class="op">=</span><span class="st">'KLK3'</span>, color_pal<span class="op">=</span><span class="st">'sunset'</span>, image<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="co">#&gt; Loading required namespace: rgeos</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kplot</span>, ncol<span class="op">=</span><span class="fl">2</span>, nrow<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/unnamed-chunk-4-1.png" width="960"></p>
<p>As expected, the expression of <em>KLK3</em>, known as prostatic antigen, is higher across the the cancerous samples (middle and bottom), compared to the normal sample (top).</p>
<p>We can also classify spots as likely tumor or likely stroma using the <code><a href="../reference/spatialDeconv.html">spatialDeconv()</a></code> function. <code>spatialGE</code> uses <code>ESTIMATE</code> to calculate tumor purity scores, followed by model-based clustering (<code>mclust</code>). Notice that we set <code>method='none'</code>, in order to avoid computing <code>xCell</code> scores for the purpose of this example. Users are encouraged to remove that argument and get cell scores.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialDeconv.html">spatialDeconv</a></span><span class="op">(</span><span class="va">prvisium</span>, method<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>
<span class="co">#&gt; This will take some time...</span>
<span class="co">#&gt; Loading required package: estimate</span>
<span class="co">#&gt; Only ESTIMATE was applied to the spatial arrays.</span></code></pre></div>
<p>We can now plot the spot classification along side the expression plot and the H&amp;E image.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">eplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">prvisium</span>, genes<span class="op">=</span><span class="st">"KLK3"</span>, color_pal<span class="op">=</span><span class="st">"sunset"</span>, purity<span class="op">=</span><span class="cn">T</span>, image<span class="op">=</span><span class="cn">T</span>, ptsize<span class="op">=</span><span class="fl">1</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">eplot</span>, ncol<span class="op">=</span><span class="fl">3</span>, nrow<span class="op">=</span><span class="fl">3</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/plot_estimate-1.png" width="1152"></p>
<p>Note that the normal prostate sample has a spots classified as tumor (dark squares). This is an artifact resulting from the model-base clustering performed by <code>spatialGE</code>, which fits <code>ESTIMATE</code> purity scores into two groups (k=2).</p>
<details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.0.3 (2020-10-10)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.7</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] mclust_5.4.7         estimate_1.0.13      Matrix_1.3-2        </span>
<span class="co">#&gt; [4] ggplot2_3.3.5        magrittr_2.0.1       spatialGE_0.0.0.9000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] sass_0.3.1        tidyr_1.1.2       jsonlite_1.7.2    carData_3.0-4    </span>
<span class="co">#&gt;  [5] bslib_0.2.4       assertthat_0.2.1  sp_1.4-5          highr_0.8        </span>
<span class="co">#&gt;  [9] cellranger_1.1.0  yaml_2.2.1        pillar_1.4.7      backports_1.2.1  </span>
<span class="co">#&gt; [13] lattice_0.20-41   glue_1.4.2        digest_0.6.27     ggsignif_0.6.1   </span>
<span class="co">#&gt; [17] snakecase_0.11.0  khroma_1.6.0      colorspace_2.0-0  cowplot_1.1.1    </span>
<span class="co">#&gt; [21] htmltools_0.5.1.1 pkgconfig_2.0.3   raster_3.4-5      broom_0.7.4      </span>
<span class="co">#&gt; [25] haven_2.3.1       purrr_0.3.4       scales_1.1.1      openxlsx_4.2.3   </span>
<span class="co">#&gt; [29] rio_0.5.16        tibble_3.0.6      generics_0.1.0    farver_2.0.3     </span>
<span class="co">#&gt; [33] car_3.0-10        ellipsis_0.3.1    ggpubr_0.4.0      cachem_1.0.1     </span>
<span class="co">#&gt; [37] withr_2.4.1       janitor_2.1.0     crayon_1.4.1      readxl_1.3.1     </span>
<span class="co">#&gt; [41] memoise_2.0.0     evaluate_0.14     fs_1.5.0          rstatix_0.7.0    </span>
<span class="co">#&gt; [45] forcats_0.5.1     foreign_0.8-81    textshaping_0.3.4 tools_4.0.3      </span>
<span class="co">#&gt; [49] data.table_1.13.6 hms_1.0.0         lifecycle_1.0.0   ggpolypath_0.1.0 </span>
<span class="co">#&gt; [53] stringr_1.4.0     V8_3.4.0          munsell_0.5.0     zip_2.1.1        </span>
<span class="co">#&gt; [57] compiler_4.0.3    pkgdown_1.6.1     jquerylib_0.1.3   concaveman_1.1.0 </span>
<span class="co">#&gt; [61] systemfonts_1.0.2 rlang_0.4.10      grid_4.0.3        labeling_0.4.2   </span>
<span class="co">#&gt; [65] rmarkdown_2.10    codetools_0.2-18  gtable_0.3.0      abind_1.4-5      </span>
<span class="co">#&gt; [69] DBI_1.1.1         curl_4.3          R6_2.5.0          lubridate_1.7.9.2</span>
<span class="co">#&gt; [73] knitr_1.31        dplyr_1.0.3       rgeos_0.5-5       fastmap_1.1.0    </span>
<span class="co">#&gt; [77] rprojroot_2.0.2   ragg_1.1.2        desc_1.2.0        stringi_1.5.3    </span>
<span class="co">#&gt; [81] parallel_4.0.3    Rcpp_1.0.6        vctrs_0.3.6       png_0.1-7        </span>
<span class="co">#&gt; [85] tidyselect_1.1.0  xfun_0.25</span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Oscar Ospina, Brooke Fridley.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
