<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>spatialGE: Kriging for Visium data sets • spatialGE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="spatialGE: Kriging for Visium data sets">
<meta property="og:description" content="spatialGE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0.0000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/melanoma_ST_vignette.html">spatialGE: Guide to spatial heterogeneity analysis in melanoma (Thrane et al. 2018)</a>
    </li>
    <li>
      <a href="../articles/prostatecancer_Visium_vignette.html">spatialGE: Kriging for Visium data sets</a>
    </li>
    <li>
      <a href="../articles/read_GeoMx_output.html">spatialGE: Reading GeoMx-DSP outputs</a>
    </li>
    <li>
      <a href="../articles/spatialGE_troubleshoot.html">Troubleshooting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>spatialGE: Kriging for Visium data sets</h1>
            
      
      
      <div class="hidden name"><code>prostatecancer_Visium_vignette.Rmd</code></div>

    </div>

    
    
<p>The package <code>spatialGE</code> provides a set of tools for the
visualization of spatially-resolved transcriptomics. In this vignette we
explore gene expression in three 10X Visium samples: A normal prostate
tissue and two prostate adenocarcinoma samples. These data sets were
generated with the Visium FFPE technology, but the instructions in this
vignette can be used with other Visium data sets. Here we introduce the
Python implementation of kriging in <code>spatialGE</code>. We use
<code>PyKrige</code> to generate transcriptomic surfaces in order to
reduce run time of the spatial interpolation. Nonetheless, kriging is a
computationally expensive procedure, and even though
<code>PyKrige</code> results in ~3X increases in speed, the
interpolation of a Visium data set (~3,000-4,000 spots) takes time.</p>
<div class="section level2">
<h2 id="setting-up-a-python-environment">Setting up a Python environment<a class="anchor" aria-label="anchor" href="#setting-up-a-python-environment"></a>
</h2>
<p><code>spatialGE</code> uses <code>reticulate</code> to bridge with
the Python module <code>PyKrige</code>. To do this, a Python environment
needs to be set up. We recommend installing <a href="https://www.anaconda.com/products/individual" class="external-link">Anaconda</a> and
create a dedicated environment (<code>r-reticulate</code> in this
example), then install <code>PyKrige</code> within that environment.
Next, the path to the Python environment needs to be set up for R to
find it. The set up is done in the <code>.Renviron</code> file. On Mac
OS, this is done by typing in the Terminal:</p>
<pre><code>echo "RETICULATE_PYTHON=\"$HOME/opt/anaconda3/envs/r-reticulate/bin/python3\"" &gt;&gt; ~/.Renviron</code></pre>
</div>
<div class="section level2">
<h2 id="data-acquisition">Data acquisition<a class="anchor" aria-label="anchor" href="#data-acquisition"></a>
</h2>
<p>The data sets are publicly available and can be downloaded from the
<a href="https://www.10xgenomics.com/resources/datasets/" class="external-link">10X Resources
website</a>. Do a search for “prostate” and click on “Spatial Gene
Expression”, and the “Visium Spatial for FFPE Demonstration (v1
Chemistry)”. For this vignette, we will use the normal,
adenocarcinoma/invasive carcinoma, and acinar cell carcinoma samples. At
this point, users may be asked to register to download the data. Before
downloading the data, create three folders on your computer’s desktop.
We have named these folders:
<code>Visium_FFPE_Human_Normal_Prostate</code>,
<code>Visium_FFPE_Human_Prostate_Acinar_Cell_Carcinoma</code>, and
<code>Visium_FFPE_Human_Prostate_Cancer</code>. Please proceed to
download the files listed in the output section of the 10X website, and
place them in their respective folders in your computer’s desktop.</p>
<p>Once the data download is completed, decompress the
<code>*_filtered_feature_bc_matrix.tar.gz</code> files from each sample.
In MacOS, double clicking each file decompresses the files and produces
the folder <code>filtered_feature_bc_matrix</code>. Repeat the procedure
for the <code>*_spatial.tar.gz</code> files (which will produce the
<code>spatial</code> folder).</p>
</div>
<div class="section level2">
<h2 id="spatialge-installation">spatialGE installation<a class="anchor" aria-label="anchor" href="#spatialge-installation"></a>
</h2>
<p>The <code>spatialGE</code> repository is available at GitHub and can
be installed via <code>devtools</code>. To install
<code>devtools</code>, in case is not installed in your R console,
please run the following code:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="st">"devtools"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>After making sure <code>devtools</code> is installed, proceed to
install <code>spatialGE</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"fridleylab/spatialGE"</span><span class="op">)</span></code></pre></div>
<p>To use <code>spatialGE</code>, load the package using the
command:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">'spatialGE'</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="creating-an-stlist-spatial-transcriptomics-list">Creating an STList (Spatial Transcriptomics List)<a class="anchor" aria-label="anchor" href="#creating-an-stlist-spatial-transcriptomics-list"></a>
</h2>
<p>Raw data and results from analyses are stored in an STList (S4 class
object). The STList can be created with the function
<code><a href="../reference/STList.html">STList()</a></code>, and in the case of Visium data, only the file
paths to the Visium folders need to be provided. Each folder must
contain at least two subfolders: <code>filtered_feature_bc_matrix</code>
and <code>spatial</code>. The <code>filtered_feature_bc_matrix</code>
contains the files <code>features.tsv.gz</code>,
<code>barcodes.tsv.gz</code>, and <code>matrix.mtx.gz</code>. The
<code>spatial</code> folder contains
<code>tissue_positions_list.csv</code> and stained tissue images (e.g.,
H&amp;E images).</p>
<p>The user also provides sample IDs partially matching names of the
folders. Optionally, the Visium folders can be provided in the second
column of the sample metadata file along with clinical/phenotype data
associated with each spatial array. We do not have metadata available
for the Visium data used in this vignette, thus, we will provide only
sample IDs.</p>
<p>Assuming the user created folders using the instructions in this
vignette (see <em>Data acquisition</em>), the output files are specified
like so:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">visiumfp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'~/Desktop/Visium_FFPE_Human_Normal_Prostate/'</span>,
              <span class="st">'~/Desktop/Visium_FFPE_Human_Prostate_Cancer/'</span>,
              <span class="st">'~/Desktop/Visium_FFPE_Human_Prostate_Acinar_Cell_Carcinoma/'</span><span class="op">)</span></code></pre></div>
<p>Appropriate sample IDs could be:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">visiumIDs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Normal_Prostate'</span>, 
               <span class="st">'Prostate_Cancer'</span>,
               <span class="st">'Prostate_Acinar_Cell'</span><span class="op">)</span> </code></pre></div>
<p>Note: Instead of sample IDs, users could also provide a file with a
comma- or tab-delimited table with minimally two columns: sample IDs and
file paths to the Visium folders, one row per sample. This method allows
users to add clinical data in subsequent columns. Please, refer to the
<a href="https://fridleylab.github.io/spatialGE/reference/STList.html" class="external-link">STList
documentation for details</a>.</p>
<p>We load the Visium data sets and sample IDs to the
<code>STList</code> function like so:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STList.html">STList</a></span><span class="op">(</span>rnacounts<span class="op">=</span><span class="va">visiumfp</span>, samples<span class="op">=</span><span class="va">visiumIDs</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #BBBB00;">Found Visium Data.</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; </span><span style="color: #00BB00; font-weight: bold;">Found 3 Visium Samples</span></span>
<span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt;   Completed!</span></span>
<span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; </span><span style="color: #BBBB00;">Requested 3 Samples</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; Cleaning Count and Coordinate Data Gene Names.</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; Converting Counts to Sparse Matrices</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; </span><span style="color: #00BB00; font-weight: bold;">Completed STList!</span></span>
<span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; </span></span></code></pre></div>
<p>The <code>prvisium</code> object is an STList.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span>
<span class="co">#&gt; Spatial Transcriptomics List (STList).</span>
<span class="co">#&gt; 3 spatial array(s):</span>
<span class="co">#&gt;  Normal_Prostate</span>
<span class="co">#&gt;      Prostate_Cancer</span>
<span class="co">#&gt;      Prostate_Acinar_Cell</span></code></pre></div>
<details><summary><strong>How to read data from a Seurat object?</strong>
</summary><p>Seurat is one of the most widely used packages for the analysis of
scRNAseq data. Recently, new methods have been added to read and analyze
10X Visium data. Seurat relies on an eponymous data structure (Seurat
object) to store data. We have implemented functionality in
<code>spatialGE</code> to read spatial transcriptomics experiments from
Seurat objects and complete analysis of spatial heterogeneity.</p>
<p>The function <code><a href="../reference/STList.html">STList()</a></code> can be provided a Seurat object
via the <code>rnacounts</code> argument, yielding an STList object. To
show this functionality, we use the <a href="https://satijalab.org/seurat/articles/spatial_vignette.html" class="external-link">mouse
brain data set</a> included in Seurat. The data set is split in two
parts (“anterior” and “posterior”).</p>
<p>First, the data is installed, and the two samples loaded.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://www.satijalab.org/seurat" class="external-link">'SeuratData'</a></span><span class="op">)</span>
<span class="co">#&gt; ── <span style="font-weight: bold;">Installed datasets</span> ───────────────────────────────────── SeuratData v0.2.1 ──</span>
<span class="co">#&gt; <span style="color: #00BB00;">✓</span> <span style="color: #0000BB;">stxBrain</span> 0.1.1</span>
<span class="co">#&gt; ────────────────────────────────────── Key ─────────────────────────────────────</span>
<span class="co">#&gt; <span style="color: #00BB00;">✓</span> Dataset loaded successfully</span>
<span class="co">#&gt; <span style="color: #BBBB00;">&gt;</span> Dataset built with a newer version of Seurat than installed</span>
<span class="co">#&gt; <span style="color: #BB0000;">❓</span> Unknown version of Seurat installed</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout<span class="op">=</span><span class="fl">600</span><span class="op">)</span> <span class="co"># Allows time for R to download data</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html" class="external-link">InstallData</a></span><span class="op">(</span><span class="st">"stxBrain"</span><span class="op">)</span>
<span class="va">brain_anterior</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span><span class="st">"stxBrain"</span>, type<span class="op">=</span><span class="st">"anterior1"</span><span class="op">)</span>
<span class="va">brain_posterior</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span><span class="st">"stxBrain"</span>, type<span class="op">=</span><span class="st">"posterior1"</span><span class="op">)</span></code></pre></div>
<p>Samples are then merged using the Seurat function
<code>merge</code></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">brain_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">brain_anterior</span>, <span class="va">brain_posterior</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: Seurat</span>
<span class="co">#&gt; Registered S3 method overwritten by 'spatstat.geom':</span>
<span class="co">#&gt;   method     from</span>
<span class="co">#&gt;   print.boxx cli</span>
<span class="co">#&gt; Attaching SeuratObject</span>
<span class="co">#&gt; Warning in CheckDuplicateCellNames(object.list = objects): Some cell names are</span>
<span class="co">#&gt; duplicated across objects provided. Renaming to enforce unique cell names.</span></code></pre></div>
<p>Now that we have a Seurat object with two samples, we proceed to
convert it to a STList.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">brain_STList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STList.html">STList</a></span><span class="op">(</span><span class="va">brain_merge</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #BBBB00;">Found Seurat object.</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; Requested 2 Samples</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; Cleaning Count and Coordinate Data Gene Names.</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; Converting Counts to Sparse Matrices</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; </span><span style="color: #00BB00; font-weight: bold;">Completed STList!</span></span>
<span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; </span></span>
<span class="va">brain_STList</span>
<span class="co">#&gt; Spatial Transcriptomics List (STList).</span>
<span class="co">#&gt; 2 spatial array(s):</span>
<span class="co">#&gt;  anterior1</span>
<span class="co">#&gt;      posterior1</span></code></pre></div>
<p>We can plot and analyze the data using spatialGE functions. For
example, we let us transform the data and plot the expression of gene
<em>Hpca</em>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">brain_STList</span> <span class="op">=</span> <span class="fu"><a href="../reference/spatialTransform.html">spatialTransform</a></span><span class="op">(</span><span class="va">brain_STList</span><span class="op">)</span>
<span class="va">brain_p</span> <span class="op">=</span> <span class="fu">spatialGE</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">brain_STList</span>, gene<span class="op">=</span><span class="st">'Hpca'</span>, image<span class="op">=</span><span class="cn">T</span>, color_pal<span class="op">=</span><span class="st">'sunset'</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">brain_p</span>, ncol<span class="op">=</span><span class="fl">2</span>, nrow<span class="op">=</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
<img src="prostatecancer_Visium_vignette_files/figure-html/seurat_data4-1.png" width="672"></details>
</div>
<div class="section level2">
<h2 id="gene--and-spot-wise-filtering">Gene- and spot-wise filtering<a class="anchor" aria-label="anchor" href="#gene--and-spot-wise-filtering"></a>
</h2>
<p>As with other omics data, examination of its quality is crucial to
extract a meaningful biological signal. Similar to scRNAseq data, we can
examine the number of total counts.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qcplots1</span> <span class="op">=</span> <span class="fu"><a href="../reference/plot_QC_stats.html">plot_QC_stats</a></span><span class="op">(</span><span class="va">prvisium</span>, qc<span class="op">=</span><span class="st">'total_reads'</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 1 rows containing missing values (geom_point).</span>
<span class="va">qcplots1</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/toral_reads-1.png" width="864"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qcplots2</span> <span class="op">=</span> <span class="fu"><a href="../reference/plot_QC_stats.html">plot_QC_stats</a></span><span class="op">(</span><span class="va">prvisium</span>, qc<span class="op">=</span><span class="st">'total_genes'</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 1 rows containing missing values (geom_point).</span>
<span class="va">qcplots2</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/total_genes-1.png" width="864"></p>
<p>We can also plot these numbers spatially, and to generate an
organized layout, we use <code>ggarrange()</code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qcplots3</span> <span class="op">=</span> <span class="fu"><a href="../reference/plot_QC_quilt.html">plot_QC_quilt</a></span><span class="op">(</span><span class="va">prvisium</span>, nreads<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">qcplots3</span>, ncol<span class="op">=</span><span class="fl">3</span>, nrow<span class="op">=</span><span class="fl">1</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/quilt_ngenes-1.png" width="864"></p>
<p>We can filter out a few spots with unusually high count in the acinar
cell cancer sample. To filter spots, we use the
<code><a href="../reference/filter_data.html">filter_data()</a></code> function.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span> <span class="op">=</span> <span class="fu"><a href="../reference/filter_data.html">filter_data</a></span><span class="op">(</span><span class="va">prvisium</span>, spot_maxreads<span class="op">=</span><span class="fl">50000</span><span class="op">)</span>

<span class="va">qcplots4</span> <span class="op">=</span> <span class="fu"><a href="../reference/plot_QC_quilt.html">plot_QC_quilt</a></span><span class="op">(</span><span class="va">prvisium</span>, nreads<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">qcplots4</span>, ncol<span class="op">=</span><span class="fl">3</span>, nrow<span class="op">=</span><span class="fl">1</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/quilt_ngenes2-1.png" width="864"></p>
</div>
<div class="section level2">
<h2 id="data-normalization-of-spatially-resolved-transcriptomic-data">Data normalization of spatially-resolved transcriptomic data<a class="anchor" aria-label="anchor" href="#data-normalization-of-spatially-resolved-transcriptomic-data"></a>
</h2>
<p>Many transformation methods are available for both bulk and
single-cell RNA-Seq count data. In <code>spatialGE</code>, the function
<code><a href="../reference/spatialTransform.html">spatialTransform()</a></code> applies log-transformation to the data,
after library size normalization. Similar to Seurat, it applies a
scaling factor (<code>scale_f=10000</code> by default).
<code>spatialGE</code> borrows methods from the field of spatial
statistics, often requiring gaussian-shaped data distributions. Hence,
users also can apply voom transformation, which could in certain
situations, provide more gaussian-shaped distributions than
log-transformed data.</p>
<p>We will apply log-transformation:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialTransform.html">spatialTransform</a></span><span class="op">(</span><span class="va">prvisium</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization-of-gene-expression">Visualization of gene expression<a class="anchor" aria-label="anchor" href="#visualization-of-gene-expression"></a>
</h2>
<p>We can plot now the gene expression levels of specific genes. When
created from Visium outputs or Seurat objects, STLists contain image
data as well. We plot the images along with quilt plots by setting
<code>image=T</code>. Let’s create plots for <em>KLK3</em>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">prvisium</span>, genes<span class="op">=</span><span class="st">'KLK3'</span>, color_pal<span class="op">=</span><span class="st">'sunset'</span>, image<span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<p>The <code>plot_gene_quilt</code> returns a list of plots. We can use
<code>ggarrange()</code> to diplay these plots in an organized
manner:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">qplot</span>, ncol<span class="op">=</span><span class="fl">2</span>, nrow<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/unnamed-chunk-3-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="spatial-interpolation-of-gene-expression">Spatial interpolation of gene expression<a class="anchor" aria-label="anchor" href="#spatial-interpolation-of-gene-expression"></a>
</h2>
<p>Transcriptomic surfaces are generated for the samples using the
<code><a href="../reference/gene_krige.html">gene_krige()</a></code> and setting <code>python=T</code> to use the
<code>PyKrige</code> implementation.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_krige.html">gene_krige</a></span><span class="op">(</span><span class="va">prvisium</span>, genes<span class="op">=</span><span class="st">'KLK3'</span>, univ<span class="op">=</span><span class="cn">F</span>, python<span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<p>As mentioned at the beginning of this vignette, we have leveraged the
<code>reticulate</code> package to benefit from the speed that python
can bring to computationally expensive functions. For Kriging, we pass
in our STList and the gene name that we are interested in
(<em>KLK3</em>). The <code>univ</code> parameter switch between
universal or ordinary kriging. Universal kriging takes spatial trends
into account, but takes much longer to perform, thus we set
<code>univ=F</code>.</p>
<p>We can now plot the transcriptomic surfaces:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_krige.html">plot_gene_krige</a></span><span class="op">(</span><span class="va">prvisium</span>, genes<span class="op">=</span><span class="st">'KLK3'</span>, color_pal<span class="op">=</span><span class="st">'sunset'</span>, image<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="co">#&gt; KLK3 kriging for subject 1 is not present in STList</span>
<span class="co">#&gt; KLK3 kriging for subject 2 is not present in STList</span>
<span class="co">#&gt; KLK3 kriging for subject 3 is not present in STList</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">kplot</span>, ncol<span class="op">=</span><span class="fl">2</span>, nrow<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/unnamed-chunk-4-1.png" width="960"></p>
<p>As expected, the expression of <em>KLK3</em>, known as prostatic
antigen, is higher across the the cancerous samples (middle and bottom),
compared to the normal sample (top).</p>
<p>We can also classify spots as likely tumor or likely stroma using the
<code><a href="../reference/spatialDeconv.html">spatialDeconv()</a></code> function. <code>spatialGE</code> uses
<code>ESTIMATE</code> to calculate tumor purity scores, followed by
model-based clustering (<code>mclust</code>). Notice that we set
<code>method='none'</code>, in order to avoid computing
<code>xCell</code> scores for the purpose of this example. Users are
encouraged to remove that argument and get cell scores to further explor
this data set.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prvisium</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialDeconv.html">spatialDeconv</a></span><span class="op">(</span><span class="va">prvisium</span>, method<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>
<span class="co">#&gt; This will take some time...</span>
<span class="co">#&gt; Loading required package: estimate</span>
<span class="co">#&gt; Only ESTIMATE was applied to the spatial arrays.</span></code></pre></div>
<p>We can now plot the spot classification along side the expression
plot and the H&amp;E image.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">eplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">prvisium</span>, genes<span class="op">=</span><span class="st">"KLK3"</span>, color_pal<span class="op">=</span><span class="st">"sunset"</span>, purity<span class="op">=</span><span class="cn">T</span>, image<span class="op">=</span><span class="cn">T</span>, ptsize<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">eplot</span>, ncol<span class="op">=</span><span class="fl">3</span>, nrow<span class="op">=</span><span class="fl">3</span>, legend<span class="op">=</span><span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="prostatecancer_Visium_vignette_files/figure-html/plot_estimate-1.png" width="1152"></p>
<p>Note that the normal prostate sample has a spots classified as tumor
(dark squares). This is an artifact resulting from the model-based
clustering performed by <code>spatialGE</code>, which fits
<code>ESTIMATE</code> purity scores into two groups (k=2).</p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.7</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] estimate_1.0.13           magrittr_2.0.2           </span>
<span class="co">#&gt; [3] SeuratObject_4.0.4        Seurat_4.1.0             </span>
<span class="co">#&gt; [5] stxBrain.SeuratData_0.1.1 SeuratData_0.2.1         </span>
<span class="co">#&gt; [7] spatialGE_1.0.0.0000     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] utf8_1.2.2                  reticulate_1.24            </span>
<span class="co">#&gt;   [3] tidyselect_1.1.1            RSQLite_2.2.9              </span>
<span class="co">#&gt;   [5] AnnotationDbi_1.56.2        htmlwidgets_1.5.4          </span>
<span class="co">#&gt;   [7] grid_4.1.2                  BiocParallel_1.28.3        </span>
<span class="co">#&gt;   [9] Rtsne_0.15                  munsell_0.5.0              </span>
<span class="co">#&gt;  [11] ScaledMatrix_1.2.0          codetools_0.2-18           </span>
<span class="co">#&gt;  [13] ragg_1.2.2                  ica_1.0-2                  </span>
<span class="co">#&gt;  [15] future_1.23.0               miniUI_0.1.1.1             </span>
<span class="co">#&gt;  [17] withr_2.4.3                 colorspace_2.0-2           </span>
<span class="co">#&gt;  [19] Biobase_2.54.0              highr_0.9                  </span>
<span class="co">#&gt;  [21] knitr_1.37                  rstudioapi_0.13            </span>
<span class="co">#&gt;  [23] stats4_4.1.2                SingleCellExperiment_1.16.0</span>
<span class="co">#&gt;  [25] ROCR_1.0-11                 ggsignif_0.6.3             </span>
<span class="co">#&gt;  [27] tensor_1.5                  listenv_0.8.0              </span>
<span class="co">#&gt;  [29] labeling_0.4.2              MatrixGenerics_1.6.0       </span>
<span class="co">#&gt;  [31] GenomeInfoDbData_1.2.7      polyclip_1.10-0            </span>
<span class="co">#&gt;  [33] farver_2.1.0                bit64_4.0.5                </span>
<span class="co">#&gt;  [35] rhdf5_2.38.0                rprojroot_2.0.2            </span>
<span class="co">#&gt;  [37] parallelly_1.30.0           vctrs_0.3.8                </span>
<span class="co">#&gt;  [39] generics_0.1.2              xfun_0.29                  </span>
<span class="co">#&gt;  [41] R6_2.5.1                    GenomeInfoDb_1.30.1        </span>
<span class="co">#&gt;  [43] khroma_1.8.0                rsvd_1.0.5                 </span>
<span class="co">#&gt;  [45] concaveman_1.1.0            bitops_1.0-7               </span>
<span class="co">#&gt;  [47] rhdf5filters_1.6.0          spatstat.utils_2.3-0       </span>
<span class="co">#&gt;  [49] cachem_1.0.6                DelayedArray_0.20.0        </span>
<span class="co">#&gt;  [51] assertthat_0.2.1            promises_1.2.0.1           </span>
<span class="co">#&gt;  [53] scales_1.1.1                gtable_0.3.0               </span>
<span class="co">#&gt;  [55] beachmat_2.10.0             globals_0.14.0             </span>
<span class="co">#&gt;  [57] goftest_1.2-3               rlang_1.0.0                </span>
<span class="co">#&gt;  [59] systemfonts_1.0.4           splines_4.1.2              </span>
<span class="co">#&gt;  [61] rstatix_0.7.0               lazyeval_0.2.2             </span>
<span class="co">#&gt;  [63] broom_0.7.12                spatstat.geom_2.3-1        </span>
<span class="co">#&gt;  [65] yaml_2.2.2                  reshape2_1.4.4             </span>
<span class="co">#&gt;  [67] abind_1.4-5                 backports_1.4.1            </span>
<span class="co">#&gt;  [69] httpuv_1.6.5                tools_4.1.2                </span>
<span class="co">#&gt;  [71] ggplot2_3.3.5               ellipsis_0.3.2             </span>
<span class="co">#&gt;  [73] spatstat.core_2.3-2         jquerylib_0.1.4            </span>
<span class="co">#&gt;  [75] RColorBrewer_1.1-2          xCell_1.1.0                </span>
<span class="co">#&gt;  [77] BiocGenerics_0.40.0         ggridges_0.5.3             </span>
<span class="co">#&gt;  [79] Rcpp_1.0.8                  plyr_1.8.6                 </span>
<span class="co">#&gt;  [81] sparseMatrixStats_1.6.0     zlibbioc_1.40.0            </span>
<span class="co">#&gt;  [83] purrr_0.3.4                 RCurl_1.98-1.5             </span>
<span class="co">#&gt;  [85] ggpubr_0.4.0                rpart_4.1-15               </span>
<span class="co">#&gt;  [87] deldir_1.0-6                pbapply_1.5-0              </span>
<span class="co">#&gt;  [89] cowplot_1.1.1               S4Vectors_0.32.3           </span>
<span class="co">#&gt;  [91] zoo_1.8-9                   SummarizedExperiment_1.24.0</span>
<span class="co">#&gt;  [93] ggrepel_0.9.1               cluster_2.1.2              </span>
<span class="co">#&gt;  [95] fs_1.5.2                    data.table_1.14.2          </span>
<span class="co">#&gt;  [97] scattermore_0.7             lmtest_0.9-39              </span>
<span class="co">#&gt;  [99] RANN_2.6.1                  fitdistrplus_1.1-6         </span>
<span class="co">#&gt; [101] matrixStats_0.61.0          patchwork_1.1.1            </span>
<span class="co">#&gt; [103] mime_0.12                   evaluate_0.14              </span>
<span class="co">#&gt; [105] GSVA_1.42.0                 xtable_1.8-4               </span>
<span class="co">#&gt; [107] XML_3.99-0.8                mclust_5.4.9               </span>
<span class="co">#&gt; [109] IRanges_2.28.0              gridExtra_2.3              </span>
<span class="co">#&gt; [111] compiler_4.1.2              tibble_3.1.6               </span>
<span class="co">#&gt; [113] V8_4.0.0                    KernSmooth_2.23-20         </span>
<span class="co">#&gt; [115] crayon_1.4.2                htmltools_0.5.2            </span>
<span class="co">#&gt; [117] mgcv_1.8-38                 later_1.3.0                </span>
<span class="co">#&gt; [119] tidyr_1.2.0                 lubridate_1.8.0            </span>
<span class="co">#&gt; [121] DBI_1.1.2                   MASS_7.3-54                </span>
<span class="co">#&gt; [123] rappdirs_0.3.3              car_3.0-12                 </span>
<span class="co">#&gt; [125] Matrix_1.3-4                cli_3.1.1                  </span>
<span class="co">#&gt; [127] parallel_4.1.2              igraph_1.2.11              </span>
<span class="co">#&gt; [129] GenomicRanges_1.46.1        pkgconfig_2.0.3            </span>
<span class="co">#&gt; [131] pkgdown_2.0.2               plotly_4.10.0              </span>
<span class="co">#&gt; [133] spatstat.sparse_2.1-0       annotate_1.72.0            </span>
<span class="co">#&gt; [135] bslib_0.3.1                 XVector_0.34.0             </span>
<span class="co">#&gt; [137] snakecase_0.11.0            stringr_1.4.0              </span>
<span class="co">#&gt; [139] digest_0.6.29               sctransform_0.3.3          </span>
<span class="co">#&gt; [141] RcppAnnoy_0.0.19            pracma_2.3.6               </span>
<span class="co">#&gt; [143] janitor_2.1.0               graph_1.72.0               </span>
<span class="co">#&gt; [145] spatstat.data_2.1-2         Biostrings_2.62.0          </span>
<span class="co">#&gt; [147] rmarkdown_2.11              leiden_0.3.9               </span>
<span class="co">#&gt; [149] uwot_0.1.11                 DelayedMatrixStats_1.16.0  </span>
<span class="co">#&gt; [151] GSEABase_1.56.0             curl_4.3.2                 </span>
<span class="co">#&gt; [153] shiny_1.7.1                 nlme_3.1-153               </span>
<span class="co">#&gt; [155] lifecycle_1.0.1             jsonlite_1.7.3             </span>
<span class="co">#&gt; [157] Rhdf5lib_1.16.0             carData_3.0-5              </span>
<span class="co">#&gt; [159] desc_1.4.0                  viridisLite_0.4.0          </span>
<span class="co">#&gt; [161] fansi_1.0.2                 pillar_1.7.0               </span>
<span class="co">#&gt; [163] lattice_0.20-45             KEGGREST_1.34.0            </span>
<span class="co">#&gt; [165] fastmap_1.1.0               httr_1.4.2                 </span>
<span class="co">#&gt; [167] survival_3.2-13             glue_1.6.1                 </span>
<span class="co">#&gt; [169] png_0.1-7                   bit_4.0.4                  </span>
<span class="co">#&gt; [171] stringi_1.7.6               sass_0.4.0                 </span>
<span class="co">#&gt; [173] HDF5Array_1.22.1            blob_1.2.2                 </span>
<span class="co">#&gt; [175] textshaping_0.3.6           BiocSingular_1.10.0        </span>
<span class="co">#&gt; [177] memoise_2.0.1               dplyr_1.0.7                </span>
<span class="co">#&gt; [179] irlba_2.3.5                 future.apply_1.8.1</span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Oscar Ospina, Brooke Fridley, Alex Soupir.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
