<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>spatialGE: Reading GeoMx-DSP outputs • spatialGE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="spatialGE: Reading GeoMx-DSP outputs">
<meta property="og:description" content="spatialGE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0.0000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/melanoma_ST_vignette.html">spatialGE: Guide to spatial heterogeneity analysis in melanoma (Thrane et al. 2018)</a>
    </li>
    <li>
      <a href="../articles/prostatecancer_Visium_vignette.html">spatialGE: Kriging for Visium data sets</a>
    </li>
    <li>
      <a href="../articles/read_GeoMx_output.html">spatialGE: Reading GeoMx-DSP outputs</a>
    </li>
    <li>
      <a href="../articles/spatialGE_troubleshoot.html">Troubleshooting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>spatialGE: Reading GeoMx-DSP outputs</h1>
            
      
      
      <div class="hidden name"><code>read_GeoMx_output.Rmd</code></div>

    </div>

    
    
<p>The package <code>spatialGE</code> provides a set of tools for the
visualization of spatially-resolved transcriptomics. In this vignette we
provide an example on how to read GeoMx-DSP data for downstream analysis
with <code>spatialGE</code>. The GeoMx platform produces data for a
series of region of interest (ROIs). In order to read GeoMx data in
<code>spatialGE</code>, three components are required: 1. Count data for
each ROI in the form of <code>.dcc</code> files (one per ROI) 2. Spatial
information within the sample annotation file (usually needs to be
exported from the GeoMx analysis suite) 3. A probe annotation file
(<code>.pkc</code>) to obtain gene names.</p>
<p>For this example, we will use the Nanostring annotated kidney data
set containing 4 diabetic kidney disease samples and 3 normal kidney
samples. Each tissue slice contains a variable number of ROIs.</p>
<div class="section level2">
<h2 id="data-acquisition">Data acquisition<a class="anchor" aria-label="anchor" href="#data-acquisition"></a>
</h2>
<p>The data set is publicly available and can be downloaded from the
Nanostring Resources <a href="https://www.nanostring.com/resources/annotated-kidney-dataset/" class="external-link">website</a>.
Users need to register to access the data at this point. Then, click the
“DOWNLOAD FILE” button, and proceed to obtain the “GeoMxTools Input
Files: Kidney_Dataset_for_GeomxTools.zip (17MB)” file. This file
contains the <code>.dcc</code> and <code>.pkc</code> files.</p>
<p>Next, click the “Kidney Dataset” link and download the “Sample
Annotations: Kidney_Sample_Annotations.txt” file (users will probably
need to download by right-cliking the link). This file contains sample
metadata, as well as the spatial information (x, y coordinates).
Optionally, download the images in “Slide images: ROI_reports.zip”
(~2.5Gb).</p>
<p>Place all the files in a folder in your desktop. In this vignette, we
have named that folder as “geomx_input”. Proceed to decompress the
zipped files (including those within the “ROI reports” folder, in case
that file was downloaded).</p>
</div>
<div class="section level2">
<h2 id="spatialge-installation">spatialGE installation<a class="anchor" aria-label="anchor" href="#spatialge-installation"></a>
</h2>
<p>The <code>spatialGE</code> repository is available at GitHub and can
be installed via <code>devtools</code>. To install
<code>devtools</code>, in case is not installed in your R console,
please run the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="st">"devtools"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>After making sure <code>devtools</code> is installed, proceed to
install <code>spatialGE</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># devtools::install_github("fridleylab/spatialGE")</span></code></pre></div>
<p>To use <code>spatialGE</code>, load the package using the
command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">'spatialGE'</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="creating-an-stlist-spatial-transcriptomics-list">Creating an STList (Spatial Transcriptomics List)<a class="anchor" aria-label="anchor" href="#creating-an-stlist-spatial-transcriptomics-list"></a>
</h2>
<p>Raw data and results from analyses are stored in an STList (S4 class
object). The STList can be created with the function
<code><a href="../reference/STList.html">STList()</a></code>, and in the case of GeoMx-DSP data,
<code>.dcc</code> files are passed to the <code>rnacounts</code>
argument. The <code>.pkc</code> file is passed to
<code>gmx_pkc</code>.</p>
<p>The annotation data is passed to <code>samples</code>, however,
additional arguments are required. The <code>gmx_slide_col</code>
indicated the column name within the annotation file with the slide ID
(each slide contains several ROIs and is considered a sample in
<code>spatialGE</code>). The <code>gmx_roi_col</code> indicates the
column with ROI IDs, matching the ROI IDs in the <code>.dcc</code>
files. <code>gmx_x_col</code> and <code>gmx_y_col</code> indicate the
columns with x and y coordinates. Additional columns from the annotation
file can be imported via the <code>gmx_meta_cols</code> argument.</p>
<p>Assuming the user created folders using the instructions in this
vignette (see <em>Data acquisition</em>), the output files are specified
like so:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dcc_files</span> <span class="op">=</span>  <span class="st">"~/Desktop/geomx_input/Kidney_Dataset/dccs"</span>
<span class="va">pkc</span> <span class="op">=</span> <span class="st">'~/Desktop/geomx_input/Kidney_Dataset/pkcs/TAP_H_WTA_v1.0.pkc'</span></code></pre></div>
<p>Annotation data is specified here:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sample_annots</span> <span class="op">=</span> <span class="st">"~/Desktop/geomx_input/Kidney_Sample_Annotations.txt"</span>
<span class="va">slide_col</span><span class="op">=</span><span class="st">"SlideName"</span>
<span class="va">id_col</span><span class="op">=</span><span class="st">"Sample_ID"</span>
<span class="va">x_col</span><span class="op">=</span><span class="st">"ROICoordinateX"</span>
<span class="va">y_col</span><span class="op">=</span><span class="st">"ROICoordinateY"</span>
<span class="va">meta_cols</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'disease_status'</span>,   <span class="st">'pathology'</span>, <span class="st">'region'</span><span class="op">)</span></code></pre></div>
<p>Then, we use these arguments with the <code>STList</code>
function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kdgeomx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STList.html">STList</a></span><span class="op">(</span>rnacounts<span class="op">=</span><span class="va">dcc_files</span>, gmx_pkc<span class="op">=</span><span class="va">pkc</span>, samples<span class="op">=</span><span class="va">sample_annots</span>,
                  gmx_slide_col<span class="op">=</span><span class="va">slide_col</span>, 
                  gmx_roi_col<span class="op">=</span><span class="va">id_col</span>, 
                  gmx_x_col<span class="op">=</span><span class="va">x_col</span>, 
                  gmx_y_col<span class="op">=</span><span class="va">y_col</span>, 
                  gmx_meta_cols<span class="op">=</span><span class="va">meta_cols</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #BBBB00;">Found GeoMx DCC output.</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; Requested 7 Samples</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; Cleaning Count and Coordinate Data Gene Names.</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; Converting Counts to Sparse Matrices</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; </span><span style="color: #00BB00; font-weight: bold;">Completed STList!</span></span>
<span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; </span></span></code></pre></div>
<p>The <code>kdgeomx</code> object is an STList.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kdgeomx</span>
<span class="co">#&gt; Spatial Transcriptomics List (STList).</span>
<span class="co">#&gt; 7 spatial array(s):</span>
<span class="co">#&gt;  disease3</span>
<span class="co">#&gt;      disease4</span>
<span class="co">#&gt;      normal3</span>
<span class="co">#&gt;      normal4</span>
<span class="co">#&gt;      disease1B</span>
<span class="co">#&gt;      disease2B</span>
<span class="co">#&gt;      normal2B</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 3 variables in sample data:</span>
<span class="co">#&gt;   disease_status, pathology, region</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-images">Loading images<a class="anchor" aria-label="anchor" href="#loading-images"></a>
</h2>
<p>If images were downloaded (“ROI reports”), users can load them into
the STList using the <code><a href="../reference/load_images.html">load_images()</a></code> function. The function
only requires two arguments: The STList to load images to, and a folder
with the images. The file names of the images must match at least
partially, the sample names in the STList (i.e. “disease3”, “disease4”,
“normal3”, etc…)</p>
<p>First, we prepare a folder with the images to load. In your
“geomx_input” folder, create a sub folder called “images”. As previously
mentioned, after decompressing the “ROI reports” file, users should
decompress the zipped files within. From each of the resulting folder,
copy the file ending in “_scan.png”. For example, from the folder
“disease1B_scan”, users should copy “disease1B_scan.png”. This file is
the image of the entire slide.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">images</span> <span class="op">=</span> <span class="st">'~/Desktop/geomx_input/images/'</span>
<span class="va">kdgeomx</span> <span class="op">=</span> <span class="fu"><a href="../reference/load_images.html">load_images</a></span><span class="op">(</span><span class="va">kdgeomx</span>, <span class="va">images</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-normalization-of-spatially-resolved-transcriptomic-data">Data normalization of spatially-resolved transcriptomic data<a class="anchor" aria-label="anchor" href="#data-normalization-of-spatially-resolved-transcriptomic-data"></a>
</h2>
<p>The suggested workflow for analysis of GeoMx includes quantile
normalization of counts. This type of normalization has not been
implemented in <code>spatialGE</code>, and thus, we will use the log
transformation in the function <code><a href="../reference/spatialTransform.html">spatialTransform()</a></code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kdgeomx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialTransform.html">spatialTransform</a></span><span class="op">(</span><span class="va">kdgeomx</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization-of-gene-expression">Visualization of gene expression<a class="anchor" aria-label="anchor" href="#visualization-of-gene-expression"></a>
</h2>
<p>We can plot now the gene expression levels of specific genes. Let’s
create plots for <em>TXNIP</em>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_quilt.html">plot_gene_quilt</a></span><span class="op">(</span><span class="va">kdgeomx</span>, image<span class="op">=</span><span class="cn">T</span>, genes<span class="op">=</span><span class="st">'TXNIP'</span>, visium<span class="op">=</span><span class="cn">F</span>, color_pal<span class="op">=</span><span class="st">'sunset'</span>, ptsize<span class="op">=</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
<p>The <code>plot_gene_quilt</code> returns a list of plots. We can use
<code>ggarrange()</code> to display these plots in an organized manner.
The tissue images in the right show a number of areas that are colored
different to the background, which correspond to the sampled ROIs. The
quilt plots on the left show the expression of <em>TXNIP</em> in those
ROIs. We have envisioned for future versions of <code>spatialGE</code>,
plotting options allowing users to visiualize gene expression directly
over the images.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">qplot</span>, ncol<span class="op">=</span><span class="fl">2</span>, nrow<span class="op">=</span><span class="fl">7</span><span class="op">)</span></code></pre></div>
<p><img src="read_GeoMx_output_files/figure-html/unnamed-chunk-4-1.png" width="768"></p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.7</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] spatialGE_1.0.0.0000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] colorspace_2.0-2            ggsignif_0.6.3             </span>
<span class="co">#&gt;   [3] rjson_0.2.21                ellipsis_0.3.2             </span>
<span class="co">#&gt;   [5] mclust_5.4.9                rprojroot_2.0.2            </span>
<span class="co">#&gt;   [7] snakecase_0.11.0            XVector_0.34.0             </span>
<span class="co">#&gt;   [9] fftwtools_0.9-11            GenomicRanges_1.46.1       </span>
<span class="co">#&gt;  [11] fs_1.5.2                    rstudioapi_0.13            </span>
<span class="co">#&gt;  [13] ggpubr_0.4.0                farver_2.1.0               </span>
<span class="co">#&gt;  [15] bit64_4.0.5                 AnnotationDbi_1.56.2       </span>
<span class="co">#&gt;  [17] fansi_1.0.2                 lubridate_1.8.0            </span>
<span class="co">#&gt;  [19] sparseMatrixStats_1.6.0     cachem_1.0.6               </span>
<span class="co">#&gt;  [21] knitr_1.37                  jsonlite_1.7.3             </span>
<span class="co">#&gt;  [23] broom_0.7.12                annotate_1.72.0            </span>
<span class="co">#&gt;  [25] png_0.1-7                   graph_1.72.0               </span>
<span class="co">#&gt;  [27] HDF5Array_1.22.1            readr_2.1.2                </span>
<span class="co">#&gt;  [29] compiler_4.1.2              httr_1.4.2                 </span>
<span class="co">#&gt;  [31] backports_1.4.1             assertthat_0.2.1           </span>
<span class="co">#&gt;  [33] Matrix_1.3-4                fastmap_1.1.0              </span>
<span class="co">#&gt;  [35] cli_3.1.1                   BiocSingular_1.10.0        </span>
<span class="co">#&gt;  [37] htmltools_0.5.2             tools_4.1.2                </span>
<span class="co">#&gt;  [39] rsvd_1.0.5                  gtable_0.3.0               </span>
<span class="co">#&gt;  [41] glue_1.6.1                  GenomeInfoDbData_1.2.7     </span>
<span class="co">#&gt;  [43] dplyr_1.0.7                 Rcpp_1.0.8                 </span>
<span class="co">#&gt;  [45] carData_3.0-5               EBImage_4.36.0             </span>
<span class="co">#&gt;  [47] estimate_1.0.13             Biobase_2.54.0             </span>
<span class="co">#&gt;  [49] jquerylib_0.1.4             pkgdown_2.0.2              </span>
<span class="co">#&gt;  [51] vctrs_0.3.8                 Biostrings_2.62.0          </span>
<span class="co">#&gt;  [53] rhdf5filters_1.6.0          xCell_1.1.0                </span>
<span class="co">#&gt;  [55] DelayedMatrixStats_1.16.0   xfun_0.29                  </span>
<span class="co">#&gt;  [57] stringr_1.4.0               beachmat_2.10.0            </span>
<span class="co">#&gt;  [59] lifecycle_1.0.1             irlba_2.3.5                </span>
<span class="co">#&gt;  [61] rstatix_0.7.0               XML_3.99-0.8               </span>
<span class="co">#&gt;  [63] zlibbioc_1.40.0             MASS_7.3-54                </span>
<span class="co">#&gt;  [65] scales_1.1.1                vroom_1.5.7                </span>
<span class="co">#&gt;  [67] ragg_1.2.2                  hms_1.1.1                  </span>
<span class="co">#&gt;  [69] MatrixGenerics_1.6.0        parallel_4.1.2             </span>
<span class="co">#&gt;  [71] SummarizedExperiment_1.24.0 rhdf5_2.38.0               </span>
<span class="co">#&gt;  [73] SingleCellExperiment_1.16.0 yaml_2.2.2                 </span>
<span class="co">#&gt;  [75] khroma_1.8.0                memoise_2.0.1              </span>
<span class="co">#&gt;  [77] ggplot2_3.3.5               sass_0.4.0                 </span>
<span class="co">#&gt;  [79] stringi_1.7.6               RSQLite_2.2.9              </span>
<span class="co">#&gt;  [81] GSVA_1.42.0                 highr_0.9                  </span>
<span class="co">#&gt;  [83] S4Vectors_0.32.3            desc_1.4.0                 </span>
<span class="co">#&gt;  [85] ScaledMatrix_1.2.0          tiff_0.1-11                </span>
<span class="co">#&gt;  [87] BiocGenerics_0.40.0         BiocParallel_1.28.3        </span>
<span class="co">#&gt;  [89] GenomeInfoDb_1.30.1         rlang_1.0.0                </span>
<span class="co">#&gt;  [91] pkgconfig_2.0.3             systemfonts_1.0.4          </span>
<span class="co">#&gt;  [93] matrixStats_0.61.0          bitops_1.0-7               </span>
<span class="co">#&gt;  [95] pracma_2.3.6                evaluate_0.14              </span>
<span class="co">#&gt;  [97] lattice_0.20-45             purrr_0.3.4                </span>
<span class="co">#&gt;  [99] Rhdf5lib_1.16.0             labeling_0.4.2             </span>
<span class="co">#&gt; [101] htmlwidgets_1.5.4           cowplot_1.1.1              </span>
<span class="co">#&gt; [103] bit_4.0.4                   tidyselect_1.1.1           </span>
<span class="co">#&gt; [105] GSEABase_1.56.0             magrittr_2.0.2             </span>
<span class="co">#&gt; [107] R6_2.5.1                    IRanges_2.28.0             </span>
<span class="co">#&gt; [109] generics_0.1.2              DelayedArray_0.20.0        </span>
<span class="co">#&gt; [111] DBI_1.1.2                   pillar_1.7.0               </span>
<span class="co">#&gt; [113] abind_1.4-5                 KEGGREST_1.34.0            </span>
<span class="co">#&gt; [115] RCurl_1.98-1.5              tibble_3.1.6               </span>
<span class="co">#&gt; [117] car_3.0-12                  janitor_2.1.0              </span>
<span class="co">#&gt; [119] crayon_1.4.2                utf8_1.2.2                 </span>
<span class="co">#&gt; [121] tzdb_0.2.0                  rmarkdown_2.11             </span>
<span class="co">#&gt; [123] jpeg_0.1-9                  locfit_1.5-9.4             </span>
<span class="co">#&gt; [125] grid_4.1.2                  blob_1.2.2                 </span>
<span class="co">#&gt; [127] digest_0.6.29               xtable_1.8-4               </span>
<span class="co">#&gt; [129] tidyr_1.2.0                 textshaping_0.3.6          </span>
<span class="co">#&gt; [131] stats4_4.1.2                munsell_0.5.0              </span>
<span class="co">#&gt; [133] bslib_0.3.1</span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Oscar Ospina, Brooke Fridley, Alex Soupir.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
