<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Detection of gene set spatial patterns and expression gradients in CosMx-SMI • spatialGE</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Detection of gene set spatial patterns and expression gradients in CosMx-SMI">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basic_functions_vignette.html">Basic functionality: Visualization, domain detection, and spatial heterogeneity</a>
    </li>
    <li>
      <a href="../articles/spatial_differential_expression.html">Spatially-informed differential gene expression</a>
    </li>
    <li>
      <a href="../articles/spatial_enrichment_gradients_smi.html">Detection of gene set spatial patterns and expression gradients in CosMx-SMI</a>
    </li>
    <li>
      <a href="../articles/troubleshooting.html">Troubleshooting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Detection of gene set spatial patterns and
expression gradients in CosMx-SMI</h1>
            
            <h4 data-toc-skip class="date">2025-04-07</h4>
      

      <div class="hidden name"><code>spatial_enrichment_gradients_smi.Rmd</code></div>

    </div>

    
    
<p><img src="img%2Fsmi_dir_structure.png" width="161"><img src="img%2Fdiagram_stenrich.png" width="1342"><img src="img%2Fspatial_gradients.png" width="1436"></p>
<p>The package <strong><em><code>spatialGE</code></em></strong> can be
used to detect spatial patterns in gene expression at the gene and gene
set level. Data from multiple spatial transcriptomics platforms can be
analyzed, as long as gene expression counts per spot or cell are
associated with spatial coordinates of those spots/cells. The
<strong><em><code>spatialGE</code></em></strong> package is also
compatible with outputs from the Visium workflow (i.e., Space Ranger) or
the CosMx-SMI spatial single-cell platform.</p>
<p>In this tutorial, the functions <code>STenrich</code>,
<code>STclust</code>, and <code>STgradient</code> will be used to test
for spatial gene set enrichment and genes showing expression gradients
on a <a href="https://nanostring.com/products/cosmx-spatial-molecular-imager/ffpe-dataset/nsclc-ffpe-dataset/" class="external-link">CosMx-SMI
data set</a> from Non-Small Cell Lung Cancer (NSCLC).</p>
<details><summary><strong>How is <code>spatialGE</code> installed? </strong>
</summary><p>The <code>spatialGE</code> repository is available at GitHub and can
be installed via <code>devtools</code>. To install <code>devtools</code>
(in case it is not already installed in your R), please run the
following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="st">"devtools"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>After making sure <code>devtools</code> is installed, proceed to
install <code>spatialGE</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># devtools::install_github("fridleylab/spatialGE")</span></span></code></pre></div>
</details><div class="section level2">
<h2 id="spatial-gene-set-enrichment-and-expression-gradients-in-non-small-cell-lung-cancer-nsclc">Spatial gene set enrichment and expression gradients in non-small
cell lung cancer (NSCLC)<a class="anchor" aria-label="anchor" href="#spatial-gene-set-enrichment-and-expression-gradients-in-non-small-cell-lung-cancer-nsclc"></a>
</h2>
<p>The CosMx-SMI platform generates single-cell level gene expression
with associated x and y locations of the cells where measurements were
taken. For the data set used in this tutorial, RNA counts were measured
for 960 genes in about 800K cells. The data was generated from eight
tissue slices, with counts obtained for a series of Fields of Vision
(FOVs) within each slice. However, in this tutorial, the data set has
been subset to 6 FOVs originating from two of the slices.</p>
<p>The complete data set can be downloaded from the links below. It is
also likely that registration in the website is necessary in order to
download the data set.</p>
<ul>
<li><a href="https://nanostring.com/resources/smi-ffpe-dataset-lung5-rep1-data/" class="external-link">Lung
5-1</a></li>
<li><a href="https://nanostring.com/resources/smi-ffpe-dataset-lung6-data/" class="external-link">Lung
6</a></li>
</ul>
<p>The data from the 6 FOVs used in this tutorial is available at the <a href="https://github.com/FridleyLab/spatialGE_Data" class="external-link">spatialGE_Data
GitHub repository</a>. Data will be downloaded and deposited in a
temporary directory. However, the download path can be changed within
the following code block:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nsclc_tmp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">nsclc_tmp</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">nsclc_tmp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">'https://github.com/FridleyLab/spatialGE_Data/raw/refs/heads/main/nanostring_nsclc.zip?download='</span>, </span>
<span>              destfile<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">nsclc_tmp</span>, <span class="st">'/'</span>, <span class="st">'nanostring_nsclc.zip'</span><span class="op">)</span>, mode<span class="op">=</span><span class="st">'wb'</span><span class="op">)</span></span>
<span><span class="va">zip_tmp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">nsclc_tmp</span>, pattern<span class="op">=</span><span class="st">'nanostring_nsclc.zip$'</span>, full.names<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span>zipfile<span class="op">=</span><span class="va">zip_tmp</span>, exdir<span class="op">=</span><span class="va">nsclc_tmp</span><span class="op">)</span></span></code></pre></div>
<p>Data for each tissue slice is stored in separate folders. The
structure of the files is shown below:</p>
<p align="center">
<img src="img%2Fsmi_dir_structure.png" height="250" width="550"></p>
<p>The key files are shown enclosed in orange (Lung 5-1) and green (Lung
6) boxes. The “_exprMat_file.csv” file of each sample contains the gene
counts. The “_metadata_file.csv” contains the cell x, y coordinates for
each sample. Optionally, the “CellComposite” folder contains images that
can be imported into the STlist (details later in this tutorial).</p>
<p>From the temporary directory, the file paths to be passed to the
<code>STlist</code> function are:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Find expression and coordinates files within the temporary folder</span></span>
<span><span class="va">exprmats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">nsclc_tmp</span>, <span class="st">'/nanostring_nsclc/Lung5_Rep1-Flat_files_and_images/Lung5_Rep1_exprMat_file_SUBSET.csv'</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">nsclc_tmp</span>, <span class="st">'/nanostring_nsclc/Lung6-Flat_files_and_images/Lung6_exprMat_file_SUBSET.csv'</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">metas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">nsclc_tmp</span>, <span class="st">'/nanostring_nsclc/Lung5_Rep1-Flat_files_and_images/Lung5_Rep1_metadata_file_SUBSET.csv'</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">nsclc_tmp</span>, <span class="st">'/nanostring_nsclc/Lung6-Flat_files_and_images/Lung6_metadata_file_SUBSET.csv'</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now, that the data is ready, load the <code>spatialGE</code>
package:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">'spatialGE'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="creating-an-stlist-spatial-transcriptomics-list">Creating an STList (Spatial Transcriptomics List)<a class="anchor" aria-label="anchor" href="#creating-an-stlist-spatial-transcriptomics-list"></a>
</h2>
<p>The starting point of a <code>spatialGE</code> analysis is the
creation of an <strong><em>STlist</em></strong> (S4 class object), which
stores raw and processed data. The STlist is created with the function
<code>STlist</code>, which can take data in multiple formats (see <a href="https://fridleylab.github.io/spatialGE/reference/STlist.html" class="external-link">here</a>
for more info or type <code><a href="../reference/STlist.html">?STlist</a></code> in the R console).</p>
<p>To load the files into an STlist, please use these commands:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lung</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STlist.html">STlist</a></span><span class="op">(</span>rnacounts<span class="op">=</span><span class="va">exprmats</span>, </span>
<span>               spotcoords<span class="op">=</span><span class="va">metas</span>, </span>
<span>               samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Lung5_Rep1'</span>, <span class="st">'Lung6'</span><span class="op">)</span>, cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Found CosMx-SMI data</span></span>
<span><span class="co">#&gt; Found 2 CosMx-SMI samples</span></span>
<span><span class="co">#&gt;  Data read completed</span></span>
<span><span class="co">#&gt; Matching gene expression and coordinate data...</span></span>
<span><span class="co">#&gt; Converting counts to sparse matrices</span></span>
<span><span class="co">#&gt; Completed STlist!</span></span></code></pre></div>
<p>To obtain count statistics, the <code>summarize_STlist</code>
function can be used:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summ_df</span> <span class="op">=</span> <span class="fu"><a href="../reference/summarize_STlist.html">summarize_STlist</a></span><span class="op">(</span><span class="va">lung</span><span class="op">)</span></span>
<span><span class="va">summ_df</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 9</span></span></span>
<span><span class="co">#&gt;   sample_name     spotscells genes min_counts_per_spotc…¹ mean_counts_per_spot…²</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Lung5_Rep1_fov…       <span style="text-decoration: underline;">3</span>744   980                      2                   326.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Lung5_Rep1_fov…       <span style="text-decoration: underline;">5</span>272   980                      0                   328.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Lung5_Rep1_fov…       <span style="text-decoration: underline;">3</span>946   980                      2                   367.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Lung6_fov_4           <span style="text-decoration: underline;">3</span>310   980                      1                   304.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Lung6_fov_6           <span style="text-decoration: underline;">3</span>325   980                      0                   523.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Lung6_fov_7           <span style="text-decoration: underline;">3</span>477   980                      0                   312.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ abbreviated names: ¹​min_counts_per_spotcell, ²​mean_counts_per_spotcell</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 4 more variables: max_counts_per_spotcell &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   min_genes_per_spotcell &lt;int&gt;, mean_genes_per_spotcell &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   max_genes_per_spotcell &lt;int&gt;</span></span></span></code></pre></div>
<p>Some cells have zero counts. We can look at the distribution of
counts per cell for the first five and last five FOVs using the
<code>distribution_plots</code> function:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/distribution_plots.html">distribution_plots</a></span><span class="op">(</span><span class="va">lung</span>, plot_type<span class="op">=</span><span class="st">'violin'</span>, plot_meta<span class="op">=</span><span class="st">'total_counts'</span>, samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">56</span><span class="op">:</span><span class="fl">60</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Sample NA does not contain total_counts.</span></span>
<span><span class="co">#&gt; Sample NA does not contain total_counts.</span></span>
<span><span class="co">#&gt; Sample NA does not contain total_counts.</span></span>
<span><span class="co">#&gt; Sample NA does not contain total_counts.</span></span>
<span><span class="co">#&gt; Sample NA does not contain total_counts.</span></span>
<span><span class="va">cp</span><span class="op">[[</span><span class="st">'total_counts'</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="spatial_enrichment_gradients_smi_files/figure-html/count_dstr-1.png" width="768"></p>
<p>Cells with zero counts, can be removed using the
<code>filter_data</code> function. The function can also be used to
remove all counts from specific genes. This option is useful in this
case, as CosMx-SMI panels include negative probe genes. While the
negative probes can be used in normalization of counts, in this tutorial
those genes will be removed. Negative probes in CosMx-SMI begin with the
token “NegPrb”:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lung</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_data.html">filter_data</a></span><span class="op">(</span><span class="va">lung</span>, spot_minreads<span class="op">=</span><span class="fl">20</span>, rm_genes_expr<span class="op">=</span><span class="st">'^NegPrb'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="transformation-of-spatially-resolved-transcriptomics-data">Transformation of spatially-resolved transcriptomics data<a class="anchor" aria-label="anchor" href="#transformation-of-spatially-resolved-transcriptomics-data"></a>
</h2>
<p>The function <code>transform_data</code> allows data transformation
using one of two possible options. The first options applies
log-transformation to the counts, after library size normalization
performed on each sample separately. The second option applies
variance-stabilizing transformation (SCT; <span class="citation">Hafemeister and Satija (2019)</span>), which is a
method increasingly used in single-cell and spatial transcriptomics
studies. To apply SCT transformation, use the following command:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lung</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transform_data.html">transform_data</a></span><span class="op">(</span><span class="va">lung</span>, method<span class="op">=</span><span class="st">'sct'</span>, cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="detecting-gene-sets-with-spatial-aggregation-patterns">Detecting gene sets with spatial aggregation patterns<a class="anchor" aria-label="anchor" href="#detecting-gene-sets-with-spatial-aggregation-patterns"></a>
</h2>
<p>An important part of gene expression analysis is the use of gene sets
to make inferences about the functional significance of changes in
expression. Normally, this is achieved by conducting a Gene Set
Enrichment Analysis (GSEA). While GSEA can be completed in a similar
fashion to scRNA-seq, it is possible with the <code>STenrich</code>
function to test for gene sets that show spatial non-uniform enrichment.
In other words, <code>STenrich</code> tests whether expression of a gene
set is concentrated in one or few areas of the tissue.</p>
<p>The algorithm in <code>STenrich</code> is depicted in the following
diagram:</p>
<p align="center">
<img src="img%2Fdiagram_stenrich.png" height="250" width="550"></p>
<p>The <code>STenrich</code> function is a modification of the method
proposed by <span class="citation">Hunter et al. (2021)</span>. The
steps in this modified version are as follows: First, Euclidean
distances are calculated among all spots/cells. Then, the gene
expression of the sample is subset to the genes within the gene set
being tested. If too few genes are left after subset
(<code>min_genes</code>), then the gene set is omitted for that sample.
The average expression and standard deviation of those genes is
calculated for each spot/cell. Next, spots/cells with gene set
expression above the average gene set expression across all spots/cells
are identified. The threshold to define these high gene set expression
spots is defined by the average gene set expression plus a number
<code>min_sds</code> of standard deviations. The sum of the Euclidean
distances between the high expression spots/cells is calculated.</p>
<p>The next step involves a permutation process, in which a null
distribution is generated in order to test if the (sum of) distances
among high expression spots are smaller than expected. To that end, a
random sample of spots/cells (regardless of expression) is selected. The
random sample has the same size as the number of high expression
spots/cells. Then, the sum of distances among the randomly sampled
spots/cells is calculated. The random selection is repeated as many
times as requested (`reps``). Finally, a p-value is calculated by noting
how many times the sum of random distances was higher than the sum of
distances among high expression spots/cells. If the sum of random
distances was most of the times higher than the sum of distances among
high expression spots/cells, then the null hypothesis of no spatial
aggregation is rejected (i.e., spots/cells with high gene set expression
are more aggregated than expected by chance).</p>
<p>A few important notes about the algorithm to consider:</p>
<ul>
<li>Notice that the metric of enrichment is currently the average
expression of the genes in a set within each ROI/spot/cell. Better
metrics of gene set enrichment are available and will be soon available
to use instead of average.</li>
<li>Special attention should be paid to the reference genome used to
annotate gene counts. The annotation of transcripts precedes any
analysis in spatialGE. Nevertheless, if transcripts are annotated with a
mouse genome (or other species), the user should use the appropriate
gene set database, as gene names will likely not match to a gene set
database with human gene names. Furthermore, but not less important,
there might be problems with gene homology.</li>
<li>The more permutations are requested, the longer the execution, but
p-value estimates are more accurate.</li>
<li>By increasing the number of standard deviations, fewer spots/cells
are selected as “high expression” (stringent analysis), however, it
becomes more challenging to detect gene sets that show evidence of
spatial aggregation. Values between 1 to 2 standard deviations are
likely suitable for most studies.</li>
<li>By adjusting the minimum number of spots (<code>min_spots</code>),
the user has some rough control over the size of the structures to be
studied. For example, small immune infiltrates could be detected with
smaller number of minimum spots. Conversely, if tissue domain-level
differences are sought, then a larger Minimum number of spots should be
set.</li>
<li>Take note of the random seed number used for the test (Seed number
parameter), in case you need to generate the same results later.
Nevertheless, running the analysis several times with different seed
numbers is advisable to check for consistency.</li>
</ul>
<p>The first step to run <code>STenrich</code> is to obtain a list of
gene sets. Here, the <code>msigdbr</code> package will be used to obtain
HALLMARK gene sets:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load msigdbr</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://igordot.github.io/msigdbr/" class="external-link">'msigdbr'</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get HALLMARK gene sets</span></span>
<span><span class="va">gene_sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr.html" class="external-link">msigdbr</a></span><span class="op">(</span>species<span class="op">=</span><span class="st">'Homo sapiens'</span><span class="op">)</span></span>
<span><span class="co"># Select the appropriate column depending on the msigdbr version</span></span>
<span><span class="va">gs_column</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">'gs_cat|gs_collection'</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span>, value<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">gene_sets</span> <span class="op">&lt;-</span> <span class="va">gene_sets</span><span class="op">[</span><span class="va">gene_sets</span><span class="op">[[</span><span class="va">gs_column</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="st">"H"</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Convert gene set dataframe to list</span></span>
<span><span class="co"># The result is a named list. The names of the list are the names of each gene set</span></span>
<span><span class="co"># The contents of each list element are the gene names within each gene set</span></span>
<span><span class="va">gene_sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">gene_sets</span><span class="op">[[</span><span class="st">'gene_symbol'</span><span class="op">]</span><span class="op">]</span>, f<span class="op">=</span><span class="va">gene_sets</span><span class="op">[[</span><span class="st">'gs_name'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>For the purpose of this example, we will only test “signaling”
pathways:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_sets</span> <span class="op">=</span> <span class="va">gene_sets</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">'SIGNALING'</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Please keep in mind that it is assumed that HUGO gene names are used.
For CosMx-SMI and Visium, normally <em>HUGO names</em> are provided. If
spatial transcriptomics data from a non-human species is used, the
appropriate database should be used. Otherwise, <code>spatialGE</code>
will not be able to identify genes within the gene set.</p>
<p>The <code>STenrich</code> function can now be used. For this
analysis, 1000 replicates (<code>reps=1000</code>) will be used to
generate the null distribution. In addition, for a gene set to be
considered as highly expressed in cell, it should be expressed at least
over the average expression plus 1.5 standard deviations
(<code>num_sds=1.5</code>). Gene sets with less than five genes in the
sample (<code>min_genes=5</code>) and gene sets with less than 10 highly
expressed cells (<code>min_units=10</code>) will not be tested. Notice a
seed was set (<code>seed=12345</code>), which critical for replication
of the results in the future.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stenrich_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STenrich.html">STenrich</a></span><span class="op">(</span><span class="va">lung</span>, gene_sets<span class="op">=</span><span class="va">gene_sets</span>, </span>
<span>                        reps<span class="op">=</span><span class="fl">1000</span>, </span>
<span>                        num_sds<span class="op">=</span><span class="fl">1.5</span>, </span>
<span>                        min_genes<span class="op">=</span><span class="fl">5</span>,</span>
<span>                        min_units<span class="op">=</span><span class="fl">10</span>, </span>
<span>                        seed<span class="op">=</span><span class="fl">12345</span>, cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Running STenrich...</span></span>
<span><span class="co">#&gt;  Calculating average gene set expression...</span></span>
<span><span class="co">#&gt; STenrich completed in 0.41 min.</span></span></code></pre></div>
<p>The result of the function is a list of data frames, with one data
frame per samples (FOV in this case). For example, this is the data
frame for the first FOV in the STlist:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">stenrich_df</span><span class="op">$</span><span class="va">Lung5_Rep1_fov_1</span><span class="op">)</span></span>
<span><span class="co">#&gt;          sample_name                         gene_set size_test size_gene_set</span></span>
<span><span class="co">#&gt; 15 Lung5_Rep1_fov_11 HALLMARK_IL6_JAK_STAT3_SIGNALING        58           103</span></span>
<span><span class="co">#&gt; 27 Lung5_Rep1_fov_11       HALLMARK_KRAS_SIGNALING_UP        55           220</span></span>
<span><span class="co">#&gt; 57 Lung5_Rep1_fov_11 HALLMARK_TNFA_SIGNALING_VIA_NFKB        76           228</span></span>
<span><span class="co">#&gt; 9  Lung5_Rep1_fov_11     HALLMARK_IL2_STAT5_SIGNALING        58           216</span></span>
<span><span class="co">#&gt; 21 Lung5_Rep1_fov_11       HALLMARK_KRAS_SIGNALING_DN        20           220</span></span>
<span><span class="co">#&gt; 33 Lung5_Rep1_fov_11        HALLMARK_MTORC1_SIGNALING        13           211</span></span>
<span><span class="co">#&gt;    prop_size_test p_value adj_p_value</span></span>
<span><span class="co">#&gt; 15          0.563   0.000      0.0000</span></span>
<span><span class="co">#&gt; 27          0.250   0.004      0.0200</span></span>
<span><span class="co">#&gt; 57          0.333   0.009      0.0300</span></span>
<span><span class="co">#&gt; 9           0.269   0.213      0.5325</span></span>
<span><span class="co">#&gt; 21          0.091   0.295      0.5900</span></span>
<span><span class="co">#&gt; 33          0.062   1.000      1.0000</span></span></code></pre></div>
<p>Each row represents a test for the null hypothesis of no spatial
aggregation in the expression of the set in the “gene_set” column. The
column “size_test” is the number of genes of a gene set that were
present in the FOV. The larger this number the better, as it indicates a
better representation of the gene set in the sample. The “adj_p_value”
is the multiple test adjusted p-value, which is the value used to decide
if a gene set shows significant indications of a spatial pattern
(adj_p_value &lt; 0.05).</p>
<p>With a few lines of code, a visual summary can be generated that
presents the gene sets with adj_p_value &lt; 0.05 across FOVs for each
tissue slide:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load tidyverse for data frame manipulation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://tidyverse.tidyverse.org" class="external-link">'tidyverse'</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine all samples in a single data frame</span></span>
<span><span class="co"># Subset to gene sets showing significant evidence of spatial pattern (adj_p_value &lt; 0.05) and</span></span>
<span><span class="co"># proportion higher or equal than 0.3 of genes from a gene set present in sample </span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">stenrich_df</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>prop_gene_set<span class="op">=</span><span class="va">size_test</span><span class="op">/</span><span class="va">size_gene_set</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">prop_gene_set</span> <span class="op">&gt;=</span> <span class="fl">0.3</span> <span class="op">&amp;</span> <span class="va">adj_p_value</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>slide<span class="op">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract</a></span><span class="op">(</span><span class="va">sample_name</span>, <span class="st">"Lung5_Rep1|Lung6"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">slide</span>, <span class="va">gene_set</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>gene_set<span class="op">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace</a></span><span class="op">(</span><span class="va">gene_set</span>, <span class="st">'HALLMARK_'</span>, <span class="st">''</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate barplot showing the number of FOVs with siginificant evidence of spatial</span></span>
<span><span class="co"># patterns for each tissue slide</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">gene_set</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle<span class="op">=</span><span class="fl">70</span>, vjust<span class="op">=</span><span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">slide</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial_enrichment_gradients_smi_files/figure-html/unnamed-chunk-1-1.png" width="768"></p>
<p>The “NOTCH signaling” pathway showed spatial patterns less frequently
in the slide “Lung5_Rep1” compared to “Lung6”. With the function
<code>STplot</code>, the average gene set expression for this pathway
can be plotted. First, FOVs with and without NOTCH signaling spatial
patterns will be identified from “Lung5_Rep1” and “Lung6”
respectively.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">stenrich_df</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>prop_gene_set<span class="op">=</span><span class="va">size_test</span><span class="op">/</span><span class="va">size_gene_set</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">prop_gene_set</span> <span class="op">&gt;=</span> <span class="fl">0.3</span> <span class="op">&amp;</span> <span class="va">adj_p_value</span> <span class="op">&gt;=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">prop_gene_set</span><span class="op">)</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">adj_p_value</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gene_set</span> <span class="op">==</span> <span class="st">'HALLMARK_NOTCH_SIGNALING'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="va">sample_name</span>, <span class="st">"Lung5_Rep1"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt;         sample_name                 gene_set size_test size_gene_set</span></span>
<span><span class="co">#&gt; 1 Lung5_Rep1_fov_11 HALLMARK_NOTCH_SIGNALING        13            34</span></span>
<span><span class="co">#&gt; 2  Lung5_Rep1_fov_2 HALLMARK_NOTCH_SIGNALING        13            34</span></span>
<span><span class="co">#&gt; 3  Lung5_Rep1_fov_6 HALLMARK_NOTCH_SIGNALING        13            34</span></span>
<span><span class="co">#&gt;   prop_size_test p_value adj_p_value prop_gene_set</span></span>
<span><span class="co">#&gt; 1          0.382   0.989       1.000     0.3823529</span></span>
<span><span class="co">#&gt; 2          0.382   0.981       0.981     0.3823529</span></span>
<span><span class="co">#&gt; 3          0.382   0.858       0.977     0.3823529</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">stenrich_df</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>prop_gene_set<span class="op">=</span><span class="va">size_test</span><span class="op">/</span><span class="va">size_gene_set</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">prop_gene_set</span> <span class="op">&gt;=</span> <span class="fl">0.3</span> <span class="op">&amp;</span> <span class="va">adj_p_value</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">prop_gene_set</span><span class="op">)</span>, <span class="va">adj_p_value</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gene_set</span> <span class="op">==</span> <span class="st">'HALLMARK_NOTCH_SIGNALING'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="va">sample_name</span>, <span class="st">"Lung6"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt;   sample_name                 gene_set size_test size_gene_set prop_size_test</span></span>
<span><span class="co">#&gt; 1 Lung6_fov_4 HALLMARK_NOTCH_SIGNALING        13            34          0.382</span></span>
<span><span class="co">#&gt; 2 Lung6_fov_6 HALLMARK_NOTCH_SIGNALING        13            34          0.382</span></span>
<span><span class="co">#&gt; 3 Lung6_fov_7 HALLMARK_NOTCH_SIGNALING        13            34          0.382</span></span>
<span><span class="co">#&gt;   p_value adj_p_value prop_gene_set</span></span>
<span><span class="co">#&gt; 1   0.000 0.000000000     0.3823529</span></span>
<span><span class="co">#&gt; 2   0.000 0.000000000     0.3823529</span></span>
<span><span class="co">#&gt; 3   0.001 0.001666667     0.3823529</span></span></code></pre></div>
<p>Plots can be generated like so:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STplot.html">STplot</a></span><span class="op">(</span><span class="va">lung</span>, genes<span class="op">=</span><span class="va">gene_sets</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span> <span class="op">==</span> <span class="st">'HALLMARK_NOTCH_SIGNALING'</span><span class="op">]</span>, </span>
<span>             samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Lung5_Rep1_fov_2'</span>, <span class="st">'Lung5_Rep1_fov_6'</span>, <span class="st">'Lung5_Rep1_fov_11'</span>,</span>
<span>                       <span class="st">'Lung6_fov_4'</span>, <span class="st">'Lung6_fov_6'</span>, <span class="st">'Lung6_fov_7'</span><span class="op">)</span>,</span>
<span>             color_pal<span class="op">=</span><span class="st">'YlOrBr'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">qp</span>, ncol<span class="op">=</span><span class="fl">3</span>, nrow<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial_enrichment_gradients_smi_files/figure-html/notch_plots-1.png" width="960"></p>
<p>Although expression of the NOTCH signaling pathway is higher overall
in “Lung6”, the expression is not uniform, but rather is arranged in
patterns within each FOV. Conversely, cells in FOVs of “Lung5_Rep1” that
show high expression of the NOTCH signaling pathway are scattered across
the tissue.</p>
<p>Since this expression of this pathway has been identified as
spatially aggregated, it might be of interest to know if there is
correspondence of NOTCH signaling with tissue domains. With the
<code>STclust</code> function, tissue domains can be identified.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lung</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STclust.html">STclust</a></span><span class="op">(</span><span class="va">lung</span>, ws<span class="op">=</span><span class="fl">0.02</span>, ks<span class="op">=</span><span class="st">'dtc'</span>, cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; STclust started...</span></span>
<span><span class="co">#&gt; Updating STlist with results...</span></span>
<span><span class="co">#&gt; STclust completed in 0.41 min.</span></span></code></pre></div>
<p>Optinally, if tissue images are available for some or all the FOVs,
users can upload those to the STlist for display next to the other plot
types available in <code>spatialGE</code>. Users can upload an entire
folder of images to the STlist, as long as image files have the same
names as the FOVs within the STlist (e.g., “Lung5_Rep1_fov_2”,
“Lung6_fov_4”, etc). Keep in mind, however, this will increase the size
of the STlist significantly, especially in CosMx experiments that
usually contain many FOVs.</p>
<p>The multiplexed immunofluorescence images have been previously
downloaded from the <a href="https://github.com/FridleyLab/spatialGE_Data" class="external-link">spatialGE_Data
GitHub repository</a>. The temporary folder created at the beggining of
this tutorial will be used to generate the file paths to the folder
containng the images. Importantly, the images need to be renamed
according to the FOV name in the STlist. This renaming was already
performed for the files in the spatialGE_Data GitHub repository. Here is
an example of how the files were renamed from the original data set:</p>
<ul>
<li>For tissue slice “Lung5_Rep1”, images were renamed from
“CellComposite_F002.jpg”, “CellComposite_F006.jpg”, and
“CellComposite_F011.jpg” to “Lung5_Rep1_fov_2.jpg”,
“Lung5_Rep1_fov_6.jpg”, and “Lung5_Rep1_fov_11.jpg” respectively.</li>
<li>For tissue slice “Lung6”, images were renamed from
“CellComposite_F004.jpg”, “CellComposite_F006.jpg”, and
“CellComposite_F007.jpg” to “Lung6_fov_4.jpg”, “Lung6_fov_6.jpg”, and
“Lung6_fov_7.jpg” respectively.</li>
</ul>
<p>Now, upload the images to the STlist with the
<code>load_images</code> function:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">img_fp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">nsclc_tmp</span>, <span class="st">'/nanostring_nsclc/Lung5_Rep1-Flat_files_and_images/CellComposite'</span><span class="op">)</span></span>
<span><span class="va">lung</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_images.html">load_images</a></span><span class="op">(</span><span class="va">lung</span>, images<span class="op">=</span><span class="va">img_fp</span><span class="op">)</span></span>
<span><span class="co">#&gt; Image for sample Lung6_fov_4 was not found.</span></span>
<span><span class="co">#&gt; Image for sample Lung6_fov_6 was not found.</span></span>
<span><span class="co">#&gt; Image for sample Lung6_fov_7 was not found.</span></span></code></pre></div>
<p>The images for slide “Lung6” were not found, which are in the
corresponsing folder. The <code>load_images</code> function is ran again
to add images from “Lung6”:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">img_fp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">nsclc_tmp</span>, <span class="st">'/nanostring_nsclc/Lung6-Flat_files_and_images/CellComposite'</span><span class="op">)</span></span>
<span><span class="va">lung</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_images.html">load_images</a></span><span class="op">(</span><span class="va">lung</span>, images<span class="op">=</span><span class="va">img_fp</span><span class="op">)</span></span>
<span><span class="co">#&gt; Image for sample Lung5_Rep1_fov_2 was not found.</span></span>
<span><span class="co">#&gt; Image for sample Lung5_Rep1_fov_6 was not found.</span></span>
<span><span class="co">#&gt; Image for sample Lung5_Rep1_fov_11 was not found.</span></span></code></pre></div>
<p>The function <code>plot_image</code> is used to generate the ggplot
objects containing the images:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ti</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_image.html">plot_image</a></span><span class="op">(</span><span class="va">lung</span><span class="op">)</span></span></code></pre></div>
<p>The domains can be plotted with the <code>STplot</code> function
too.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dom_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STplot.html">STplot</a></span><span class="op">(</span><span class="va">lung</span>, ks<span class="op">=</span><span class="st">'dtc'</span>, ws<span class="op">=</span><span class="fl">0.02</span>, deepSplit<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>                color_pal<span class="op">=</span><span class="st">'discreterainbow'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span><span class="va">dom_p</span><span class="op">$</span><span class="va">Lung5_Rep1_fov_2_stclust_spw0.02_dsplFalse</span>,</span>
<span>                  <span class="va">ti</span><span class="op">$</span><span class="va">image_Lung5_Rep1_fov_2</span>,</span>
<span>                  <span class="va">dom_p</span><span class="op">$</span><span class="va">Lung5_Rep1_fov_6_stclust_spw0.02_dsplFalse</span>,</span>
<span>                  <span class="va">ti</span><span class="op">$</span><span class="va">image_Lung5_Rep1_fov_6</span>,</span>
<span>                  <span class="va">dom_p</span><span class="op">$</span><span class="va">Lung5_Rep1_fov_11_stclust_spw0.02_dsplFalse</span>,</span>
<span>                  <span class="va">ti</span><span class="op">$</span><span class="va">image_Lung5_Rep1_fov_11</span>,</span>
<span>                  <span class="va">dom_p</span><span class="op">$</span><span class="va">Lung6_fov_4_stclust_spw0.02_dsplFalse</span>,</span>
<span>                  <span class="va">ti</span><span class="op">$</span><span class="va">image_Lung6_fov_4</span>,</span>
<span>                  <span class="va">dom_p</span><span class="op">$</span><span class="va">Lung6_fov_6_stclust_spw0.02_dsplFalse</span>,</span>
<span>                  <span class="va">ti</span><span class="op">$</span><span class="va">image_Lung6_fov_6</span>,</span>
<span>                  <span class="va">dom_p</span><span class="op">$</span><span class="va">Lung6_fov_7_stclust_spw0.02_dsplFalse</span>,</span>
<span>                  <span class="va">ti</span><span class="op">$</span><span class="va">image_Lung6_fov_7</span>,</span>
<span>                  ncol<span class="op">=</span><span class="fl">4</span>, nrow<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial_enrichment_gradients_smi_files/figure-html/plot_domains-1.png" width="1152"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># If no tissue images are available:</span></span>
<span><span class="co"># ggpubr::ggarrange(plotlist=dom_p, ncol=3, nrow=2)</span></span></code></pre></div>
<p>Now, to identify the domains, differential gene expression can be
performed:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STdiff.html">STdiff</a></span><span class="op">(</span><span class="va">lung</span>, w<span class="op">=</span><span class="fl">0.02</span>, k<span class="op">=</span><span class="st">'dtc'</span>, deepSplit<span class="op">=</span><span class="cn">FALSE</span>, test_type<span class="op">=</span><span class="st">'wilcoxon'</span>, cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Testing STclust assignment (w=0.02, dtc deepSplit=False)...</span></span>
<span><span class="co">#&gt;  Running Wilcoxon's tests...</span></span>
<span><span class="co">#&gt;  Completed Wilcoxon's tests (5.77 min).</span></span>
<span><span class="co">#&gt; STdiff completed in 5.77 min.</span></span></code></pre></div>
<p>Then, the top 3 DE genes per cluster in the samples showing NOTCH
signaling spatial patterns are identified:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">deg</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">adj_p_val</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">avg_log2fc</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Lung6_fov_4'</span>, <span class="st">'Lung6_fov_6'</span>, <span class="st">'Lung6_fov_7'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">cluster_1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 57 × 6</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   sample, cluster_1 [19]</span></span></span>
<span><span class="co">#&gt;    sample      gene     avg_log2fc cluster_1 wilcox_p_val adj_p_val</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Lung6_fov_4 KRT15         0.895 1            7.13<span style="color: #949494;">e</span><span style="color: #BB0000;">-194</span> 8.56<span style="color: #949494;">e</span><span style="color: #BB0000;">-191</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Lung6_fov_4 IL1R1         0.361 1            1.24<span style="color: #949494;">e</span><span style="color: #BB0000;">- 82</span> 2.29<span style="color: #949494;">e</span><span style="color: #BB0000;">- 80</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Lung6_fov_4 KRT17         0.416 1            1.58<span style="color: #949494;">e</span><span style="color: #BB0000;">- 76</span> 2.23<span style="color: #949494;">e</span><span style="color: #BB0000;">- 74</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Lung6_fov_4 H4C3          0.574 2            2.87<span style="color: #949494;">e</span><span style="color: #BB0000;">-105</span> 8.12<span style="color: #949494;">e</span><span style="color: #BB0000;">-103</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Lung6_fov_4 HSP90AA1      0.473 2            4.61<span style="color: #949494;">e</span><span style="color: #BB0000;">- 94</span> 1.05<span style="color: #949494;">e</span><span style="color: #BB0000;">- 91</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Lung6_fov_4 H2AZ1         0.424 2            1.37<span style="color: #949494;">e</span><span style="color: #BB0000;">- 87</span> 2.74<span style="color: #949494;">e</span><span style="color: #BB0000;">- 85</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Lung6_fov_4 CD74          0.697 3            4.70<span style="color: #949494;">e</span><span style="color: #BB0000;">-226</span> 2.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-222</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Lung6_fov_4 HLA-DRB1      0.790 3            7.49<span style="color: #949494;">e</span><span style="color: #BB0000;">-211</span> 1.20<span style="color: #949494;">e</span><span style="color: #BB0000;">-207</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Lung6_fov_4 VIM           0.533 3            2.84<span style="color: #949494;">e</span><span style="color: #BB0000;">-180</span> 2.72<span style="color: #949494;">e</span><span style="color: #BB0000;">-177</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Lung6_fov_4 KRT13         1.06  4            1.59<span style="color: #949494;">e</span><span style="color: #BB0000;">-131</span> 7.62<span style="color: #949494;">e</span><span style="color: #BB0000;">-129</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 47 more rows</span></span></span></code></pre></div>
<p>Expression of KRT genes in domains 1 and 2 of FOVs “Lung6_fov_4” and
“Lung6_fov_7”, and domains 2 and 4 of “Lung6_fov_6” are indicative of
tumor cells. In those domains (with the exception of 4 in
“Lung6_fov_6”), there seems to be an overlap with the areas of of NOTCH
signaling, which indicates that in slide “Lung6”, the tumor compartment
is spatially enriched for NOTCH signaling.</p>
</div>
<div class="section level2">
<h2 id="identifcation-of-genes-presenting-expression-spatial-gradients">Identifcation of genes presenting expression spatial gradients<a class="anchor" aria-label="anchor" href="#identifcation-of-genes-presenting-expression-spatial-gradients"></a>
</h2>
<p>Since the tumor areas have been visually identified, it is possible
now to ascertain which genes show expression that varies with distance
within the tumor compartment. In the FOVs “Lung6_fov_4”, “Lung6_fov_7”,
and “Lung6_fov_6”, the domain 3 seems to be surrounding the tumor
compartment. Hence, domain 3 will be used as reference domain to test
for genes for which expression increases or decreases with distance from
domain 3. This test is performed by the <code>STgradient</code>
function. The following diagram depicts the <code>STgradient</code>
algorithm:</p>
<p align="center">
<img src="img%2Fspatial_gradients.png" height="250" width="550"></p>
<p>In <code>STgradient</code>, the spatial coordinates are used to
calculate Euclidean distances of each cell to each of the other cells.
The reference domain is the domain from which the distances will be
calculated (domain 3 in this case). The non-reference domain(s) are the
domain(s) on which the user wants to study gradients. Next, removes
isolated spots/cells in the reference domain (i.e., have a number of
immediate neighbors smaller than parameter <code>min_nb</code>). The
<code>min_nb</code> parameter intends to reduce the effects of small
“pockets” of the reference domain on the correlation coefficient. At
this point, either the minimum or average distance of each non-reference
cell to each reference cell is calculated (<code>distsumm</code>). The
minimum distance is easier to interpret, however, it is more sensitive
to highly fragmented reference domains. On the other hand, the
interpretability of the average distance is not as easy as the minimum
distance, but it might detect in a better way the gradients when the
reference domain is distributed across the sample. Next, the gene
expression data is subset to the non-reference spots/cells, and the top
variable genes (<code>topgenes</code>) are selected. Results will be
produced only for those genes. Depending on the user’s selection, the
ordinary least squares (OLS) or robust regression is used to calculate
the slope of the regression line between the minimum/average distance
and the expression of a gene. If OLS was selected, users might opt to
remove outliers via the interquartile method. Finally, the Spearman’s
coefficient is calculated in addition to the regression line
coefficient. The slope indicates the direction of the correlation. If
positive, the gene expression tends to be higher when farther from the
reference domain. If negative, gene expression is higher when closer to
the reference domain. The interpretation of the sign for the Spearman’s
coefficient is similar, however its magnitude indicates how strong the
correlation is between gene expression and distance to the reference
domain.</p>
<p>The <code>STgradient</code> function can be executed with these
commands, in order to test for gradients within the tumor compartment.
First, it is necessary to find the name of the column in
<code>lung_subset@spatial_meta</code> containing the domain
assigments:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lung</span><span class="op">@</span><span class="va">spatial_meta</span><span class="op">$</span><span class="va">Lung5_Rep1_fov_2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "libname"                   "xpos"                     </span></span>
<span><span class="co">#&gt; [3] "ypos"                      "total_counts"             </span></span>
<span><span class="co">#&gt; [5] "total_genes"               "stclust_spw0.02_dsplFalse"</span></span></code></pre></div>
<p>The name of the column is <code>stclust_spw0.02_dsplFalse</code>
(STclust results using spatial weight of 0.02 and DynamicTreeCuts with
no deepSplit). This name is used in the <code>annot</code> argument of
<code>STgradient</code>:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STgradient.html">STgradient</a></span><span class="op">(</span><span class="va">lung</span>, </span>
<span>                  topgenes<span class="op">=</span><span class="fl">1000</span>, </span>
<span>                  annot<span class="op">=</span><span class="st">'stclust_spw0.02_dsplFalse'</span>,</span>
<span>                  ref<span class="op">=</span><span class="fl">3</span>,</span>
<span>                  samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Lung6_fov_4'</span>, <span class="st">'Lung6_fov_6'</span>, <span class="st">'Lung6_fov_7'</span><span class="op">)</span>, </span>
<span>                  distsumm<span class="op">=</span><span class="st">'min'</span>, </span>
<span>                  robust<span class="op">=</span><span class="cn">TRUE</span>, cores<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; STgradient completed in 0.1 min.</span></span></code></pre></div>
<p>Here are the first few rows of the <code>STgradient</code> result for
sample “Lung6_fov_4”:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">stg</span><span class="op">$</span><span class="va">Lung6_fov_4</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 8</span></span></span>
<span><span class="co">#&gt;   sample_name gene   min_lm_coef min_lm_pval min_spearman_r min_spearman_r_pval</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Lung6_fov_4 HLA-B    -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">213</span>    2.63<span style="color: #949494;">e</span><span style="color: #BB0000;">-28</span>         -<span style="color: #BB0000;">0.208</span>            2.73<span style="color: #949494;">e</span><span style="color: #BB0000;">-28</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Lung6_fov_4 S100A2    0.000<span style="text-decoration: underline;">222</span>    8.32<span style="color: #949494;">e</span><span style="color: #BB0000;">-31</span>          0.208            1.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-28</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Lung6_fov_4 EGFR      0.000<span style="text-decoration: underline;">133</span>    3.19<span style="color: #949494;">e</span><span style="color: #BB0000;">-18</span>          0.200            2.36<span style="color: #949494;">e</span><span style="color: #BB0000;">-26</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Lung6_fov_4 MT1X      0.000<span style="text-decoration: underline;">132</span>    2.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-17</span>          0.192            2.82<span style="color: #949494;">e</span><span style="color: #BB0000;">-24</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Lung6_fov_4 KRT6A     0.000<span style="text-decoration: underline;">231</span>    2.47<span style="color: #949494;">e</span><span style="color: #BB0000;">-29</span>          0.180            1.43<span style="color: #949494;">e</span><span style="color: #BB0000;">-21</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Lung6_fov_4 GPNMB     0.000<span style="text-decoration: underline;">125</span>    6.22<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span>          0.164            4.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-18</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: min_spearman_r_pval_adj &lt;dbl&gt;, min_pval_comment &lt;chr&gt;</span></span></span></code></pre></div>
<p>The key columns to look ar are <code>min_spearman_r</code> and
<code>min_spearman_r_pval_adj</code>. Now, to find genes with
significant spatial gradients across the three samples, these commands
can be used:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">stg</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">min_spearman_r</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">sample_name</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">5</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10 × 8</span></span></span>
<span><span class="co">#&gt;    sample_name gene   min_lm_coef min_lm_pval min_spearman_r min_spearman_r_pval</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Lung6_fov_4 EGFR     0.000<span style="text-decoration: underline;">133</span>     3.19<span style="color: #949494;">e</span><span style="color: #BB0000;">-18</span>          0.200            2.36<span style="color: #949494;">e</span><span style="color: #BB0000;">-26</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Lung6_fov_6 FOS     -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">062</span><span style="color: #BB0000;">7</span>    3.86<span style="color: #949494;">e</span><span style="color: #BB0000;">-23</span>         -<span style="color: #BB0000;">0.201</span>            1.44<span style="color: #949494;">e</span><span style="color: #BB0000;">-26</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Lung6_fov_4 HLA-B   -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">213</span>     2.63<span style="color: #949494;">e</span><span style="color: #BB0000;">-28</span>         -<span style="color: #BB0000;">0.208</span>            2.73<span style="color: #949494;">e</span><span style="color: #BB0000;">-28</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Lung6_fov_6 HSP90…   0.000<span style="text-decoration: underline;">112</span>     6.96<span style="color: #949494;">e</span><span style="color: #BB0000;">-29</span>          0.216            9.82<span style="color: #949494;">e</span><span style="color: #BB0000;">-31</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Lung6_fov_4 KRT6A    0.000<span style="text-decoration: underline;">231</span>     2.47<span style="color: #949494;">e</span><span style="color: #BB0000;">-29</span>          0.180            1.43<span style="color: #949494;">e</span><span style="color: #BB0000;">-21</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Lung6_fov_6 KRT6B   -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">089</span><span style="color: #BB0000;">1</span>    1.91<span style="color: #949494;">e</span><span style="color: #BB0000;">-19</span>         -<span style="color: #BB0000;">0.180</span>            9.61<span style="color: #949494;">e</span><span style="color: #BB0000;">-22</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Lung6_fov_4 MT1X     0.000<span style="text-decoration: underline;">132</span>     2.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-17</span>          0.192            2.82<span style="color: #949494;">e</span><span style="color: #BB0000;">-24</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Lung6_fov_6 MT1X    -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">181</span>     2.52<span style="color: #949494;">e</span><span style="color: #BB0000;">-46</span>         -<span style="color: #BB0000;">0.285</span>            6.02<span style="color: #949494;">e</span><span style="color: #BB0000;">-53</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Lung6_fov_6 NTRK2    0.000<span style="text-decoration: underline;">128</span>     1.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-32</span>          0.231            7.62<span style="color: #949494;">e</span><span style="color: #BB0000;">-35</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Lung6_fov_4 S100A2   0.000<span style="text-decoration: underline;">222</span>     8.32<span style="color: #949494;">e</span><span style="color: #BB0000;">-31</span>          0.208            1.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-28</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: min_spearman_r_pval_adj &lt;dbl&gt;, min_pval_comment &lt;chr&gt;</span></span></span></code></pre></div>
<p>It can be seen from these results that multiple keratin genes show
spatial gradients, all of them with negative Spearman’s coefficients
(cells close to domain 3 tend to have higher keratin expression). Other
such <strong>HLA-B</strong> and <strong>NDRG1</strong> have higher
expression in cells far from domain 3:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STplot.html">STplot</a></span><span class="op">(</span><span class="va">lung</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'KRT6A'</span>, <span class="st">'HLA-B'</span><span class="op">)</span>, samples<span class="op">=</span><span class="st">'Lung6_fov_4'</span>, color_pal<span class="op">=</span><span class="st">'YlOrBr'</span><span class="op">)</span></span>
<span><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">qp2</span>, common.legend<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial_enrichment_gradients_smi_files/figure-html/stg_plots-1.png" width="768"></p>
<p>Here are the results for genes within the NOTCH signaling
pathway:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">stg</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">min_spearman_r</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">gene_sets</span><span class="op">[[</span><span class="st">'HALLMARK_NOTCH_SIGNALING'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">min_spearman_r_pval_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 8</span></span></span>
<span><span class="co">#&gt;   sample_name gene    min_lm_coef min_lm_pval min_spearman_r min_spearman_r_pval</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Lung6_fov_4 JAG1        5.97<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span>    0.001<span style="text-decoration: underline;">45</span>          0.070<span style="text-decoration: underline;">2</span>            0.000<span style="text-decoration: underline;">221</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Lung6_fov_6 NOTCH1     -<span style="color: #BB0000;">2.72</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-8</span>    0.000<span style="text-decoration: underline;">250</span>        -<span style="color: #BB0000;">0.066</span><span style="color: #BB0000; text-decoration: underline;">5</span>            0.000<span style="text-decoration: underline;">456</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Lung6_fov_6 CCND1      -<span style="color: #BB0000;">1.90</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-8</span>    0.004<span style="text-decoration: underline;">21</span>         -<span style="color: #BB0000;">0.054</span><span style="color: #BB0000; text-decoration: underline;">0</span>            0.004<span style="text-decoration: underline;">44</span> </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: min_spearman_r_pval_adj &lt;dbl&gt;, min_pval_comment &lt;chr&gt;</span></span></span></code></pre></div>
<p>Although the correlations in genes from this pathway are not as large
as and those for <strong>KRT6A</strong> and <strong>HLA-B</strong>
(i.e., lower absolute Spearman’s coefficient), it can be seen that the
genes <strong>NOTCH1</strong> and <strong>NOTCH3</strong> show evidence
of expression gradients in FOVs ‘Lung6_fov_4’ and ‘Lung6_fov_6’.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-hafemeister_2019" class="csl-entry">
Hafemeister, Christoph, and Rahul Satija. 2019. <span>“Normalization and
Variance Stabilization of Single-Cell RNA-Seq Data Using Regularized
Negative Binomial Regression.”</span> <em>Genome Biology</em> 20: 296.
<a href="https://doi.org/10.1186/s13059-019-1874-1" class="external-link">https://doi.org/10.1186/s13059-019-1874-1</a>.
</div>
<div id="ref-hunter2021spatially" class="csl-entry">
Hunter, Miranda V, Reuben Moncada, Joshua M Weiss, Itai Yanai, and
Richard M White. 2021. <span>“Spatially Resolved Transcriptomics Reveals
the Architecture of the Tumor-Microenvironment Interface.”</span>
<em>Nature Communications</em> 12 (1): 6278.
</div>
</div>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Sequoia 15.3.2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/New_York</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    </span></span>
<span><span class="co">#&gt;  [5] purrr_1.0.2     readr_2.1.5     tidyr_1.3.1     tibble_3.2.1   </span></span>
<span><span class="co">#&gt;  [9] ggplot2_3.5.1   tidyverse_2.0.0 msigdbr_7.5.1   spatialGE_1.2.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.1      EBImage_4.46.0        khroma_1.13.0        </span></span>
<span><span class="co">#&gt;  [4] farver_2.1.2          bitops_1.0-8          RCurl_1.98-1.16      </span></span>
<span><span class="co">#&gt;  [7] fastmap_1.2.0         tweenr_2.0.3          digest_0.6.37        </span></span>
<span><span class="co">#&gt; [10] timechange_0.3.0      lifecycle_1.0.4       magrittr_2.0.3       </span></span>
<span><span class="co">#&gt; [13] compiler_4.4.2        rlang_1.1.4           sass_0.4.9           </span></span>
<span><span class="co">#&gt; [16] tools_4.4.2           utf8_1.2.4            yaml_2.3.10          </span></span>
<span><span class="co">#&gt; [19] knitr_1.48            ggsignif_0.6.4        S4Arrays_1.4.1       </span></span>
<span><span class="co">#&gt; [22] labeling_0.4.3        htmlwidgets_1.6.4     DelayedArray_0.30.1  </span></span>
<span><span class="co">#&gt; [25] abind_1.4-5           babelgene_22.9        withr_3.0.1          </span></span>
<span><span class="co">#&gt; [28] BiocGenerics_0.50.0   desc_1.4.3            grid_4.4.2           </span></span>
<span><span class="co">#&gt; [31] polyclip_1.10-7       stats4_4.4.2          fansi_1.0.6          </span></span>
<span><span class="co">#&gt; [34] ggpubr_0.6.0          colorspace_2.1-1      scales_1.3.0         </span></span>
<span><span class="co">#&gt; [37] MASS_7.3-61           ggpolypath_0.3.0      cli_3.6.3            </span></span>
<span><span class="co">#&gt; [40] rmarkdown_2.28        crayon_1.5.3          ragg_1.3.2           </span></span>
<span><span class="co">#&gt; [43] generics_0.1.3        rstudioapi_0.16.0     tzdb_0.4.0           </span></span>
<span><span class="co">#&gt; [46] cachem_1.1.0          ggforce_0.4.2         zlibbioc_1.50.0      </span></span>
<span><span class="co">#&gt; [49] parallel_4.4.2        tiff_0.1-12           XVector_0.44.0       </span></span>
<span><span class="co">#&gt; [52] matrixStats_1.3.0     vctrs_0.6.5           Matrix_1.7-1         </span></span>
<span><span class="co">#&gt; [55] carData_3.0-5         jsonlite_1.8.8        car_3.1-2            </span></span>
<span><span class="co">#&gt; [58] fftwtools_0.9-11      IRanges_2.38.1        hms_1.1.3            </span></span>
<span><span class="co">#&gt; [61] S4Vectors_0.42.1      rstatix_0.7.2         jpeg_0.1-10          </span></span>
<span><span class="co">#&gt; [64] systemfonts_1.1.0     locfit_1.5-9.10       jquerylib_0.1.4      </span></span>
<span><span class="co">#&gt; [67] glue_1.7.0            pkgdown_2.1.1         cowplot_1.1.3        </span></span>
<span><span class="co">#&gt; [70] stringi_1.8.4         gtable_0.3.5          munsell_0.5.1        </span></span>
<span><span class="co">#&gt; [73] pillar_1.9.0          htmltools_0.5.8.1     R6_2.5.1             </span></span>
<span><span class="co">#&gt; [76] textshaping_0.4.0     evaluate_0.24.0       lattice_0.22-6       </span></span>
<span><span class="co">#&gt; [79] highr_0.11            backports_1.5.0       png_0.1-8            </span></span>
<span><span class="co">#&gt; [82] broom_1.0.6           bslib_0.8.0           Rcpp_1.0.13          </span></span>
<span><span class="co">#&gt; [85] gridExtra_2.3         SparseArray_1.4.8     xfun_0.47            </span></span>
<span><span class="co">#&gt; [88] fs_1.6.4              MatrixGenerics_1.16.0 pkgconfig_2.0.3</span></span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Oscar Ospina, Alex Soupir, Brooke Fridley.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
