<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Detection of gene set spatial patterns and expression gradients in CosMx-SMI • spatialGE</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Detection of gene set spatial patterns and expression gradients in CosMx-SMI">
<meta property="og:description" content="spatialGE">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basic_functions_vignette.html">Basic functionality: Visualization, domain detection, and spatial heterogeneity</a>
    </li>
    <li>
      <a href="../articles/spatial_differential_expression.html">Spatially-informed differential gene expression</a>
    </li>
    <li>
      <a href="../articles/spatial_enrichment_gradients_smi.html">Detection of gene set spatial patterns and expression gradients in CosMx-SMI</a>
    </li>
    <li>
      <a href="../articles/troubleshooting.html">Troubleshooting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Detection of gene set spatial patterns and
expression gradients in CosMx-SMI</h1>
            
            <h4 data-toc-skip class="date">2023-09-02</h4>
      
      
      <div class="hidden name"><code>spatial_enrichment_gradients_smi.Rmd</code></div>

    </div>

    
    
<p>The package <strong><em><code>spatialGE</code></em></strong> can be
used to detect spatial patterns in gene expression at the gene and gene
set level. Data from multiple spatial transcriptomics platforms can be
analyzed, as long as gene expression counts per spot or cell are
associated with spatial coordinates of those spots/cells. The
<strong><em><code>spatialGE</code></em></strong> package is compatible
with outputs from the Visium workflow or the CosMx-SMI spatial
single-cell platform.</p>
<p>In this tutorial, the functions <code>STenrich</code> and
<code>STgradient</code> will be used to test for spatial gene set
enrichment and genes with gradients expression on a <a href="https://nanostring.com/products/cosmx-spatial-molecular-imager/nsclc-ffpe-dataset/" class="external-link">CosMx-SMI
data set</a> from Non-Small Cell Lung Cancer (NSCLC).</p>
<details><summary><strong>How is <code>spatialGE</code> installed? </strong>
</summary><p>The <code>spatialGE</code> repository is available at GitHub and can
be installed via <code>devtools</code>. To install <code>devtools</code>
(in case it is not already installed in your R), please run the
following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="st">"devtools"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>After making sure <code>devtools</code> is installed, proceed to
install <code>spatialGE</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># devtools::install_github("fridleylab/spatialGE")</span></code></pre></div>
</details><div class="section level2">
<h2 id="spatial-gene-set-enrichment-and-expression-gradients-in-non-small-cell-lung-cancer-nsclc">Spatial gene set enrichment and expression gradients in non-small
cell lung cancer (NSCLC)<a class="anchor" aria-label="anchor" href="#spatial-gene-set-enrichment-and-expression-gradients-in-non-small-cell-lung-cancer-nsclc"></a>
</h2>
<p>The CosMx-SMI platform generates single-cell level gene expression
with associated x and y locations of the cells where measurements were
taken. In specific for this data set, counts were generated for 960
genes for about 800K cells. The data was generated from eight tissue
slices, however, in this tutorial only two slides will be analyzed.</p>
<p>To start, data can be downloaded by accessing the links below.
Please, note that each sample is about 1.5G. It is also likely that
registration in the website is necessary in order to download the data
set.</p>
<ul>
<li><a href="https://nanostring.com/resources/smi-ffpe-dataset-lung5-rep1-data/" class="external-link">Lung
5-1</a></li>
<li><a href="https://nanostring.com/resources/smi-ffpe-dataset-lung6-data/" class="external-link">Lung
6</a></li>
</ul>
<p>Create a folder in your computer’s Desktop and name it
“lung_cancer_smi”. Place the downloaded files within the folder created
in the Desktop. The downloaded files are compressed. Please decompress
the files by double clicking on each file. The decompression might take
a few seconds, and then, two directories should be generated. The
structure of the files is shown below:</p>
<p align="center">
<img src="../smi_dir_structure.png" height="250" width="550"></p>
<p>The key files are shown enclosed in orange (Lung 5-1) and green (Lung
6) boxes. The “_exprMat_file.csv” file of each sample contains the gene
counts. The “_metadata_file.csv” contains the cell x, y coordinates for
each sample. Optionally, the “CellComposite” folder contains images that
can be imported into the STlist (details later in this tutorial).</p>
<p>Now, that the data is ready, load the <code>spatialGE</code>
package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">'spatialGE'</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="creating-an-stlist-spatial-transcriptomics-list">Creating an STList (Spatial Transcriptomics List)<a class="anchor" aria-label="anchor" href="#creating-an-stlist-spatial-transcriptomics-list"></a>
</h2>
<p>The starting point of a <code>spatialGE</code> analysis is the
creation of an <strong><em>STlist</em></strong> (S4 class object), which
stores raw and processed data. The STlist is created with the function
<code>STlist</code>, which can take data in multiple formats (see <a href="https://fridleylab.github.io/spatialGE/reference/STlist.html" class="external-link">here</a>
for more info or type <code><a href="../reference/STlist.html">?STlist</a></code> in the R console). In this
tutorial we will provide the paths to the <code>exprMat</code> and
<code>metadata</code> files for each of the two slides.</p>
<p>To load the files into an STlist, please use these commands:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">exprmats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'~/Desktop/lung_cancer_smi/Lung5_Rep1/Lung5_Rep1-Flat_files_and_images/Lung5_Rep1_exprMat_file.csv'</span>,
             <span class="st">'~/Desktop/lung_cancer_smi/Lung6/Lung6-Flat_files_and_images/Lung6_exprMat_file.csv'</span><span class="op">)</span>
<span class="va">metas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'~/Desktop/lung_cancer_smi/Lung5_Rep1/Lung5_Rep1-Flat_files_and_images/Lung5_Rep1_metadata_file.csv'</span>,
             <span class="st">'~/Desktop/lung_cancer_smi/Lung6/Lung6-Flat_files_and_images/Lung6_metadata_file.csv'</span><span class="op">)</span>

<span class="va">lung</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STlist.html">STlist</a></span><span class="op">(</span>rnacounts<span class="op">=</span><span class="va">exprmats</span>, spotcoords<span class="op">=</span><span class="va">metas</span>, samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Lung5_Rep1'</span>, <span class="st">'Lung6'</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #00BB00;">Found CosMx-SMI data</span></span>
<span class="co"><span style="color: #00BB00;">#&gt; </span><span style="color: #00BB00; font-weight: bold;">Found 2 CosMx-SMI samples</span></span>
<span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt;   Data read completed</span></span>
<span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; </span><span style="color: #BBBB00;">Matching gene expression and coordinate data...</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; Converting counts to sparse matrices</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; </span><span style="color: #00BB00; font-weight: bold;">Completed STlist!</span></span>
<span class="co"><span style="color: #00BB00; font-weight: bold;">#&gt; </span></span></code></pre></div>
<p>To obtain count statistics, the <code>summarize_STlist</code>
function can be used:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">summ_df</span> <span class="op">=</span> <span class="fu"><a href="../reference/summarize_STlist.html">summarize_STlist</a></span><span class="op">(</span><span class="va">lung</span><span class="op">)</span>
<span class="va">summ_df</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 60 × 9</span></span>
<span class="co">#&gt;    sample_name     spots…¹ genes min_c…² mean_…³ max_c…⁴ min_g…⁵ mean_…⁶ max_g…⁷</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Lung5_Rep1_fov…    <span style="text-decoration: underline;">2</span>408   980       1    196.    <span style="text-decoration: underline;">1</span>069       1    100.     315</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Lung5_Rep1_fov…    <span style="text-decoration: underline;">3</span>744   980       2    326.    <span style="text-decoration: underline;">1</span>872       2    136.     402</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Lung5_Rep1_fov…    <span style="text-decoration: underline;">2</span>942   980       0    374.    <span style="text-decoration: underline;">2</span>475       0    152.     473</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Lung5_Rep1_fov…    <span style="text-decoration: underline;">3</span>617   980       0    363.    <span style="text-decoration: underline;">2</span>246       0    149.     405</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Lung5_Rep1_fov…    <span style="text-decoration: underline;">3</span>556   980       0    359.    <span style="text-decoration: underline;">1</span>818       0    148.     389</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Lung5_Rep1_fov…    <span style="text-decoration: underline;">5</span>272   980       0    328.    <span style="text-decoration: underline;">1</span>721       0    142.     416</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Lung5_Rep1_fov…    <span style="text-decoration: underline;">3</span>705   980       2    420.    <span style="text-decoration: underline;">1</span>954       2    165.     489</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Lung5_Rep1_fov…    <span style="text-decoration: underline;">3</span>120   980       0    467.    <span style="text-decoration: underline;">1</span>891       0    182.     440</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Lung5_Rep1_fov…    <span style="text-decoration: underline;">3</span>304   980       1    278.    <span style="text-decoration: underline;">1</span>570       1    120.     402</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Lung5_Rep1_fov…    <span style="text-decoration: underline;">3</span>174   980       0    354.    <span style="text-decoration: underline;">2</span>002       0    141.     438</span>
<span class="co">#&gt; <span style="color: #949494;"># … with 50 more rows, and abbreviated variable names ¹​spotscells,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   ²​min_counts_per_spotcell, ³​mean_counts_per_spotcell,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   ⁴​max_counts_per_spotcell, ⁵​min_genes_per_spotcell,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   ⁶​mean_genes_per_spotcell, ⁷​max_genes_per_spotcell</span></span></code></pre></div>
<p>Some cells have zero counts. We can look at the distribution of
counts per cell for the first five and last five FOVs using the
<code>distribution_plots</code> function:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/distribution_plots.html">distribution_plots</a></span><span class="op">(</span><span class="va">lung</span>, plot_type<span class="op">=</span><span class="st">'violin'</span>, plot_meta<span class="op">=</span><span class="st">'total_counts'</span>, samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">56</span><span class="op">:</span><span class="fl">60</span><span class="op">)</span><span class="op">)</span>
<span class="va">cp</span><span class="op">[[</span><span class="st">'total_counts'</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="spatial_enrichment_gradients_smi_files/figure-html/count_dstr-1.png" width="768"></p>
<p>Cells with zero counts, can be removed using the
<code>filter_data</code> function. The function can also be used to
remove all counts from specific genes. This option is useful in this
case, as CosMx-SMI panels include negative probe genes. While the
negative probes can be used in normalization of counts, in this tutorial
those genes will be removed. Negative probes in CosMx-SMI begin with the
token “NegPrb”:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lung</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_data.html">filter_data</a></span><span class="op">(</span><span class="va">lung</span>, spot_minreads<span class="op">=</span><span class="fl">20</span>, rm_genes_expr<span class="op">=</span><span class="st">'^NegPrb'</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="transformation-of-spatially-resolved-transcriptomics-data">Transformation of spatially-resolved transcriptomics data<a class="anchor" aria-label="anchor" href="#transformation-of-spatially-resolved-transcriptomics-data"></a>
</h2>
<p>The function <code>transform_data</code> allows data transformation
using one of two possible options. The first options applies
log-transformation to the counts, after library size normalization
performed on each sample separately. The second option applies
variance-stabilizing transformation (SCT; <span class="citation">Hafemeister and Satija (2019)</span>), which is a
method increasingly used in single-cell and spatial transcriptomics
studies. To apply SCT transformation, use the following command:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lung</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transform_data.html">transform_data</a></span><span class="op">(</span><span class="va">lung</span>, method<span class="op">=</span><span class="st">'sct'</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="detecting-gene-sets-with-spatial-aggregation-patterns">Detecting gene sets with spatial aggregation patterns<a class="anchor" aria-label="anchor" href="#detecting-gene-sets-with-spatial-aggregation-patterns"></a>
</h2>
<p>An important part of gene expression analysis is the use of gene sets
to make inferences about the functional significance of changes in
expression. Normally, this is achieved by conducting a Gene Set
Enrichment Analysis (GSEA). While GSEA can be completed in a similar
fashion to scRNA-seq, it is possible with the <code>STenrich</code>
function to test for gene sets that show spatial non-uniform enrichment.
In other words, <code>STenrich</code> tests whether expression of a gene
set is concentrated in one or few areas of the tissue.</p>
<p>The algorithm in <code>STenrich</code> is depicted in the following
diagram:</p>
<p align="center">
<img src="../diagram_stenrich.png" height="250" width="550"></p>
<p>The <code>STenrich</code> function is a modification of the method
proposed by <span class="citation">Hunter et al. (2021)</span>. The
steps in this modified version are as follows: First, Euclidean
distances are calculated among all spots/cells. Then, the gene
expression of the sample is subset to the genes within the gene set
being tested. If too few genes are left after subset
(<code>min_genes</code>), then the gene set is omitted for that sample.
The average expression and standard deviation of those genes is
calculated for each spot/cell. Next, spots/cells with gene set
expression above the average gene set expression across all spots/cells
are identified. The threshold to define these high gene set expression
spots is defined by the average gene set expression plus a number
<code>min_sds</code> of standard deviations. The sum of the Euclidean
distances between the high expression spots/cells is calculated.</p>
<p>The next step involves a permutation process, in which a null
distribution is generated in order to test if the (sum of) distances
among high expression spots are smaller than expected. To that end, a
random sample of spots/cells (regardless of expression) is selected. The
random sample has the same size as the number of high expression
spots/cells. Then, the sum of distances among the randomly sampled
spots/cells is calculated. The random selection is repeated as many
times as requested (`reps``). Finally, a p-value is calculated by noting
how many times the sum of random distances was higher than the sum of
distances among high expression spots/cells. If the sum of random
distances was most of the times higher than the sum of distances among
high expression spots/cells, then the null hypothesis of no spatial
aggregation is rejected (i.e., spots/cells with high gene set expression
are more aggregated than expected by chance).</p>
<p>A few important notes about the algorithm to consider:</p>
<ul>
<li>Notice that the metric of enrichment is currently the average
expression of the genes in a set within each ROI/spot/cell. Better
metrics of gene set enrichment are available and will be soon available
to use instead of average.</li>
<li>Special attention should be paid to the reference genome used to
annotate gene counts. The annotation of transcripts precedes any
analysis in spatialGE. Nevertheless, if transcripts are annotated with a
mouse genome (or other species), the user should use the appropriate
gene set database, as gene names will likely not match to a gene set
database with human gene names. Furthermore, but not less important,
there might be problems with gene homology.</li>
<li>The more permutations are requested, the longer the execution, but
p-value estimates are more accurate.</li>
<li>By increasing the number of standard deviations, fewer spots/cells
are selected as “high expression” (stringent analysis), however, it
becomes more challenging to detect gene sets that show evidence of
spatial aggregation. Values between 1 to 2 standard deviations are
likely suitable for most studies.</li>
<li>By adjusting the minimum number of spots (<code>min_spots</code>),
the user has some rough control over the size of the structures to be
studied. For example, small immune infiltrates could be detected with
smaller number of minimum spots. Conversely, if tissue domain-level
differences are sought, then a larger Minimum number of spots should be
set.</li>
<li>Take note of the random seed number used for the test (Seed number
parameter), in case you need to generate the same results later.
Nevertheless, running the analysis several times with different seed
numbers is advisable to check for consistency.</li>
</ul>
<p>The first step to run <code>STenrich</code> is to obtain a list of
gene sets. Here, the <code>msigdbr</code> package will be used to obtain
HALLMARK gene sets:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load msigdbr</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://igordot.github.io/msigdbr/" class="external-link">'msigdbr'</a></span><span class="op">)</span>

<span class="co"># Get HALLMARK gene sets</span>
<span class="va">gene_sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr.html" class="external-link">msigdbr</a></span><span class="op">(</span>species<span class="op">=</span><span class="st">'Homo sapiens'</span><span class="op">)</span>
<span class="va">gene_sets</span> <span class="op">&lt;-</span> <span class="va">gene_sets</span><span class="op">[</span><span class="va">gene_sets</span><span class="op">[[</span><span class="st">'gs_cat'</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="st">"H"</span>, <span class="op">]</span>

<span class="co"># Convert gene set dataframe to list</span>
<span class="co"># The result is a named list. The names of the list are the names of each gene set</span>
<span class="co"># The contents of each list element are the gene names within each gene set</span>
<span class="va">gene_sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">gene_sets</span><span class="op">[[</span><span class="st">'gene_symbol'</span><span class="op">]</span><span class="op">]</span>, f<span class="op">=</span><span class="va">gene_sets</span><span class="op">[[</span><span class="st">'gs_name'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p>Please keep in mind that it is assumed that HUGO gene names are used.
For CosMx-SMI and Visium, normally <em>HUGO names</em> are provided. If
spatial transcriptomics data from a non-human species is used, the
appropriate database should be used. Otherwise, <code>spatialGE</code>
will not be able to identify genes within the gene set.</p>
<p>The <code>STenrich</code> function can now be used. For this
analysis, 1000 replicates (<code>reps=1000</code>) will be used to
generate the null distribution. In addition, for a gene set to be
considered as highly expressed in cell, it should be expressed at least
over the average expression plus 1.5 standard deviations
(<code>num_sds=1.5</code>). Gene sets with less than five genes in the
sample (<code>min_genes=5</code>) and gene sets with less than 10 highly
expressed cells (<code>min_units=10</code>) will not be tested. Notice a
seed was set (<code>seed=12345</code>), which critical for replication
of the results in the future.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">stenrich_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STenrich.html">STenrich</a></span><span class="op">(</span><span class="va">lung</span>, gene_sets<span class="op">=</span><span class="va">gene_sets</span>, 
                        reps<span class="op">=</span><span class="fl">1000</span>, 
                        num_sds<span class="op">=</span><span class="fl">1.5</span>, 
                        min_genes<span class="op">=</span><span class="fl">5</span>,
                        min_units<span class="op">=</span><span class="fl">10</span>, 
                        seed<span class="op">=</span><span class="fl">12345</span><span class="op">)</span> 
<span class="co">#&gt; <span style="color: #BBBB00;">Running STenrich...</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; </span><span style="color: #00BB00;">STenrich completed in 39.88 min.</span></span>
<span class="co"><span style="color: #00BB00;">#&gt; </span></span></code></pre></div>
<p>The result of the function is a list of data frames, with one data
frame per samples (FOV in this case). For example, this is the data
frame for the first FOV in the STlist:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">stenrich_df</span><span class="op">$</span><span class="va">Lung5_Rep1_fov_1</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 6</span></span>
<span class="co">#&gt;   sample_name      gene_set                      size_…¹ size_…² p_value adj_p…³</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Lung5_Rep1_fov_1 HALLMARK_ANDROGEN_RESPONSE         16     102       0       0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Lung5_Rep1_fov_1 HALLMARK_CHOLESTEROL_HOMEOST…      13      77       0       0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Lung5_Rep1_fov_1 HALLMARK_COMPLEMENT                47     236       0       0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Lung5_Rep1_fov_1 HALLMARK_ESTROGEN_RESPONSE_E…      30     216       0       0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Lung5_Rep1_fov_1 HALLMARK_ESTROGEN_RESPONSE_L…      36     218       0       0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Lung5_Rep1_fov_1 HALLMARK_HYPOXIA                   43     213       0       0</span>
<span class="co">#&gt; <span style="color: #949494;"># … with abbreviated variable names ¹​size_test, ²​size_gene_set, ³​adj_p_value</span></span></code></pre></div>
<p>Each row represents a test for the null hypothesis of no spatial
aggregation in the expression of the set in the “gene_set” column. The
column “size_test” is the number of genes of a gene set that were
present in the FOV. The larger this number the better, as it indicates a
better representation of the gene set in the sample. The “adj_p_value”
is the multiple test adjusted p-value, which is the value used to decide
if a gene set shows significant indications of a spatial pattern
(adj_p_value &lt; 0.05).</p>
<p>With a few lines of code, a visual summary can be generated that
presents the gene sets with adj_p_value &lt; 0.05 across FOVs for each
tissue slide:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load tidyverse for data frame manipulation</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://tidyverse.tidyverse.org" class="external-link">'tidyverse'</a></span><span class="op">)</span>

<span class="co"># Combine all samples in a single data frame</span>
<span class="co"># Subset to gene sets showing significant evidence of spatial pattern (adj_p_value &lt; 0.05) and</span>
<span class="co"># proportion higher or equal than 0.3 of genes from a gene set present in sample </span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">stenrich_df</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>prop_gene_set<span class="op">=</span><span class="va">size_test</span><span class="op">/</span><span class="va">size_gene_set</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">prop_gene_set</span> <span class="op">&gt;=</span> <span class="fl">0.3</span> <span class="op">&amp;</span> <span class="va">adj_p_value</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>slide<span class="op">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract</a></span><span class="op">(</span><span class="va">sample_name</span>, <span class="st">"Lung5_Rep1|Lung6"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">slide</span>, <span class="va">gene_set</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>gene_set<span class="op">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace</a></span><span class="op">(</span><span class="va">gene_set</span>, <span class="st">'HALLMARK_'</span>, <span class="st">''</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Generate barplot showing the number of FOVs with siginificant evidence of spatial</span>
<span class="co"># patterns for each tissue slide</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">gene_set</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle<span class="op">=</span><span class="fl">70</span>, vjust<span class="op">=</span><span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">slide</span><span class="op">)</span></code></pre></div>
<p><img src="spatial_enrichment_gradients_smi_files/figure-html/unnamed-chunk-1-1.png" width="768"></p>
<p>The “NOTCH signaling” pathway showed spatial patterns less frequently
in the slide “Lung5_Rep1” compared to “Lung6”. With the function
<code>STplot</code>, the average gene set expression for this pathway
can be plotted. First, FOVs with and without NOTCH signaling spatial
patterns will be identified from “Lung5_Rep1” and “Lung6”
respectively.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">stenrich_df</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>prop_gene_set<span class="op">=</span><span class="va">size_test</span><span class="op">/</span><span class="va">size_gene_set</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">prop_gene_set</span> <span class="op">&gt;=</span> <span class="fl">0.3</span> <span class="op">&amp;</span> <span class="va">adj_p_value</span> <span class="op">&gt;=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">prop_gene_set</span><span class="op">)</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">adj_p_value</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gene_set</span> <span class="op">==</span> <span class="st">'HALLMARK_NOTCH_SIGNALING'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="va">sample_name</span>, <span class="st">"Lung5_Rep1"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 7</span></span>
<span class="co">#&gt;   sample_name       gene_set             size_…¹ size_…² p_value adj_p…³ prop_…⁴</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Lung5_Rep1_fov_2  HALLMARK_NOTCH_SIGN…      14      34   0.982       1   0.412</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Lung5_Rep1_fov_6  HALLMARK_NOTCH_SIGN…      14      34   0.849       1   0.412</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Lung5_Rep1_fov_11 HALLMARK_NOTCH_SIGN…      14      34   0.988       1   0.412</span>
<span class="co">#&gt; <span style="color: #949494;"># … with abbreviated variable names ¹​size_test, ²​size_gene_set, ³​adj_p_value,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   ⁴​prop_gene_set</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">stenrich_df</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>prop_gene_set<span class="op">=</span><span class="va">size_test</span><span class="op">/</span><span class="va">size_gene_set</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">prop_gene_set</span> <span class="op">&gt;=</span> <span class="fl">0.3</span> <span class="op">&amp;</span> <span class="va">adj_p_value</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">prop_gene_set</span><span class="op">)</span>, <span class="va">adj_p_value</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gene_set</span> <span class="op">==</span> <span class="st">'HALLMARK_NOTCH_SIGNALING'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="va">sample_name</span>, <span class="st">"Lung6"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 7</span></span>
<span class="co">#&gt;   sample_name gene_set                 size_test size_…¹ p_value adj_p…² prop_…³</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Lung6_fov_4 HALLMARK_NOTCH_SIGNALING        14      34       0       0   0.412</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Lung6_fov_6 HALLMARK_NOTCH_SIGNALING        14      34       0       0   0.412</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Lung6_fov_7 HALLMARK_NOTCH_SIGNALING        14      34       0       0   0.412</span>
<span class="co">#&gt; <span style="color: #949494;"># … with abbreviated variable names ¹​size_gene_set, ²​adj_p_value,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   ³​prop_gene_set</span></span></code></pre></div>
<p>Plots can be generated like so:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STplot.html">STplot</a></span><span class="op">(</span><span class="va">lung</span>, genes<span class="op">=</span><span class="va">gene_sets</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span> <span class="op">==</span> <span class="st">'HALLMARK_NOTCH_SIGNALING'</span><span class="op">]</span>, 
             samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Lung5_Rep1_fov_2'</span>, <span class="st">'Lung5_Rep1_fov_6'</span>, <span class="st">'Lung5_Rep1_fov_11'</span>,
                       <span class="st">'Lung6_fov_4'</span>, <span class="st">'Lung6_fov_6'</span>, <span class="st">'Lung6_fov_7'</span><span class="op">)</span>,
             color_pal<span class="op">=</span><span class="st">'YlOrBr'</span><span class="op">)</span>

<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">qp</span>, ncol<span class="op">=</span><span class="fl">3</span>, nrow<span class="op">=</span><span class="fl">2</span>, common.legend<span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<p><img src="spatial_enrichment_gradients_smi_files/figure-html/notch_plots-1.png" width="960"></p>
<p>Although expression of the NOTCH signaling pathway is higher overall
in “Lung6”, the expression is not uniform, but rather is arranged in
patterns within each FOV. Conversely, cells in FOVs of “Lung5_Rep1” that
show high expression of the NOTCH signaling pathway are scattered across
the tissue.</p>
<p>Since this expression of this pathway has been identified as
spatially aggregated, it might be of interest to know if there is
correspondence of NOTCH signaling with tissue domains. With the
<code>STclust</code> function, tissue domains can be identified.</p>
<p>To speed up computations, the STlist will be subset to only the FOVs
in the plots</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">keep_fovs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Lung5_Rep1_fov_2'</span>, <span class="st">'Lung5_Rep1_fov_6'</span>, <span class="st">'Lung5_Rep1_fov_11'</span>,
              <span class="st">'Lung6_fov_4'</span>, <span class="st">'Lung6_fov_6'</span>, <span class="st">'Lung6_fov_7'</span><span class="op">)</span>
<span class="va">rm_fovs</span> <span class="op">=</span> <span class="va">summ_df</span><span class="op">$</span><span class="va">sample_name</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="va">summ_df</span><span class="op">$</span><span class="va">sample_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">keep_fovs</span><span class="op">)</span><span class="op">]</span> 
<span class="va">lung_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_data.html">filter_data</a></span><span class="op">(</span><span class="va">lung</span>, rm_tissue<span class="op">=</span><span class="va">rm_fovs</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lung_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STclust.html">STclust</a></span><span class="op">(</span><span class="va">lung_subset</span>, ws<span class="op">=</span><span class="fl">0.02</span>, ks<span class="op">=</span><span class="st">'dtc'</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #00BB00;">STclust started.</span></span>
<span class="co"><span style="color: #00BB00;">#&gt; STclust completed in 1.22 min.</span></span>
<span class="co"><span style="color: #00BB00;">#&gt; </span></span></code></pre></div>
<p>Optinally, if tissue images are available for some or all the FOVs,
users can upload those to the STlist for display next to the other plot
types available in <code>spatialGE</code>. Users can upload an entire
folder of images to the STlist, as long as image files have the same
names as the FOVs within the STlist (e.g., “Lung5_Rep1_fov_2”,
“Lung6_fov_4”, etc). Keep in mind, however, this will increase the size
of the STlist significantly, especially in CosMx experiment that usually
contain many FOVs. Here, only the images for the FOVs on which domain
prediction was done will be uploaded. To do that, the corresponding
images need to be renamed. Please, locate in the downloaded data the
folder named “CellComposite” (there is one for each slide). The files
are named “CellComposite_F001.jpg”, “CellComposite_F002.jpg”, and so on.
The consecutive numbers indicate the FOV id. Now please rename the files
as follows:</p>
<ul>
<li>For the “CellComposite” folder within the folder of slide
“Lung5_Rep1”, please change: “CellComposite_F002.jpg”,
“CellComposite_F006.jpg”, and “CellComposite_F011.jpg” to
“Lung5_Rep1_fov_2.jpg”, “Lung5_Rep1_fov_6.jpg”, and
“Lung5_Rep1_fov_11.jpg” respectively.</li>
<li>For the “CellComposite” folder within the folder of slide “Lung6”,
please change, please rename “CellComposite_F004.jpg”,
“CellComposite_F006.jpg”, and “CellComposite_F007.jpg” to
“Lung6_fov_4.jpg”, “Lung6_fov_6.jpg”, and “Lung6_fov_7.jpg”
respectively.</li>
</ul>
<p>Now, upload the images to the STlist with the
<code>load_images</code> function:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lung_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_images.html">load_images</a></span><span class="op">(</span><span class="va">lung_subset</span>, images<span class="op">=</span><span class="st">'~/Desktop/lung_cancer_smi/Lung5_Rep1/Lung5_Rep1-Flat_files_and_images/CellComposite'</span><span class="op">)</span>
<span class="co">#&gt; Image for sample Lung6_fov_4 was not found.</span>
<span class="co">#&gt; Image for sample Lung6_fov_6 was not found.</span>
<span class="co">#&gt; Image for sample Lung6_fov_7 was not found.</span></code></pre></div>
<p>The images for slide “Lung6” were not found, which are in the
corresponsing folder. The <code>load_images</code> function is ran again
to add images from “Lung6”:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lung_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_images.html">load_images</a></span><span class="op">(</span><span class="va">lung_subset</span>, images<span class="op">=</span><span class="st">'~/Desktop/lung_cancer_smi/Lung6/Lung6-Flat_files_and_images/CellComposite'</span><span class="op">)</span>
<span class="co">#&gt; Image for sample Lung5_Rep1_fov_2 was not found.</span>
<span class="co">#&gt; Image for sample Lung5_Rep1_fov_6 was not found.</span>
<span class="co">#&gt; Image for sample Lung5_Rep1_fov_11 was not found.</span></code></pre></div>
<p>The function <code>plot_image</code> is used to generate the ggplot
objects containing the images:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ti</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_image.html">plot_image</a></span><span class="op">(</span><span class="va">lung_subset</span><span class="op">)</span></code></pre></div>
<p>The domains can be plotted with the <code>STplot</code> function
too.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dom_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STplot.html">STplot</a></span><span class="op">(</span><span class="va">lung_subset</span>, ks<span class="op">=</span><span class="st">'dtc'</span>, ws<span class="op">=</span><span class="fl">0.02</span>, deepSplit<span class="op">=</span><span class="cn">F</span>,
                color_pal<span class="op">=</span><span class="st">'discreterainbow'</span><span class="op">)</span>

<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span><span class="va">dom_p</span><span class="op">$</span><span class="va">Lung5_Rep1_fov_2_stclust_spw0.02_dsplFalse</span>,
                  <span class="va">ti</span><span class="op">$</span><span class="va">image_Lung5_Rep1_fov_2</span>,
                  <span class="va">dom_p</span><span class="op">$</span><span class="va">Lung5_Rep1_fov_6_stclust_spw0.02_dsplFalse</span>,
                  <span class="va">ti</span><span class="op">$</span><span class="va">image_Lung5_Rep1_fov_6</span>,
                  <span class="va">dom_p</span><span class="op">$</span><span class="va">Lung5_Rep1_fov_11_stclust_spw0.02_dsplFalse</span>,
                  <span class="va">ti</span><span class="op">$</span><span class="va">image_Lung5_Rep1_fov_11</span>,
                  <span class="va">dom_p</span><span class="op">$</span><span class="va">Lung6_fov_4_stclust_spw0.02_dsplFalse</span>,
                  <span class="va">ti</span><span class="op">$</span><span class="va">image_Lung6_fov_4</span>,
                  <span class="va">dom_p</span><span class="op">$</span><span class="va">Lung6_fov_6_stclust_spw0.02_dsplFalse</span>,
                  <span class="va">ti</span><span class="op">$</span><span class="va">image_Lung6_fov_6</span>,
                  <span class="va">dom_p</span><span class="op">$</span><span class="va">Lung6_fov_7_stclust_spw0.02_dsplFalse</span>,
                  <span class="va">ti</span><span class="op">$</span><span class="va">image_Lung6_fov_7</span>,
                  ncol<span class="op">=</span><span class="fl">4</span>, nrow<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="spatial_enrichment_gradients_smi_files/figure-html/plot_domains-1.png" width="960"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># If no tissue images are available:</span>
<span class="co"># ggpubr::ggarrange(plotlist=dom_p, ncol=3, nrow=2)</span></code></pre></div>
<p>Now, to identify the domains, differential gene expression can be
performed:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STdiff.html">STdiff</a></span><span class="op">(</span><span class="va">lung_subset</span>, w<span class="op">=</span><span class="fl">0.02</span>, k<span class="op">=</span><span class="st">'dtc'</span>, deepSplit<span class="op">=</span><span class="cn">F</span>, test_type<span class="op">=</span><span class="st">'wilcoxon'</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #BBBB00;">Testing STclust assignment (w=0.02, dtc deepSplit=False)...</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt;    Running Wilcoxon's tests...</span></span>
<span class="co"><span style="color: #BBBB00;">#&gt; </span><span style="color: #00BB00;">   Completed Wilcoxon's tests (12.51 min).</span></span>
<span class="co"><span style="color: #00BB00;">#&gt; STdiff completed in 12.52 min.</span></span>
<span class="co"><span style="color: #00BB00;">#&gt; </span></span></code></pre></div>
<p>Then, the top 3 DE genes per cluster in the samples showing NOTCH
signaling spatial patterns are identified:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">deg</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">adj_p_val</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">avg_log2fc</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Lung6_fov_4'</span>, <span class="st">'Lung6_fov_6'</span>, <span class="st">'Lung6_fov_7'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">cluster_1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 30 × 6</span></span>
<span class="co">#&gt; <span style="color: #949494;"># Groups:   sample, cluster_1 [10]</span></span>
<span class="co">#&gt;    sample      gene     avg_log2fc cluster_1 wilcox_p_val adj_p_val</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Lung6_fov_4 EGFR          0.738 1            2.62<span style="color: #949494;">e</span><span style="color: #BB0000;">-171</span> 1.08<span style="color: #949494;">e</span><span style="color: #BB0000;">-168</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Lung6_fov_4 GPNMB         0.705 1            1.29<span style="color: #949494;">e</span><span style="color: #BB0000;">-136</span> 3.10<span style="color: #949494;">e</span><span style="color: #BB0000;">-134</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Lung6_fov_4 KRT5          0.613 1            5.86<span style="color: #949494;">e</span><span style="color: #BB0000;">-126</span> 9.92<span style="color: #949494;">e</span><span style="color: #BB0000;">-124</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Lung6_fov_4 CD74          0.645 2            2.48<span style="color: #949494;">e</span><span style="color: #BB0000;">-220</span> 3.56<span style="color: #949494;">e</span><span style="color: #BB0000;">-217</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Lung6_fov_4 VIM           0.565 2            7.80<span style="color: #949494;">e</span><span style="color: #BB0000;">-218</span> 7.49<span style="color: #949494;">e</span><span style="color: #BB0000;">-215</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Lung6_fov_4 HLA-DRB1      0.712 2            5.33<span style="color: #949494;">e</span><span style="color: #BB0000;">-194</span> 3.21<span style="color: #949494;">e</span><span style="color: #BB0000;">-191</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Lung6_fov_4 KRT13         1.11  3            1.61<span style="color: #949494;">e</span><span style="color: #BB0000;">-117</span> 2.32<span style="color: #949494;">e</span><span style="color: #BB0000;">-115</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Lung6_fov_4 KRT16         0.993 3            4.89<span style="color: #949494;">e</span><span style="color: #BB0000;">-107</span> 5.42<span style="color: #949494;">e</span><span style="color: #BB0000;">-105</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Lung6_fov_4 ANXA1         0.594 3            6.77<span style="color: #949494;">e</span><span style="color: #BB0000;">- 78</span> 4.75<span style="color: #949494;">e</span><span style="color: #BB0000;">- 76</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Lung6_fov_6 HMGB2         0.273 1            2.52<span style="color: #949494;">e</span><span style="color: #BB0000;">- 55</span> 2.69<span style="color: #949494;">e</span><span style="color: #BB0000;">- 53</span></span>
<span class="co">#&gt; <span style="color: #949494;"># … with 20 more rows</span></span></code></pre></div>
<p>Expression of KRT genes in domains 1 and 2 of FOVs “Lung6_fov_4” and
“Lung6_fov_7”, and domains 2 and 4 of “Lung6_fov_6” are indicative of
tumor cells. In those domains (with the exception of 4 in
“Lung6_fov_6”), there seems to be an overlap with the areas of of NOTCH
signaling, which indicates that in slide “Lung6”, the tumor compartment
is spatially enriched for NOTCH signaling.</p>
</div>
<div class="section level2">
<h2 id="identifcation-of-genes-presenting-expression-spatial-gradients">Identifcation of genes presenting expression spatial gradients<a class="anchor" aria-label="anchor" href="#identifcation-of-genes-presenting-expression-spatial-gradients"></a>
</h2>
<p>Since the tumor areas have been visually identified, it is possible
now to ascertain which genes show expression that varies with distance
within the tumor compartment. In the FOVs “Lung6_fov_4”, “Lung6_fov_7”,
and “Lung6_fov_6”, the domain 3 seems to be surrounding the tumor
compartment. Hence, domain 3 will be used as reference domain to test
for genes for which expression increases or decreases with distance from
domain 3. This test is performed by the <code>STgradient</code>
function. The following diagram depicts the <code>STgradient</code>
algorithm:</p>
<p align="center">
<img src="../spatial_gradients.png" height="250" width="550"></p>
<p>In <code>STgradient</code>, the spatial coordinates are used to
calculate Euclidean distances of each cell to each of the other cells.
The reference domain is the domain from which the distances will be
calculated (domain 3 in this case). The non-reference domain(s) are the
domain(s) on which the user wants to study gradients. Next, removes
isolated spots/cells in the reference domain (i.e., have a number of
immediate neighbors smaller than parameter <code>min_nb</code>). The
<code>min_nb</code> parameter intends to reduce the effects of small
“pockets” of the reference domain on the correlation coefficient. At
this point, either the minimum or average distance of each non-reference
cell to each reference cell is calculated (<code>distsumm</code>). The
minimum distance is easier to interpret, however, it is more sensitive
to highly fragmented reference domains. On the other hand, the
interpretability of the average distance is not as easy as the minimum
distance, but it might detect in a better way the gradients when the
reference domain is distributed across the sample. Next, the gene
expression data is subset to the non-reference spots/cells, and the top
variable genes (<code>topgenes</code>) are selected. Results will be
produced only for those genes. Depending on the user’s selection, the
ordinary least squares (OLS) or robust regression is used to calculate
the slope of the regression line between the minimum/average distance
and the expression of a gene. If OLS was selected, users might opt to
remove outliers via the interquartile method. Finally, the Spearman’s
coefficient is calculated in addition to the regression line
coefficient. The slope indicates the direction of the correlation. If
positive, the gene expression tends to be higher when farther from the
reference domain. If negative, gene expression is higher when closer to
the reference domain. The interpretation of the sign for the Spearman’s
coefficient is similar, however its magnitude indicates how strong the
correlation is between gene expression and distance to the reference
domain.</p>
<p>The <code>STgradient</code> function can be executed with these
commands, in order to test for gradients within the tumor compartment.
First, it is necessary to find the name of the column in
<code>lung_subset@spatial_meta</code> containing the domain
assigments:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lung_subset</span><span class="op">@</span><span class="va">spatial_meta</span><span class="op">$</span><span class="va">Lung5_Rep1_fov_2</span><span class="op">)</span>
<span class="co">#&gt; [1] "libname"                   "xpos"                     </span>
<span class="co">#&gt; [3] "ypos"                      "total_counts"             </span>
<span class="co">#&gt; [5] "total_genes"               "stclust_spw0.02_dsplFalse"</span></code></pre></div>
<p>The name of the column is <code>stclust_spw0.02_dsplFalse</code>
(STclust results using spatial weight of 0.02 and DynamicTreeCuts with
no deepSplit). This name is used in the <code>annot</code> argument of
<code>STgradient</code>:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">stg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STgradient.html">STgradient</a></span><span class="op">(</span><span class="va">lung_subset</span>, 
                  topgenes<span class="op">=</span><span class="fl">1000</span>, 
                  annot<span class="op">=</span><span class="st">'stclust_spw0.02_dsplFalse'</span>,
                  ref<span class="op">=</span><span class="fl">3</span>,
                  samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Lung6_fov_4'</span>, <span class="st">'Lung6_fov_6'</span>, <span class="st">'Lung6_fov_7'</span><span class="op">)</span>, 
                  distsumm<span class="op">=</span><span class="st">'min'</span>, 
                  robust<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #00BB00;">STgradient completed in 0.42 min.</span></span>
<span class="co"><span style="color: #00BB00;">#&gt; </span></span></code></pre></div>
<p>Here are the first few rows of the <code>STgradient</code> result for
sample “Lung6_fov_4”:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">stg</span><span class="op">$</span><span class="va">Lung6_fov_4</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 8</span></span>
<span class="co">#&gt;   sample_name gene   min_lm_coef min_lm_pval min_spe…¹ min_sp…² min_sp…³ min_p…⁴</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Lung6_fov_4 EGFR     -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">234</span>    1.69<span style="color: #949494;">e</span><span style="color: #BB0000;">-42</span>    -<span style="color: #BB0000;">0.255</span> 5.49<span style="color: #949494;">e</span><span style="color: #BB0000;">-44</span> 5.24<span style="color: #949494;">e</span><span style="color: #BB0000;">-41</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Lung6_fov_4 KRT6A    -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">242</span>    1.70<span style="color: #949494;">e</span><span style="color: #BB0000;">-40</span>    -<span style="color: #BB0000;">0.250</span> 2.21<span style="color: #949494;">e</span><span style="color: #BB0000;">-42</span> 1.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-39</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Lung6_fov_4 HLA-B     0.000<span style="text-decoration: underline;">265</span>    1.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-42</span>     0.249 8.34<span style="color: #949494;">e</span><span style="color: #BB0000;">-42</span> 2.66<span style="color: #949494;">e</span><span style="color: #BB0000;">-39</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Lung6_fov_4 S100A2   -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">172</span>    4.16<span style="color: #949494;">e</span><span style="color: #BB0000;">-28</span>    -<span style="color: #BB0000;">0.207</span> 4.10<span style="color: #949494;">e</span><span style="color: #BB0000;">-29</span> 9.80<span style="color: #949494;">e</span><span style="color: #BB0000;">-27</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Lung6_fov_4 KRT19    -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">197</span>    9.85<span style="color: #949494;">e</span><span style="color: #BB0000;">-27</span>    -<span style="color: #BB0000;">0.190</span> 9.66<span style="color: #949494;">e</span><span style="color: #BB0000;">-25</span> 1.85<span style="color: #949494;">e</span><span style="color: #BB0000;">-22</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Lung6_fov_4 SEC61G   -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">168</span>    7.60<span style="color: #949494;">e</span><span style="color: #BB0000;">-24</span>    -<span style="color: #BB0000;">0.185</span> 1.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-23</span> 1.69<span style="color: #949494;">e</span><span style="color: #BB0000;">-21</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #949494;"># … with abbreviated variable names ¹​min_spearman_r, ²​min_spearman_r_pval,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   ³​min_spearman_r_pval_adj, ⁴​min_pval_comment</span></span></code></pre></div>
<p>The key columns to look ar are <code>min_spearman_r</code> and
<code>min_spearman_r_pval_adj</code>. Now, to find genes with
significant spatial gradients across the three samples, these commands
can be used:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">stg</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">min_spearman_r</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">sample_name</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">5</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 15 × 8</span></span>
<span class="co">#&gt;    sample_name gene    min_lm_coef min_lm_pval min_s…¹ min_sp…² min_sp…³ min_p…⁴</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Lung6_fov_4 EGFR     -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">234</span>     1.69<span style="color: #949494;">e</span><span style="color: #BB0000;">-42</span>  -<span style="color: #BB0000;">0.255</span> 5.49<span style="color: #949494;">e</span><span style="color: #BB0000;">-44</span> 5.24<span style="color: #949494;">e</span><span style="color: #BB0000;">-41</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Lung6_fov_6 FOS      -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">078</span><span style="color: #BB0000;">5</span>    7.82<span style="color: #949494;">e</span><span style="color: #BB0000;">-42</span>  -<span style="color: #BB0000;">0.252</span> 1.38<span style="color: #949494;">e</span><span style="color: #BB0000;">-44</span> 6.59<span style="color: #949494;">e</span><span style="color: #BB0000;">-42</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Lung6_fov_7 G6PD     -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">107</span>     9.58<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span>  -<span style="color: #BB0000;">0.111</span> 8.64<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span> 2.07<span style="color: #949494;">e</span><span style="color: #BB0000;">- 7</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Lung6_fov_4 HLA-B     0.000<span style="text-decoration: underline;">265</span>     1.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-42</span>   0.249 8.34<span style="color: #949494;">e</span><span style="color: #BB0000;">-42</span> 2.66<span style="color: #949494;">e</span><span style="color: #BB0000;">-39</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Lung6_fov_6 KRT16    -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">119</span>     1.24<span style="color: #949494;">e</span><span style="color: #BB0000;">-32</span>  -<span style="color: #BB0000;">0.221</span> 2.78<span style="color: #949494;">e</span><span style="color: #BB0000;">-34</span> 5.30<span style="color: #949494;">e</span><span style="color: #BB0000;">-32</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Lung6_fov_4 KRT19    -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">197</span>     9.85<span style="color: #949494;">e</span><span style="color: #BB0000;">-27</span>  -<span style="color: #BB0000;">0.190</span> 9.66<span style="color: #949494;">e</span><span style="color: #BB0000;">-25</span> 1.85<span style="color: #949494;">e</span><span style="color: #BB0000;">-22</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Lung6_fov_4 KRT6A    -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">242</span>     1.70<span style="color: #949494;">e</span><span style="color: #BB0000;">-40</span>  -<span style="color: #BB0000;">0.250</span> 2.21<span style="color: #949494;">e</span><span style="color: #BB0000;">-42</span> 1.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-39</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Lung6_fov_6 KRT6B    -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">114</span>     3.83<span style="color: #949494;">e</span><span style="color: #BB0000;">-40</span>  -<span style="color: #BB0000;">0.241</span> 1.09<span style="color: #949494;">e</span><span style="color: #BB0000;">-40</span> 2.60<span style="color: #949494;">e</span><span style="color: #BB0000;">-38</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Lung6_fov_6 MT1X     -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">169</span>     1.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-53</span>  -<span style="color: #BB0000;">0.278</span> 4.65<span style="color: #949494;">e</span><span style="color: #BB0000;">-54</span> 4.43<span style="color: #949494;">e</span><span style="color: #BB0000;">-51</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Lung6_fov_7 NDRG1     0.000<span style="text-decoration: underline;">064</span>3    8.81<span style="color: #949494;">e</span><span style="color: #BB0000;">- 6</span>   0.122 1.22<span style="color: #949494;">e</span><span style="color: #BB0000;">-11</span> 3.90<span style="color: #949494;">e</span><span style="color: #BB0000;">- 9</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">11</span> Lung6_fov_7 NEAT1     0.000<span style="text-decoration: underline;">141</span>     3.55<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span>   0.125 5.32<span style="color: #949494;">e</span><span style="color: #BB0000;">-12</span> 2.55<span style="color: #949494;">e</span><span style="color: #BB0000;">- 9</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">12</span> Lung6_fov_6 NTRK2     0.000<span style="text-decoration: underline;">132</span>     6.21<span style="color: #949494;">e</span><span style="color: #BB0000;">-45</span>   0.248 3.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-43</span> 1.25<span style="color: #949494;">e</span><span style="color: #BB0000;">-40</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">13</span> Lung6_fov_7 S100A10  -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">166</span>     4.43<span style="color: #949494;">e</span><span style="color: #BB0000;">-17</span>  -<span style="color: #BB0000;">0.151</span> 5.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-17</span> 4.84<span style="color: #949494;">e</span><span style="color: #BB0000;">-14</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">14</span> Lung6_fov_4 S100A2   -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">172</span>     4.16<span style="color: #949494;">e</span><span style="color: #BB0000;">-28</span>  -<span style="color: #BB0000;">0.207</span> 4.10<span style="color: #949494;">e</span><span style="color: #BB0000;">-29</span> 9.80<span style="color: #949494;">e</span><span style="color: #BB0000;">-27</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">15</span> Lung6_fov_7 TYK2     -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">072</span><span style="color: #BB0000;">9</span>    1.43<span style="color: #949494;">e</span><span style="color: #BB0000;">- 7</span>  -<span style="color: #BB0000;">0.105</span> 5.91<span style="color: #949494;">e</span><span style="color: #BB0000;">- 9</span> 1.13<span style="color: #949494;">e</span><span style="color: #BB0000;">- 6</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #949494;"># … with abbreviated variable names ¹​min_spearman_r, ²​min_spearman_r_pval,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   ³​min_spearman_r_pval_adj, ⁴​min_pval_comment</span></span></code></pre></div>
<p>It can be seen from these results that multiple keratin genes show
spatial gradients, all of them with negative Spearman’s coefficients
(cells close to domain 3 tend to have higher keratin expression). Other
such <strong>HLA-B</strong> and <strong>NDRG1</strong> have higher
expression in cells far from domain 3:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/STplot.html">STplot</a></span><span class="op">(</span><span class="va">lung_subset</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'KRT6A'</span>, <span class="st">'HLA-B'</span><span class="op">)</span>, samples<span class="op">=</span><span class="st">'Lung6_fov_4'</span>, color_pal<span class="op">=</span><span class="st">'YlOrBr'</span><span class="op">)</span>
<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist<span class="op">=</span><span class="va">qp2</span>, common.legend<span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<p><img src="spatial_enrichment_gradients_smi_files/figure-html/stg_plots-1.png" width="768"></p>
<p>Here are the results for genes within the NOTCH signaling
pathway:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">stg</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">min_spearman_r</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">gene_sets</span><span class="op">[[</span><span class="st">'HALLMARK_NOTCH_SIGNALING'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">min_spearman_r_pval_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 12 × 8</span></span>
<span class="co">#&gt;    sample_name gene   min_lm_coef  min_lm_pval min_spe…¹ min_s…² min_s…³ min_p…⁴</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Lung6_fov_4 JAG1      -<span style="color: #BB0000;">8.61</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> 0.000<span style="text-decoration: underline;">000</span>032<span style="text-decoration: underline;">3</span>   -<span style="color: #BB0000;">0.108</span>  6.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-9</span> 1.66<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Lung6_fov_4 FZD7      -<span style="color: #BB0000;">3.54</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-8</span> 0.000<span style="text-decoration: underline;">001</span>44     -<span style="color: #BB0000;">0.093</span><span style="color: #BB0000; text-decoration: underline;">7</span> 4.77<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span> 7.60<span style="color: #949494;">e</span><span style="color: #BB0000;">-6</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Lung6_fov_6 NOTCH3    -<span style="color: #BB0000;">2.60</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> 0.000<span style="text-decoration: underline;">001</span>27     -<span style="color: #BB0000;">0.090</span><span style="color: #BB0000; text-decoration: underline;">9</span> 6.51<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span> 1.38<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Lung6_fov_4 NOTCH1    -<span style="color: #BB0000;">9.88</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-9</span> 0.000<span style="text-decoration: underline;">072</span>0      -<span style="color: #BB0000;">0.076</span><span style="color: #BB0000; text-decoration: underline;">5</span> 4.01<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> 4.40<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Lung6_fov_6 FZD7      -<span style="color: #BB0000;">2.10</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> 0.000<span style="text-decoration: underline;">089</span>7      -<span style="color: #BB0000;">0.069</span><span style="color: #BB0000; text-decoration: underline;">8</span> 1.34<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span> 1.54<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Lung6_fov_6 WNT5A      2.51<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> 0.000<span style="text-decoration: underline;">388</span>        0.066<span style="text-decoration: underline;">4</span> 2.85<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span> 2.92<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Lung6_fov_7 WNT5A     -<span style="color: #BB0000;">1.06</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span> 0.000<span style="text-decoration: underline;">070</span>6      -<span style="color: #BB0000;">0.065</span><span style="color: #BB0000; text-decoration: underline;">7</span> 2.83<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span> 5.77<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Lung6_fov_7 FZD1      -<span style="color: #BB0000;">5.03</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-9</span> 0.000<span style="text-decoration: underline;">364</span>       -<span style="color: #BB0000;">0.063</span><span style="color: #BB0000; text-decoration: underline;">5</span> 4.52<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span> 7.87<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Lung6_fov_7 MAML2     -<span style="color: #BB0000;">3.64</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-9</span> 0.000<span style="text-decoration: underline;">494</span>       -<span style="color: #BB0000;">0.061</span><span style="color: #BB0000; text-decoration: underline;">2</span> 7.29<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span> 1.07<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Lung6_fov_4 NOTCH3    -<span style="color: #BB0000;">1.20</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-8</span> 0.000<span style="text-decoration: underline;">783</span>       -<span style="color: #BB0000;">0.059</span><span style="color: #BB0000; text-decoration: underline;">8</span> 1.33<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> 9.02<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">11</span> Lung6_fov_6 NOTCH1    -<span style="color: #BB0000;">7.64</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-8</span> 0.000<span style="text-decoration: underline;">929</span>       -<span style="color: #BB0000;">0.058</span><span style="color: #BB0000; text-decoration: underline;">4</span> 1.41<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> 1.07<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">12</span> Lung6_fov_7 PPARD     -<span style="color: #BB0000;">1.65</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-9</span> 0.004<span style="text-decoration: underline;">01</span>        -<span style="color: #BB0000;">0.050</span><span style="color: #BB0000; text-decoration: underline;">1</span> 5.65<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> 4.51<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> <span style="color: #BB0000;">NA</span>     </span>
<span class="co">#&gt; <span style="color: #949494;"># … with abbreviated variable names ¹​min_spearman_r, ²​min_spearman_r_pval,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   ³​min_spearman_r_pval_adj, ⁴​min_pval_comment</span></span></code></pre></div>
<p>Although the correlations in genes from this pathway are not as large
as and those for <strong>KRT6A</strong> and <strong>HLA-B</strong>
(i.e., lower absolute Spearman’s coefficient), it can be seen that the
genes <strong>NOTCH1</strong> and <strong>NOTCH3</strong> show evidence
of expression gradients in FOVs ‘Lung6_fov_4’ and ‘Lung6_fov_6’.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-hafemeister_2019" class="csl-entry">
Hafemeister, Christoph, and Rahul Satija. 2019. <span>“Normalization and
Variance Stabilization of Single-Cell RNA-Seq Data Using Regularized
Negative Binomial Regression.”</span> <em>Genome Biology</em> 20: 296.
<a href="https://doi.org/10.1186/s13059-019-1874-1" class="external-link">https://doi.org/10.1186/s13059-019-1874-1</a>.
</div>
<div id="ref-hunter2021spatially" class="csl-entry">
Hunter, Miranda V, Reuben Moncada, Joshua M Weiss, Itai Yanai, and
Richard M White. 2021. <span>“Spatially Resolved Transcriptomics Reveals
the Architecture of the Tumor-Microenvironment Interface.”</span>
<em>Nature Communications</em> 12 (1): 6278.
</div>
</div>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Big Sur 10.16</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] forcats_0.5.1   stringr_1.5.0   dplyr_1.1.0     purrr_1.0.1    </span>
<span class="co">#&gt;  [5] readr_2.1.2     tidyr_1.3.0     tibble_3.2.1    ggplot2_3.3.5  </span>
<span class="co">#&gt;  [9] tidyverse_1.3.1 msigdbr_7.4.1   spatialGE_1.2.0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] utf8_1.2.2            reticulate_1.24       tidyselect_1.2.0     </span>
<span class="co">#&gt;   [4] htmlwidgets_1.5.4     grid_4.1.2            Rtsne_0.15           </span>
<span class="co">#&gt;   [7] ggpolypath_0.1.0      munsell_0.5.0         codetools_0.2-18     </span>
<span class="co">#&gt;  [10] ragg_1.2.5            ica_1.0-2             future_1.23.0        </span>
<span class="co">#&gt;  [13] miniUI_0.1.1.1        withr_2.5.0           colorspace_2.0-2     </span>
<span class="co">#&gt;  [16] highr_0.9             knitr_1.37            rstudioapi_0.13      </span>
<span class="co">#&gt;  [19] Seurat_4.1.0          ROCR_1.0-11           ggsignif_0.6.3       </span>
<span class="co">#&gt;  [22] tensor_1.5            listenv_0.8.0         labeling_0.4.2       </span>
<span class="co">#&gt;  [25] polyclip_1.10-0       farver_2.1.0          rprojroot_2.0.2      </span>
<span class="co">#&gt;  [28] parallelly_1.30.0     vctrs_0.6.2           generics_0.1.2       </span>
<span class="co">#&gt;  [31] xfun_0.31             R6_2.5.1              khroma_1.8.0         </span>
<span class="co">#&gt;  [34] locfit_1.5-9.4        fields_14.1           bitops_1.0-7         </span>
<span class="co">#&gt;  [37] spatstat.utils_3.0-1  cachem_1.0.6          assertthat_0.2.1     </span>
<span class="co">#&gt;  [40] promises_1.2.0.1      scales_1.2.0          gtable_0.3.0         </span>
<span class="co">#&gt;  [43] globals_0.14.0        goftest_1.2-3         spam_2.9-1           </span>
<span class="co">#&gt;  [46] rlang_1.1.1           systemfonts_1.0.4     splines_4.1.2        </span>
<span class="co">#&gt;  [49] rstatix_0.7.0         lazyeval_0.2.2        spatstat.geom_3.0-6  </span>
<span class="co">#&gt;  [52] broom_0.7.12          yaml_2.2.2            reshape2_1.4.4       </span>
<span class="co">#&gt;  [55] abind_1.4-5           modelr_0.1.8          backports_1.4.1      </span>
<span class="co">#&gt;  [58] httpuv_1.6.5          tools_4.1.2           ellipsis_0.3.2       </span>
<span class="co">#&gt;  [61] spatstat.core_2.3-2   jquerylib_0.1.4       RColorBrewer_1.1-2   </span>
<span class="co">#&gt;  [64] BiocGenerics_0.40.0   dynamicTreeCut_1.63-1 wordspace_0.2-6      </span>
<span class="co">#&gt;  [67] ggridges_0.5.3        Rcpp_1.0.10           plyr_1.8.6           </span>
<span class="co">#&gt;  [70] RCurl_1.98-1.5        ggpubr_0.4.0          rpart_4.1-15         </span>
<span class="co">#&gt;  [73] deldir_1.0-6          pbapply_1.5-0         viridis_0.6.2        </span>
<span class="co">#&gt;  [76] cowplot_1.1.1         zoo_1.8-9             SeuratObject_4.0.4   </span>
<span class="co">#&gt;  [79] haven_2.4.3           ggrepel_0.9.1         cluster_2.1.2        </span>
<span class="co">#&gt;  [82] fs_1.5.2              magrittr_2.0.2        data.table_1.14.2    </span>
<span class="co">#&gt;  [85] scattermore_0.7       lmtest_0.9-39         reprex_2.0.1         </span>
<span class="co">#&gt;  [88] RANN_2.6.1            fitdistrplus_1.1-6    matrixStats_0.61.0   </span>
<span class="co">#&gt;  [91] fftwtools_0.9-11      hms_1.1.1             patchwork_1.1.1      </span>
<span class="co">#&gt;  [94] mime_0.12             evaluate_0.14         xtable_1.8-4         </span>
<span class="co">#&gt;  [97] jpeg_0.1-9            sparsesvd_0.2         readxl_1.3.1         </span>
<span class="co">#&gt; [100] gridExtra_2.3         compiler_4.1.2        maps_3.4.0           </span>
<span class="co">#&gt; [103] KernSmooth_2.23-20    crayon_1.4.2          htmltools_0.5.2      </span>
<span class="co">#&gt; [106] tiff_0.1-11           mgcv_1.8-38           later_1.3.0          </span>
<span class="co">#&gt; [109] tzdb_0.3.0            lubridate_1.8.0       DBI_1.1.2            </span>
<span class="co">#&gt; [112] tweenr_1.0.2          dbplyr_2.1.1          MASS_7.3-54          </span>
<span class="co">#&gt; [115] babelgene_21.4        Matrix_1.4-1          car_3.0-12           </span>
<span class="co">#&gt; [118] cli_3.6.0             parallel_4.1.2        dotCall64_1.0-1      </span>
<span class="co">#&gt; [121] igraph_1.2.11         pkgconfig_2.0.3       pkgdown_2.0.7        </span>
<span class="co">#&gt; [124] plotly_4.10.0         spatstat.sparse_3.0-0 xml2_1.3.3           </span>
<span class="co">#&gt; [127] bslib_0.3.1           iotools_0.3-2         rvest_1.0.2          </span>
<span class="co">#&gt; [130] digest_0.6.29         sctransform_0.3.3     RcppAnnoy_0.0.19     </span>
<span class="co">#&gt; [133] spatstat.data_3.0-0   rmarkdown_2.14        cellranger_1.1.0     </span>
<span class="co">#&gt; [136] leiden_0.3.9          uwot_0.1.11           EBImage_4.36.0       </span>
<span class="co">#&gt; [139] shiny_1.7.1           lifecycle_1.0.3       nlme_3.1-153         </span>
<span class="co">#&gt; [142] jsonlite_1.7.3        carData_3.0-5         desc_1.4.2           </span>
<span class="co">#&gt; [145] viridisLite_0.4.0     fansi_1.0.2           pillar_1.8.1         </span>
<span class="co">#&gt; [148] lattice_0.20-45       fastmap_1.1.0         httr_1.4.2           </span>
<span class="co">#&gt; [151] survival_3.4-0        glue_1.6.2            png_0.1-7            </span>
<span class="co">#&gt; [154] ggforce_0.3.3         stringi_1.7.6         sass_0.4.0           </span>
<span class="co">#&gt; [157] textshaping_0.3.6     memoise_2.0.1         irlba_2.3.5          </span>
<span class="co">#&gt; [160] future.apply_1.8.1</span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Oscar Ospina, Brooke Fridley, Alex Soupir.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
