<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>filter_data: Filters cells/spots, genes, or samples — filter_data • spatialGE</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="filter_data: Filters cells/spots, genes, or samples — filter_data"><meta property="og:description" content="Filtering of spots/cells, genes or samples, as well as count-based
filtering"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basic_functions_vignette.html">Basic functionality: Visualization, domain detection, and spatial heterogeneity</a>
    </li>
    <li>
      <a href="../articles/spatial_differential_expression.html">Spatially-informed differential gene expression</a>
    </li>
    <li>
      <a href="../articles/spatial_enrichment_gradients_smi.html">Detection of gene set spatial patterns and expression gradients in CosMx-SMI</a>
    </li>
    <li>
      <a href="../articles/troubleshooting.html">Troubleshooting</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>filter_data: Filters cells/spots, genes, or samples</h1>

    <div class="hidden name"><code>filter_data.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Filtering of spots/cells, genes or samples, as well as count-based
filtering</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">filter_data</span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  spot_minreads <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  spot_maxreads <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  spot_mingenes <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  spot_maxgenes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  spot_minpct <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  spot_maxpct <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_minreads <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  gene_maxreads <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_minspots <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  gene_maxspots <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_minpct <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  gene_maxpct <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rm_tissue <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rm_spots <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rm_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rm_genes_expr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  spot_pct_expr <span class="op">=</span> <span class="st">"^MT-"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>an STlist</p></dd>


<dt id="arg-spot-minreads">spot_minreads<a class="anchor" aria-label="anchor" href="#arg-spot-minreads"></a></dt>
<dd><p>the minimum number of total reads for a spot to be retained</p></dd>


<dt id="arg-spot-maxreads">spot_maxreads<a class="anchor" aria-label="anchor" href="#arg-spot-maxreads"></a></dt>
<dd><p>the maximum number of total reads for a spot to be retained</p></dd>


<dt id="arg-spot-mingenes">spot_mingenes<a class="anchor" aria-label="anchor" href="#arg-spot-mingenes"></a></dt>
<dd><p>the minimum number of non-zero counts for a spot to be retained</p></dd>


<dt id="arg-spot-maxgenes">spot_maxgenes<a class="anchor" aria-label="anchor" href="#arg-spot-maxgenes"></a></dt>
<dd><p>the maximum number of non-zero counts for a spot to be retained</p></dd>


<dt id="arg-spot-minpct">spot_minpct<a class="anchor" aria-label="anchor" href="#arg-spot-minpct"></a></dt>
<dd><p>the minimum percentage of counts for features defined by <code>spot_pct_expr</code> for a spot to be retained.</p></dd>


<dt id="arg-spot-maxpct">spot_maxpct<a class="anchor" aria-label="anchor" href="#arg-spot-maxpct"></a></dt>
<dd><p>the maximum percentage of counts for features defined by <code>spot_pct_expr</code> for a spot to be retained.</p></dd>


<dt id="arg-gene-minreads">gene_minreads<a class="anchor" aria-label="anchor" href="#arg-gene-minreads"></a></dt>
<dd><p>the minimum number of total reads for a gene to be retained</p></dd>


<dt id="arg-gene-maxreads">gene_maxreads<a class="anchor" aria-label="anchor" href="#arg-gene-maxreads"></a></dt>
<dd><p>the maximum number of total reads for a gene to be retained</p></dd>


<dt id="arg-gene-minspots">gene_minspots<a class="anchor" aria-label="anchor" href="#arg-gene-minspots"></a></dt>
<dd><p>he minimum number of spots with non-zero counts for a gene to be retained</p></dd>


<dt id="arg-gene-maxspots">gene_maxspots<a class="anchor" aria-label="anchor" href="#arg-gene-maxspots"></a></dt>
<dd><p>the maximum number of spots with non-zero counts for a gene to be retained</p></dd>


<dt id="arg-gene-minpct">gene_minpct<a class="anchor" aria-label="anchor" href="#arg-gene-minpct"></a></dt>
<dd><p>the minimum percentage of spots with non-zero counts for a gene to be retained</p></dd>


<dt id="arg-gene-maxpct">gene_maxpct<a class="anchor" aria-label="anchor" href="#arg-gene-maxpct"></a></dt>
<dd><p>the maximum percentage of spots with non-zero counts for a gene to be retained</p></dd>


<dt id="arg-samples">samples<a class="anchor" aria-label="anchor" href="#arg-samples"></a></dt>
<dd><p>samples (as in <code>names(x@counts)</code>) to perform filtering.</p></dd>


<dt id="arg-rm-tissue">rm_tissue<a class="anchor" aria-label="anchor" href="#arg-rm-tissue"></a></dt>
<dd><p>sample (as in <code>names(x@counts)</code>) to remove from STlist. Removes samples in <code>x@counts</code>, <code>x@tr_counts</code>, <code>x@spatial_meta</code>, <code>x@gene_meta</code>, and <code>x@sample_meta</code></p></dd>


<dt id="arg-rm-spots">rm_spots<a class="anchor" aria-label="anchor" href="#arg-rm-spots"></a></dt>
<dd><p>vector of spot/cell IDs to remove. Removes spots/cells in <code>x@counts</code>, <code>x@tr_counts</code>, and <code>x@spatial_meta</code></p></dd>


<dt id="arg-rm-genes">rm_genes<a class="anchor" aria-label="anchor" href="#arg-rm-genes"></a></dt>
<dd><p>vector of gene names to remove from STlist. Removes genes in <code>x@counts</code>, <code>x@tr_counts</code>, and <code>x@gene_meta</code></p></dd>


<dt id="arg-rm-genes-expr">rm_genes_expr<a class="anchor" aria-label="anchor" href="#arg-rm-genes-expr"></a></dt>
<dd><p>a regular expression that matches genes to remove. Removes genes in <code>x@counts</code>, <code>x@tr_counts</code>, and <code>x@gene_meta</code></p></dd>


<dt id="arg-spot-pct-expr">spot_pct_expr<a class="anchor" aria-label="anchor" href="#arg-spot-pct-expr"></a></dt>
<dd><p>a expression to use with <code>spot_minpct</code> and <code>spot_maxpct</code>. By default '^MT-'.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>an STlist containing the filtered data</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function provides options to filter elements in an STlist. It can remove
cells/spots or genes based on raw counts (<code>x@counts</code>). Users can input an
regular expression to query gene names and calculate percentages (for example %
mtDNA genes). The function also can filter entire samples. Note that the function
removes cells/spots, genes, and/or samples in the raw counts, transformed counts,
spatial variables, gene variables, and sample metadata. Also note that the function
filters in the following order:</p><ol><li><p>Samples (<code>rm_tissue</code>)</p></li>
<li><p>Spots (<code>rm_spots</code>)</p></li>
<li><p>Genes (<code>rm_genes</code>)</p></li>
<li><p>Genes matching <code>rm_genes_expr</code></p></li>
<li><p>Min and max counts</p></li>
</ol></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Using included melanoma example (Thrane et al.)</span></span></span>
<span class="r-in"><span><span class="co"># Download example data set from spatialGE_Data</span></span></span>
<span class="r-in"><span><span class="va">thrane_tmp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">thrane_tmp</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">thrane_tmp</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">lk</span><span class="op">=</span><span class="st">'https://github.com/FridleyLab/spatialGE_Data/raw/refs/heads/main/melanoma_thrane.zip?download='</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">lk</span>, destfile<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">thrane_tmp</span>, <span class="st">'/'</span>, <span class="st">'melanoma_thrane.zip'</span><span class="op">)</span>, mode<span class="op">=</span><span class="st">'wb'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">zip_tmp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">thrane_tmp</span>, pattern<span class="op">=</span><span class="st">'melanoma_thrane.zip$'</span>, full.names<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span>zipfile<span class="op">=</span><span class="va">zip_tmp</span>, exdir<span class="op">=</span><span class="va">thrane_tmp</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Generate the file paths to be passed to the STlist function</span></span></span>
<span class="r-in"><span><span class="va">count_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">thrane_tmp</span>, <span class="st">'/melanoma_thrane'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                          full.names<span class="op">=</span><span class="cn">TRUE</span>, pattern<span class="op">=</span><span class="st">'counts'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">coord_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">thrane_tmp</span>, <span class="st">'/melanoma_thrane'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                          full.names<span class="op">=</span><span class="cn">TRUE</span>, pattern<span class="op">=</span><span class="st">'mapping'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">clin_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">thrane_tmp</span>, <span class="st">'/melanoma_thrane'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                       full.names<span class="op">=</span><span class="cn">TRUE</span>, pattern<span class="op">=</span><span class="st">'clinical'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Create STlist</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">'spatialGE'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="STlist.html">STlist</a></span><span class="op">(</span>rnacounts<span class="op">=</span><span class="va">count_files</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                   spotcoords<span class="op">=</span><span class="va">coord_files</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                   samples<span class="op">=</span><span class="va">clin_file</span><span class="op">)</span> <span class="co"># Only first two samples</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Sample ST_mel3_rep1 was not found among the count/coordinate files.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Sample ST_mel4_rep2 was not found among the count/coordinate files.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Found matrix data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Matching gene expression and coordinate data...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Converting counts to sparse matrices</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Completed STlist!</span>
<span class="r-in"><span><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu">filter_data</span><span class="op">(</span><span class="va">melanoma</span>, spot_minreads<span class="op">=</span><span class="fl">2000</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Oscar Ospina, Alex Soupir, Brooke Fridley.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

