<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ spatialGE</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0.0000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/melanoma_ST_vignette.html">spatialGE: Guide to spatial heterogeneity analysis in melanoma (Thrane et al. 2018)</a>
    </li>
    <li>
      <a href="../articles/prostatecancer_Visium_vignette.html">spatialGE: Kriging for Visium data sets</a>
    </li>
    <li>
      <a href="../articles/read_GeoMx_output.html">spatialGE: Reading GeoMx-DSP outputs</a>
    </li>
    <li>
      <a href="../articles/spatialGE_troubleshoot.html">Troubleshooting</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="STList-class.html">STList-class</a></code> </p>
        </td>
        <td><p>Definition of an STList object class.</p></td>
      </tr><tr><td>
          <p><code><a href="STList.html">STList()</a></code> </p>
        </td>
        <td><p>STList: Creation of STList objects</p></td>
      </tr><tr><td>
          <p><code><a href="STclust.html">STclust()</a></code> </p>
        </td>
        <td><p>STclust: Detect clusters on ST arrays</p></td>
      </tr><tr><td>
          <p><code><a href="bulk_pca.html">bulk_pca()</a></code> </p>
        </td>
        <td><p>bulk_pca: Performs PCA on "bulk" spatial array</p></td>
      </tr><tr><td>
          <p><code><a href="cluster_purity.html">cluster_purity()</a></code> </p>
        </td>
        <td><p>cluster_purity</p></td>
      </tr><tr><td>
          <p><code><a href="color_parse.html">color_parse()</a></code> </p>
        </td>
        <td><p>color_parse: Creates a color palette</p></td>
      </tr><tr><td>
          <p><code><a href="count_cores.html">count_cores()</a></code> </p>
        </td>
        <td><p>Get the number of cores to use in parallel as a function of the number of
data to process. It will not yield a higher number of cores than half of the
total available cores. Will default to 1 core if not Unix</p></td>
      </tr><tr><td>
          <p><code><a href="create_listw.html">create_listw()</a></code> </p>
        </td>
        <td><p>create_listw</p></td>
      </tr><tr><td>
          <p><code><a href="deconv_krige.html">deconv_krige()</a></code> </p>
        </td>
        <td><p>deconv_krige: Spatial interpolation of gene expression deconvolution scores</p></td>
      </tr><tr><td>
          <p><code><a href="detect_input.html">detect_input()</a></code> </p>
        </td>
        <td><p>detect_input: Determine what is being provided to STList</p></td>
      </tr><tr><td>
          <p><code><a href="dim-STList-method.html">dim(<i>&lt;STList&gt;</i>)</a></code> </p>
        </td>
        <td><p>dim: Prints the dimensions of count arrays within an STList object.</p></td>
      </tr><tr><td>
          <p><code><a href="filter_data.html">filter_data()</a></code> </p>
        </td>
        <td><p>filter_data: Removes spots, genes, or samples</p></td>
      </tr><tr><td>
          <p><code><a href="gene_geary_C.html">gene_geary_C()</a></code> </p>
        </td>
        <td><p>gene_geary_C</p></td>
      </tr><tr><td>
          <p><code><a href="gene_getis_Gi.html">gene_getis_Gi()</a></code> </p>
        </td>
        <td><p>gene_getis_Gi</p></td>
      </tr><tr><td>
          <p><code><a href="gene_krige.html">gene_krige()</a></code> </p>
        </td>
        <td><p>gene_krige: Spatial interpolation of gene expression</p></td>
      </tr><tr><td>
          <p><code><a href="gene_moran_I.html">gene_moran_I()</a></code> </p>
        </td>
        <td><p>gene_moran_I</p></td>
      </tr><tr><td>
          <p><code><a href="get_gene_stats.html">get_gene_stats()</a></code> </p>
        </td>
        <td><p>get_gene_stats</p></td>
      </tr><tr><td>
          <p><code><a href="krige_p.html">krige_p()</a></code> </p>
        </td>
        <td><p>krige_p</p></td>
      </tr><tr><td>
          <p><code><a href="krige_p_purity.html">krige_p_purity()</a></code> </p>
        </td>
        <td><p>krige_p_purity</p></td>
      </tr><tr><td>
          <p><code><a href="krige_p_pvals.html">krige_p_pvals()</a></code> </p>
        </td>
        <td><p>krige_p_pvals</p></td>
      </tr><tr><td>
          <p><code><a href="load_images.html">load_images()</a></code> </p>
        </td>
        <td><p>load_images: Place tissue images within STList</p></td>
      </tr><tr><td>
          <p><code><a href="log_transf.html">log_transf()</a></code> </p>
        </td>
        <td><p>log_transf: Log-transformation of spatial arrays</p></td>
      </tr><tr><td>
          <p><code><a href="plot_QC_quilt.html">plot_QC_quilt()</a></code> </p>
        </td>
        <td><p>plot_QC_quilt: Plot UMI counts at each spot from an spatial array</p></td>
      </tr><tr><td>
          <p><code><a href="plot_QC_stats.html">plot_QC_stats()</a></code> </p>
        </td>
        <td><p>plot_QCstats: Plots distribution of counts, gene number, or percents per sample</p></td>
      </tr><tr><td>
          <p><code><a href="plot_STclusters.html">plot_STclusters()</a></code> </p>
        </td>
        <td><p>plot_STclusters: Plot cluster memberships of ST spots</p></td>
      </tr><tr><td>
          <p><code><a href="plot_deconv_krige.html">plot_deconv_krige()</a></code> </p>
        </td>
        <td><p>plot_deconv_krige: Visualize spatial interpolation of deconvolution scores</p></td>
      </tr><tr><td>
          <p><code><a href="plot_deconv_quilt.html">plot_deconv_quilt()</a></code> </p>
        </td>
        <td><p>plot_deconv_quilt: Deconvolution scores at each spot</p></td>
      </tr><tr><td>
          <p><code><a href="plot_gene_krige.html">plot_gene_krige()</a></code> </p>
        </td>
        <td><p>plot_gene_krige: Visualize transcriptomic surfaces</p></td>
      </tr><tr><td>
          <p><code><a href="plot_gene_quilt.html">plot_gene_quilt()</a></code> </p>
        </td>
        <td><p>plot_gene_quilt: Gene expression at each spot from an spatial array</p></td>
      </tr><tr><td>
          <p><code><a href="quilt_p.html">quilt_p()</a></code> </p>
        </td>
        <td><p>quilt_p</p></td>
      </tr><tr><td>
          <p><code><a href="quilt_p_purity.html">quilt_p_purity()</a></code> </p>
        </td>
        <td><p>quilt_p_purity</p></td>
      </tr><tr><td>
          <p><code><a href="quilt_p_purity_bw.html">quilt_p_purity_bw()</a></code> </p>
        </td>
        <td><p>quilt_p_purity_bw</p></td>
      </tr><tr><td>
          <p><code><a href="show-STList-method.html">show(<i>&lt;STList&gt;</i>)</a></code> </p>
        </td>
        <td><p>show: Prints overview of STList oject.</p></td>
      </tr><tr><td>
          <p><code><a href="spatialDeconv.html">spatialDeconv()</a></code> </p>
        </td>
        <td><p>spatialDeconv: Deconvolution of ST data</p></td>
      </tr><tr><td>
          <p><code><a href="spatialTransform.html">spatialTransform()</a></code> </p>
        </td>
        <td><p>spatialTransform: Data transformation of spatial transcriptomics data</p></td>
      </tr><tr><td>
          <p><code><a href="spatial_purity.html">spatial_purity()</a></code> </p>
        </td>
        <td><p>spatial_purity</p></td>
      </tr><tr><td>
          <p><code><a href="spatial_stat_plot.html">spatial_stat_plot()</a></code> </p>
        </td>
        <td><p>spatial_stat_plot: Plots association of samples variable and spatial statistics</p></td>
      </tr><tr><td>
          <p><code><a href="spatial_stat_plot_gene.html">spatial_stat_plot_gene()</a></code> </p>
        </td>
        <td><p>spatial_stat_plot_gene</p></td>
      </tr><tr><td>
          <p><code><a href="spatial_xcell.html">spatial_xcell()</a></code> </p>
        </td>
        <td><p>spatial_xcell</p></td>
      </tr><tr><td>
          <p><code><a href="summary-STList-method.html">summary(<i>&lt;STList&gt;</i>)</a></code> </p>
        </td>
        <td><p>summary: Prints overview of STList oject.</p></td>
      </tr><tr><td>
          <p><code><a href="voom_norm.html">voom_norm()</a></code> </p>
        </td>
        <td><p>voom_norm: voom transformation of ST arrays</p></td>
      </tr><tr><td>
          <p><code><a href="xcell_names.html">xcell_names()</a></code> </p>
        </td>
        <td><p>xcell_names: Prints xCell cell type names</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Oscar Ospina, Brooke Fridley, Alex Soupir.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

