<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>STdiff: Differential gene expression analysis for spatial transcriptomics data — STdiff • spatialGE</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="STdiff: Differential gene expression analysis for spatial transcriptomics data — STdiff"><meta property="og:description" content="Tests for differentially expressed genes using linear models with or
without spatial covariance structures"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spatialGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basic_functions_vignette.html">Basic functionality: Visualization, domain detection, and spatial heterogeneity</a>
    </li>
    <li>
      <a href="../articles/spatial_differential_expression.html">Spatially-informed differential gene expression</a>
    </li>
    <li>
      <a href="../articles/spatial_enrichment_gradients_smi.html">Detection of gene set spatial patterns and expression gradients in CosMx-SMI</a>
    </li>
    <li>
      <a href="../articles/troubleshooting.html">Troubleshooting</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>STdiff: Differential gene expression analysis for spatial transcriptomics data</h1>

    <div class="hidden name"><code>STdiff.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Tests for differentially expressed genes using linear models with or
without spatial covariance structures</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">STdiff</span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annot <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  w <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  k <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  deepSplit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  topgenes <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  pval_thr <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  pval_adj <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>  test_type <span class="op">=</span> <span class="st">"mm"</span>,</span>
<span>  sp_topgenes <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  clusters <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pairwise <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  cores <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>an STlist</p></dd>


<dt id="arg-samples">samples<a class="anchor" aria-label="anchor" href="#arg-samples"></a></dt>
<dd><p>an integer indicating the spatial samples to be included in the DE tests.
Numbers follow the order in <code>names(x@counts)</code>. Sample names are also allowed.
If NULL, performs tests on all samples</p></dd>


<dt id="arg-annot">annot<a class="anchor" aria-label="anchor" href="#arg-annot"></a></dt>
<dd><p>a column name in <code>x@spatial_meta</code> containing the groups/clusters to
be tested. Required if <code>k</code> and <code>w</code> are empty.</p></dd>


<dt id="arg-w">w<a class="anchor" aria-label="anchor" href="#arg-w"></a></dt>
<dd><p>the spatial weight used in STclust. Required if <code>annot</code> is empty.</p></dd>


<dt id="arg-k">k<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>the k value used in STclust, or <code>dtc</code> for dynamicTreeCut clusters. Required if <code>annot</code> is empty.</p></dd>


<dt id="arg-deepsplit">deepSplit<a class="anchor" aria-label="anchor" href="#arg-deepsplit"></a></dt>
<dd><p>the deepSplit value if used in STclust. Required if <code>k='dtc'</code>.</p></dd>


<dt id="arg-topgenes">topgenes<a class="anchor" aria-label="anchor" href="#arg-topgenes"></a></dt>
<dd><p>an integer indicating the top variable genes to select from each sample
based on variance (default=5000). If NULL, all genes are selected.</p></dd>


<dt id="arg-pval-thr">pval_thr<a class="anchor" aria-label="anchor" href="#arg-pval-thr"></a></dt>
<dd><p>cut-off of adjusted p-values to define differentially expressed genes from
non-spatial linear models. A proportion of genes (<code>sp_topgenes</code>) under this cut-off
will be applied the spatial models. Default=0.05</p></dd>


<dt id="arg-pval-adj">pval_adj<a class="anchor" aria-label="anchor" href="#arg-pval-adj"></a></dt>
<dd><p>Method to adjust p-values. Defaults to <code>FDR</code>. Other options as
available from <code>p.adjust</code></p></dd>


<dt id="arg-test-type">test_type<a class="anchor" aria-label="anchor" href="#arg-test-type"></a></dt>
<dd><p>one of <code>mm</code>, <code>t_test</code>, or <code>wilcoxon</code>. Specifies the type of
test performed.</p></dd>


<dt id="arg-sp-topgenes">sp_topgenes<a class="anchor" aria-label="anchor" href="#arg-sp-topgenes"></a></dt>
<dd><p>Proportion of differentially expressed genes from non-spatial
linear models (and controlled by <code>pval_thr</code>) to use in differential gene expression
analysis with spatial linear models. If 0 (zero), no spatial models are fit. Default=0.2</p></dd>


<dt id="arg-clusters">clusters<a class="anchor" aria-label="anchor" href="#arg-clusters"></a></dt>
<dd><p>cluster name(s) to test DE genes, as opposed to all clusters.</p></dd>


<dt id="arg-pairwise">pairwise<a class="anchor" aria-label="anchor" href="#arg-pairwise"></a></dt>
<dd><p>whether or not to carry tests on a pairwise manner. The default is
<code>pairwise=F</code>, meaning that DE genes are tested by comparing each cluster to the
rest of the pooled cell/spots.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>output progress indicators. If <code>verbose=0</code>, no text is shown in console.
Other values are 1 and 2 indicating increasing level of verbosity. Default is
<code>verbose=1</code></p></dd>


<dt id="arg-cores">cores<a class="anchor" aria-label="anchor" href="#arg-cores"></a></dt>
<dd><p>Number of cores to use in parallelization. If <code>NULL</code>, the number of
cores to use is detected automatically</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>a list with one data frame per sample with results of differential gene
expression analysis</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The method tests for differentially expressed genes between groups of spots/cells
(e.g., clusters) in a spatial transcriptomics sample. Specifically, the function
tests for genes with significantly higher or lower gene expression in one group of
spots/cells with respect to the rest of spots/cells in the sample. The method first
runs non-spatial linear models on the genes to detect differentially expressed genes.
Then spatial linear models with exponential covariance structure are fit on a
subset of genes detected as differentially expressed by the non-linear models (<code>sp_topgenes</code>).
If running on clusters detected via STclust, the user can specify the assignments
using the same parameters (<code>w</code>, <code>k</code>, <code>deepSplit</code>). Otherwise, the assignments are
specified by indicating one of the column names in <code>x@spatial_meta</code>. The function
uses <code><a href="https://rdrr.io/pkg/spaMM/man/fitme.html" class="external-link">spaMM::fitme</a></code> and is computationally expensive even on HPC environments.
To run the STdiff using the non-spatial approach (faster), set <code>sp_topgenes=0</code>.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Oscar Ospina, Alex Soupir, Brooke Fridley.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

