<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Creates an STlist object from one or multiple spatial transcriptomic samples."><title>STlist: Creation of STlist objects for spatial transcriptomics analysis — STlist • spatialGE</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="STlist: Creation of STlist objects for spatial transcriptomics analysis — STlist"><meta property="og:description" content="Creates an STlist object from one or multiple spatial transcriptomic samples."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">spatialGE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/basic_functions_vignette.html">spatialGE: Basic functionality</a>
    <a class="dropdown-item" href="../articles/spatial_diff_expression_vignette.html">Spatially-informed differential gene expression</a>
    <a class="dropdown-item" href="../articles/spatial_enrichment_gradients_smi.html">Detecting gene set spatial patterns expression gradients in CosMx-SMI</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>STlist: Creation of STlist objects for spatial transcriptomics analysis</h1>
      
      <div class="d-none name"><code>STlist.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Creates an STlist object from one or multiple spatial transcriptomic samples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">STlist</span><span class="op">(</span>rnacounts <span class="op">=</span> <span class="cn">NULL</span>, spotcoords <span class="op">=</span> <span class="cn">NULL</span>, samples <span class="op">=</span> <span class="cn">NULL</span>, cores <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>rnacounts</dt>
<dd><p>the count data which can be provided in one of these formats:</p><ul><li><p>File paths to comma- or tab-delimited files containing raw gene counts, one file
for each spatial sample. The first column contains gene names and subsequent columns
contain the expression data for each cell/spot. Duplicate gene names will be
modified using <code>make.unique</code>. Requires <code>spotcoords</code> and <code>samples</code></p></li>
<li><p>File paths to Visium output directories (one per spatial sample). The directory
should follow the structure resulting from <code>spaceranger count</code>. The directory contains
the <code>.h5</code> and <code>spatial</code> sub-directory. If no <code>.h5</code> file is available, sparse
matrices (MEX) from <code>spaceranger count</code>. In that case a second sub-directory
called <code>filtered_feature_bc_matrix</code> should contain contain the <code>barcodes.tsv.gz</code>,
<code>features.tsv.gz</code>, and <code>matrix.mtx.gz</code> files. The <code>spatial</code> sub-directory minimally
contains the coordinates (<code>tissue_positions_list.csv</code>), and optionally the high
resolution PNG image and accompanying scaling factors (<code>scalefactors_json.json</code>).
Requires <code>samples</code></p></li>
<li><p>The <code>exprMat</code> file for each slide of a CosMx-SMI output. The file must contain
the "fov" and "cell_ID" columns. The <code>STlist</code> function will separate data from each
FOV, since analysis in spatialGE is conducted at the FOV level. Requires <code>samples</code> and
<code>spotcoords</code></p></li>
<li><p>A named list of data frames with raw gene counts (one data frame per spatial
sample). Requires <code>spotcoords</code></p></li>
</ul></dd>


<dt>spotcoords</dt>
<dd><p>the cell/spot coordinates. Not required if inputs are Visium
space ranger outputs</p><ul><li><p>File paths to comma- or tab-delimited files containing cell/spot coordinates, one
for each spatial sample. The files must contain three columns: cell/spot IDs, Y positions, and
X positions. The cell/spot IDs must match the column names for each cells/spots (columns) in
the gene count files</p></li>
<li><p>The <code>metadata</code> file for each slide of a CosMx-SMI output. The file must contain
the "fov", "cell_ID", "CenterX_local_px", and "CenterY_local_px" columns. The <code>STlist</code>
function will separate data from each FOV, since analysis in spatialGE is conducted at
the FOV level. Requires <code>samples</code> and <code>rnacounts</code></p></li>
<li><p>A named list of data frames with cell/spot coordinates. The list names must
match list names of the gene counts list</p></li>
</ul></dd>


<dt>samples</dt>
<dd><p>the sample names/IDs and (optionally) metadata associated with
each spatial sample. This file can also include files paths to gene counts and
cell/spot coordinate files, bypassing the need to specify <code>rnacounts</code> and <code>spotcoords</code>.
The following options are available for <code>samples</code>:</p><ul><li><p>A vector with sample names, which will be used to partially match gene
counts and cell/spot coordinates file paths. A sample name must not match file
paths for two different samples</p></li>
<li><p>A path to a file containing a table with metadata. This file is a comma- or
tab-separated table with one sample per row and sample names/IDs in the first
column. Paths to gene counts and coordinate files can be placed in the second and
third columns respectively (while leaving the <code>rnacounts</code> and <code>spotcoords</code> arguments
empty). If Visium directories are provided, only the second column with paths to
<code>spaceranger count</code> directories are expected. Subsequent columns can contain
variables associated with each spatial sample</p></li>
</ul></dd>


<dt>cores</dt>
<dd><p>integer indicating the number of cores to use during parallelization.
If NULL, the function uses half of the available cores at a maximum. The parallelization
uses <code><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">parallel::mclapply</a></code> and works only in Unix systems.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>x an STlist object containing the counts and coordinates, and optionally
the sample metadata, which can be used for downstream analysis with <code>spatialGE</code></p>


    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Objects of the S4 class STlist are the starting point of analyses in <strong><code>spatialGE</code></strong>.
The STlist contains data from one or multiple samples (i.e., tissue slices), and
results from most <code>spatialGE</code>'s functions are stored within the object.</p><ul><li><p>Raw gene counts and spatial coordinates. Gene count data have genes in rows and
sampling units (e.g., cells, spots, ROIs) in columns. Spatial coordinates have
sampling units in rows and three columns: sample unit IDs, Y position, and X position.</p></li>
<li><p>Visium outputs from <em>space ranger</em>. The Visium directory should generally have
the file structure resulting from <code>spaceranger count</code>, with either a count matrix
represented in MEX files or a h5 file. The directory should also contain a <code>spatial</code>
sub-directory, with the spatial coordinates (<code>tissue_positions_list.csv</code>), and
optionally the high resolution tissue image and scaling factor file <code>scalefactors_json.json</code>.</p></li>
<li><p>CosMx-SMI outputs. Two files are required to process SMI outputs: The <code>exprMat</code> and
<code>metadata</code> files. Both files must contain the "fov" and "cell_ID" columns. In addition,
the <code>metadata</code> files must contain the "CenterX_local_px" and "CenterY_local_px" columns.</p></li>
</ul><p>Optionally, the user can input a path to a file containing a table of sample-level
metadata (e.g., clinical outcomes, tissue type, age). This sample metadata file
should contain sample IDs in the first column partially matching the file names of
the count/coordinate file paths or Visium directories.</p>
<p>The function uses parallelization if run in a unix system. Windows users
will experience longer times depending on the number of samples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># Using included melanoma example (Thrane et al.)</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">'spatialGE'</span><span class="op">)</span></span>
<span class="r-in"><span class="va">data_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"spatialGE"</span><span class="op">)</span>, recursive<span class="op">=</span><span class="cn">T</span>, full.names<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span class="r-in"><span class="va">count_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"counts"</span>, <span class="va">data_files</span>, value<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span class="r-in"><span class="va">coord_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"mapping"</span>, <span class="va">data_files</span>, value<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span class="r-in"><span class="va">clin_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"thrane_clinical"</span>, <span class="va">data_files</span>, value<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span class="r-in"><span class="va">melanoma</span> <span class="op">&lt;-</span> <span class="fu">STlist</span><span class="op">(</span>rnacounts<span class="op">=</span><span class="va">count_files</span>, spotcoords<span class="op">=</span><span class="va">coord_files</span>, samples<span class="op">=</span><span class="va">clin_file</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Found matrix data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">Matching gene expression and coordinate data...</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">Converting counts to sparse matrices</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00; font-weight: bold;">Completed STlist!</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span class="va">melanoma</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Spatial Transcriptomics List (STlist).</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8 spatial array(s):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	ST_mel1_rep1 (279 ROIs|spots|cells x 15666 genes)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	ST_mel1_rep2 (293 ROIs|spots|cells x 16148 genes)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	ST_mel2_rep1 (383 ROIs|spots|cells x 16831 genes)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	ST_mel2_rep2 (380 ROIs|spots|cells x 16605 genes)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	ST_mel3_rep1 (256 ROIs|spots|cells x 15653 genes)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	ST_mel3_rep2 (294 ROIs|spots|cells x 15769 genes)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	ST_mel4_rep1 (212 ROIs|spots|cells x 14408 genes)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	ST_mel4_rep2 (248 ROIs|spots|cells x 15991 genes)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	NA ( ROIs|spots|cells x  genes)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	NA ( ROIs|spots|cells x  genes)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9 variables in sample-level data:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	 patient, slice, gender, BRAF_status, NRAS_status, CDKN2A_status, survival, survival_months, RIN </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Oscar Ospina, Brooke Fridley, Alex Soupir.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

