##
#' @title pseudobulk_pca: Performs PCA on "pseudo bulk" spatial sample
#' @description Perform and plot a PCA of "pseudo bulk" RNA-Seq from spatial
#' transcriptomics data.
#' @details
#' This function takes an STlist, and optionally the name of a clinical or sample-associated
#' metadata variable, and performs PCA on "pseudo bulk" RNA-Seq from ST data. The "pseudo bulk"
#' counts are generated by summing all counts from all cell/spots for each gene.
#'
#' @param x an STList.
#' @param plot_meta a string indicating the name of the variable in the sample metadata.
#' @param n_genes Number of most variable genes (standard deviation) to use in PCA
#' @param color_pal a string of a color palette from khroma or RColorBrewer, or a
#' vector with colors with enough elements to plot categories.
#' @param ptsize the size of the points in the plot. Passed to `size` aesthetic from `ggplot2`.
#'
#' @examples
#' # In this example, melanoma is an STList.
#' # bulk_pca(melanoma, samplemeta='gender')
#'
#' @export
#'
#' @import Matrix
#' @import ggplot2
#' @importFrom magrittr %>%
#' @importFrom methods as is new
#
#
pseudobulk_pca <- function(x=NULL, plot_meta=NULL, n_genes=5000, color_pal="muted", ptsize=5) {

  #require('magrittr')
  #require('Matrix')
  #require('ggplot2')

  tr_method='log'

  if(!is.null(plot_meta) && !(plot_meta %in% colnames(x@sample_meta))){
    stop('Variable not in sample metadata. Verify that input matches variable name in sample data')
  }

  # Test if an STList has been input.
  if(is.null(x) | !is(x, 'STlist')){
    stop("The input must be an STList.")
  }

  # Stop function if only one sample provided.
  if(length(x@counts) < 2){
    stop('Refusing to plot a PCA containing a single sample!')
  }

  # Extract clinical/metadata from specified variable. If none specified, use the
  # array IDs from the first column of clinical data.
  # Also get labels for PCA points.
  if(!is.null(plot_meta)){
    clinvar_vals <- as.character(x@sample_meta[[plot_meta]])
    clinvar_vals <- tibble::as_tibble_col(clinvar_vals, column_name=as.character(plot_meta))
    pca_labs <- as.character(x@sample_meta[[1]])
  }else{
    plot_meta <- 'ST_sample'
    clinvar_vals = names(x@counts)
    clinvar_vals <- tibble::as_tibble_col(clinvar_vals, column_name='ST_sample')
    pca_labs = as.character(clinvar_vals[[1]])
  }

  # Create data frame to store "bulk counts".
  bulkcounts_df <- tibble::tibble(gene=rownames(x@counts[[1]]))

  # Loop through count matrices, get gene-wise sums, and collate samples
  for(i in 1:length(x@counts)){
    expanded_mtx = expandSparse(x@counts[[i]])

    bulk_expr <- rowSums(expanded_mtx) %>%
      tibble::as_tibble_col(., column_name=paste0("st_", i)) %>%
      tibble::add_column(gene=rownames(x@counts[[i]]), .before=1)

    bulkcounts_df <- dplyr::inner_join(bulkcounts_df, bulk_expr, by='gene')
  }

  # Apply voom or log transform
  if(tr_method == 'voom'){
    # Get normalization factors from the "bulk" libraries.
    norm_factors <- edgeR::calcNormFactors(bulkcounts_df[, -1], method='TMM', lib.size=NULL)
    # Create new object to store the size-corrected counts.
    normcounts_df <- c()
    # Divide each count value by their respective column (sample) normalization factor.
    for(raw_col in names(bulkcounts_df[, -1])){
      normcounts_df <- dplyr::bind_cols(normcounts_df, tibble::as_tibble(bulkcounts_df[raw_col] / norm_factors[raw_col]))
    }
    # Apply limma-voom to lib-size normalized "bulk" libraries.
    tr_df <- limma::voom(normcounts_df, design=NULL, lib.size=colSums(bulkcounts_df[, -1]), normalize.method='none', plot=F)
    # Put back gene names to matrix and store in object.
    tr_df = tibble::as_tibble(tr_df$E) %>%
      tibble::add_column(gene=bulkcounts_df[[1]], .before=1)
  } else if(tr_method == 'log'){
    # Calculate (spot) library sizes. Then, add 1 to each library size.
    libsizes = colSums(bulkcounts_df[, -1])
    # Check that there are not zero-count spots
    if(any(libsizes == 0)){
      stop('Please, remove spots containing zero reads...')
    }
    # Divide each count value by their respective column (spot) normalization factor.
    normcounts_df = sweep(bulkcounts_df[, -1], 2, libsizes, '/')
    # Then multiply by scaling factor
    # df = df * scale_f
    # Apply log transformation to count data.
    tr_df = log1p(normcounts_df)
    # Put back gene names to matrix and store in object.
    tr_df = tibble::as_tibble(tr_df) %>%
      tibble::add_column(gene=bulkcounts_df[[1]], .before=1) %>%
      tibble::column_to_rownames('gene')
  }

  # Turn transformed counts to a transposed matrix.
  tr_df <- t(as.matrix(tr_df))
  # Get variable genes and subset
  vargenes = sort(apply(tr_df, 2, sd), decreasing=T)
  tr_df = tr_df[, names(vargenes[1:n_genes])]

  # Perform PCA on transoposed matrix.
  pca_expr <- prcomp(tr_df, scale=TRUE)

  # Get PCA coordinates and add clinical variable data.
  pca_tbl <- tibble::as_tibble(pca_expr$x)
  pca_tbl <- pca_tbl %>%
    tibble::add_column(clinvar_vals, .before=1)

  # Get number of categories from selected variable.
  n_cats <- nlevels(as.factor(pca_tbl[[plot_meta]]))

  # Create color palette.
  cat_cols = color_parse(color_pal, n_cats=n_cats)

  # Assocuate colors with categories.
  names(cat_cols) <- levels(as.factor(pca_tbl[[plot_meta]]))

  # Define shapes of points according to clinical variable.
  cat_shapes <- (16:25)[1:n_cats]
  names(cat_shapes) <- levels(as.factor(pca_tbl[[plot_meta]]))

# NOTE: INSTEAD OF NUMBERS, WOULD BE GREAT TO HAVE SAMPLE ID PLOTTED
  ggplot(pca_tbl) +
    geom_point(aes(x=PC1, y=PC2, shape=get(plot_meta), color=get(plot_meta)), size=ptsize) +
    ggrepel::geom_text_repel(aes(x=PC1, y=PC2, label=pca_labs)) +
    scale_color_manual(plot_meta, values=cat_cols) +
    scale_shape_manual(plot_meta, values=cat_shapes) +
    ggtitle('PCA of "pseudobulk" samples') +
    coord_fixed() +
    theme_bw()

}
